{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56e63404",
   "metadata": {},
   "source": [
    "This is going to be so amazing you guys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "edf0a38a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Briefgarde\\anaconda3\\envs\\semesterproject\\lib\\site-packages\\tqdm\\auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "Some weights of the model checkpoint at bert-large-uncased-whole-word-masking-finetuned-squad were not used when initializing BertModel: ['qa_outputs.bias', 'qa_outputs.weight']\n",
      "- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pdf2image\n",
    "import pytesseract\n",
    "import nltk\n",
    "import hashlib\n",
    "from neo4j import GraphDatabase\n",
    "import pandas as pd\n",
    "from transformers import AutoTokenizer\n",
    "from zero_shot_re import RelTaggerModel, RelationExtractor\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a752c08",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package punkt to\n",
      "[nltk_data]     C:\\Users\\Briefgarde\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package punkt is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nltk.download('punkt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8b2a50c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"arX1v:2312.09317v1 [q-bio.TO] 14 Dec 2023\\n\\nA parallelized cellular Potts model that enables simulations at\\ntissue scale\\n\\nShabaz Sultan!?, Sapna Devi?, Scott N. Mueller?, Johannes Textor!?\\n\\n! Data Science group, Institute for Computing and Information Sciences, Radboud University, Nijmegen,\\nThe Netherlands\\n\\n? Department of Medical BioSciences, Radboud University Medical Center, Nijmegen, The Netherlands\\n\\n3 Department of Microbiology and Immunology, The University of Melbourne, at the Peter Doherty\\nInstitute for Infection and Immunity, Melbourne, Victoria, Australia\\n\\nSummary\\n\\nThe Cellular Potts Model (CPM) is a widely used simulation paradigm for systems of interacting cells that\\nhas been used to study scenarios ranging from plant development to morphogenesis, tumor growth and\\ncell migration. Despite their wide use, CPM simulations are considered too computationally intensive\\nfor three-dimensional (3D) models at organ scale. CPMs have been difficult to parallelise because of\\ntheir inherently sequential update scheme. Here, we present a Graphical Processing Unit (GPU)-based\\nparallelisation scheme that preserves local update statistics and is up to 3-4 orders of magnitude faster\\nthan serial implementations. We show several examples where our scheme preserves simulation behaviors\\nthat are drastically altered by existing parallelisation methods. We use our framework to construct tissue-\\nscale models of liver and lymph node environments containing millions of cells that are directly based on\\nmicroscopy-imaged tissue structures. Thus, our GPU-based CPM framework enables in silico studies of\\nmulticellular systems of unprecedented scale.\\n Introduction\\n\\nThe CPM! is a powerful simulation method that is known for its ability to simulate a wide range\\nof biological scenarios. When modelling tissue environments, many researchers wish to consider both\\nhe complex morphology of single cells, and the behavior of these cells as part of larger collectives.\\nThe appeal of the CPM lies partly in its ability to integrate both these aspects in a single model.\\nAs such, it has been applied to diverse problems including tumor growth,” vascularization,® wound\\nhealing,* and morphogenesis.>*® Many extensions and hybrid model variants are available in the literature\\n‘0, for example, add support for cell motility,” cell signalling,!° and finite-element based mechanical\\ninteractions.!! Thus, the CPM can be described as a generic platform upon which a wide range of models\\ncan be built, possibly explaining its popularity in the tissue modelling field.\\n\\nFor computational reasons, CPM models are often built using a small number of cells (perhaps in the\\nhousands) in simplified two-dimensional (2D) geometries. Some research questions in tissue modelling\\n\\nhowever demand a realistically large number of cells, and accurate 3D representations of tissues. For\\ninstance, it is widely appreciated that the efficiency of space exploration of motile cells depends criti-\\ncally on the dimensionality of the environment, leading researchers to use 3D CPM simulations.® !?:13\\nAssessment of tumor growth behavior and invasive patterns of tumors likewise changes as a function of\\ndimension, resulting in the use of 3D CPM simulations in the tumor modelling field as well.? While there\\nis thus a clear research demand for realistically large 3D CPM simulations, scaling up such simulations is\\ngenerally considered computationally infeasible.4 1° Removing this computational barrier would allow\\na host of CPM based modelling research to be scaled to what is biologically realistic, instead of scaling\\nto what technical constraints allow.\\n\\nThe CPM is an on-lattice method: simulations are executed on a grid. A common strategy to speed up\\ngrid-based models is to cut the simulation into smaller subsections that are run in parallel. However, for\\nthe CPM, model behavior is intimately tied to locations on the grid being updated sequentially, in a single-\\nlocation update Markov chain Monte Carlo process. As such, parallelisation has been cautioned against,\\nas it risks drastic changes in CPM “system kinetics”.° Nevertheless, several attempts have been made to\\nemploy subsection-based parallelisation. Chen et al. proposed a checkerboard scheme that divides the\\nCPM lattice up over a computer cluster, with sections chosen much larger than cells, preventing cell-level\\nsynchronization issues.!” Tapia et al. propose a similar scheme for GPUs:!® because the sections are much\\nsmaller than a cell on the GPU, they use a per-cell locking mechanism to prevent simultaneous updates of\\na cell. Neither approach has been extensively checked for correct model behavior. To our knowledge, no\\npublicly available implementation of subsection-based parallelisation exists for the CPM. The software\\n“Morpheus” !® includes a non-subsection-based Central Processing Unit (CPU) parallelisation strategy;\\nthis method is less suited to fully utilise parallel accelerator hardware such as GPUs.\\n\\nIn this paper, we aim to make several contributions: First, we introduce a highly optimized parallelized\\nimplementation of the CPM using several optimization strategies not previously explored in the literature.\\nSecond, we conduct an empirical analysis of potential pitfalls in parallelising a model dependent on serial\\nupdates of single locations and show that our approach mitigates these issues while preserving performance\\nbenefits. Third, we demonstrate that our approach retains the behavior of cell motility simulations that\\nare particularly prone to changed model kinetics upon parallelisation. And finally, we build two large 3D\\n\\nimage-based simulations of T cell migration in liver and lymph nodes, demonstrating how complex tissue\\nenvironments can be analysed at a larger scale and realism than was possible before.\\n\\nResults\\n\\nSerial Cellular Potts Models are limited to sub-millimetre scale\\n\\nThe CPM’s serial update schedule — only one location at a time is updated on a grid — is central to\\nits behavior, and thus not trivially changed. We therefore first aimed to quantify the computational\\nlimitations for models based on the serial CPM: how big can our computational experiments be, and\\nwhat timescale can they span, without making fundamental changes to the CPM?\\n\\nWe set up 3D CPM tissue experiments at different spatial and temporal scales and measured the required\\nwall-clock time. We choose spatial and temporal simulation resolutions that allow reasonably accurate\\nrepresentations of highly motile cells within tissues. Lymphocytes are among the smallest nucleated cells\\nin the mammalian body at 8-10um diameter for B cells and T cells. At speeds of up to a few body\\n oO\\n£\\nx\\n(2)\\n2\\n2\\noS\\n=\\n64° 256° 1024°\\nsimulation volume (um’)\\nB\\nial lyr °\\nlo | 30 values/param.\\n> - oO\\n2 € 1wk 7\\n5 c= a = , -\\n[sj x e Pa\\n£ jw || 8 ow\\n6 cS ihr 4—_-\\n2 L | 2\\noO ras o\\n° y =\\nOS Tmin 10 values/param.\\nL 4\\npolarisation strength replicates 1 2 3 4 5\\n\\nparameters\\n\\nFigure 1: Serial CPM implementations do not scale well to large simulations and big en-\\nsembles. (A) Computational cost of simulating a 256°m* volume (1m® per voxel) for 10 simulated\\nminutes (1s per simulation step). With a high-performance CPU implementation, this takes 1.2 hours of\\nwall-clock time. Extrapolation to a full day of simulated time for a 1mm? simulation estimates 1.4 years\\nof wall-clock time. Time measurements (black) are averages of 5 simulations, projections are based on\\ninear fits of simulations > 64°m* in log-log space (grey). (B) Building CPMs typically requires tuning\\nparameters. Here we perform a multidimensional grid search for parameters that yield motile cells using\\nthe Act extension of the CPM;’ a modest single parameter grid search where 10 values are tried takes\\n5.9 seconds; larger grid searches take exponentially more time as function of parameters searched, with a\\nfive parameter search with 30 values is extrapolated to 1.1 years of wall-clock time. Time measurements\\n(black) are averaged over 5 simulations, projections fitted on linear trends in linear-log space. CPM\\nparameters are shown in Table 1 and Table 2.\\n\\nengths per minute, they are also among the fastest movers within tissues. We chose lum? per voxel and\\nsecond per simulation step, or Monte Carlo Step (MCS), as our simulation resolutions, corresponding\\nto Beltman et al.9\\n\\nTo create realistically large models of, for example, a complete murine Lymph Node (LN), or scale a\\ntumor model to the size of an avascular tumor spheroid,?° experiments at mm® scale would be needed.\\nCertain organ scale dynamics require timescales of at least a day, such as T cell transit through LNs.?4\\nMore modest simulation experiments at 256°um? (Figure 1A) at 10 simulated minutes require ~1h of\\nwall-clock time on a highly optimized CPU implementation of the CPM (Methods). Such experiments are\\nsufficient to see minimal movement of motile cells such as T cells and can capture organ compartments\\nsuch as a LN paracortical T cell zone, or a micrometastasis for cancer modelling. For mm? scale, one-day\\nexperiments however, researchers would have to wait at least a year — making them impractical for all\\nbut the most patient of researchers.\\n\\nA single experiment is rarely sufficient, in both empirical and simulation research. Beyond the need\\n‘or experimental replicates, simulations can have parameters for which associated model behavior is non-\\nobvious, requiring parameterized simulation ensembles. In Figure 1B, we explore a cell motility extension\\nto the CPM that coarsely models actin dynamics.” A single two-dimension simulation is relatively cheap,\\nbut when exploring even a modest number of parameters the execution time goes up exponentially.\\n\\nThus, serial 3D CPM simulations at a scale of 1 mm® are currently not feasible, and even 100°um? scale\\n\\nsimulations were extremely challenging in our experiments. While ensemble runs can be accelerated with\\nparallel simulations without changing the CPM, the high computational costs of running large ensembles\\n‘urther motivates the need for a faster CPM.\\n\\n Designing a parallel execution scheme for the CPM\\n\\nTo make large scale simulations practical, we propose a parallelisation scheme for the CPM that uses\\n\\nhighly parallel GPU hardware. A natural way\\n\\ndivide the simulation grid in\\nsupport up to thousands of\\n\\nsmal\\ncalcu\\n\\n‘0 parallelise on-lattice models, such as the CPM, is to\\ner sections; each section runs a local simulation in parallel. As GPUs\\nations in parallel, these sections will be quite small. The CPM also\\n\\nupdates information outside of the grid, such as sizes of simulated cells. We first consider three issues\\nwhen designing our parallelisation scheme: adjacent sections with active simulation interfering with\\n\\neach other during updates o:\\nsimultaneously, and stale in:\\nupdated by another simulati\\n\\nA CPM simulation proceeds\\n\\ncopying the position’s state to a ne\\n\\nsimulation running, we run t.\\nboundary. Such “race condi\\nalmost simultaneous update\\nstandard problem in paralle\\n\\nf the\\non.\\nby cl\\n\\nhe ris\\nions”\\n— wi\\n\\nisatio\\n\\nto be active at the same time. As\\n\\nsections are sufficiently spaced to prevent overlappin,\\nbetween active and dormant state.\\n\\normal\\n\\nsimulation gri\\nion being used by one paralle’\\n\\nhoosing a random position on\\n\\nk of simulations “racing” to wri\\nrun the risk of information ge\\n\\nforward solu\\noard” patte\\n\\nm, with a straight-\\naced out “checker\\n\\nThe CPM also tracks information involving more than one gri\\n\\nsimulated cells. We face the same “race condition” problem here,\\nupdates touch a single cell (Figure 2B, left column).\\nsmall active subsections, making such concurrent up\\nscheme is proposed, that rejects all but the first concurrent update attempt on a varia\\nmiddle column). We instead opt to use atomic operations (Figure 2B, right column), which equa!\\nprevents update information from getting lost, as the hardware will eventually integrate all updates\\n\\nsuch operations.\\n\\nIn serial\\nevery W\\n\\nmay be\\nTapia e\\nof-date\\nmay to.\\nbeing s\\n\\ndesigned an additional scheme that sti\\n\\nGPU pai\\n\\nCPMs, global information travels instantaneously: as only one grid position is w\\n\\nhout incorporating information of the overwritten update. This\\n\\ng updates (Figure 2A). Regions cycle perio\\n\\nates more likely. In Tapia et a\\n\\n; non-grid information such as cell sizes being updated\\n\\nsimulation after said information got\\n\\nthe simulation grid, and stochastically\\n\\nighbour. If adjacent subsections of the grid each have an active parallel\\n\\ne to the same location on the sections’\\nting lost, as one section overwri\\n\\nion: never allow two adjacent sections\\nrn of sections is only ever active; active\\nically\\n\\nof\\nel\\nof\\n\\nlocation, such as the current sizes\\nas it is possible that multiple paral\\nrallelisation requires a high number\\n1.18 a per-cell locking\\nble (Figure 2B,\\nly\\nor\\n\\ndated at a time,\\n\\ndate attempt has access to the results of all previous updates. This instantaneous information\\naccess is relaxed in our proposed scheme: when information such as cell size is accessed,\\nout-of-date due to still in-flight updates to the accessed information. The cell-loc!\\nt al. prevents this from happening, by outright rejecting any updates that cou\\ninformation; we explored a less strict scheme, reasoning that the stochastic CPM copy attempts\\nerate out-of-date information. For CPM extensions that do not tolerate their global variables\\nightly out sync with each other (e.g. variables used in center-of-mass “centroid” tracking) we\\n1 will keep variables in syne with each other, and sits on top of our\\n\\nthis information\\ning scheme from\\nd have used out-\\n\\natomics based scheme (Figure 2C, Methods). Otherwise, due to atomic updating, our scheme still gives\\nstrong guarantees on tracked data being correct — for example, cell sizes will be tracked correctly — and\\n\\nthe added flexibili\\n\\nImplementation o\\n\\nwe cache local da\\nsimulation calcula\\n\\nF our scheme on rea\\nGPUs, memory access almost always\\nprevents easy use of a GPU’s caching hierarchy.!® To maximize performan\\nscheme that carefully operates in loc\\n\\nbottlenecks performance —\\n\\nhardware requires consideration o:\\n\\ny allows us to explore the effects of allowing the use of slightly outdated information.\\n\\nOn\\nCPM’s random access pattern\\nce, we designed a data transfer\\n\\nf several technical details.\\nbut the\\n\\nk-step with checkerboard work splitting (Figure 2D, Methods):\\na needed by the regions that are becoming active, an\\ntions to fully utilise\\n\\noverlap memory transfer and\\n\\nGPU hardware. To increase simulation volume, and gain further\\n\\nspeed ups, we also prototyped a multi-GPU scheme (Figure 2E, Methods).\\n\\nIn short, our proposed parallel CPM sc!\\n\\nin running paralle\\n\\nheme is tailored to make opt\\n\\ncode. It still gives strong guarantees on simulation sta‘\\n\\nimal use of GPUs: hardware specialized\\n\\ne consistency — for example, a\\n\\nglobal variable tracking a cell’s size will always equal the number of pixels on the grid occupied by that\\n\\ncell. Yet, we do all\\n\\nflexibility and further maximise performance gains.\\n\\now the use of slightly outdated values of such global variables in copy attempts to gain\\n ba)\\n\\n?-\\n\\nB race condition mutex locking atomic updates\\n(Tapia et al.) (our method)\\n(lel=12) (lla i= 12 |}\\n\\nold old *\\n\\n| ASP—$$ $l new = 13)- - - [le] = 13 &)- - - {le] = +18)\\nnO OO Oo Hii --(=tie---(41=19\\n@©008000\\neee0\\n\\nreese ew. = 13)- - - (lel := 13 )- - - {\\n? | ——{l*],.4 = 13)- - -(lo = 13 w)- - - |\\n\\nc I*]s:= 4 I#lnou = 5\\n© Xu Yous *= (2-8, 3.3} © Xen? Voow = (3; 3} (Kumn= 11) (Yaun= 13} (lel= 4)\\n5\\n4 ee [Xeon 2= +4! Yeu = +2 | lel= +1 | Be]\\nXon= 15) (Van= 15) (Iel=5)\\n[new = Xeun/@l = 3; Yoow = Yeun/l#] = 3)\\n\\nE_ per GPU region border sync\\n\\n~-7--- global simulation barrier\\nSS caput sizessyne pu\\n\\n2\\n|e] = 23 je| = 29 |e| = 23 |e| = 29\\n[Je|:= +4 Jel: +14=he! = +0 |el:=-5 |\\n\\n|e] = 27 |e|=25 |e] = 27 je|=25\\n\\nFigure 2: Designing a parallel CPM based on checkerboarding and atomic updates for\\nGPUs. (A) We divide simulation space into a subgrid, with alternating regions being active in a\\n“checkerboard” pattern, preventing update conflicts on the grid. (B) We implement two strategies to\\nprevent global variable update conflicts (e.g. cell size tracking): a mutex locking strategy (Tapia et\\nal.'8), and a novel atomic operation strategy. (C) Combining interleaved storage and atomic operations\\non adjacent data allows variables to be kept in sync with each other. (D) A four-step execution scheme,\\nwhere active checkerboard areas are transferred to fast shared memory, accelerates copy attempts. Thread\\nbarriers synchronize active threads after transfers. (E) A multi-GPU execution scheme, where regions\\nof the simulation are assigned to different GPUs. Borders are synchronized between relevant adjacent\\nregions, and global variables are synchronized through delta tracking.\\n\\nglobal memory (slow)\\n\\non\\n A T, : time in region time divergence thread barriers\\n\\nmm : time for one copy attempt ———-—— T,\\nid ‘. :\\n0 Oe\\ne\\n\\ne%\\nSys\\n\\nT,\\n5 checkerboard size thread size\\n\\n' loa! —~“—_.\\n\\ne ? ee : :\\n<@ e ‘|e eee ‘ eco\\né } ° 1|0@ : eee\\nao @ : : ee\\n\\nT, T. : :\\n\\n<>\\ncopy attempts before checherboard switch\\n\\nsingle update parallel update\\noo\\n\\neoeoeeen\\ntime since last copy attempt\\nee®@ r ) 4 copy attempts per switch, 1 copy attempt per switch,\\nnumber of copy attempts 4x4 area/thread, 8x8 area/thread,\\n2x2 area/checkerboard 4x4 area/checkerboard\\n| | | lh __\\n0123 45 0123 45 0123 4 5\\n\\ncopy attempts per location\\nbinomial distribution —\\n\\n0 1 2 0 1 2 0 1 2\\nwaiting times between copy attempts (MCS)\\n\\ngeometric distribution —\\n\\nFigure 3: Checkerboarding can drastically alter local update statistics. (A) To prevent threads\\nfrom racing ahead of others too much, thread barriers are imposed after each update. (B) Checkerboard\\nsize and switching frequencies determine the distribution of waiting times between consecutive updates\\nof the same locations. (C) Small checkerboard areas or infrequent switches can alter the waiting time\\ndistribution substantially compared to the single-update scheme. Update frequency histograms were\\ngenerated by simulating one MCS (256 updates) on a 16x16 grid. Waiting time histograms were simulated\\non the same grid using 100 simulated MCSs.\\n cell boundary type:\\n\\n@-@ dark-dark\\n\\n60% 1 @-© dark-light\\n-© light-light\\n\\n40%\\n\\n20%\\n\\n% of cell boundaries\\n\\n0%\\n\\n10 1000 100000\\ntime (MCS)\\n\\nFigure 4: Atomic updates lead to correct behavior in a simulation of cell sorting. A cell sorting\\nsimulation based on Graner et al.! is implemented with two strategies to prevent race conditions for cell\\nsize variables. The mutex locking strategy proposed by Tapia et al.!® leads to severe starvation in our\\ntests. Using an atomic update approach recovers the desired sorting behavior, closely matching sorting\\nspeed of the original CPM. The simulation was run with a 4x4 area per thread and 2x2 checkerboards,\\nbut other sizes were tested and did not make a significant difference for this simulation. Sorting trends\\nof different CPM implementations are averaged over 5 simulations. Simulation parameters can be found\\nin Table 1.\\n\\nAligning the lattice update statistics of parallel CPMs to serial CPMs\\n\\nOur novel parallel execution scheme guarantees simulation data consistency, but drastically changes the\\nway the simulation lattice is updated. We investigated statistical properties of copy attempts on the\\nsimulation lattice for different lattice positions; specifically, we analysed the relative ordering of copy\\nattempts, frequency of copy attempts, and waiting times between copy attempts. This allowed us to\\nquantify the divergence of parallel lattice update kinetics from the serial CPM. Such divergence could\\nlead to changed simulation behavior in complex models of cells and tissues.\\n\\nWithout guardrails, performing independent CPM updates in parallel would allow different parts of a\\nsimulated tissue to have been alive for different lengths of simulated time (Figure 3A). Modern GPUs use\\nthread barriers to synchronize parallel threads without strong computational performance implications.\\nWe set thread barriers after ever single copy attempt, so for every currently active checkerboard region\\nthere is only a maximum time divergence of one copy attempt between active regions.\\n\\nOur parallel execution scheme allows us to freely choose the size of subregions and frequency of active\\nregion switching (Figure 3B), without impacting data consistency guarantees. We measure frequency\\nand waiting times between copy attempts per position on the lattice in the standard, serial implementation\\nof the CPM (Figure 3C, left column). For our initial parameter choice, we found that underlying update\\nstatistics are drastically changed (Figure 3C, middle column): waiting times between copy attempts\\nbecame much more predictable due to the much more even distribution of copy attempts compared to\\nthe serial CPM. Using large parallel subsections with frequent active region switching allows non-parallel\\nupdate statistics to be closely recapitulated in our parallel scheme (Figure 3C, right column).\\n\\nWe thus find that parallelisation parameters have major influence on copy attempt statistics on the\\nlattice. By switching active regions of the checkerboard frequently, and avoiding too small areas per\\nthread, we can obtain statistics that closely resemble the serial CPM.\\n\\nEmergent cell and tissue behavior is maintained in a parallel CPM\\n\\nOur goal is to build parallel CPMs that are much faster than previous implementations but maintain cell\\nand tissue kinetics consistent with existing serial CPMs. We hypothesized that aligning the waiting times\\nbetween update attempts of our parallel CPM to the serial version should suffice to achieve such consis-\\ntency. To investigate this, we ran several simulation scenarios with different CPM variants, measuring\\nemergent behaviors on the cell and tissue levels and comparing these to serial CPM versions.\\n\\nWe first investigated the classic cell sorting simulation:! two mixed populations of cells with differential\\n A single update\\n4 copy attempts per switch,\\n4x4 area/thread,\\ntime 2x2 area/checkerboard\\nTeoo00\\ne\\ne\\ne\\ne\\ne\\nTy\\nay ay\\n= 0.3 £03 2\\n202 5 0.2 5\\n50.1 & 0.1 &\\na a\\n0.0 8 0.0 8\\n2000 6000 10000 2000 6000 10000\\ntime (MCS) time (MCS)\\nB » low polarisation strength\\n! I\\ni i\\n& ied |\\n2 ee qe 4S\\n' high polarisation strength '\\n. «) f 4 i\\nc c c\\n& & &\\n3 08 3 08 3\\n£ 0.4 £ 0.4 £\\n8 8 8\\n8 0.0 8 0.0 8\\n& & &\\n200 400 600 800 200 400 600 800\\npolarisation polarisation\\nCc 8 copy attempts per switch,\\n4x4x4 area/thread,\\n2x2x2 area/checkerboard\\n\\nparallel update\\n\\nlow density crowd\\n\\n“Ip\\n\\n~ “it ze\\n\\n+\\n\\n0.8\\n& 0.6 Gy\\n204 z\\n© 0.2 fe)\\n0.0 A\\n0.25 0.50 0.75 1.00 0.25 0.50 0.75 1.00\\ndensity density\\n\\neso\\noun\\n\\n1 copy attempt per switch,\\n8x8 area/thread,\\n4x4 area/checkerboard\\n\\nS\\nwo\\n\\noo\\n\\n2000 6000 10000\\ntime (MCS)\\n\\n0.8\\n\\n00 » awa\\n0.0\\n\\n200 400 600 800\\npolarisation\\n\\n1 copy attempt per switch,\\n8x8x8 area/thread,\\n4x4x4 area/checkerboard\\n\\n0.25 0.50 0.75 1.00\\ndensity\\n\\nFigure 5: Motility behavior matches serial CPM behavior when parallel CPM local statistics\\nare similar. (A) Cells in a circular chemokine gradient simulation move highly regular in a serial CPM\\n(low per-pixel entropy, shaded: 1SD); parallel simulations diverge slowly from reference behavior when\\ncopy attempt statistics are preserved (low per-pixel cross-entropy). (B) In the Act model of intra-cell\\npolarization,” stronger polarization strength leads to higher directional autocorrelation (measured for 750\\n\\nMCSs at 50 MCS intervals, averaged over 50 simulations, shading: 4\\n\\nemerges when update waiting time distributions are preserved. (C)\\n\\nt1SD); directional persistence only\\nPhase transition simulation in 3D\\n\\nwhere synchronous motility emerges at a density around 0.5. Cell order indices after 2000 MCSs are\\n\\nshown (averaged over 5 simulation runs, shading: 4\\n\\nt1SD). The phase\\n\\nransition only occurs when update\\n\\nwaiting time statistics are preserved. Simulation parameters can be found in Table 1 and Table 2.\\n adhesion (i.e., cells of the same type have stronger\\n\\nadhesion to each other than cells of different types)\\n\\nsort themselves in two homogeneous regions over time (Figure 4, grey lines). The mixing kinetics can be\\n\\nmeasured by tracking homogeneous and heterogene:\\n\\nbased parallelisation approach proposed by Tapia et a\\n\\nleads to a lack of sorting behavior; the simulation\\nconcurrent copy attempt is allowed per cell un\\nscheme allows expected sorting behavior to re\\n\\nremained consistent for all parallelisation paramete:\\n\\n‘ous cell neighbours over time.! Strikingly, the mutex\\n1.18 — while still being able to run a simulation —\\nencounters so-called “starvation”, because only one\\n\\ner this scheme (Figure 4, red lines). Our less restrictive\\n-emerge (Figure 4, black lines). The sor\\n\\ning behavior\\nrs we examined (Supplementary Figure $1).\\n\\nWe next examined three motility model variants of the CPM, where additional biases are placed on\\n\\no obtain cells that move persis\\nfunction of random membrane fluctuations. Bec:\\nneed to roughly maintain their target size, copy\\nsections for movement to occur. We therefore s\\nby changes to the waiting times between copy at\\n\\ncopy attempts\\n\\nThe first motili\\nikely to succeed when aligned with the underlyi\\nbias is fixed in this model — the field does not change\\nsimulations with the same initial conditions closely\\nWe quantified\\ncontains a cell in multiple simulations, where low\\nbetween simulations, and low cross-entropy indica:\\nwith reference serial runs. In simulations with\\nincreased quickly (Figure 5A, middle column); w.\\nrises much slower (Figure 5A, right column).\\n\\nNext, we analysed a model that simulates localize\\n\\nborder of a cell makes further copy attempts involving those\\ntimes between copy attempts are very important in this mo\\nin a small neighbourhood successfully expand at the same time; this is less\\n\\ny model we consider adds a chemotactic vector field to the CPM; copy attempts are more\\nng vector field (see Methods). Because the directional\\n\\nhis alignment using pixel-wise (cross-)entropy: we measure the probability that a pixe\\nentropy of\\n\\nhighly chan\\nhen the w\\n\\nently; in the standard CPM movement only occurs as a\\nause motile cells have a leading and a retrac\\nattempts need to sufficiently alternate between these cel\\npeculated that movement models might be more affecte\\nempts.\\n\\ning edge an\\n\\nover time — cells move in a highly regular fashion, an\\nalign even at a pixel level (Figure 5A, left column).\\n\\npixel probabilities indicated high agreement\\nes a high level of agreement of parallel simulations\\nged lattice update statistics, cross-entropy\\npdate statistics are preserved cross-entropy\\n\\ncell polarization,’ where recent pixel updates at the\\npixels more likely to succeed. The waiting\\nel, as cells start to move when many pixels\\nlikely when copy attempts are\\n\\nspaced out evenly. A polarization strength parame\\n\\ncell; varying the polarization strength interpolates motility be\\ncolumn). In paral\\n\\nmotility modes (Figure 5B, lef\\nwaiting time statistics this full range of behaviors\\nthey re-emerge when update sta’\\n\\nFinally, we examined a simulatio\\n\\nwith this vector, and the vector\\nmotility model involves informat\\n\\nis periodically upd\\nion tracked on the\\n\\nto the simulation grid, placing additional demands on a para!\\n\\ndensity cell collective exhibit Brownian movement,\\nas density is increased;?? this characteristic phase\\n\\nleft column). In a parallel CPM this phase transition disa,\\n(Figure 5C, middle column), but re-emerges when update\\n\\ncolumn).\\n\\nTaken together, these results demonstrate that a strict updat\\n\\nistics are preserved\\n\\nn of a tightly packe\\na preferred movement direction vector; copy attempts for a ce\\n\\ns how much polarization bias is added to a\\ntween Brownian and directionally persistent\\nel simulations with changed update frequency and\\noes not appear (Figure 5B, middle column), but\\n(Figure 5B, right column).\\n\\n3D col\\n\\ner contro\\n\\nlective of immune cells,° where each cell has\\n1 are more likely to succeed when they align\\nated to align with recent movement of the cell. This\\nattice, and additional information tracked external\\nel CPM. In a similar simulation model, low\\ntransitioning sharply to ordered collective movement\\nransition is reproduced in the CPM (Figure 5C,\\npears when update statistics are changed\\nstatistics are preserved (Figure 5C, right\\n\\ne scheme preventing parallel updates of the\\n\\nsame cell is not necessary for maintaining valid CPM behavior and can even lead to severe “starvation”\\n\\nwhere simulations do not progress anymore. On\\n\\nhe other hand, preserving waiting times between copy\\n\\nattempts involving the same pixel can be very important to maintain model behavior, and was critical\\n\\nin all three motility scenarios we examined.\\n\\nGPU-based CPM parallelisation accelerates models by several orders of mag-\\n\\nnitude\\n\\nHaving confirmed that our parallel CPM indeed pro\\n\\nduced correct model behavior in a range of examples,\\n\\nwe next investigated the performance improvements resulting from our implementation choices when\\n\\ncompared to a serial CPM.\\n % 1000 2\\n10 g 2\\n< 100 oO\\ns 8 g 8 & 19547x13805x\\nBe 3 3 10 3\\n®\\n2 3 < 1 g\\n24 a & 2\\n& 2 =\\n2 eg ol °\\n= 0.01 a\\nkernel specialisation ( ae ) 128° 5127 20487 8192! implementation\\nGPUs dimension\\nshared memory @ @ e@\\nD E area/thread: @ 4x4 A8x8 mi 16x16\\n> ° 03 +0\\n3 10000 3 9786x 2\\n®\\nS 1000 g = 02 = 0.9 >\\n& 400 ® § g 8\\n& 2 8 308 3\\noo «10 ® — 2902x ZO 8 8\\n£ 1 s G 07\\n0.0\\n642 256° 10242 implementation 1 4 16 64 1 4 16 64 0.0 0.1 0.20.3\\ndimension updates per switch updates per switch simulation time\\n\\nFigure 6: Parallel CPMs outperform serial implementations by orders of magnitude. (A)\\nPerformance benefits are obtained by a shared memory scheme, and a pre-compiled kernel specialization\\nscheme, based on a sorting simulation extended to densely fill simulation space (Figure 4, all sorting\\nbenchmarks are average of 5 simulations of 100 MCSs, 100 MCS burn-in). (B) Distributing the lattice\\nover 4 GPUs speeds up the simulation further. (C) Performance of sorting simulations are significantly\\nhigher relative to two serial implementations; relative speedup shown for 2048x2048 simulations. (D)\\nPerformance of 3D cell collective simulations is also improved (Figure 5C, average of 2 simulations of\\n100 MCSs, 100 MCS burn-in); relative speedup shown for 256° simulations. (E) Trade-offs between\\nspeed and accuracy for chemokine field simulations (Figure 5A, accuracy: 1 - cross-entropy of parallel\\nensemble relative to serial ensemble, 250 simulations per ensemble). CPM parameter settings can be\\nfound in Table 1 and Table 2.\\n\\nWe first examined the performance gains resulting from our use of a shared memory caching scheme\\n(Figure 2D, Methods) using a 2D sorting simulation benchmark (Figure 4). On its own, shared memory\\ncaching achieved only minimal speedup (Figure 6A). However, for further performance improvement,\\nwe pre-generated different versions of our code based on which CPM extensions were in use (Methods).\\nCombining these specialized kernels with the caching scheme achieved a significant 10-fold speedup, more\\nthan expected from either optimization alone (Figure 6A). Testing our multi-GPU scheme (Figure 2E,\\nMethods) in the same scenario achieved a further 2.3x speedup by distributing the simulation across\\nfour GPUs (Figure 6B). Hence, careful consideration of GPU-specific implementation choices resulted\\nin significant performance benefits.\\n\\nWe next compared our parallel CPM (on a single GPU) to our own optimized CPU implementation\\n(Methods), and to the state-of-the-art software “Morpheus”.!® Using the same 2D cell sorting bench-\\nmark, we observed a {3,500-fold speedup compared to our own CPU implementation, and a j25,000-fold\\nimprovement relative to Morpheus (Figure 6C). On the 3D cell collective simulation (Figure 5C), a\\n{,600-fold speedup was achieved relative to serial CPMs (Figure 6D). It is important to note that our\\nbenchmark hardware did not permit use of shared memory caching for the 3D model; future hardware\\nwith more shared memory may improve 3D performance.\\n\\nAs shown in Figure 3 and Figure 5, checkerboard size and switching frequency influences model behav-\\nior: larger divisions and more frequent switching leads to behavior concordant with serial CPMs. These\\nupdate patterns also interact with caching, further impacting performance. As the sorting simulation has\\nlow sensitivity to parallelisation parameters (Supplementary Figure $1), we instead benchmarked our\\n2D chemotaxis simulation (Figure 5A). We found that larger subdivisions indeed slowed simulations\\ndown considerably, whereas frequent switching led to a less marked slowdown (Figure 6E). These results\\nsuggested a trade-off between performance and model accuracy.\\n\\nHence, our parallel CPM represents a significant speed improvement over even well optimized serial CPMs\\n\\n10\\n such as Morpheus. This improvement was to a substantial extent a result of GPU-specific optimizations\\nand led to an overall speedup of several orders of magnitude.\\n\\nBuilding an image-based model of T cell surveillance of the liver\\n\\nOur parallel CPM was motivated by the need for tissue simulations at realistic scale with realistic cell\\nmorphology. While our approach increases engineering complexity of the model, it can also simplify\\nmodel building, as it enables direct use of imaging data to initialize the simulation environment. To\\ndetermine the feasibility of directly using real-world imaging data in our system, we constructed a model\\nof T cells moving in the sinusoidal regions of the liver.\\n\\nTissue resident memory T cells (Trms) are confined to specific organs, where they sample their envi-\\nronment for signs of disease”+ — in particular, liver Trms are key in providing sterilising immunity for\\nmalaria.?° Trms display a distinct motility pattern in the liver, and efficacy of malaria clearing depends\\non the number of Trms present.?® We constructed a 3D simulation environment of liver sinusoids — capil-\\nlaries that Trms reside in and migrate through — and combined it with a polarized cell motility model” to\\nrepresent the Trms (Figure 7B). Trm motility parameters were derived from a LN model constructed in\\n\\nparallel, and our parameter choice will be explained in more detail with the LN data. These parameter\\nsettings lead to motility modes also observed in intravital liver data: intermittent movement and pausing\\nof Trms. The goal of our simulation was to predict how much of the liver volume would be explored by a\\ngiven number of T cells within a set amount of time; this same question was addressed by Fernandez-Ruiz\\net al. using a more abstract mathematical model.?°\\n\\nAs can be expected, we found that liver exploration by the Trm population increased with the number of\\n\\nexploring cells, and with the amount of exploration time (Figure 7C). Remarkably, the sinusoid volume\\nexplored by an individual cell increased dramatically as the total Trm population was raized from 1,000\\nto 10,000 cells (Figure 7D). Further investigation revealed the source of this phase transition: cells in\\nour simulations can be in an unpolarized state, in which they do not move persistently; collisions can\\ntrigger polarization of stationary cells. Since such collisions occur more frequently at higher densities,\\nthis causes a larger fraction of cells to be motile (Figure 7E). Thus, if Trms can help each other to\\nmove, a larger Trm population would screen the environment more efficiently than expected from the\\npopulation size alone.\\n\\nNotably, the mean speed of Trms peak at ~10.5 um/min (Figure 7F), matching average speeds in our\\nlymph node simulations — but only at densities where the arrest coefficient is close to 0. In two-photon\\nexperiments, liver Trm speed is reported to be significantly lower than T cells in lymph nodes;\\nreproduce this in our simulations with moderate cell density, which leads to both a realistic slowdown,\\n\\nand higher arrest coefficient. The simulated liver Trm speed and arrest coefficient closely matches previ-\\nously reported values in literature,?? where they were used to quantify observed characteristic patrolling\\nbehavior of Trms.\\n\\nThese liver experiments demonstrated that our system allowed us to combine 3D imaging data with a\\nrelatively realistic model of polarizing cell motility. The resulting simulation produced an interesting\\nphase transition of Trm screening efficiency. While we do not claim that such an effect is present in the\\n\\nreal liver — more detailed tuning of the motility model to match the characteristics of real sinusoidal Trm\\ncells would be out of scope for this simple proof of concept — this does illustrate how detailed models such\\nas the CPM generate emergent behavior that is not an immediately obvious consequence of the model\\ningredients and can generate interesting biological hypotheses worthy of further exploration.\\n\\nA T cell collective moves robustly in a realistic, organ-scale model of the lymph\\nnode\\n\\nIn LNs, millions of T cells move in search of antigen-presenting cells,?!°° and finding the cognate antigen\\n\\nis a key early step in adaptive immune responses.*! T cell motility in the LN is shaped by the high\\ndeformability of the cells®? and their interactions with a complex stromal environment, the fibroblastic\\nreticular cell network.*3:°4 Given the LN’s unique structure as an organ consisting mainly of moving\\ncells, it presents a challenging test case for any tissue simulation system. Indeed, many models of LN\\nT cell dynamics already exist, but they were often limited in either cell count and simulation volume, or in\\nrealism of cell morphology and LN environment. For example, T cell interactions with a synthetic stromal\\nstructure have been investigated using 3D CPM simulations of thousands of cells,° while collective motility\\n\\n11\\n at a more realistic scale of hundreds of thousands of cells has been investigated with non-morphological\\nagent-based models.*°:3° We were therefore interested in our system’s ability to simulate a realistically\\nlarge collective of millions of realistically shaped T cells. As in the previous use case, we used microscopy\\ndata to initialize — and parameterize — our model.\\n\\nWe constructed a LN simulation using the work of Beltman et al.9 as a starting point. To improve\\nrealism, we used an imaged LN section to initialize the stromal structure (Figure 8A), we used the\\nAct-CPM extension for cell motility,” and we fit the two Act motility parameters directly to intravital\\nimaging data (Figure 8B-E). When fitting these parameters, we aimed to reproduce the intermittent\\nstop-and-go motion®” observed in real data whilst also quantitatively approximate the real speed distri-\\nbution. Our system allowed fitting to be performed on realistically large simulations (256um diameter\\nsimulation volume, containing 50,135 T cells, 20 minutes simulated time). Crucially, we did not use\\nperiodic boundaries which, in dense simulations, can lead to artificial global synchronisation of motion\\nin the cell collective.\\n\\nWe used the KL divergence to compare simulated to real speed distributions (Figure 8B). Increased\\nadhesion of T cells to the stromal structure made the speed distribution more right-skewed (Figure 8C,\\ntop), which was more consistent with real data (Figure 8C bottom). For most examined parameter\\nsettings, we found the pronounced variations in cell speed along cell trajectories that are also observed\\nin real data (Figure 8D). When we performed grid searches of motility parameters at two different\\nstromal adhesion strengths (Figure 8E), both grid searches had broadly similar fit landscapes, and the\\noverall lowest KL divergence values were very similar. Qualitatively, we found more right-skewed and\\nbimodal speed distributions in high-adhesion than low-adhesion simulations for a subset of the motility\\nparameters (Supplementary Figure $2). Based on these fitting results, we chose the best fit motility\\nparameters of the high adhesion setting for our further simulations.\\n\\nWe next investigated organ-scale collective T cell motility by scaling up our simulated volume to 1mm\\ndiameter, now containing 3,359,752 T cells (Figure 9A). While this is not entirely realistic — a real LN\\nhas a more complex compartmental structure, while we simulate only a T cell zone — our main intention\\nwas to determine the spatial range of collective behavior that can now be investigated with our system.\\nWe simulated one virtual day in 1.9 days of wall clock time — where serial implementations were projected\\n‘o have taken a full 1.4 years (Figure 1A).\\n\\nWe used a mixing simulation of two differentially labelled populations as an easily observable proxy for\\ncollective search (Figure 9B). In the simulation, it takes about 1 hour to reach a state where 50% of the\\ncells are next to at least one cell of the other population (Figure 9C). While stromal adhesion influences\\nintra-cellular stop-and-go motility, it has minimal influence on collective motion: high and low stromal\\nadhesion simulations have broadly similar mixing speeds. While collective motility is slightly ordered at\\ncell-scale distances, no ordering persists at organ-scale (Figure 9D).\\n\\nIn summary, we used an organ-scale LN simulation as a challenging test case: an environment known for\\ncomplex, large scale collective motility dynamics. Our system allowed us to integrate of microscopy data\\ndirectly into our simulation, and to fit motility parameters directly on two-photon data. We investigated\\nthe behavior of the fitted cell population at organ-scale (millions of cells). We hope that this will inspire\\nfurther research on large cell collectives — integrating various sources of empirical data, at a high level of\\nrealism, and at realistic scale.\\n\\nDiscussion\\n\\nWe have implemented a parallel version of the CPM that, in some experiments, was 3 to 4 orders\\nof magnitude faster on GPUs than a state-of-the-art serial CPM. To achieve this, we designed novel\\nschemes for lattice work splitting and synchronization, maintaining global state, efficient GPU memory\\ntransfer, and multi-GPU execution. We found that preservation of local lattice update statistics was\\nsufficient to maintain high-level system dynamics, despite removing the sequential nature of the original\\nCPM algorithm.\\n\\nIn our work we aimed to ensure that important properties of CPM simulation trajectories were main-\\ntained. Some previous CPM parallelisation efforts instead focused on equilibrium state distributions as\\nimportant quantities. For example, Chen et al.!? argue that their parallel CPM maintains certain equi-\\nlibrium properties. While such arguments are frequently applied in the analysis of energy-based models\\n\\n12\\n in physics, Voss-Béhme et al. showe\\n\\nWhile we have shown that the same\\nCPM in our benchmark scenarios, thi\\nparameters that were originally fitte\\nto re-fit their simulation parameters\\n\\nespecially with com\\n\\nplex models such\\n\\nthat these are less relevant for the CPM because the Metropolis\\nalgorithm used in the CPM drastically alters longtime behavior of the mode\\n\\nrange of behaviors is present in our parallel CPM as in a serial\\ns would still need to be carefully checked when re-using simulation\\non a serial implementation. Specifically, researchers might want\\nwhen using our system instead of a serial\\nis sensible in general when moving between different implementations of the s\\nas the CPM where implementation details might anyway differ\\n\\nversion — advice that\\name model formalism,\\n\\nsubtly. Furthermore, the observed relation between model kinetics and performance (Figure 6E) also\\nsuggests it might be fruitful to explore more aggressive parallelisation settings that might yield further\\n\\nspeed-up whilst maintaining desired simulated behavior.\\n\\nThrough accelerating the CPM on accessible GPU hardware, we have increased the range of feasible\\nsimulation sizes — because our scheme fully utilises all GPU cores, simulation size is now mostly limited\\n\\nby the available GPU memory. As memory per GPU increases, our scheme wi\\nti-GPU prototype scheme will allow for further increase of simulation volume. Our\\nto include inter-GPU communication over MPI,*° allowing\\nding block for a GPU cluster-based version of the CPM —\\nher to lcm? scale.\\n\\nsimulations; our mul\\n\\nmulti-GPU scheme could be further extende\\n\\nour system to be us\\npotentially scaling o\\n\\nLarge-scale multicel\\nenvironments.\\nare morphologically\\ncell shape into acco\\nour example simula’\\n\\n14, 15, 35, 36\\n\\ned as a fundamental bui\\n\\nur Imm? simulations fur\\n\\n1 easily scale to larger\\n\\nular simulations are often performed using more coarse-grained agent-based model\\nOur high-performance CPM now allows researchers to perform simulations that\\nmuch more detailed and allow the investigation of new questions that require taking\\nunt, such as migration in complex environments or morphogenesis. We hope that\\nions of liver and LN environments will serve as inspiration for many such spatially\\n\\nresolved tissue and organ simulations in the future. Similar approaches to the one shown here could be\\nuseful for parallelisation of CPM variants or other sequential lattice-based models.®\\n\\n13\\n A B mmm 100 yum explored space @ T cells\\n\\nfe)\\n\\n3 g 16\\n\\nSs Sa\\n\\na ao\\n\\ng 38 1.0\\n\\noO oO\\n\\n°° oa\\n\\n3 % 0.5\\n\\ncy Cy\\n\\nS 10 1000 100000\\nE i F cells\\n\\n_ 1 10.0\\n\\n5 0.8 =\\n\\n2 € 75\\n\\n@ 0.6 E so\\n\\n204 3\\n\\n3 3\\n\\n0.2 Q 25\\n\\n0.0 0.0\\n10 1000 100000 10 1000 100000\\ncells cells\\n\\nFigure 7: T cells in a liver sinusoid simulation exhibit a transition from a non-motile to\\na motile state. (A) A 1024x1024x182 um® scan of a liver sinusoid is directly imported into our\\nsimulation environment (64ym slice shown). (B) Simulated T cells explore liver sinusoids (5000 cell\\nsimulation shown). (C) Simulations at different cell densities explore the environment at different speeds\\n(single simulation of 24hr, with 1 min snapshots). (D) An increase of explored volume per cell occurs\\nwhen cell density rises from 1,000 to 10,000 cells; 100,000 cells fill ~1/3 of sinusoid volume, causing\\nexploration efficiency to saturate. (E) At low cell densities most simulated cells are in an un-polarized,\\nnon-motile state; at higher densities collisions cause cells to transition in to a polarized, motile state (1\\nsimulation of 30 min per density, arrest coefficient: fraction cells ; 2um/min,?? shaded: Wilson score\\ninterval based on the number of 2-minute subtracks). CPM parameter settings can be found in Table 3.\\n(F) Average speed of the simulated T cells at different densities.\\n\\n14\\n 1-15 ym 16-30 pm 31-45 im 46-60 pm 61-75 um 76-90 um =: 91-110 um\\n\\nB simufation =e D jowadhesion E — towadhesion low adhesion\\n2 2 ey\\n5 5 Ey\\n: :\\n8 8\\na a s\\n2 2 Da\\n3 3 a 5\\n0 10 20 30 0 10 20 30 SOHm =m 3\\nspeed (\\\\um/min) speed (um/min) high adhesion & high adhesion high adhesion\\noO\\nEy\\no oS\\n8 0.40 @ 0.25 ee eh cc\\n5, = 0.20 a\\n® 0.05 20.15\\nES $ 0.10\\n3 0.00 ge 3 os\\nZ 0.00\\n10 20 30 adhesion o 4 4. ° % polarisation decay 10 30\\nspeed (um/min) speed (um/min)\\n\\nKL divergence\\n\\nFigure 8: Reproducing in vivo T cell motility characteristics in a CPM. We compare intravital\\ndata from Mandl et al.?!. to the motion of 50,185 CPM cells moving in a spherical structure (256um\\ndiameter). (A) A 130x130x110 um* scan of the stromal structure (fibroblastic reticular cells) in the\\n\\nparacortical T cell zone of a murine LN. The images show average intensity projections of diffe\\nranges. After segmentation and mirroring, these data are used to fill a spherical simulation vo\\n\\nrent depth\\nlume. (B)\\n\\nSpeeds observed in simulated data are used to construct a probability distribution (upper panel, solid),\\nwhich we compare to the empirical distribution (upper panel, dashed) using the Kullback-Leibler (KL)\\ndivergence metric. To this end, we bin the speed distributions, compute the divergence in each bin, and\\nsum the values for each bin (lower panel). (C) Increased adhesion of T cells to stromal cells causes\\n\\nthe speed distributions to be more right-skewed (top) and match real data more closely (bot:\\n\\nom). (D)\\n\\nIncreased adhesion can slow T cell paths non-uniformly by creating longer and/or more frequent pauses\\n\\n(red) but does not prevent rapid motion altogether. Two representative tracks are shown. (\\n\\nE) A grid\\n\\nsearch of motility parameters for cell polarization reveals which parameter ranges best fit the intravital\\n\\ndata. Both low and high adhesion simulations have similar fit landscapes; the best fit is\\n\\nound in a\\n\\nhigh adhesion simulation, although the difference to the low-adhesion best fit is small (1 simulation per\\n\\nparameter combination, best fit: red box/red line). CPM parameter settings can be found in\\n\\n15\\n\\nTable 3.\\n 3,359,752 simulated cells\\n1pm simulation resolution\\n\\nO mins 40 mins 80 mins 120 mins\\n\\nLe)\\ni=]\\n\\n0.8\\nD 06 5 0.6\\n& 0.4 304\\nE02 20.2\\n0.0 0.0\\n\\n0 30 60 90 120 30 100 300 1000\\n\\ntime (in min) scale (in um)\\n\\nhigh adhesion low adhesion\\n\\nFigure 9: Mixing of motile T cell collectives in a realistic-scale LN simulation. (A) Spherical\\nLN simulation at 1mm diameter containing 3,359,752 T cells and stromal cells (grey; Figure 8A). For\\nclearer visualization, only 10% of the simulated cells are shown, and red and green cells are equally param-\\neterized although coloured differently. (B) The simulation is initialized with two equally parameterized\\npopulations each starting at one half of the sphere. Mixing is quantified by tracking the fraction of cells\\nhaving at least one neighbour of the other population. (C) Global mixing is slower in the high-adhesion\\nsimulation (red) than at low adhesion (black), although the difference is minor. Both simulations reach\\n50% mixing in ~1h. (D) Cells do not synchronize their directionality much at longer spatial scales, and\\nthis absence of long-range synchronization is the same for low and high adhesion. CPM parameter set-\\ntings can be found in Table 3.\\n\\n16\\n Tables\\n\\nTable 1: Model settings for 2D simulations. For all 2D simulations we use Viarget = 2500, Atarget =\\n600, Aperimeter = 2; Jcett,ecm = 10 and T = 10. The vector field in the 2D circular chemotaxis simulations\\n(Figure 5A) consists of vectors at a 90 degree angle to radial vectors from points in the field to the\\nsimulation center, with all vectors in the field normalized to length 1.\\n\\n| Avotume | Aaet | Maxact | Achemotaxis\\nFig 1B 20, 30, 40 | 200, 250, 300, 350, 400 50 —\\nFig 5A, Fig 6E | 20 = — 10\\nFig 5B 20 50-800, in increments of 50 | 50 —\\n\\nTable 2: Model settings for 3D simulations. For all simulations Viargee = 150, Avolume = 25,\\nAtarget = 1500, Aperimeter = 0.2, Scett,ce = 10 and T = 20. The time resolution is 1 second per Monte\\nCarlo step.\\n\\n| uw _| At | Apersistence | Jcell,BCM\\nFig 1A — f—|[— 10\\nFig 5C, Fig 6D | 0.95 | 15 | 40 0\\n\\nTable 3: Model settings for liver and lymph node simulations. For all simulations Viarge, = 150,\\nAvolume = 25, Atarget = 1500, Aperimeter = 0-2, Jcett,cett = 10, Jeet,ncm = 0, and T = 20. The time\\nresolution is 1 second per Monte Carlo step.\\n| Aact | Maxact Teell, FRC\\n\\nFig 7 2600 10 0\\n\\nFig 8B | 2600 10 0\\n\\nFig 8C | 2600 10 -400, -300, -200, -100, 0\\n\\nFig 8D | 2600 10 -400, 10\\n\\nFig 8E | 200-3000, in increments of 200 | 10, 15, 20, 25, 30, 40, 50, 60, 70 | -3000, 10\\n\\nFig 9C | 2600 10 -800, 0\\n\\nFig 9D | 2600 10 -800\\n\\n17\\n Acknowledgments\\n\\nThis research was supported by Young Investigator Grant 10620 from KWF Kankerbestrijding and HFSP\\nprogram grant RGP0053/2020 (to JT). JT was also supported by NWO Vidi grant VI.Vidi.192.084.\\nWe thank Franka Buytenhuijs for giving valuable feedback on this manuscript, and we thank L.W.A.\\nStuurman and Pedro Aroca Lara for testing early versions of this simulation system. We also gratefully\\nacknowledge the support of NVIDIA Corporation with the donation of the Titan V GPU used for this\\nresearch. The funders had no role in design, execution, and interpretation of the research reported in\\nthis manuscript.\\n\\nAuthor contributions\\n\\nSS and JT conceived the study. SS designed the parallel CPM algorithms. SS performed the research and\\nanalysed the data. SS and JT wrote the manuscript. JT supervized the project. SM provided critical\\nfeedback and provided the liver and lymph node imaging data. SD performed image acquisition of the\\nliver and LN data. All authors critically revized the manuscript for intellectual content.\\n\\nDeclaration of interests\\n\\nThe authors declare no competing interests.\\n\\n18\\n Materials and Methods\\n\\nCellular Potts model and model variants\\n\\nThe CPM is a lattice based formalism for simulating biological cells. Simulations proceeds with a Markov\\nchain Monte Carlo process, where a single position at a time can transition to one of the states of its\\nneighbours. The transition probability is based on the change of a Hamiltonian energy function. The\\noriginal CPM has\\n\\nH =~ Ip(ox),r(0))(1 — 84:,0;) + ¥ > Avotume(Te)(V (0) — Veargot(Te))? (1)\\nig o\\n\\nas its Hamiltonian, with i, 7 being possible neighbour locations in the lattice, 7; the cell present at location\\ni, T is the type of a particular cell, 5 is the Kronecker delta, V(c) is the number of lattice sites a cell o\\noccupies, and Ayolume(T) how strongly a volume constraint contributes to the Hamiltonian for a particular\\ncell type. In all our simulations a Moore neighbourhood was used. Transitions stochastically succeed\\nbased on the change in energy AH, using a Boltzmann probability distribution\\n\\n1 if AH <0,\\nP(AH) = {laws if AH > 0. (2)\\n\\nBy adding extra terms to the Hamiltonian, constraints can be added to cells and new types of behavior\\nintroduced to the model. When cell shape needs to be parameterized beyond just cell size, a perimeter\\nconstraint*® is often introduced,\\n\\nHperimeter = Ss Aperimeter (To )(A(a) _ Atarget (To). (3)\\n\\nHere A(c) represents the number of interfaces between lattice sites of a cell with sites not part of the\\nsame cell.\\n\\nWe also implemented several motility models. The first allows a vector field to bias the energy change\\nAH during a copy attempt\\n\\nAH chemotaxis = Achemotaxis (To )#i+7 * Ci. (4)\\n\\nHere #;_,; is the unit vector pointing in the direction of a copy attempt from position i to 7, and C; the\\nvalue of a vector field at position 7. In this paper, C; is always a field of unit vectors.\\n\\nFor the model where cells exhibit persistent movement based on one vector per cell, we add\\nAHpersistence = Apersistence (To )#i+5 * Po,t (5)\\n\\nto the Hamiltonian delta. The vector ,,, is the persistence direction of cell o at time t, which is updated\\niteratively at every MCS with\\n\\ney Xo ~~ X, t—At a\\nDot = (1 = pr) —S =— + flr Pot—1s (6)\\n|Xot — Xot-atll\\n~ Dot\\nPot = 7° (7)\\n\\\\|Do.tll\\n\\nHere ju, is the strength of directional persistence for cell type T, Xot is the centroid location of cell o at\\ntime t and At is the number of MCSs we look back for to determine the current direction of a cell.\\nFinally, we implement the Act model’ that allows for localized polarization. For each position i we track\\nan activity value each time said location is updated, here labelled as Act(i). This localized ‘memory’ is\\ninitialized at Maxact, and decays by 1 each timestep until it reaches 0. The Hamiltonian delta is adapted\\nby\\n\\nact (To)\\nMaxact (To)\\n\\nAHact (GMact (?) — GMact (J)), (8)\\n\\n19\\n making copies that originate from areas of a cell with a high activity value more likely. The activity is\\ncalculated for a full Moore neighbourhood using a geometric mean\\n\\nGMaci(i) = ( T] Act(n)) Ol. (9)\\n\\nn€N(i)\\n\\nN(i) are all neighbours of pixel or voxel i.\\n\\nOptimized CPU implementation of CPM\\n\\nBesides a high-performance GPU implementation, we also developed our own CPU implementation. This\\nserved both as a reference implementation for model behavior and internal implementation details for\\nus, and as a reference performance benchmark. We strived to reduce the major performance bottlenecks\\nwe could see in profiling. The implementation tracks lattice sites on cell boundaries, and only tries\\ncopy attempts involving such positions; a common optimisation, which is also implemented in Morpheus.\\nThe tracking implementation uses a high-performance hashmap implementation.*! We use a custom\\nhigh-performance pseudorandom number generator, xoshiro256.4?_ The lattice data is stored with a Z-\\norder space filling curve, allowing data that is close in the simulation space also to be close in the\\nlinear memory space, maximizing cache efficiency. Memory addresses are calculated using intrinsic bit\\ninterleaving instructions from the x86 bit manipulation instruction set extension.\\n\\nGPU implementation details\\n\\nOur CPM was implemented using CUDA; concepts equivalent to ones described here exist in non-CUDA\\nGPU hardware. A GPU consists of multiple streaming multiprocessors (SMs). Each SM runs multiple\\nthreads in parallel; contrasting with CPU threads, threads on the same SM execute in lockstep, in a\\nSingle Instruction Multiple Threads approach. A single GPU process — called a kernel — can consist of\\nup to millions of threads.\\n\\nGPUs have relatively slow global memory available to all threads, and much faster shared memory local\\nto SMs. Global memory transfer speed is often the main performance bottleneck for GPU programs.\\nThreads are bundled in thread blocks (TBs); each TB is bound to a specific SM. By caching data on\\ntheir SM in shared memory, TBs can greatly reduce global memory transfer bottlenecks.\\n\\nMemory transfer bottlenecks can be further reduced by two strategies: read coalescing and high SM\\noccupancy. Read coalescing has sequential threads in a TB do contiguous reads of global memory;\\ncontiguous reads are needed to fully utilise the memory bus bandwidth. High occupancy is achieved when\\na high number of threads are scheduled to execute per SM; the SM scheduler will switch out threads\\nwhen they are blocked on memory transfers, ensuring a high number of in-flight memory instructions at\\na time.\\n\\nWe propose a scheme of memory transfer, local SM execution and synchronization (Figure 2D). We\\nfocus on caching just the grid of cell ids, as this data is always needed, and can be used to quickly reject\\ncopy attempts internal to a cell. The simulation grid is divided in subsections, with regions larger than\\nearlier checkerboard subdivisions; each subsection is assigned to a TB, subsection size is chosen so active\\nregions fit in shared memory. Each TB determines which checkerboard sections are active in its region;\\nactive sections are cached in shared memory using coalesced reads.\\n\\nAfter caching — intended to speed up reads — the simulation is run, and successful copy attempts write\\nnew values to both shared and global memory. Alternatively, changes could be cached in shared memory\\nand written back coalesced to global memory; in our testing, an immediate shared/global write yielded\\nbetter performance. If a TB is done with simulation execution, it immediately starts a memory request\\nfor the next active region, ensuring maximum occupancy. All threads in the kernel are synchronized after\\nnew memory transfers are finished, ensuring maximum overlap of code execution and memory transfer.\\n\\nCPM literature offers a wide range of model variants and extensions, of which we implement a mod-\\nest subset. This comes with potential performance issues: even when extensions are disabled through\\nbranching statements, on GPUs such branching code still incurs a significant performance loss compared\\nto CPUs. We therefore pre-generated specialized versions of the kernel for a range of different extensions.\\n\\nWe use a custom high-performance random number generator on the GPU.*%\\n\\n20\\n Composite variable tracking\\n\\nSeveral CPM extensions introduce additional global variables. Some of these variable need to be kept\\nin-syne with each other — one variable being slightly out of date with others is not tolerated. One\\nexample is tracking of a center-of-mass centroid; for 2D this can be tracked as a sum of x and y values,\\nand corresponding count of lattice locations occupied (Figure 2C). These are integer values: tracking\\ncentroid locations as floating-point would cause values to drift due to rounding errors. The centroid value\\nis calculated by\\n\\n1\\n{a,y} = = X {asum, Ysum }- (10)\\n\\nOnline updates change both a cell’s centroid position and size. When a new lattice position {Xposition, Yposition }\\nis added to a cell, its centroid’s x value can be updated with\\n\\nLoia X |® lola + Lposition\\n\\n(11)\\n\\nZnew lel\\nnew\\n\\nCell sizes are stored interleaved with sum values; sizes and sums are both stored as 32-bit unsigned\\nintegers. The interleaved storage enables use of 64-bit operations to update both: a copy-and-swap\\n(CAS) operation is used, which atomically updates sum and cell size values simultaneously. This scheme\\nalso handles periodic boundaries, by adding or subtracting dimension x |e| to the sum values before\\nexecuting the CAS as needed, to keep centroids within simulation bounds.\\n\\nMulti-GPU execution\\n\\nWe devized a scheme to execute our CPM on multiple GPUs (Figure 2E); this scheme is intended as a\\nprototype, and as such not as optimized as other parts of our CPM. Simulation space is cut into equal\\nregions, each region is assigned to a GPU. A region’s lattice data, such as the cell ids, are sent to each\\nGPU, with an additional border layer. Global variables such as cell sizes are tracked on each GPU as\\ndeltas of the value since last inter-GPU data synchronisation. Every four MCSs, these deltas are gathered\\nfrom all GPUs, deltas are integrated, and updated values distributed to each GPU. Multi-GPU speedup\\nis mostly hampered by frequent or slow inter-GPU communication; inter-GPU synchronization after every\\nMCS negates most speedup, and a faster GPU-to-GPU interlink would likely increase performance.\\n\\nHardware\\n\\nThe CPU-based simulations, using our custom optimized CPU implementation and Morpheus 2.2.6 (the\\nlast non-parallelized version of Morpheus), and most GPU simulations were run on a system with a\\nNvidia Titan V, an AMD 2950x CPU and 128 GB of RAM at 2133 MHz. The only exceptions to this\\nare the multi-GPU simulations (Figure 2E), which were run on a system with four Nvidia Geforce RTX\\n2080 Ti GPUs, an Intel i9-9820x CPU at 3.3 GHz, and 128 GB of RAM at 2133 MHz.\\n\\nImaging of thick lymph node sections\\n\\nNaive inguinal LNs were harvested from CCL19-EYFP mice“ and fixed with 4% paraformaldehyde (PFA,\\nElectron Microscopy Sciences) for 1 hour at 4°C. Following a brief wash with PBS, LNs were embedded in\\n2% agarose and 250-300 ym tissue sections were generated using a vibratome (Leica). LN sections were\\nsubsequently immersed in a Special Wash Buffer (Baxter water, 0.2% BSA, 10% 10X PBS, 0.2% Tween-\\n20, 0.2% Triton X-100, 0.2% 10% SDS) at 4°C overnight, followed by staining with a rabbit anti-GFP Ab\\n(Thermo Scientific, 1:100, diluted in blocking buffer made with Baxter water, 0.2% BSA, 10% 10X PBS,\\n0.1% Tween-20, 0.2% 10% SDS) at 4°C for three days. After incubation, LN sections were thoroughly\\nwashed in PBS and mounted in 85% glycerol in PBS with glass coverslips. Thick LN sections were imaged\\nusing a LSM880 AiryscanFast confocal microscope (Carl Zeiss) with a x63/1.4NA oil immersion lens and\\nimages processed with ZEN 2010 software (Carl Zeiss).\\n\\nImaging of liver sinusoids\\n\\nFor whole-liver imaging, AlexaFluor 647 conjugated anti-CD31 rat monoclonal antibody (Biolegend) was\\nintravenously injected into wild-type C57BL/6 mice at 5ug for 15 minutes prior to harvesting the liver.\\n\\n21\\n\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdf = requests.get('https://arxiv.org/pdf/2312.09317.pdf') #A parallelized cellular Potts model that enables simulations at\n",
    "# tissue scale\n",
    "doc = pdf2image.convert_from_bytes(pdf.content)\n",
    "# Get the article text\n",
    "article = []\n",
    "for page_number, page_data in enumerate(doc):\n",
    "    txt = pytesseract.image_to_string(page_data).encode(\"utf-8\")\n",
    "    # Sixth page are only references\n",
    "    if page_number < 21:\n",
    "        article.append(txt.decode(\"utf-8\"))\n",
    "article_txt = \" \".join(article)\n",
    "article_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6d54bfa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['The CPM!',\n",
       " 'is a powerful simulation method that is known for its ability to simulate a wide range\\nof biological scenarios.',\n",
       " 'When modelling tissue environments, many researchers wish to consider both\\nhe complex morphology of single cells, and the behavior of these cells as part of larger collectives.',\n",
       " 'The appeal of the CPM lies partly in its ability to integrate both these aspects in a single model.',\n",
       " 'As such, it has been applied to diverse problems including tumor growth,” vascularization,® wound\\nhealing,* and morphogenesis.>*® Many extensions and hybrid model variants are available in the literature\\n‘0, for example, add support for cell motility,” cell signalling,!° and finite-element based mechanical\\ninteractions.!!',\n",
       " 'Thus, the CPM can be described as a generic platform upon which a wide range of models\\ncan be built, possibly explaining its popularity in the tissue modelling field.',\n",
       " 'For computational reasons, CPM models are often built using a small number of cells (perhaps in the\\nhousands) in simplified two-dimensional (2D) geometries.',\n",
       " 'Some research questions in tissue modelling\\nhowever demand a realistically large number of cells, and accurate 3D representations of tissues.',\n",
       " 'For\\ninstance, it is widely appreciated that the efficiency of space exploration of motile cells depends criti-\\ncally on the dimensionality of the environment, leading researchers to use 3D CPM simulations.® !?',\n",
       " ':13\\nAssessment of tumor growth behavior and invasive patterns of tumors likewise changes as a function of\\ndimension, resulting in the use of 3D CPM simulations in the tumor modelling field as well.?',\n",
       " 'While there\\nis thus a clear research demand for realistically large 3D CPM simulations, scaling up such simulations is\\ngenerally considered computationally infeasible.4 1° Removing this computational barrier would allow\\na host of CPM based modelling research to be scaled to what is biologically realistic, instead of scaling\\nto what technical constraints allow.',\n",
       " 'The CPM is an on-lattice method: simulations are executed on a grid.',\n",
       " 'A common strategy to speed up\\ngrid-based models is to cut the simulation into smaller subsections that are run in parallel.',\n",
       " 'However, for\\nthe CPM, model behavior is intimately tied to locations on the grid being updated sequentially, in a single-\\nlocation update Markov chain Monte Carlo process.',\n",
       " 'As such, parallelisation has been cautioned against,\\nas it risks drastic changes in CPM “system kinetics”.° Nevertheless, several attempts have been made to\\nemploy subsection-based parallelisation.',\n",
       " 'Chen et al.',\n",
       " 'proposed a checkerboard scheme that divides the\\nCPM lattice up over a computer cluster, with sections chosen much larger than cells, preventing cell-level\\nsynchronization issues.',\n",
       " '!” Tapia et al.',\n",
       " 'propose a similar scheme for GPUs:!® because the sections are much\\nsmaller than a cell on the GPU, they use a per-cell locking mechanism to prevent simultaneous updates of\\na cell.',\n",
       " 'Neither approach has been extensively checked for correct model behavior.',\n",
       " 'To our knowledge, no\\npublicly available implementation of subsection-based parallelisation exists for the CPM.',\n",
       " 'The software\\n“Morpheus” !® includes a non-subsection-based Central Processing Unit (CPU) parallelisation strategy;\\nthis method is less suited to fully utilise parallel accelerator hardware such as GPUs.',\n",
       " 'In this paper, we aim to make several contributions: First, we introduce a highly optimized parallelized\\nimplementation of the CPM using several optimization strategies not previously explored in the literature.',\n",
       " 'Second, we conduct an empirical analysis of potential pitfalls in parallelising a model dependent on serial\\nupdates of single locations and show that our approach mitigates these issues while preserving performance\\nbenefits.',\n",
       " 'Third, we demonstrate that our approach retains the behavior of cell motility simulations that\\nare particularly prone to changed model kinetics upon parallelisation.',\n",
       " 'And finally, we build two large 3D\\nimage-based simulations of T cell migration in liver and lymph nodes, demonstrating how complex tissue\\nenvironments can be analysed at a larger scale and realism than was possible before.',\n",
       " 'Serial Cellular Potts Models are limited to sub-millimetre scale\\nThe CPM’s serial update schedule — only one location at a time is updated on a grid — is central to\\nits behavior, and thus not trivially changed.',\n",
       " 'We therefore first aimed to quantify the computational\\nlimitations for models based on the serial CPM: how big can our computational experiments be, and\\nwhat timescale can they span, without making fundamental changes to the CPM?',\n",
       " 'We set up 3D CPM tissue experiments at different spatial and temporal scales and measured the required\\nwall-clock time.',\n",
       " 'We choose spatial and temporal simulation resolutions that allow reasonably accurate\\nrepresentations of highly motile cells within tissues.',\n",
       " 'Lymphocytes are among the smallest nucleated cells\\nin the mammalian body at 8-10um diameter for B cells and T cells.',\n",
       " 'At speeds of up to a few body\\nlo | 30 values/param.',\n",
       " '2 € 1wk 7\\n5 c= a = , -\\n[sj x e Pa\\n£ jw || 8 ow\\n6 cS ihr 4—_-\\n2 L | 2\\nOS Tmin 10 values/param.',\n",
       " 'polarisation strength replicates 1 2 3 4 5\\nsembles.',\n",
       " '(A) Computational cost of simulating a 256°m* volume (1m® per voxel) for 10 simulated\\nminutes (1s per simulation step).',\n",
       " 'With a high-performance CPU implementation, this takes 1.2 hours of\\nwall-clock time.',\n",
       " 'Extrapolation to a full day of simulated time for a 1mm?',\n",
       " 'simulation estimates 1.4 years\\nof wall-clock time.',\n",
       " 'Time measurements (black) are averages of 5 simulations, projections are based on\\ninear fits of simulations > 64°m* in log-log space (grey).',\n",
       " '(B) Building CPMs typically requires tuning\\nparameters.',\n",
       " 'Here we perform a multidimensional grid search for parameters that yield motile cells using\\nthe Act extension of the CPM;’ a modest single parameter grid search where 10 values are tried takes\\n5.9 seconds; larger grid searches take exponentially more time as function of parameters searched, with a\\nfive parameter search with 30 values is extrapolated to 1.1 years of wall-clock time.',\n",
       " 'Time measurements\\n(black) are averaged over 5 simulations, projections fitted on linear trends in linear-log space.',\n",
       " 'CPM\\nparameters are shown in Table 1 and Table 2.\\nengths per minute, they are also among the fastest movers within tissues.',\n",
       " 'We chose lum?',\n",
       " 'per voxel and\\nsecond per simulation step, or Monte Carlo Step (MCS), as our simulation resolutions, corresponding\\nto Beltman et al.9\\nTo create realistically large models of, for example, a complete murine Lymph Node (LN), or scale a\\ntumor model to the size of an avascular tumor spheroid,?° experiments at mm® scale would be needed.',\n",
       " 'Certain organ scale dynamics require timescales of at least a day, such as T cell transit through LNs.',\n",
       " '?4\\nMore modest simulation experiments at 256°um?',\n",
       " '(Figure 1A) at 10 simulated minutes require ~1h of\\nwall-clock time on a highly optimized CPU implementation of the CPM (Methods).',\n",
       " 'Such experiments are\\nsufficient to see minimal movement of motile cells such as T cells and can capture organ compartments\\nsuch as a LN paracortical T cell zone, or a micrometastasis for cancer modelling.',\n",
       " 'For mm?',\n",
       " 'scale, one-day\\nexperiments however, researchers would have to wait at least a year — making them impractical for all\\nbut the most patient of researchers.',\n",
       " 'A single experiment is rarely sufficient, in both empirical and simulation research.',\n",
       " 'Beyond the need\\n‘or experimental replicates, simulations can have parameters for which associated model behavior is non-\\nobvious, requiring parameterized simulation ensembles.',\n",
       " 'In Figure 1B, we explore a cell motility extension\\nto the CPM that coarsely models actin dynamics.” A single two-dimension simulation is relatively cheap,\\nbut when exploring even a modest number of parameters the execution time goes up exponentially.',\n",
       " 'Thus, serial 3D CPM simulations at a scale of 1 mm® are currently not feasible, and even 100°um?',\n",
       " 'scale\\nsimulations were extremely challenging in our experiments.',\n",
       " 'While ensemble runs can be accelerated with\\nparallel simulations without changing the CPM, the high computational costs of running large ensembles\\n‘urther motivates the need for a faster CPM.',\n",
       " 'Designing a parallel execution scheme for the CPM\\nTo make large scale simulations practical, we propose a parallelisation scheme for the CPM that uses\\nhighly parallel GPU hardware.',\n",
       " 'A natural way\\ndivide the simulation grid in\\nsupport up to thousands of\\n‘0 parallelise on-lattice models, such as the CPM, is to\\ner sections; each section runs a local simulation in parallel.',\n",
       " 'As GPUs\\nations in parallel, these sections will be quite small.',\n",
       " 'The CPM also\\nupdates information outside of the grid, such as sizes of simulated cells.',\n",
       " 'We first consider three issues\\nwhen designing our parallelisation scheme: adjacent sections with active simulation interfering with\\neach other during updates o:\\nsimultaneously, and stale in:\\nupdated by another simulati\\nA CPM simulation proceeds\\ncopying the position’s state to a ne\\nsimulation running, we run t.\\nboundary.',\n",
       " 'Such “race condi\\nstandard problem in paralle\\nto be active at the same time.',\n",
       " 'As\\nsections are sufficiently spaced to prevent overlappin,\\nbetween active and dormant state.',\n",
       " 'ion being used by one paralle’\\nhoosing a random position on\\nk of simulations “racing” to wri\\nrun the risk of information ge\\nm, with a straight-\\nThe CPM also tracks information involving more than one gri\\nsimulated cells.',\n",
       " 'We face the same “race condition” problem here,\\nupdates touch a single cell (Figure 2B, left column).',\n",
       " 'small active subsections, making such concurrent up\\nscheme is proposed, that rejects all but the first concurrent update attempt on a varia\\nmiddle column).',\n",
       " 'We instead opt to use atomic operations (Figure 2B, right column), which equa!',\n",
       " 'prevents update information from getting lost, as the hardware will eventually integrate all updates\\ndesigned an additional scheme that sti\\nCPMs, global information travels instantaneously: as only one grid position is w\\nhout incorporating information of the overwritten update.',\n",
       " 'This\\ng updates (Figure 2A).',\n",
       " 'Regions cycle perio\\nates more likely.',\n",
       " 'In Tapia et a\\n; non-grid information such as cell sizes being updated\\nsimulation after said information got\\nthe simulation grid, and stochastically\\nighbour.',\n",
       " 'If adjacent subsections of the grid each have an active parallel\\ne to the same location on the sections’\\nting lost, as one section overwri\\nion: never allow two adjacent sections\\nrn of sections is only ever active; active\\nlocation, such as the current sizes\\nas it is possible that multiple paral\\nrallelisation requires a high number\\n1.18 a per-cell locking\\ndated at a time,\\ndate attempt has access to the results of all previous updates.',\n",
       " 'This instantaneous information\\naccess is relaxed in our proposed scheme: when information such as cell size is accessed,\\nout-of-date due to still in-flight updates to the accessed information.',\n",
       " 'The cell-loc!',\n",
       " 't al.',\n",
       " 'prevents this from happening, by outright rejecting any updates that cou\\ninformation; we explored a less strict scheme, reasoning that the stochastic CPM copy attempts\\nerate out-of-date information.',\n",
       " 'For CPM extensions that do not tolerate their global variables\\nightly out sync with each other (e.g.',\n",
       " 'variables used in center-of-mass “centroid” tracking) we\\n1 will keep variables in syne with each other, and sits on top of our\\nd have used out-\\natomics based scheme (Figure 2C, Methods).',\n",
       " 'Otherwise, due to atomic updating, our scheme still gives\\nstrong guarantees on tracked data being correct — for example, cell sizes will be tracked correctly — and\\nwe cache local da\\nF our scheme on rea\\nGPUs, memory access almost always\\nprevents easy use of a GPU’s caching hierarchy.',\n",
       " '!® To maximize performan\\nscheme that carefully operates in loc\\nhardware requires consideration o:\\ny allows us to explore the effects of allowing the use of slightly outdated information.',\n",
       " 'CPM’s random access pattern\\nce, we designed a data transfer\\nf several technical details.',\n",
       " 'k-step with checkerboard work splitting (Figure 2D, Methods):\\na needed by the regions that are becoming active, an\\ntions to fully utilise\\noverlap memory transfer and\\nGPU hardware.',\n",
       " 'To increase simulation volume, and gain further\\nspeed ups, we also prototyped a multi-GPU scheme (Figure 2E, Methods).',\n",
       " 'In short, our proposed parallel CPM sc!',\n",
       " 'heme is tailored to make opt\\ncode.',\n",
       " 'It still gives strong guarantees on simulation sta‘\\nimal use of GPUs: hardware specialized\\ne consistency — for example, a\\nglobal variable tracking a cell’s size will always equal the number of pixels on the grid occupied by that\\ncell.',\n",
       " 'Yet, we do all\\nflexibility and further maximise performance gains.',\n",
       " 'ow the use of slightly outdated values of such global variables in copy attempts to gain\\nB race condition mutex locking atomic updates\\n(Tapia et al.)',\n",
       " '(our method)\\n(lel=12) (lla i= 12 |}\\n| ASP—$$ $l new = 13)- - - [le] = 13 &)- - - {le] = +18)\\nnO OO Oo Hii --(=tie---(41=19\\nreese ew.',\n",
       " '= 13)- - - (lel := 13 )- - - {\\n?',\n",
       " '| ——{l*],.4 = 13)- - -(lo = 13 w)- - - |\\nc I*]s:= 4 I#lnou = 5\\n© Xu Yous *= (2-8, 3.3} © Xen?',\n",
       " 'Voow = (3; 3} (Kumn= 11) (Yaun= 13} (lel= 4)\\n4 ee [Xeon 2= +4!',\n",
       " 'Yeu = +2 | lel= +1 | Be]\\nXon= 15) (Van= 15) (Iel=5)\\n[new = Xeun/@l = 3; Yoow = Yeun/l#] = 3)\\nE_ per GPU region border sync\\n~-7--- global simulation barrier\\nSS caput sizessyne pu\\n|e] = 23 je| = 29 |e| = 23 |e| = 29\\n[Je|:= +4 Jel: +14=he!',\n",
       " '= +0 |el:=-5 |\\n|e] = 27 |e|=25 |e] = 27 je|=25\\nGPUs.',\n",
       " '(A) We divide simulation space into a subgrid, with alternating regions being active in a\\n“checkerboard” pattern, preventing update conflicts on the grid.',\n",
       " '(B) We implement two strategies to\\nprevent global variable update conflicts (e.g.',\n",
       " 'cell size tracking): a mutex locking strategy (Tapia et\\nal.',\n",
       " \"'8), and a novel atomic operation strategy.\",\n",
       " '(C) Combining interleaved storage and atomic operations\\non adjacent data allows variables to be kept in sync with each other.',\n",
       " '(D) A four-step execution scheme,\\nwhere active checkerboard areas are transferred to fast shared memory, accelerates copy attempts.',\n",
       " 'Thread\\nbarriers synchronize active threads after transfers.',\n",
       " '(E) A multi-GPU execution scheme, where regions\\nof the simulation are assigned to different GPUs.',\n",
       " 'Borders are synchronized between relevant adjacent\\nregions, and global variables are synchronized through delta tracking.',\n",
       " 'A T, : time in region time divergence thread barriers\\nmm : time for one copy attempt ———-—— T,\\n5 checkerboard size thread size\\ne ?',\n",
       " 'ee : :\\n<@ e ‘|e eee ‘ eco\\né } ° 1|0@ : eee\\nao @ : : ee\\nT, T. : :\\ncopy attempts before checherboard switch\\nsingle update parallel update\\ntime since last copy attempt\\nee®@ r ) 4 copy attempts per switch, 1 copy attempt per switch,\\nnumber of copy attempts 4x4 area/thread, 8x8 area/thread,\\n2x2 area/checkerboard 4x4 area/checkerboard\\n| | | lh __\\n0123 45 0123 45 0123 4 5\\ncopy attempts per location\\n0 1 2 0 1 2 0 1 2\\nwaiting times between copy attempts (MCS)\\nfrom racing ahead of others too much, thread barriers are imposed after each update.',\n",
       " '(B) Checkerboard\\nsize and switching frequencies determine the distribution of waiting times between consecutive updates\\nof the same locations.',\n",
       " '(C) Small checkerboard areas or infrequent switches can alter the waiting time\\ndistribution substantially compared to the single-update scheme.',\n",
       " 'Update frequency histograms were\\ngenerated by simulating one MCS (256 updates) on a 16x16 grid.',\n",
       " 'Waiting time histograms were simulated\\non the same grid using 100 simulated MCSs.',\n",
       " 'cell boundary type:\\n60% 1 @-© dark-light\\n% of cell boundaries\\nsimulation based on Graner et al.!',\n",
       " 'is implemented with two strategies to prevent race conditions for cell\\nsize variables.',\n",
       " 'The mutex locking strategy proposed by Tapia et al.',\n",
       " '!® leads to severe starvation in our\\ntests.',\n",
       " 'Using an atomic update approach recovers the desired sorting behavior, closely matching sorting\\nspeed of the original CPM.',\n",
       " 'The simulation was run with a 4x4 area per thread and 2x2 checkerboards,\\nbut other sizes were tested and did not make a significant difference for this simulation.',\n",
       " 'Sorting trends\\nof different CPM implementations are averaged over 5 simulations.',\n",
       " 'Simulation parameters can be found\\nAligning the lattice update statistics of parallel CPMs to serial CPMs\\nOur novel parallel execution scheme guarantees simulation data consistency, but drastically changes the\\nway the simulation lattice is updated.',\n",
       " 'We investigated statistical properties of copy attempts on the\\nsimulation lattice for different lattice positions; specifically, we analysed the relative ordering of copy\\nattempts, frequency of copy attempts, and waiting times between copy attempts.',\n",
       " 'This allowed us to\\nquantify the divergence of parallel lattice update kinetics from the serial CPM.',\n",
       " 'Such divergence could\\nlead to changed simulation behavior in complex models of cells and tissues.',\n",
       " 'Without guardrails, performing independent CPM updates in parallel would allow different parts of a\\nsimulated tissue to have been alive for different lengths of simulated time (Figure 3A).',\n",
       " 'Modern GPUs use\\nthread barriers to synchronize parallel threads without strong computational performance implications.',\n",
       " 'We set thread barriers after ever single copy attempt, so for every currently active checkerboard region\\nthere is only a maximum time divergence of one copy attempt between active regions.',\n",
       " 'Our parallel execution scheme allows us to freely choose the size of subregions and frequency of active\\nregion switching (Figure 3B), without impacting data consistency guarantees.',\n",
       " 'We measure frequency\\nand waiting times between copy attempts per position on the lattice in the standard, serial implementation\\nof the CPM (Figure 3C, left column).',\n",
       " 'For our initial parameter choice, we found that underlying update\\nstatistics are drastically changed (Figure 3C, middle column): waiting times between copy attempts\\nbecame much more predictable due to the much more even distribution of copy attempts compared to\\nthe serial CPM.',\n",
       " 'Using large parallel subsections with frequent active region switching allows non-parallel\\nupdate statistics to be closely recapitulated in our parallel scheme (Figure 3C, right column).',\n",
       " 'We thus find that parallelisation parameters have major influence on copy attempt statistics on the\\nlattice.',\n",
       " 'By switching active regions of the checkerboard frequently, and avoiding too small areas per\\nthread, we can obtain statistics that closely resemble the serial CPM.',\n",
       " 'Emergent cell and tissue behavior is maintained in a parallel CPM\\nOur goal is to build parallel CPMs that are much faster than previous implementations but maintain cell\\nand tissue kinetics consistent with existing serial CPMs.',\n",
       " 'We hypothesized that aligning the waiting times\\nbetween update attempts of our parallel CPM to the serial version should suffice to achieve such consis-\\ntency.',\n",
       " 'To investigate this, we ran several simulation scenarios with different CPM variants, measuring\\nemergent behaviors on the cell and tissue levels and comparing these to serial CPM versions.',\n",
       " 'We first investigated the classic cell sorting simulation:!',\n",
       " \"two mixed populations of cells with differential\\n A single update\\n4 copy attempts per switch,\\n= 0.3 £03 2\\n202 5 0.2 5\\n50.1 & 0.1 &\\n0.0 8 0.0 8\\n2000 6000 10000 2000 6000 10000\\ntime (MCS) time (MCS)\\nB » low polarisation strength\\n2 ee qe 4S\\n' high polarisation strength '\\n.\",\n",
       " '«) f 4 i\\n3 08 3 08 3\\n£ 0.4 £ 0.4 £\\n8 0.0 8 0.0 8\\n200 400 600 800 200 400 600 800\\nCc 8 copy attempts per switch,\\n0.25 0.50 0.75 1.00 0.25 0.50 0.75 1.00\\n1 copy attempt per switch,\\n200 400 600 800\\n1 copy attempt per switch,\\n0.25 0.50 0.75 1.00\\nare similar.',\n",
       " '(A) Cells in a circular chemokine gradient simulation move highly regular in a serial CPM\\n(low per-pixel entropy, shaded: 1SD); parallel simulations diverge slowly from reference behavior when\\ncopy attempt statistics are preserved (low per-pixel cross-entropy).',\n",
       " '(B) In the Act model of intra-cell\\npolarization,” stronger polarization strength leads to higher directional autocorrelation (measured for 750\\nMCSs at 50 MCS intervals, averaged over 50 simulations, shading: 4\\nemerges when update waiting time distributions are preserved.',\n",
       " '(C)\\nt1SD); directional persistence only\\nPhase transition simulation in 3D\\nwhere synchronous motility emerges at a density around 0.5.',\n",
       " 'Cell order indices after 2000 MCSs are\\nshown (averaged over 5 simulation runs, shading: 4\\nransition only occurs when update\\nwaiting time statistics are preserved.',\n",
       " 'Simulation parameters can be found in Table 1 and Table 2.\\n adhesion (i.e., cells of the same type have stronger\\nadhesion to each other than cells of different types)\\nsort themselves in two homogeneous regions over time (Figure 4, grey lines).',\n",
       " 'The mixing kinetics can be\\nmeasured by tracking homogeneous and heterogene:\\nbased parallelisation approach proposed by Tapia et a\\nleads to a lack of sorting behavior; the simulation\\nconcurrent copy attempt is allowed per cell un\\nscheme allows expected sorting behavior to re\\nremained consistent for all parallelisation paramete:\\n‘ous cell neighbours over time.!',\n",
       " 'Strikingly, the mutex\\n1.18 — while still being able to run a simulation —\\nencounters so-called “starvation”, because only one\\ner this scheme (Figure 4, red lines).',\n",
       " 'Our less restrictive\\n-emerge (Figure 4, black lines).',\n",
       " 'The sor\\nrs we examined (Supplementary Figure $1).',\n",
       " 'We next examined three motility model variants of the CPM, where additional biases are placed on\\no obtain cells that move persis\\nfunction of random membrane fluctuations.',\n",
       " 'Bec:\\nneed to roughly maintain their target size, copy\\nsections for movement to occur.',\n",
       " 'We therefore s\\nby changes to the waiting times between copy at\\nikely to succeed when aligned with the underlyi\\nbias is fixed in this model — the field does not change\\nsimulations with the same initial conditions closely\\ncontains a cell in multiple simulations, where low\\nbetween simulations, and low cross-entropy indica:\\nwith reference serial runs.',\n",
       " 'In simulations with\\nincreased quickly (Figure 5A, middle column); w.\\nrises much slower (Figure 5A, right column).',\n",
       " 'Next, we analysed a model that simulates localize\\nborder of a cell makes further copy attempts involving those\\ntimes between copy attempts are very important in this mo\\nin a small neighbourhood successfully expand at the same time; this is less\\ny model we consider adds a chemotactic vector field to the CPM; copy attempts are more\\nng vector field (see Methods).',\n",
       " 'Because the directional\\nhis alignment using pixel-wise (cross-)entropy: we measure the probability that a pixe\\nently; in the standard CPM movement only occurs as a\\nause motile cells have a leading and a retrac\\nattempts need to sufficiently alternate between these cel\\npeculated that movement models might be more affecte\\nover time — cells move in a highly regular fashion, an\\nalign even at a pixel level (Figure 5A, left column).',\n",
       " 'pixel probabilities indicated high agreement\\nes a high level of agreement of parallel simulations\\nged lattice update statistics, cross-entropy\\npdate statistics are preserved cross-entropy\\ncell polarization,’ where recent pixel updates at the\\npixels more likely to succeed.',\n",
       " 'The waiting\\nel, as cells start to move when many pixels\\nlikely when copy attempts are\\nspaced out evenly.',\n",
       " 'A polarization strength parame\\ncell; varying the polarization strength interpolates motility be\\nmotility modes (Figure 5B, lef\\nwaiting time statistics this full range of behaviors\\nthey re-emerge when update sta’\\nFinally, we examined a simulatio\\nwith this vector, and the vector\\nmotility model involves informat\\nion tracked on the\\nto the simulation grid, placing additional demands on a para!',\n",
       " 'density cell collective exhibit Brownian movement,\\nas density is increased;??',\n",
       " 'this characteristic phase\\nleft column).',\n",
       " 'In a parallel CPM this phase transition disa,\\n(Figure 5C, middle column), but re-emerges when update\\nTaken together, these results demonstrate that a strict updat\\nn of a tightly packe\\na preferred movement direction vector; copy attempts for a ce\\ns how much polarization bias is added to a\\ntween Brownian and directionally persistent\\nel simulations with changed update frequency and\\noes not appear (Figure 5B, middle column), but\\n(Figure 5B, right column).',\n",
       " 'lective of immune cells,° where each cell has\\n1 are more likely to succeed when they align\\nated to align with recent movement of the cell.',\n",
       " 'This\\nattice, and additional information tracked external\\nel CPM.',\n",
       " 'In a similar simulation model, low\\ntransitioning sharply to ordered collective movement\\nransition is reproduced in the CPM (Figure 5C,\\npears when update statistics are changed\\nstatistics are preserved (Figure 5C, right\\ne scheme preventing parallel updates of the\\nsame cell is not necessary for maintaining valid CPM behavior and can even lead to severe “starvation”\\nwhere simulations do not progress anymore.',\n",
       " 'On\\nhe other hand, preserving waiting times between copy\\nattempts involving the same pixel can be very important to maintain model behavior, and was critical\\nin all three motility scenarios we examined.',\n",
       " 'GPU-based CPM parallelisation accelerates models by several orders of mag-\\nHaving confirmed that our parallel CPM indeed pro\\nduced correct model behavior in a range of examples,\\nwe next investigated the performance improvements resulting from our implementation choices when\\ncompared to a serial CPM.',\n",
       " '% 1000 2\\ns 8 g 8 & 19547x13805x\\nBe 3 3 10 3\\n2 3 < 1 g\\n24 a & 2\\n2 eg ol °\\nkernel specialisation ( ae ) 128° 5127 20487 8192!',\n",
       " 'implementation\\nshared memory @ @ e@\\nD E area/thread: @ 4x4 A8x8 mi 16x16\\n> ° 03 +0\\n3 10000 3 9786x 2\\nS 1000 g = 02 = 0.9 >\\n& 400 ® § g 8\\n& 2 8 308 3\\noo «10 ® — 2902x ZO 8 8\\n£ 1 s G 07\\n642 256° 10242 implementation 1 4 16 64 1 4 16 64 0.0 0.1 0.20.3\\ndimension updates per switch updates per switch simulation time\\nPerformance benefits are obtained by a shared memory scheme, and a pre-compiled kernel specialization\\nscheme, based on a sorting simulation extended to densely fill simulation space (Figure 4, all sorting\\nbenchmarks are average of 5 simulations of 100 MCSs, 100 MCS burn-in).',\n",
       " '(B) Distributing the lattice\\nover 4 GPUs speeds up the simulation further.',\n",
       " '(C) Performance of sorting simulations are significantly\\nhigher relative to two serial implementations; relative speedup shown for 2048x2048 simulations.',\n",
       " '(D)\\nPerformance of 3D cell collective simulations is also improved (Figure 5C, average of 2 simulations of\\n100 MCSs, 100 MCS burn-in); relative speedup shown for 256° simulations.',\n",
       " '(E) Trade-offs between\\nspeed and accuracy for chemokine field simulations (Figure 5A, accuracy: 1 - cross-entropy of parallel\\nensemble relative to serial ensemble, 250 simulations per ensemble).',\n",
       " 'CPM parameter settings can be\\nfound in Table 1 and Table 2.',\n",
       " 'We first examined the performance gains resulting from our use of a shared memory caching scheme\\n(Figure 2D, Methods) using a 2D sorting simulation benchmark (Figure 4).',\n",
       " 'On its own, shared memory\\ncaching achieved only minimal speedup (Figure 6A).',\n",
       " 'However, for further performance improvement,\\nwe pre-generated different versions of our code based on which CPM extensions were in use (Methods).',\n",
       " 'Combining these specialized kernels with the caching scheme achieved a significant 10-fold speedup, more\\nthan expected from either optimization alone (Figure 6A).',\n",
       " 'Testing our multi-GPU scheme (Figure 2E,\\nMethods) in the same scenario achieved a further 2.3x speedup by distributing the simulation across\\nfour GPUs (Figure 6B).',\n",
       " 'Hence, careful consideration of GPU-specific implementation choices resulted\\nin significant performance benefits.',\n",
       " 'We next compared our parallel CPM (on a single GPU) to our own optimized CPU implementation\\n(Methods), and to the state-of-the-art software “Morpheus”.',\n",
       " '!® Using the same 2D cell sorting bench-\\nmark, we observed a {3,500-fold speedup compared to our own CPU implementation, and a j25,000-fold\\nimprovement relative to Morpheus (Figure 6C).',\n",
       " 'On the 3D cell collective simulation (Figure 5C), a\\n{,600-fold speedup was achieved relative to serial CPMs (Figure 6D).',\n",
       " 'It is important to note that our\\nbenchmark hardware did not permit use of shared memory caching for the 3D model; future hardware\\nwith more shared memory may improve 3D performance.',\n",
       " 'As shown in Figure 3 and Figure 5, checkerboard size and switching frequency influences model behav-\\nior: larger divisions and more frequent switching leads to behavior concordant with serial CPMs.',\n",
       " 'These\\nupdate patterns also interact with caching, further impacting performance.',\n",
       " 'As the sorting simulation has\\nlow sensitivity to parallelisation parameters (Supplementary Figure $1), we instead benchmarked our\\n2D chemotaxis simulation (Figure 5A).',\n",
       " 'We found that larger subdivisions indeed slowed simulations\\ndown considerably, whereas frequent switching led to a less marked slowdown (Figure 6E).',\n",
       " 'These results\\nsuggested a trade-off between performance and model accuracy.',\n",
       " 'Hence, our parallel CPM represents a significant speed improvement over even well optimized serial CPMs\\n such as Morpheus.',\n",
       " 'This improvement was to a substantial extent a result of GPU-specific optimizations\\nand led to an overall speedup of several orders of magnitude.',\n",
       " 'Building an image-based model of T cell surveillance of the liver\\nOur parallel CPM was motivated by the need for tissue simulations at realistic scale with realistic cell\\nmorphology.',\n",
       " 'While our approach increases engineering complexity of the model, it can also simplify\\nmodel building, as it enables direct use of imaging data to initialize the simulation environment.',\n",
       " 'To\\ndetermine the feasibility of directly using real-world imaging data in our system, we constructed a model\\nof T cells moving in the sinusoidal regions of the liver.',\n",
       " 'Tissue resident memory T cells (Trms) are confined to specific organs, where they sample their envi-\\nronment for signs of disease”+ — in particular, liver Trms are key in providing sterilising immunity for\\nmalaria.',\n",
       " '?° Trms display a distinct motility pattern in the liver, and efficacy of malaria clearing depends\\non the number of Trms present.',\n",
       " '?® We constructed a 3D simulation environment of liver sinusoids — capil-\\nlaries that Trms reside in and migrate through — and combined it with a polarized cell motility model” to\\nrepresent the Trms (Figure 7B).',\n",
       " 'Trm motility parameters were derived from a LN model constructed in\\nparallel, and our parameter choice will be explained in more detail with the LN data.',\n",
       " 'These parameter\\nsettings lead to motility modes also observed in intravital liver data: intermittent movement and pausing\\nof Trms.',\n",
       " 'The goal of our simulation was to predict how much of the liver volume would be explored by a\\ngiven number of T cells within a set amount of time; this same question was addressed by Fernandez-Ruiz\\net al.',\n",
       " 'using a more abstract mathematical model.',\n",
       " '?°\\nAs can be expected, we found that liver exploration by the Trm population increased with the number of\\nexploring cells, and with the amount of exploration time (Figure 7C).',\n",
       " 'Remarkably, the sinusoid volume\\nexplored by an individual cell increased dramatically as the total Trm population was raized from 1,000\\nto 10,000 cells (Figure 7D).',\n",
       " 'Further investigation revealed the source of this phase transition: cells in\\nour simulations can be in an unpolarized state, in which they do not move persistently; collisions can\\ntrigger polarization of stationary cells.',\n",
       " 'Since such collisions occur more frequently at higher densities,\\nthis causes a larger fraction of cells to be motile (Figure 7E).',\n",
       " 'Thus, if Trms can help each other to\\nmove, a larger Trm population would screen the environment more efficiently than expected from the\\nNotably, the mean speed of Trms peak at ~10.5 um/min (Figure 7F), matching average speeds in our\\nlymph node simulations — but only at densities where the arrest coefficient is close to 0.',\n",
       " 'In two-photon\\nexperiments, liver Trm speed is reported to be significantly lower than T cells in lymph nodes;\\nreproduce this in our simulations with moderate cell density, which leads to both a realistic slowdown,\\nand higher arrest coefficient.',\n",
       " 'The simulated liver Trm speed and arrest coefficient closely matches previ-\\nously reported values in literature,??',\n",
       " 'where they were used to quantify observed characteristic patrolling\\nThese liver experiments demonstrated that our system allowed us to combine 3D imaging data with a\\nrelatively realistic model of polarizing cell motility.',\n",
       " 'The resulting simulation produced an interesting\\nphase transition of Trm screening efficiency.',\n",
       " 'While we do not claim that such an effect is present in the\\nreal liver — more detailed tuning of the motility model to match the characteristics of real sinusoidal Trm\\ncells would be out of scope for this simple proof of concept — this does illustrate how detailed models such\\nas the CPM generate emergent behavior that is not an immediately obvious consequence of the model\\ningredients and can generate interesting biological hypotheses worthy of further exploration.',\n",
       " 'A T cell collective moves robustly in a realistic, organ-scale model of the lymph\\nIn LNs, millions of T cells move in search of antigen-presenting cells,?',\n",
       " '!°° and finding the cognate antigen\\nis a key early step in adaptive immune responses.*!',\n",
       " 'T cell motility in the LN is shaped by the high\\ndeformability of the cells®?',\n",
       " 'and their interactions with a complex stromal environment, the fibroblastic\\nreticular cell network.',\n",
       " '*3:°4 Given the LN’s unique structure as an organ consisting mainly of moving\\ncells, it presents a challenging test case for any tissue simulation system.',\n",
       " 'Indeed, many models of LN\\nT cell dynamics already exist, but they were often limited in either cell count and simulation volume, or in\\nrealism of cell morphology and LN environment.',\n",
       " 'For example, T cell interactions with a synthetic stromal\\nstructure have been investigated using 3D CPM simulations of thousands of cells,° while collective motility\\n at a more realistic scale of hundreds of thousands of cells has been investigated with non-morphological\\nagent-based models.',\n",
       " '*°:3° We were therefore interested in our system’s ability to simulate a realistically\\nlarge collective of millions of realistically shaped T cells.',\n",
       " 'As in the previous use case, we used microscopy\\ndata to initialize — and parameterize — our model.',\n",
       " 'We constructed a LN simulation using the work of Beltman et al.9 as a starting point.',\n",
       " 'To improve\\nrealism, we used an imaged LN section to initialize the stromal structure (Figure 8A), we used the\\nAct-CPM extension for cell motility,” and we fit the two Act motility parameters directly to intravital\\nimaging data (Figure 8B-E).',\n",
       " 'When fitting these parameters, we aimed to reproduce the intermittent\\nstop-and-go motion®” observed in real data whilst also quantitatively approximate the real speed distri-\\nbution.',\n",
       " 'Our system allowed fitting to be performed on realistically large simulations (256um diameter\\nsimulation volume, containing 50,135 T cells, 20 minutes simulated time).',\n",
       " 'Crucially, we did not use\\nperiodic boundaries which, in dense simulations, can lead to artificial global synchronisation of motion\\nin the cell collective.',\n",
       " 'We used the KL divergence to compare simulated to real speed distributions (Figure 8B).',\n",
       " 'Increased\\nadhesion of T cells to the stromal structure made the speed distribution more right-skewed (Figure 8C,\\ntop), which was more consistent with real data (Figure 8C bottom).',\n",
       " 'For most examined parameter\\nsettings, we found the pronounced variations in cell speed along cell trajectories that are also observed\\nin real data (Figure 8D).',\n",
       " 'When we performed grid searches of motility parameters at two different\\nstromal adhesion strengths (Figure 8E), both grid searches had broadly similar fit landscapes, and the\\noverall lowest KL divergence values were very similar.',\n",
       " 'Qualitatively, we found more right-skewed and\\nbimodal speed distributions in high-adhesion than low-adhesion simulations for a subset of the motility\\nparameters (Supplementary Figure $2).',\n",
       " 'Based on these fitting results, we chose the best fit motility\\nparameters of the high adhesion setting for our further simulations.',\n",
       " 'We next investigated organ-scale collective T cell motility by scaling up our simulated volume to 1mm\\ndiameter, now containing 3,359,752 T cells (Figure 9A).',\n",
       " 'While this is not entirely realistic — a real LN\\nhas a more complex compartmental structure, while we simulate only a T cell zone — our main intention\\nwas to determine the spatial range of collective behavior that can now be investigated with our system.',\n",
       " 'We simulated one virtual day in 1.9 days of wall clock time — where serial implementations were projected\\n‘o have taken a full 1.4 years (Figure 1A).',\n",
       " 'We used a mixing simulation of two differentially labelled populations as an easily observable proxy for\\ncollective search (Figure 9B).',\n",
       " 'In the simulation, it takes about 1 hour to reach a state where 50% of the\\ncells are next to at least one cell of the other population (Figure 9C).',\n",
       " 'While stromal adhesion influences\\nintra-cellular stop-and-go motility, it has minimal influence on collective motion: high and low stromal\\nadhesion simulations have broadly similar mixing speeds.',\n",
       " 'While collective motility is slightly ordered at\\ncell-scale distances, no ordering persists at organ-scale (Figure 9D).',\n",
       " 'In summary, we used an organ-scale LN simulation as a challenging test case: an environment known for\\ncomplex, large scale collective motility dynamics.',\n",
       " 'Our system allowed us to integrate of microscopy data\\ndirectly into our simulation, and to fit motility parameters directly on two-photon data.',\n",
       " 'We investigated\\nthe behavior of the fitted cell population at organ-scale (millions of cells).',\n",
       " 'We hope that this will inspire\\nfurther research on large cell collectives — integrating various sources of empirical data, at a high level of\\nrealism, and at realistic scale.',\n",
       " 'We have implemented a parallel version of the CPM that, in some experiments, was 3 to 4 orders\\nof magnitude faster on GPUs than a state-of-the-art serial CPM.',\n",
       " 'To achieve this, we designed novel\\nschemes for lattice work splitting and synchronization, maintaining global state, efficient GPU memory\\ntransfer, and multi-GPU execution.',\n",
       " 'We found that preservation of local lattice update statistics was\\nsufficient to maintain high-level system dynamics, despite removing the sequential nature of the original\\nIn our work we aimed to ensure that important properties of CPM simulation trajectories were main-\\ntained.',\n",
       " 'Some previous CPM parallelisation efforts instead focused on equilibrium state distributions as\\nimportant quantities.',\n",
       " 'For example, Chen et al.!?',\n",
       " 'argue that their parallel CPM maintains certain equi-\\nlibrium properties.',\n",
       " 'While such arguments are frequently applied in the analysis of energy-based models\\n in physics, Voss-Béhme et al.',\n",
       " 'showe\\nWhile we have shown that the same\\nCPM in our benchmark scenarios, thi\\nparameters that were originally fitte\\nto re-fit their simulation parameters\\nthat these are less relevant for the CPM because the Metropolis\\nalgorithm used in the CPM drastically alters longtime behavior of the mode\\nrange of behaviors is present in our parallel CPM as in a serial\\ns would still need to be carefully checked when re-using simulation\\non a serial implementation.',\n",
       " 'Specifically, researchers might want\\nwhen using our system instead of a serial\\nis sensible in general when moving between different implementations of the s\\nas the CPM where implementation details might anyway differ\\nversion — advice that\\nsubtly.',\n",
       " 'Furthermore, the observed relation between model kinetics and performance (Figure 6E) also\\nsuggests it might be fruitful to explore more aggressive parallelisation settings that might yield further\\nspeed-up whilst maintaining desired simulated behavior.',\n",
       " 'Through accelerating the CPM on accessible GPU hardware, we have increased the range of feasible\\nsimulation sizes — because our scheme fully utilises all GPU cores, simulation size is now mostly limited\\nby the available GPU memory.',\n",
       " 'As memory per GPU increases, our scheme wi\\nti-GPU prototype scheme will allow for further increase of simulation volume.',\n",
       " 'Our\\nto include inter-GPU communication over MPI,*° allowing\\nding block for a GPU cluster-based version of the CPM —\\nher to lcm?',\n",
       " 'scale.',\n",
       " 'multi-GPU scheme could be further extende\\nour system to be us\\ncell shape into acco\\n14, 15, 35, 36\\ned as a fundamental bui\\nur Imm?',\n",
       " 'simulations fur\\n1 easily scale to larger\\nular simulations are often performed using more coarse-grained agent-based model\\nOur high-performance CPM now allows researchers to perform simulations that\\nmuch more detailed and allow the investigation of new questions that require taking\\nunt, such as migration in complex environments or morphogenesis.',\n",
       " 'We hope that\\nions of liver and LN environments will serve as inspiration for many such spatially\\nresolved tissue and organ simulations in the future.',\n",
       " 'Similar approaches to the one shown here could be\\nuseful for parallelisation of CPM variants or other sequential lattice-based models.®\\n A B mmm 100 yum explored space @ T cells\\nS 10 1000 100000\\nE i F cells\\n@ 0.6 E so\\n10 1000 100000 10 1000 100000\\na motile state.',\n",
       " '(A) A 1024x1024x182 um® scan of a liver sinusoid is directly imported into our\\nsimulation environment (64ym slice shown).',\n",
       " '(B) Simulated T cells explore liver sinusoids (5000 cell\\nsimulation shown).',\n",
       " '(C) Simulations at different cell densities explore the environment at different speeds\\n(single simulation of 24hr, with 1 min snapshots).',\n",
       " '(D) An increase of explored volume per cell occurs\\nwhen cell density rises from 1,000 to 10,000 cells; 100,000 cells fill ~1/3 of sinusoid volume, causing\\nexploration efficiency to saturate.',\n",
       " '(E) At low cell densities most simulated cells are in an un-polarized,\\nnon-motile state; at higher densities collisions cause cells to transition in to a polarized, motile state (1\\nsimulation of 30 min per density, arrest coefficient: fraction cells ; 2um/min,??',\n",
       " 'shaded: Wilson score\\ninterval based on the number of 2-minute subtracks).',\n",
       " 'CPM parameter settings can be found in Table 3.',\n",
       " '(F) Average speed of the simulated T cells at different densities.',\n",
       " '1-15 ym 16-30 pm 31-45 im 46-60 pm 61-75 um 76-90 um =: 91-110 um\\nB simufation =e D jowadhesion E — towadhesion low adhesion\\n3 3 a 5\\n0 10 20 30 0 10 20 30 SOHm =m 3\\nspeed (\\\\um/min) speed (um/min) high adhesion & high adhesion high adhesion\\n8 0.40 @ 0.25 ee eh cc\\n5, = 0.20 a\\n3 0.00 ge 3 os\\n10 20 30 adhesion o 4 4.',\n",
       " '° % polarisation decay 10 30\\nspeed (um/min) speed (um/min)\\ndata from Mandl et al.?!.',\n",
       " 'to the motion of 50,185 CPM cells moving in a spherical structure (256um\\ndiameter).',\n",
       " '(A) A 130x130x110 um* scan of the stromal structure (fibroblastic reticular cells) in the\\nparacortical T cell zone of a murine LN.',\n",
       " 'The images show average intensity projections of diffe\\nranges.',\n",
       " 'After segmentation and mirroring, these data are used to fill a spherical simulation vo\\nSpeeds observed in simulated data are used to construct a probability distribution (upper panel, solid),\\nwhich we compare to the empirical distribution (upper panel, dashed) using the Kullback-Leibler (KL)\\ndivergence metric.',\n",
       " 'To this end, we bin the speed distributions, compute the divergence in each bin, and\\nsum the values for each bin (lower panel).',\n",
       " '(C) Increased adhesion of T cells to stromal cells causes\\nthe speed distributions to be more right-skewed (top) and match real data more closely (bot:\\nIncreased adhesion can slow T cell paths non-uniformly by creating longer and/or more frequent pauses\\n(red) but does not prevent rapid motion altogether.',\n",
       " 'Two representative tracks are shown.',\n",
       " '(\\nsearch of motility parameters for cell polarization reveals which parameter ranges best fit the intravital\\ndata.',\n",
       " 'Both low and high adhesion simulations have similar fit landscapes; the best fit is\\nhigh adhesion simulation, although the difference to the low-adhesion best fit is small (1 simulation per\\nparameter combination, best fit: red box/red line).',\n",
       " 'CPM parameter settings can be found in\\n 3,359,752 simulated cells\\nO mins 40 mins 80 mins 120 mins\\nD 06 5 0.6\\n0 30 60 90 120 30 100 300 1000\\ntime (in min) scale (in um)\\nhigh adhesion low adhesion\\nLN simulation at 1mm diameter containing 3,359,752 T cells and stromal cells (grey; Figure 8A).',\n",
       " 'For\\nclearer visualization, only 10% of the simulated cells are shown, and red and green cells are equally param-\\neterized although coloured differently.',\n",
       " '(B) The simulation is initialized with two equally parameterized\\npopulations each starting at one half of the sphere.',\n",
       " 'Mixing is quantified by tracking the fraction of cells\\nhaving at least one neighbour of the other population.',\n",
       " '(C) Global mixing is slower in the high-adhesion\\nsimulation (red) than at low adhesion (black), although the difference is minor.',\n",
       " 'Both simulations reach\\n50% mixing in ~1h.',\n",
       " '(D) Cells do not synchronize their directionality much at longer spatial scales, and\\nthis absence of long-range synchronization is the same for low and high adhesion.',\n",
       " 'CPM parameter set-\\ntings can be found in Table 3.',\n",
       " 'Table 1: Model settings for 2D simulations.',\n",
       " 'For all 2D simulations we use Viarget = 2500, Atarget =\\n600, Aperimeter = 2; Jcett,ecm = 10 and T = 10.',\n",
       " 'The vector field in the 2D circular chemotaxis simulations\\n(Figure 5A) consists of vectors at a 90 degree angle to radial vectors from points in the field to the\\nsimulation center, with all vectors in the field normalized to length 1.',\n",
       " '| Avotume | Aaet | Maxact | Achemotaxis\\nFig 1B 20, 30, 40 | 200, 250, 300, 350, 400 50 —\\nFig 5A, Fig 6E | 20 = — 10\\nFig 5B 20 50-800, in increments of 50 | 50 —\\nTable 2: Model settings for 3D simulations.',\n",
       " 'For all simulations Viargee = 150, Avolume = 25,\\nAtarget = 1500, Aperimeter = 0.2, Scett,ce = 10 and T = 20.',\n",
       " 'The time resolution is 1 second per Monte\\n| uw _| At | Apersistence | Jcell,BCM\\nFig 1A — f—|[— 10\\nFig 5C, Fig 6D | 0.95 | 15 | 40 0\\nTable 3: Model settings for liver and lymph node simulations.',\n",
       " 'For all simulations Viarge, = 150,\\nAvolume = 25, Atarget = 1500, Aperimeter = 0-2, Jcett,cett = 10, Jeet,ncm = 0, and T = 20.',\n",
       " 'The time\\nresolution is 1 second per Monte Carlo step.',\n",
       " '| Aact | Maxact Teell, FRC\\nFig 7 2600 10 0\\nFig 8B | 2600 10 0\\nFig 8C | 2600 10 -400, -300, -200, -100, 0\\nFig 8D | 2600 10 -400, 10\\nFig 8E | 200-3000, in increments of 200 | 10, 15, 20, 25, 30, 40, 50, 60, 70 | -3000, 10\\nFig 9C | 2600 10 -800, 0\\nFig 9D | 2600 10 -800\\nThis research was supported by Young Investigator Grant 10620 from KWF Kankerbestrijding and HFSP\\nprogram grant RGP0053/2020 (to JT).',\n",
       " 'JT was also supported by NWO Vidi grant VI.Vidi.192.084.',\n",
       " 'We thank Franka Buytenhuijs for giving valuable feedback on this manuscript, and we thank L.W.A.',\n",
       " 'Stuurman and Pedro Aroca Lara for testing early versions of this simulation system.',\n",
       " 'We also gratefully\\nacknowledge the support of NVIDIA Corporation with the donation of the Titan V GPU used for this\\nresearch.',\n",
       " 'The funders had no role in design, execution, and interpretation of the research reported in\\nSS and JT conceived the study.',\n",
       " 'SS designed the parallel CPM algorithms.',\n",
       " 'SS performed the research and\\nanalysed the data.',\n",
       " 'SS and JT wrote the manuscript.',\n",
       " 'JT supervized the project.',\n",
       " 'SM provided critical\\nfeedback and provided the liver and lymph node imaging data.',\n",
       " 'SD performed image acquisition of the\\nliver and LN data.',\n",
       " 'All authors critically revized the manuscript for intellectual content.',\n",
       " 'The authors declare no competing interests.',\n",
       " 'Materials and Methods\\nCellular Potts model and model variants\\nThe CPM is a lattice based formalism for simulating biological cells.',\n",
       " 'Simulations proceeds with a Markov\\nchain Monte Carlo process, where a single position at a time can transition to one of the states of its\\nneighbours.',\n",
       " 'The transition probability is based on the change of a Hamiltonian energy function.',\n",
       " 'The\\nH =~ Ip(ox),r(0))(1 — 84:,0;) + ¥ > Avotume(Te)(V (0) — Veargot(Te))?',\n",
       " '(1)\\nas its Hamiltonian, with i, 7 being possible neighbour locations in the lattice, 7; the cell present at location\\ni, T is the type of a particular cell, 5 is the Kronecker delta, V(c) is the number of lattice sites a cell o\\noccupies, and Ayolume(T) how strongly a volume constraint contributes to the Hamiltonian for a particular\\ncell type.',\n",
       " 'In all our simulations a Moore neighbourhood was used.',\n",
       " 'Transitions stochastically succeed\\nbased on the change in energy AH, using a Boltzmann probability distribution\\n1 if AH <0,\\nP(AH) = {laws if AH > 0.',\n",
       " '(2)\\nBy adding extra terms to the Hamiltonian, constraints can be added to cells and new types of behavior\\nintroduced to the model.',\n",
       " 'When cell shape needs to be parameterized beyond just cell size, a perimeter\\nconstraint*® is often introduced,\\nHperimeter = Ss Aperimeter (To )(A(a) _ Atarget (To).',\n",
       " '(3)\\nHere A(c) represents the number of interfaces between lattice sites of a cell with sites not part of the\\nWe also implemented several motility models.',\n",
       " 'The first allows a vector field to bias the energy change\\nAH during a copy attempt\\nAH chemotaxis = Achemotaxis (To )#i+7 * Ci.',\n",
       " '(4)\\nHere #;_,; is the unit vector pointing in the direction of a copy attempt from position i to 7, and C; the\\nvalue of a vector field at position 7.',\n",
       " 'In this paper, C; is always a field of unit vectors.',\n",
       " 'For the model where cells exhibit persistent movement based on one vector per cell, we add\\nAHpersistence = Apersistence (To )#i+5 * Po,t (5)\\nto the Hamiltonian delta.',\n",
       " 'The vector ,,, is the persistence direction of cell o at time t, which is updated\\niteratively at every MCS with\\ney Xo ~~ X, t—At a\\nDot = (1 = pr) —S =— + flr Pot—1s (6)\\nPot = 7° (7)\\nHere ju, is the strength of directional persistence for cell type T, Xot is the centroid location of cell o at\\ntime t and At is the number of MCSs we look back for to determine the current direction of a cell.',\n",
       " 'Finally, we implement the Act model’ that allows for localized polarization.',\n",
       " 'For each position i we track\\nan activity value each time said location is updated, here labelled as Act(i).',\n",
       " 'This localized ‘memory’ is\\ninitialized at Maxact, and decays by 1 each timestep until it reaches 0.',\n",
       " 'The Hamiltonian delta is adapted\\nAHact (GMact (?)',\n",
       " '— GMact (J)), (8)\\n making copies that originate from areas of a cell with a high activity value more likely.',\n",
       " 'The activity is\\ncalculated for a full Moore neighbourhood using a geometric mean\\nGMaci(i) = ( T] Act(n)) Ol.',\n",
       " '(9)\\nN(i) are all neighbours of pixel or voxel i.\\nOptimized CPU implementation of CPM\\nBesides a high-performance GPU implementation, we also developed our own CPU implementation.',\n",
       " 'This\\nserved both as a reference implementation for model behavior and internal implementation details for\\nus, and as a reference performance benchmark.',\n",
       " 'We strived to reduce the major performance bottlenecks\\nwe could see in profiling.',\n",
       " 'The implementation tracks lattice sites on cell boundaries, and only tries\\ncopy attempts involving such positions; a common optimisation, which is also implemented in Morpheus.',\n",
       " 'The tracking implementation uses a high-performance hashmap implementation.*!',\n",
       " 'We use a custom\\nhigh-performance pseudorandom number generator, xoshiro256.4?_ The lattice data is stored with a Z-\\norder space filling curve, allowing data that is close in the simulation space also to be close in the\\nlinear memory space, maximizing cache efficiency.',\n",
       " 'Memory addresses are calculated using intrinsic bit\\ninterleaving instructions from the x86 bit manipulation instruction set extension.',\n",
       " 'Our CPM was implemented using CUDA; concepts equivalent to ones described here exist in non-CUDA\\nGPU hardware.',\n",
       " 'A GPU consists of multiple streaming multiprocessors (SMs).',\n",
       " 'Each SM runs multiple\\nthreads in parallel; contrasting with CPU threads, threads on the same SM execute in lockstep, in a\\nSingle Instruction Multiple Threads approach.',\n",
       " 'A single GPU process — called a kernel — can consist of\\nup to millions of threads.',\n",
       " 'GPUs have relatively slow global memory available to all threads, and much faster shared memory local\\nto SMs.',\n",
       " 'Global memory transfer speed is often the main performance bottleneck for GPU programs.',\n",
       " 'Threads are bundled in thread blocks (TBs); each TB is bound to a specific SM.',\n",
       " 'By caching data on\\ntheir SM in shared memory, TBs can greatly reduce global memory transfer bottlenecks.',\n",
       " 'Memory transfer bottlenecks can be further reduced by two strategies: read coalescing and high SM\\noccupancy.',\n",
       " 'Read coalescing has sequential threads in a TB do contiguous reads of global memory;\\ncontiguous reads are needed to fully utilise the memory bus bandwidth.',\n",
       " 'High occupancy is achieved when\\na high number of threads are scheduled to execute per SM; the SM scheduler will switch out threads\\nwhen they are blocked on memory transfers, ensuring a high number of in-flight memory instructions at\\nWe propose a scheme of memory transfer, local SM execution and synchronization (Figure 2D).',\n",
       " 'We\\nfocus on caching just the grid of cell ids, as this data is always needed, and can be used to quickly reject\\ncopy attempts internal to a cell.',\n",
       " 'The simulation grid is divided in subsections, with regions larger than\\nearlier checkerboard subdivisions; each subsection is assigned to a TB, subsection size is chosen so active\\nregions fit in shared memory.',\n",
       " 'Each TB determines which checkerboard sections are active in its region;\\nactive sections are cached in shared memory using coalesced reads.',\n",
       " 'After caching — intended to speed up reads — the simulation is run, and successful copy attempts write\\nnew values to both shared and global memory.',\n",
       " 'Alternatively, changes could be cached in shared memory\\nand written back coalesced to global memory; in our testing, an immediate shared/global write yielded\\nbetter performance.',\n",
       " 'If a TB is done with simulation execution, it immediately starts a memory request\\nfor the next active region, ensuring maximum occupancy.',\n",
       " 'All threads in the kernel are synchronized after\\nnew memory transfers are finished, ensuring maximum overlap of code execution and memory transfer.',\n",
       " 'CPM literature offers a wide range of model variants and extensions, of which we implement a mod-\\nest subset.',\n",
       " 'This comes with potential performance issues: even when extensions are disabled through\\nbranching statements, on GPUs such branching code still incurs a significant performance loss compared\\nto CPUs.',\n",
       " 'We therefore pre-generated specialized versions of the kernel for a range of different extensions.',\n",
       " 'We use a custom high-performance random number generator on the GPU.',\n",
       " '*%\\n Composite variable tracking\\nSeveral CPM extensions introduce additional global variables.',\n",
       " 'Some of these variable need to be kept\\nin-syne with each other — one variable being slightly out of date with others is not tolerated.',\n",
       " 'One\\nexample is tracking of a center-of-mass centroid; for 2D this can be tracked as a sum of x and y values,\\nand corresponding count of lattice locations occupied (Figure 2C).',\n",
       " 'These are integer values: tracking\\ncentroid locations as floating-point would cause values to drift due to rounding errors.',\n",
       " 'The centroid value\\n{a,y} = = X {asum, Ysum }- (10)\\nOnline updates change both a cell’s centroid position and size.',\n",
       " 'When a new lattice position {Xposition, Yposition }\\nis added to a cell, its centroid’s x value can be updated with\\nLoia X |® lola + Lposition\\nCell sizes are stored interleaved with sum values; sizes and sums are both stored as 32-bit unsigned\\nintegers.',\n",
       " 'The interleaved storage enables use of 64-bit operations to update both: a copy-and-swap\\n(CAS) operation is used, which atomically updates sum and cell size values simultaneously.',\n",
       " 'This scheme\\nalso handles periodic boundaries, by adding or subtracting dimension x |e| to the sum values before\\nexecuting the CAS as needed, to keep centroids within simulation bounds.',\n",
       " 'We devized a scheme to execute our CPM on multiple GPUs (Figure 2E); this scheme is intended as a\\nprototype, and as such not as optimized as other parts of our CPM.',\n",
       " 'Simulation space is cut into equal\\nregions, each region is assigned to a GPU.',\n",
       " 'A region’s lattice data, such as the cell ids, are sent to each\\nGPU, with an additional border layer.',\n",
       " 'Global variables such as cell sizes are tracked on each GPU as\\ndeltas of the value since last inter-GPU data synchronisation.',\n",
       " 'Every four MCSs, these deltas are gathered\\nfrom all GPUs, deltas are integrated, and updated values distributed to each GPU.',\n",
       " 'Multi-GPU speedup\\nis mostly hampered by frequent or slow inter-GPU communication; inter-GPU synchronization after every\\nMCS negates most speedup, and a faster GPU-to-GPU interlink would likely increase performance.',\n",
       " 'The CPU-based simulations, using our custom optimized CPU implementation and Morpheus 2.2.6 (the\\nlast non-parallelized version of Morpheus), and most GPU simulations were run on a system with a\\nNvidia Titan V, an AMD 2950x CPU and 128 GB of RAM at 2133 MHz.',\n",
       " 'The only exceptions to this\\nare the multi-GPU simulations (Figure 2E), which were run on a system with four Nvidia Geforce RTX\\n2080 Ti GPUs, an Intel i9-9820x CPU at 3.3 GHz, and 128 GB of RAM at 2133 MHz.',\n",
       " 'Imaging of thick lymph node sections\\nNaive inguinal LNs were harvested from CCL19-EYFP mice“ and fixed with 4% paraformaldehyde (PFA,\\nElectron Microscopy Sciences) for 1 hour at 4°C.',\n",
       " 'Following a brief wash with PBS, LNs were embedded in\\n2% agarose and 250-300 ym tissue sections were generated using a vibratome (Leica).',\n",
       " 'LN sections were\\nsubsequently immersed in a Special Wash Buffer (Baxter water, 0.2% BSA, 10% 10X PBS, 0.2% Tween-\\n20, 0.2% Triton X-100, 0.2% 10% SDS) at 4°C overnight, followed by staining with a rabbit anti-GFP Ab\\n(Thermo Scientific, 1:100, diluted in blocking buffer made with Baxter water, 0.2% BSA, 10% 10X PBS,\\n0.1% Tween-20, 0.2% 10% SDS) at 4°C for three days.',\n",
       " 'After incubation, LN sections were thoroughly\\nwashed in PBS and mounted in 85% glycerol in PBS with glass coverslips.',\n",
       " 'Thick LN sections were imaged\\nusing a LSM880 AiryscanFast confocal microscope (Carl Zeiss) with a x63/1.4NA oil immersion lens and\\nimages processed with ZEN 2010 software (Carl Zeiss).',\n",
       " 'Imaging of liver sinusoids\\nFor whole-liver imaging, AlexaFluor 647 conjugated anti-CD31 rat monoclonal antibody (Biolegend) was\\nintravenously injected into wild-type C57BL/6 mice at 5ug for 15 minutes prior to harvesting the liver.']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def clean_text(text):\n",
    "    \"\"\"Remove section titles and figure descriptions from text\"\"\"\n",
    "    clean = \"\\n\".join([row for row in text.split(\"\\n\") if\n",
    "                      (len(row.split(\" \"))) > 3 and not (row.startswith(\"(a)\"))\n",
    "                      and not row.startswith(\"Figure\")])\n",
    "    return clean\n",
    "\n",
    "text = article_txt.split(\"Introduction\")[1]\n",
    "ctext = clean_text(text)\n",
    "sentences = nltk.tokenize.sent_tokenize(ctext)\n",
    "sentences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "19aa071d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_plain(text, url=\"http://bern2.korea.ac.kr/plain\"):\n",
    " \"\"\"Biomedical entity linking API\"\"\"\n",
    " return requests.post(url, json={'text':str(text)}).json()\n",
    "\n",
    "\n",
    "def query_plain2(text, url=\"http://bern2.korea.ac.kr/plain\"):\n",
    "    \"\"\"Biomedical entity linking API\"\"\"\n",
    "    try:\n",
    "        response = requests.post(url, json={'text': str(text)})\n",
    "        response.raise_for_status()  # Raise an error for bad responses (4xx and 5xx)\n",
    "        return response.json()\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(f\"Error making request: {e}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4df47184",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "376\n",
      "doing sentence : 0\n",
      "doing sentence : 1\n",
      "doing sentence : 2\n",
      "doing sentence : 3\n",
      "doing sentence : 4\n",
      "doing sentence : 5\n",
      "doing sentence : 6\n",
      "doing sentence : 7\n",
      "doing sentence : 8\n",
      "doing sentence : 9\n",
      "doing sentence : 10\n",
      "doing sentence : 11\n",
      "doing sentence : 12\n",
      "doing sentence : 13\n",
      "doing sentence : 14\n",
      "doing sentence : 15\n",
      "doing sentence : 16\n",
      "doing sentence : 17\n",
      "doing sentence : 18\n",
      "doing sentence : 19\n",
      "doing sentence : 20\n",
      "doing sentence : 21\n",
      "doing sentence : 22\n",
      "doing sentence : 23\n",
      "doing sentence : 24\n",
      "doing sentence : 25\n",
      "doing sentence : 26\n",
      "doing sentence : 27\n",
      "doing sentence : 28\n",
      "doing sentence : 29\n",
      "doing sentence : 30\n",
      "doing sentence : 31\n",
      "doing sentence : 32\n",
      "doing sentence : 33\n",
      "doing sentence : 34\n",
      "doing sentence : 35\n",
      "doing sentence : 36\n",
      "doing sentence : 37\n",
      "doing sentence : 38\n",
      "doing sentence : 39\n",
      "doing sentence : 40\n",
      "doing sentence : 41\n",
      "doing sentence : 42\n",
      "doing sentence : 43\n",
      "doing sentence : 44\n",
      "doing sentence : 45\n",
      "doing sentence : 46\n",
      "doing sentence : 47\n",
      "doing sentence : 48\n",
      "doing sentence : 49\n",
      "doing sentence : 50\n",
      "doing sentence : 51\n",
      "Error making request: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "doing sentence : 52\n",
      "doing sentence : 53\n",
      "doing sentence : 54\n",
      "doing sentence : 55\n",
      "doing sentence : 56\n",
      "doing sentence : 57\n",
      "doing sentence : 58\n",
      "doing sentence : 59\n",
      "doing sentence : 60\n",
      "doing sentence : 61\n",
      "doing sentence : 62\n",
      "doing sentence : 63\n",
      "Error making request: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "doing sentence : 64\n",
      "doing sentence : 65\n",
      "doing sentence : 66\n",
      "doing sentence : 67\n",
      "doing sentence : 68\n",
      "doing sentence : 69\n",
      "doing sentence : 70\n",
      "doing sentence : 71\n",
      "doing sentence : 72\n",
      "doing sentence : 73\n",
      "doing sentence : 74\n",
      "doing sentence : 75\n",
      "doing sentence : 76\n",
      "doing sentence : 77\n",
      "doing sentence : 78\n",
      "doing sentence : 79\n",
      "doing sentence : 80\n",
      "doing sentence : 81\n",
      "doing sentence : 82\n",
      "doing sentence : 83\n",
      "doing sentence : 84\n",
      "doing sentence : 85\n",
      "doing sentence : 86\n",
      "doing sentence : 87\n",
      "doing sentence : 88\n",
      "doing sentence : 89\n",
      "doing sentence : 90\n",
      "doing sentence : 91\n",
      "doing sentence : 92\n",
      "doing sentence : 93\n",
      "doing sentence : 94\n",
      "doing sentence : 95\n",
      "doing sentence : 96\n",
      "doing sentence : 97\n",
      "doing sentence : 98\n",
      "doing sentence : 99\n",
      "doing sentence : 100\n",
      "doing sentence : 101\n",
      "doing sentence : 102\n",
      "doing sentence : 103\n",
      "doing sentence : 104\n",
      "doing sentence : 105\n",
      "doing sentence : 106\n",
      "doing sentence : 107\n",
      "doing sentence : 108\n",
      "doing sentence : 109\n",
      "doing sentence : 110\n",
      "doing sentence : 111\n",
      "doing sentence : 112\n",
      "doing sentence : 113\n",
      "doing sentence : 114\n",
      "doing sentence : 115\n",
      "doing sentence : 116\n",
      "doing sentence : 117\n",
      "doing sentence : 118\n",
      "doing sentence : 119\n",
      "doing sentence : 120\n",
      "doing sentence : 121\n",
      "doing sentence : 122\n",
      "doing sentence : 123\n",
      "doing sentence : 124\n",
      "doing sentence : 125\n",
      "doing sentence : 126\n",
      "doing sentence : 127\n",
      "doing sentence : 128\n",
      "doing sentence : 129\n",
      "doing sentence : 130\n",
      "doing sentence : 131\n",
      "doing sentence : 132\n",
      "doing sentence : 133\n",
      "doing sentence : 134\n",
      "doing sentence : 135\n",
      "doing sentence : 136\n",
      "doing sentence : 137\n",
      "doing sentence : 138\n",
      "doing sentence : 139\n",
      "doing sentence : 140\n",
      "doing sentence : 141\n",
      "doing sentence : 142\n",
      "doing sentence : 143\n",
      "doing sentence : 144\n",
      "doing sentence : 145\n",
      "doing sentence : 146\n",
      "doing sentence : 147\n",
      "doing sentence : 148\n",
      "doing sentence : 149\n",
      "doing sentence : 150\n",
      "doing sentence : 151\n",
      "doing sentence : 152\n",
      "doing sentence : 153\n",
      "doing sentence : 154\n",
      "doing sentence : 155\n",
      "doing sentence : 156\n",
      "doing sentence : 157\n",
      "doing sentence : 158\n",
      "doing sentence : 159\n",
      "doing sentence : 160\n",
      "Error making request: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "doing sentence : 161\n",
      "doing sentence : 162\n",
      "doing sentence : 163\n",
      "doing sentence : 164\n",
      "doing sentence : 165\n",
      "doing sentence : 166\n",
      "doing sentence : 167\n",
      "doing sentence : 168\n",
      "doing sentence : 169\n",
      "doing sentence : 170\n",
      "doing sentence : 171\n",
      "doing sentence : 172\n",
      "doing sentence : 173\n",
      "doing sentence : 174\n",
      "doing sentence : 175\n",
      "doing sentence : 176\n",
      "doing sentence : 177\n",
      "doing sentence : 178\n",
      "doing sentence : 179\n",
      "doing sentence : 180\n",
      "doing sentence : 181\n",
      "doing sentence : 182\n",
      "doing sentence : 183\n",
      "doing sentence : 184\n",
      "doing sentence : 185\n",
      "doing sentence : 186\n",
      "doing sentence : 187\n",
      "doing sentence : 188\n",
      "doing sentence : 189\n",
      "doing sentence : 190\n",
      "doing sentence : 191\n",
      "doing sentence : 192\n",
      "doing sentence : 193\n",
      "doing sentence : 194\n",
      "doing sentence : 195\n",
      "doing sentence : 196\n",
      "doing sentence : 197\n",
      "doing sentence : 198\n",
      "doing sentence : 199\n",
      "doing sentence : 200\n",
      "doing sentence : 201\n",
      "doing sentence : 202\n",
      "doing sentence : 203\n",
      "doing sentence : 204\n",
      "doing sentence : 205\n",
      "doing sentence : 206\n",
      "doing sentence : 207\n",
      "doing sentence : 208\n",
      "doing sentence : 209\n",
      "doing sentence : 210\n",
      "doing sentence : 211\n",
      "doing sentence : 212\n",
      "doing sentence : 213\n",
      "doing sentence : 214\n",
      "doing sentence : 215\n",
      "doing sentence : 216\n",
      "doing sentence : 217\n",
      "doing sentence : 218\n",
      "doing sentence : 219\n",
      "doing sentence : 220\n",
      "doing sentence : 221\n",
      "doing sentence : 222\n",
      "doing sentence : 223\n",
      "doing sentence : 224\n",
      "doing sentence : 225\n",
      "doing sentence : 226\n",
      "doing sentence : 227\n",
      "doing sentence : 228\n",
      "doing sentence : 229\n",
      "doing sentence : 230\n",
      "doing sentence : 231\n",
      "doing sentence : 232\n",
      "doing sentence : 233\n",
      "doing sentence : 234\n",
      "doing sentence : 235\n",
      "doing sentence : 236\n",
      "doing sentence : 237\n",
      "doing sentence : 238\n",
      "doing sentence : 239\n",
      "doing sentence : 240\n",
      "doing sentence : 241\n",
      "doing sentence : 242\n",
      "doing sentence : 243\n",
      "doing sentence : 244\n",
      "doing sentence : 245\n",
      "doing sentence : 246\n",
      "doing sentence : 247\n",
      "doing sentence : 248\n",
      "doing sentence : 249\n",
      "doing sentence : 250\n",
      "doing sentence : 251\n",
      "doing sentence : 252\n",
      "doing sentence : 253\n",
      "doing sentence : 254\n",
      "doing sentence : 255\n",
      "doing sentence : 256\n",
      "doing sentence : 257\n",
      "doing sentence : 258\n",
      "doing sentence : 259\n",
      "doing sentence : 260\n",
      "doing sentence : 261\n",
      "doing sentence : 262\n",
      "doing sentence : 263\n",
      "doing sentence : 264\n",
      "doing sentence : 265\n",
      "doing sentence : 266\n",
      "doing sentence : 267\n",
      "doing sentence : 268\n",
      "doing sentence : 269\n",
      "doing sentence : 270\n",
      "doing sentence : 271\n",
      "doing sentence : 272\n",
      "doing sentence : 273\n",
      "doing sentence : 274\n",
      "doing sentence : 275\n",
      "doing sentence : 276\n",
      "doing sentence : 277\n",
      "doing sentence : 278\n",
      "doing sentence : 279\n",
      "doing sentence : 280\n",
      "doing sentence : 281\n",
      "doing sentence : 282\n",
      "doing sentence : 283\n",
      "doing sentence : 284\n",
      "doing sentence : 285\n",
      "doing sentence : 286\n",
      "doing sentence : 287\n",
      "doing sentence : 288\n",
      "doing sentence : 289\n",
      "doing sentence : 290\n",
      "doing sentence : 291\n",
      "doing sentence : 292\n",
      "doing sentence : 293\n",
      "doing sentence : 294\n",
      "doing sentence : 295\n",
      "doing sentence : 296\n",
      "doing sentence : 297\n",
      "doing sentence : 298\n",
      "doing sentence : 299\n",
      "doing sentence : 300\n",
      "doing sentence : 301\n",
      "doing sentence : 302\n",
      "doing sentence : 303\n",
      "doing sentence : 304\n",
      "doing sentence : 305\n",
      "doing sentence : 306\n",
      "doing sentence : 307\n",
      "doing sentence : 308\n",
      "doing sentence : 309\n",
      "doing sentence : 310\n",
      "doing sentence : 311\n",
      "doing sentence : 312\n",
      "doing sentence : 313\n",
      "doing sentence : 314\n",
      "doing sentence : 315\n",
      "doing sentence : 316\n",
      "doing sentence : 317\n",
      "doing sentence : 318\n",
      "doing sentence : 319\n",
      "doing sentence : 320\n",
      "doing sentence : 321\n",
      "doing sentence : 322\n",
      "Error making request: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "doing sentence : 323\n",
      "doing sentence : 324\n",
      "doing sentence : 325\n",
      "doing sentence : 326\n",
      "doing sentence : 327\n",
      "doing sentence : 328\n",
      "doing sentence : 329\n",
      "doing sentence : 330\n",
      "doing sentence : 331\n",
      "doing sentence : 332\n",
      "doing sentence : 333\n",
      "doing sentence : 334\n",
      "doing sentence : 335\n",
      "doing sentence : 336\n",
      "doing sentence : 337\n",
      "doing sentence : 338\n",
      "doing sentence : 339\n",
      "doing sentence : 340\n",
      "doing sentence : 341\n",
      "doing sentence : 342\n",
      "doing sentence : 343\n",
      "doing sentence : 344\n",
      "doing sentence : 345\n",
      "doing sentence : 346\n",
      "doing sentence : 347\n",
      "doing sentence : 348\n",
      "doing sentence : 349\n",
      "doing sentence : 350\n",
      "doing sentence : 351\n",
      "doing sentence : 352\n",
      "doing sentence : 353\n",
      "doing sentence : 354\n",
      "doing sentence : 355\n",
      "doing sentence : 356\n",
      "doing sentence : 357\n",
      "doing sentence : 358\n",
      "doing sentence : 359\n",
      "doing sentence : 360\n",
      "doing sentence : 361\n",
      "doing sentence : 362\n",
      "doing sentence : 363\n",
      "doing sentence : 364\n",
      "doing sentence : 365\n",
      "doing sentence : 366\n",
      "doing sentence : 367\n",
      "doing sentence : 368\n",
      "doing sentence : 369\n",
      "doing sentence : 370\n",
      "doing sentence : 371\n",
      "doing sentence : 372\n",
      "doing sentence : 373\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "doing sentence : 374\n",
      "all calls done\n",
      "[{'annotations': [], 'text': 'The CPM!', 'timestamp': 'Fri Dec 22 09:30:40 +0000 2023'}, {'annotations': [], 'text': 'is a powerful simulation method that is known for its ability to simulate a wide range of biological scenarios.', 'timestamp': 'Fri Dec 22 09:30:41 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'single cells', 'obj': 'cell_type', 'prob': 0.7551743984222412, 'span': {'begin': 100, 'end': 112}}], 'text': 'When modelling tissue environments, many researchers wish to consider both he complex morphology of single cells, and the behavior of these cells as part of larger collectives.', 'timestamp': 'Fri Dec 22 09:30:43 +0000 2023'}, {'annotations': [], 'text': 'The appeal of the CPM lies partly in its ability to integrate both these aspects in a single model.', 'timestamp': 'Fri Dec 22 09:30:44 +0000 2023'}, {'annotations': [{'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumor', 'obj': 'disease', 'prob': 0.9999828338623047, 'span': {'begin': 59, 'end': 64}}], 'text': 'As such, it has been applied to diverse problems including tumor growth, vascularization, wound healing,* and morphogenesis.>* Many extensions and hybrid model variants are available in the literature 0, for example, add support for cell motility, cell signalling,! and finite-element based mechanical interactions.!!', 'timestamp': 'Fri Dec 22 09:30:45 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.4906628131866455, 'span': {'begin': 10, 'end': 13}}], 'text': 'Thus, the CPM can be described as a generic platform upon which a wide range of models can be built, possibly explaining its popularity in the tissue modelling field.', 'timestamp': 'Fri Dec 22 09:30:46 +0000 2023'}, {'annotations': [], 'text': 'For computational reasons, CPM models are often built using a small number of cells (perhaps in the housands) in simplified two-dimensional (2D) geometries.', 'timestamp': 'Fri Dec 22 09:30:47 +0000 2023'}, {'annotations': [], 'text': 'Some research questions in tissue modelling however demand a realistically large number of cells, and accurate 3D representations of tissues.', 'timestamp': 'Fri Dec 22 09:30:48 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'motile cells', 'obj': 'cell_type', 'prob': 0.8783984184265137, 'span': {'begin': 83, 'end': 95}}], 'text': 'For instance, it is widely appreciated that the efficiency of space exploration of motile cells depends criti- cally on the dimensionality of the environment, leading researchers to use 3D CPM simulations. !?', 'timestamp': 'Fri Dec 22 09:30:50 +0000 2023'}, {'annotations': [{'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumor', 'obj': 'disease', 'prob': 0.9999728202819824, 'span': {'begin': 18, 'end': 23}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumors', 'obj': 'disease', 'prob': 0.9999914169311523, 'span': {'begin': 65, 'end': 71}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumor', 'obj': 'disease', 'prob': 0.9998588562011719, 'span': {'begin': 167, 'end': 172}}], 'text': ':13 Assessment of tumor growth behavior and invasive patterns of tumors likewise changes as a function of dimension, resulting in the use of 3D CPM simulations in the tumor modelling field as well.?', 'timestamp': 'Fri Dec 22 09:30:51 +0000 2023'}, {'annotations': [], 'text': 'While there is thus a clear research demand for realistically large 3D CPM simulations, scaling up such simulations is generally considered computationally infeasible.4 1 Removing this computational barrier would allow a host of CPM based modelling research to be scaled to what is biologically realistic, instead of scaling to what technical constraints allow.', 'timestamp': 'Fri Dec 22 09:30:52 +0000 2023'}, {'annotations': [], 'text': 'The CPM is an on-lattice method: simulations are executed on a grid.', 'timestamp': 'Fri Dec 22 09:30:54 +0000 2023'}, {'annotations': [], 'text': 'A common strategy to speed up grid-based models is to cut the simulation into smaller subsections that are run in parallel.', 'timestamp': 'Fri Dec 22 09:30:55 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.5344692468643188, 'span': {'begin': 17, 'end': 20}}], 'text': 'However, for the CPM, model behavior is intimately tied to locations on the grid being updated sequentially, in a single- location update Markov chain Monte Carlo process.', 'timestamp': 'Fri Dec 22 09:30:56 +0000 2023'}, {'annotations': [], 'text': 'As such, parallelisation has been cautioned against, as it risks drastic changes in CPM system kinetics. Nevertheless, several attempts have been made to employ subsection-based parallelisation.', 'timestamp': 'Fri Dec 22 09:30:57 +0000 2023'}, {'annotations': [], 'text': 'Chen et al.', 'timestamp': 'Fri Dec 22 09:30:58 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.6286525130271912, 'span': {'begin': 48, 'end': 51}}], 'text': 'proposed a checkerboard scheme that divides the CPM lattice up over a computer cluster, with sections chosen much larger than cells, preventing cell-level synchronization issues.', 'timestamp': 'Fri Dec 22 09:30:59 +0000 2023'}, {'annotations': [], 'text': '! Tapia et al.', 'timestamp': 'Fri Dec 22 09:31:00 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.7008001208305359, 'span': {'begin': 29, 'end': 33}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.618698239326477, 'span': {'begin': 93, 'end': 96}}], 'text': 'propose a similar scheme for GPUs:! because the sections are much smaller than a cell on the GPU, they use a per-cell locking mechanism to prevent simultaneous updates of a cell.', 'timestamp': 'Fri Dec 22 09:31:02 +0000 2023'}, {'annotations': [], 'text': 'Neither approach has been extensively checked for correct model behavior.', 'timestamp': 'Fri Dec 22 09:31:03 +0000 2023'}, {'annotations': [], 'text': 'To our knowledge, no publicly available implementation of subsection-based parallelisation exists for the CPM.', 'timestamp': 'Fri Dec 22 09:31:04 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.4314751923084259, 'span': {'begin': 194, 'end': 198}}], 'text': 'The software Morpheus ! includes a non-subsection-based Central Processing Unit (CPU) parallelisation strategy; this method is less suited to fully utilise parallel accelerator hardware such as GPUs.', 'timestamp': 'Fri Dec 22 09:31:05 +0000 2023'}, {'annotations': [], 'text': 'In this paper, we aim to make several contributions: First, we introduce a highly optimized parallelized implementation of the CPM using several optimization strategies not previously explored in the literature.', 'timestamp': 'Fri Dec 22 09:31:07 +0000 2023'}, {'annotations': [], 'text': 'Second, we conduct an empirical analysis of potential pitfalls in parallelising a model dependent on serial updates of single locations and show that our approach mitigates these issues while preserving performance benefits.', 'timestamp': 'Fri Dec 22 09:31:08 +0000 2023'}, {'annotations': [], 'text': 'Third, we demonstrate that our approach retains the behavior of cell motility simulations that are particularly prone to changed model kinetics upon parallelisation.', 'timestamp': 'Fri Dec 22 09:31:10 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.8993971943855286, 'span': {'begin': 62, 'end': 68}}], 'text': 'And finally, we build two large 3D image-based simulations of T cell migration in liver and lymph nodes, demonstrating how complex tissue environments can be analysed at a larger scale and realism than was possible before.', 'timestamp': 'Fri Dec 22 09:31:11 +0000 2023'}, {'annotations': [], 'text': 'Serial Cellular Potts Models are limited to sub-millimetre scale The CPMs serial update schedule  only one location at a time is updated on a grid  is central to its behavior, and thus not trivially changed.', 'timestamp': 'Fri Dec 22 09:31:12 +0000 2023'}, {'annotations': [], 'text': 'We therefore first aimed to quantify the computational limitations for models based on the serial CPM: how big can our computational experiments be, and what timescale can they span, without making fundamental changes to the CPM?', 'timestamp': 'Fri Dec 22 09:31:15 +0000 2023'}, {'annotations': [], 'text': 'We set up 3D CPM tissue experiments at different spatial and temporal scales and measured the required wall-clock time.', 'timestamp': 'Fri Dec 22 09:31:17 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'highly', 'obj': 'cell_line', 'prob': 0.5670343637466431, 'span': {'begin': 104, 'end': 110}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'highly motile cells', 'obj': 'cell_type', 'prob': 0.6405029892921448, 'span': {'begin': 104, 'end': 123}}], 'text': 'We choose spatial and temporal simulation resolutions that allow reasonably accurate representations of highly motile cells within tissues.', 'timestamp': 'Fri Dec 22 09:31:18 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Lymphocytes', 'obj': 'cell_type', 'prob': 0.793220579624176, 'span': {'begin': 0, 'end': 11}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'nucleated cells', 'obj': 'cell_type', 'prob': 0.7073931694030762, 'span': {'begin': 35, 'end': 50}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'B cells', 'obj': 'cell_type', 'prob': 0.9133166074752808, 'span': {'begin': 96, 'end': 103}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.89553302526474, 'span': {'begin': 108, 'end': 115}}], 'text': 'Lymphocytes are among the smallest nucleated cells in the mammalian body at 8-10um diameter for B cells and T cells.', 'timestamp': 'Fri Dec 22 09:31:21 +0000 2023'}, {'annotations': [], 'text': 'At speeds of up to a few body lo | 30 values/param.', 'timestamp': 'Fri Dec 22 09:31:22 +0000 2023'}, {'annotations': [], 'text': '2  1wk 7 5 c= a = , - [sj x e Pa  jw || 8 ow 6 cS ihr 4_- 2 L | 2 OS Tmin 10 values/param.', 'timestamp': 'Fri Dec 22 09:31:23 +0000 2023'}, {'annotations': [], 'text': 'polarisation strength replicates 1 2 3 4 5 sembles.', 'timestamp': 'Fri Dec 22 09:31:25 +0000 2023'}, {'annotations': [], 'text': '(A) Computational cost of simulating a 256m* volume (1m per voxel) for 10 simulated minutes (1s per simulation step).', 'timestamp': 'Fri Dec 22 09:31:26 +0000 2023'}, {'annotations': [], 'text': 'With a high-performance CPU implementation, this takes 1.2 hours of wall-clock time.', 'timestamp': 'Fri Dec 22 09:31:29 +0000 2023'}, {'annotations': [], 'text': 'Extrapolation to a full day of simulated time for a 1mm?', 'timestamp': 'Fri Dec 22 09:31:31 +0000 2023'}, {'annotations': [], 'text': 'simulation estimates 1.4 years of wall-clock time.', 'timestamp': 'Fri Dec 22 09:31:33 +0000 2023'}, {'annotations': [], 'text': 'Time measurements (black) are averages of 5 simulations, projections are based on inear fits of simulations > 64m* in log-log space (grey).', 'timestamp': 'Fri Dec 22 09:31:35 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPMs', 'obj': 'cell_line', 'prob': 0.4444953203201294, 'span': {'begin': 13, 'end': 17}}], 'text': '(B) Building CPMs typically requires tuning parameters.', 'timestamp': 'Fri Dec 22 09:31:38 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'motile cells', 'obj': 'cell_type', 'prob': 0.8557302951812744, 'span': {'begin': 73, 'end': 85}}, {'id': ['NCBIGene:12', '71', '72', '9457', '345651', '11332', '440915', '641455'], 'is_neural_normalized': True, 'mention': 'Act', 'obj': 'gene', 'prob': 0.8408151268959045, 'span': {'begin': 96, 'end': 99}}], 'text': 'Here we perform a multidimensional grid search for parameters that yield motile cells using the Act extension of the CPM; a modest single parameter grid search where 10 values are tried takes 5.9 seconds; larger grid searches take exponentially more time as function of parameters searched, with a five parameter search with 30 values is extrapolated to 1.1 years of wall-clock time.', 'timestamp': 'Fri Dec 22 09:31:41 +0000 2023'}, {'annotations': [], 'text': 'Time measurements (black) are averaged over 5 simulations, projections fitted on linear trends in linear-log space.', 'timestamp': 'Fri Dec 22 09:31:42 +0000 2023'}, {'annotations': [], 'text': 'CPM parameters are shown in Table 1 and Table 2. engths per minute, they are also among the fastest movers within tissues.', 'timestamp': 'Fri Dec 22 09:31:49 +0000 2023'}, {'annotations': [], 'text': 'We chose lum?', 'timestamp': 'Fri Dec 22 09:31:51 +0000 2023'}, {'annotations': [{'id': ['NCBITaxon:10090'], 'is_neural_normalized': False, 'mention': 'murine', 'obj': 'species', 'prob': 0.9930986762046814, 'span': {'begin': 198, 'end': 204}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'tumor', 'obj': 'disease', 'prob': 0.9999860525131226, 'span': {'begin': 233, 'end': 238}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'avascular', 'obj': 'cell_line', 'prob': 0.5243708491325378, 'span': {'begin': 263, 'end': 272}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'avascular tumor spheroid', 'obj': 'cell_type', 'prob': 0.5380792021751404, 'span': {'begin': 263, 'end': 287}}], 'text': 'per voxel and second per simulation step, or Monte Carlo Step (MCS), as our simulation resolutions, corresponding to Beltman et al.9 To create realistically large models of, for example, a complete murine Lymph Node (LN), or scale a tumor model to the size of an avascular tumor spheroid,? experiments at mm scale would be needed.', 'timestamp': 'Fri Dec 22 09:31:53 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.8515430688858032, 'span': {'begin': 75, 'end': 81}}], 'text': 'Certain organ scale dynamics require timescales of at least a day, such as T cell transit through LNs.', 'timestamp': 'Fri Dec 22 09:31:54 +0000 2023'}, {'annotations': [], 'text': '?4 More modest simulation experiments at 256um?', 'timestamp': 'Fri Dec 22 09:31:58 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPU', 'obj': 'DNA', 'prob': 0.4887787103652954, 'span': {'begin': 89, 'end': 92}}], 'text': '(Figure 1A) at 10 simulated minutes require ~1h of wall-clock time on a highly optimized CPU implementation of the CPM (Methods).', 'timestamp': 'Fri Dec 22 09:31:59 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'motile cells', 'obj': 'cell_type', 'prob': 0.7375972270965576, 'span': {'begin': 59, 'end': 71}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9229081869125366, 'span': {'begin': 80, 'end': 87}}, {'id': ['mesh:D009369'], 'is_neural_normalized': False, 'mention': 'cancer', 'obj': 'disease', 'prob': 0.9997592568397522, 'span': {'begin': 187, 'end': 193}}], 'text': 'Such experiments are sufficient to see minimal movement of motile cells such as T cells and can capture organ compartments such as a LN paracortical T cell zone, or a micrometastasis for cancer modelling.', 'timestamp': 'Fri Dec 22 09:32:00 +0000 2023'}, {'annotations': [], 'text': 'For mm?', 'timestamp': 'Fri Dec 22 09:32:01 +0000 2023'}, {'annotations': [{'id': ['NCBITaxon:9606'], 'is_neural_normalized': False, 'mention': 'patient', 'obj': 'species', 'prob': 0.8306940793991089, 'span': {'begin': 129, 'end': 136}}], 'text': 'scale, one-day experiments however, researchers would have to wait at least a year  making them impractical for all but the most patient of researchers.', 'timestamp': 'Fri Dec 22 09:32:02 +0000 2023'}, None, {'annotations': [], 'text': 'Beyond the need or experimental replicates, simulations can have parameters for which associated model behavior is non- obvious, requiring parameterized simulation ensembles.', 'timestamp': 'Fri Dec 22 09:32:04 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:148709', '402221'], 'is_neural_normalized': True, 'mention': 'actin', 'obj': 'gene', 'prob': 0.9955233335494995, 'span': {'begin': 83, 'end': 88}}], 'text': 'In Figure 1B, we explore a cell motility extension to the CPM that coarsely models actin dynamics. A single two-dimension simulation is relatively cheap, but when exploring even a modest number of parameters the execution time goes up exponentially.', 'timestamp': 'Fri Dec 22 09:32:06 +0000 2023'}, {'annotations': [], 'text': 'Thus, serial 3D CPM simulations at a scale of 1 mm are currently not feasible, and even 100um?', 'timestamp': 'Fri Dec 22 09:32:07 +0000 2023'}, {'annotations': [], 'text': 'scale simulations were extremely challenging in our experiments.', 'timestamp': 'Fri Dec 22 09:32:08 +0000 2023'}, {'annotations': [], 'text': 'While ensemble runs can be accelerated with parallel simulations without changing the CPM, the high computational costs of running large ensembles urther motivates the need for a faster CPM.', 'timestamp': 'Fri Dec 22 09:32:09 +0000 2023'}, {'annotations': [], 'text': 'Designing a parallel execution scheme for the CPM To make large scale simulations practical, we propose a parallelisation scheme for the CPM that uses highly parallel GPU hardware.', 'timestamp': 'Fri Dec 22 09:32:10 +0000 2023'}, {'annotations': [], 'text': 'A natural way divide the simulation grid in support up to thousands of 0 parallelise on-lattice models, such as the CPM, is to er sections; each section runs a local simulation in parallel.', 'timestamp': 'Fri Dec 22 09:32:12 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.5757036805152893, 'span': {'begin': 3, 'end': 7}}], 'text': 'As GPUs ations in parallel, these sections will be quite small.', 'timestamp': 'Fri Dec 22 09:32:13 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated', 'obj': 'cell_line', 'prob': 0.49996381998062134, 'span': {'begin': 71, 'end': 80}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated cells', 'obj': 'cell_type', 'prob': 0.7239546775817871, 'span': {'begin': 71, 'end': 86}}], 'text': 'The CPM also updates information outside of the grid, such as sizes of simulated cells.', 'timestamp': 'Fri Dec 22 09:32:14 +0000 2023'}, {'annotations': [], 'text': 'We first consider three issues when designing our parallelisation scheme: adjacent sections with active simulation interfering with each other during updates o: simultaneously, and stale in: updated by another simulati A CPM simulation proceeds copying the positions state to a ne simulation running, we run t. boundary.', 'timestamp': 'Fri Dec 22 09:32:16 +0000 2023'}, {'annotations': [], 'text': 'Such race condi standard problem in paralle to be active at the same time.', 'timestamp': 'Fri Dec 22 09:32:17 +0000 2023'}, None, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated cells', 'obj': 'cell_type', 'prob': 0.6656293869018555, 'span': {'begin': 201, 'end': 216}}], 'text': 'ion being used by one paralle hoosing a random position on k of simulations racing to wri run the risk of information ge m, with a straight- The CPM also tracks information involving more than one gri simulated cells.', 'timestamp': 'Fri Dec 22 09:32:19 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'single cell', 'obj': 'cell_type', 'prob': 0.566253125667572, 'span': {'begin': 62, 'end': 73}}], 'text': 'We face the same race condition problem here, updates touch a single cell (Figure 2B, left column).', 'timestamp': 'Fri Dec 22 09:32:20 +0000 2023'}, {'annotations': [], 'text': 'small active subsections, making such concurrent up scheme is proposed, that rejects all but the first concurrent update attempt on a varia middle column).', 'timestamp': 'Fri Dec 22 09:32:21 +0000 2023'}, {'annotations': [], 'text': 'We instead opt to use atomic operations (Figure 2B, right column), which equa!', 'timestamp': 'Fri Dec 22 09:32:22 +0000 2023'}, {'annotations': [], 'text': 'prevents update information from getting lost, as the hardware will eventually integrate all updates designed an additional scheme that sti CPMs, global information travels instantaneously: as only one grid position is w hout incorporating information of the overwritten update.', 'timestamp': 'Fri Dec 22 09:32:23 +0000 2023'}, {'annotations': [], 'text': 'This g updates (Figure 2A).', 'timestamp': 'Fri Dec 22 09:32:24 +0000 2023'}, {'annotations': [], 'text': 'Regions cycle perio ates more likely.', 'timestamp': 'Fri Dec 22 09:32:25 +0000 2023'}, {'annotations': [], 'text': 'In Tapia et a ; non-grid information such as cell sizes being updated simulation after said information got the simulation grid, and stochastically ighbour.', 'timestamp': 'Fri Dec 22 09:32:26 +0000 2023'}, {'annotations': [], 'text': 'If adjacent subsections of the grid each have an active parallel e to the same location on the sections ting lost, as one section overwri ion: never allow two adjacent sections rn of sections is only ever active; active location, such as the current sizes as it is possible that multiple paral rallelisation requires a high number 1.18 a per-cell locking dated at a time, date attempt has access to the results of all previous updates.', 'timestamp': 'Fri Dec 22 09:32:28 +0000 2023'}, {'annotations': [], 'text': 'This instantaneous information access is relaxed in our proposed scheme: when information such as cell size is accessed, out-of-date due to still in-flight updates to the accessed information.', 'timestamp': 'Fri Dec 22 09:32:29 +0000 2023'}, {'annotations': [], 'text': 'The cell-loc!', 'timestamp': 'Fri Dec 22 09:32:30 +0000 2023'}, {'annotations': [], 'text': 't al.', 'timestamp': 'Fri Dec 22 09:32:31 +0000 2023'}, {'annotations': [], 'text': 'prevents this from happening, by outright rejecting any updates that cou information; we explored a less strict scheme, reasoning that the stochastic CPM copy attempts erate out-of-date information.', 'timestamp': 'Fri Dec 22 09:32:32 +0000 2023'}, {'annotations': [], 'text': 'For CPM extensions that do not tolerate their global variables ightly out sync with each other (e.g.', 'timestamp': 'Fri Dec 22 09:32:33 +0000 2023'}, {'annotations': [], 'text': 'variables used in center-of-mass centroid tracking) we 1 will keep variables in syne with each other, and sits on top of our d have used out- atomics based scheme (Figure 2C, Methods).', 'timestamp': 'Fri Dec 22 09:32:34 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.42912566661834717, 'span': {'begin': 200, 'end': 204}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.6695387363433838, 'span': {'begin': 257, 'end': 261}}], 'text': 'Otherwise, due to atomic updating, our scheme still gives strong guarantees on tracked data being correct  for example, cell sizes will be tracked correctly  and we cache local da F our scheme on rea GPUs, memory access almost always prevents easy use of a GPUs caching hierarchy.', 'timestamp': 'Fri Dec 22 09:32:35 +0000 2023'}, {'annotations': [], 'text': '! To maximize performan scheme that carefully operates in loc hardware requires consideration o: y allows us to explore the effects of allowing the use of slightly outdated information.', 'timestamp': 'Fri Dec 22 09:32:36 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPMs', 'obj': 'DNA', 'prob': 0.4738963544368744, 'span': {'begin': 0, 'end': 4}}], 'text': 'CPMs random access pattern ce, we designed a data transfer f several technical details.', 'timestamp': 'Fri Dec 22 09:32:38 +0000 2023'}, {'annotations': [], 'text': 'k-step with checkerboard work splitting (Figure 2D, Methods): a needed by the regions that are becoming active, an tions to fully utilise overlap memory transfer and GPU hardware.', 'timestamp': 'Fri Dec 22 09:32:39 +0000 2023'}, {'annotations': [], 'text': 'To increase simulation volume, and gain further speed ups, we also prototyped a multi-GPU scheme (Figure 2E, Methods).', 'timestamp': 'Fri Dec 22 09:32:40 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.477714866399765, 'span': {'begin': 32, 'end': 35}}], 'text': 'In short, our proposed parallel CPM sc!', 'timestamp': 'Fri Dec 22 09:32:41 +0000 2023'}, {'annotations': [{'id': ['mesh:D006418'], 'is_neural_normalized': False, 'mention': 'heme', 'obj': 'drug', 'prob': 0.6413710117340088, 'span': {'begin': 0, 'end': 4}}], 'text': 'heme is tailored to make opt code.', 'timestamp': 'Fri Dec 22 09:32:43 +0000 2023'}, {'annotations': [], 'text': 'It still gives strong guarantees on simulation sta imal use of GPUs: hardware specialized e consistency  for example, a global variable tracking a cells size will always equal the number of pixels on the grid occupied by that cell.', 'timestamp': 'Fri Dec 22 09:32:44 +0000 2023'}, {'annotations': [], 'text': 'Yet, we do all flexibility and further maximise performance gains.', 'timestamp': 'Fri Dec 22 09:32:46 +0000 2023'}, {'annotations': [], 'text': 'ow the use of slightly outdated values of such global variables in copy attempts to gain B race condition mutex locking atomic updates (Tapia et al.)', 'timestamp': 'Fri Dec 22 09:32:47 +0000 2023'}, {'annotations': [], 'text': '(our method) (lel=12) (lla i= 12 |} | ASP$$ $l new = 13)- - - [le] = 13 &)- - - {le] = +18) nO OO Oo Hii --(=tie---(41=19 reese ew.', 'timestamp': 'Fri Dec 22 09:32:48 +0000 2023'}, {'annotations': [], 'text': '= 13)- - - (lel := 13 )- - - { ?', 'timestamp': 'Fri Dec 22 09:32:49 +0000 2023'}, {'annotations': [], 'text': '| {l*],.4 = 13)- - -(lo = 13 w)- - - | c I*]s:= 4 I#lnou = 5  Xu Yous *= (2-8, 3.3}  Xen?', 'timestamp': 'Fri Dec 22 09:32:51 +0000 2023'}, {'annotations': [{'id': ['CHEBI:74879'], 'is_neural_normalized': True, 'mention': 'Yaun', 'obj': 'drug', 'prob': 0.826878547668457, 'span': {'begin': 26, 'end': 30}}, {'id': ['mesh:D003520'], 'is_neural_normalized': True, 'mention': 'Xeon 2', 'obj': 'drug', 'prob': 0.6321195363998413, 'span': {'begin': 51, 'end': 57}}], 'text': 'Voow = (3; 3} (Kumn= 11) (Yaun= 13} (lel= 4) 4 ee [Xeon 2= +4!', 'timestamp': 'Fri Dec 22 09:32:53 +0000 2023'}, {'annotations': [{'id': ['CHEBI:74883'], 'is_neural_normalized': True, 'mention': 'Yeu', 'obj': 'drug', 'prob': 0.9442379474639893, 'span': {'begin': 0, 'end': 3}}, {'id': ['CHEBI:30503'], 'is_neural_normalized': False, 'mention': 'Be', 'obj': 'drug', 'prob': nan, 'span': {'begin': 21, 'end': 23}}, {'id': ['CHEBI:49957'], 'is_neural_normalized': True, 'mention': 'Xon', 'obj': 'drug', 'prob': 0.9343575835227966, 'span': {'begin': 25, 'end': 28}}, {'id': ['CHEBI:27698'], 'is_neural_normalized': True, 'mention': 'Van', 'obj': 'drug', 'prob': 0.9993822574615479, 'span': {'begin': 35, 'end': 38}}, {'id': ['CHEBI:51468'], 'is_neural_normalized': True, 'mention': 'Iel', 'obj': 'drug', 'prob': 0.9977790117263794, 'span': {'begin': 45, 'end': 48}}, {'id': ['CHEBI:49956'], 'is_neural_normalized': True, 'mention': 'Xeun', 'obj': 'drug', 'prob': 0.9991557598114014, 'span': {'begin': 59, 'end': 63}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5404716730117798, 'span': {'begin': 100, 'end': 103}}], 'text': 'Yeu = +2 | lel= +1 | Be] Xon= 15) (Van= 15) (Iel=5) [new = Xeun/@l = 3; Yoow = Yeun/l#] = 3) E_ per GPU region border sync ~-7--- global simulation barrier SS caput sizessyne pu |e] = 23 je| = 29 |e| = 23 |e| = 29 [Je|:= +4 Jel: +14=he!', 'timestamp': 'Fri Dec 22 09:32:55 +0000 2023'}, {'annotations': [], 'text': '= +0 |el:=-5 | |e] = 27 |e|=25 |e] = 27 je|=25 GPUs.', 'timestamp': 'Fri Dec 22 09:32:56 +0000 2023'}, {'annotations': [], 'text': '(A) We divide simulation space into a subgrid, with alternating regions being active in a checkerboard pattern, preventing update conflicts on the grid.', 'timestamp': 'Fri Dec 22 09:32:57 +0000 2023'}, {'annotations': [], 'text': '(B) We implement two strategies to prevent global variable update conflicts (e.g.', 'timestamp': 'Fri Dec 22 09:32:58 +0000 2023'}, {'annotations': [], 'text': 'cell size tracking): a mutex locking strategy (Tapia et al.', 'timestamp': 'Fri Dec 22 09:32:59 +0000 2023'}, {'annotations': [], 'text': \"'8), and a novel atomic operation strategy.\", 'timestamp': 'Fri Dec 22 09:33:01 +0000 2023'}, {'annotations': [], 'text': '(C) Combining interleaved storage and atomic operations on adjacent data allows variables to be kept in sync with each other.', 'timestamp': 'Fri Dec 22 09:33:02 +0000 2023'}, {'annotations': [], 'text': '(D) A four-step execution scheme, where active checkerboard areas are transferred to fast shared memory, accelerates copy attempts.', 'timestamp': 'Fri Dec 22 09:33:03 +0000 2023'}, {'annotations': [], 'text': 'Thread barriers synchronize active threads after transfers.', 'timestamp': 'Fri Dec 22 09:33:04 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6252509355545044, 'span': {'begin': 12, 'end': 15}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.505109429359436, 'span': {'begin': 92, 'end': 96}}], 'text': '(E) A multi-GPU execution scheme, where regions of the simulation are assigned to different GPUs.', 'timestamp': 'Fri Dec 22 09:33:05 +0000 2023'}, {'annotations': [], 'text': 'Borders are synchronized between relevant adjacent regions, and global variables are synchronized through delta tracking.', 'timestamp': 'Fri Dec 22 09:33:06 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'thread', 'obj': 'DNA', 'prob': 0.4236859083175659, 'span': {'begin': 38, 'end': 44}}], 'text': 'A T, : time in region time divergence thread barriers mm : time for one copy attempt - T, 5 checkerboard size thread size e ?', 'timestamp': 'Fri Dec 22 09:33:07 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lh __ 0123 45', 'obj': 'DNA', 'prob': 0.4811079800128937, 'span': {'begin': 332, 'end': 345}}], 'text': 'ee : : <@ e |e eee  eco  }  1|0@ : eee ao @ : : ee T, T. : : copy attempts before checherboard switch single update parallel update time since last copy attempt ee@ r ) 4 copy attempts per switch, 1 copy attempt per switch, number of copy attempts 4x4 area/thread, 8x8 area/thread, 2x2 area/checkerboard 4x4 area/checkerboard | | | lh __ 0123 45 0123 45 0123 4 5 copy attempts per location 0 1 2 0 1 2 0 1 2 waiting times between copy attempts (MCS) from racing ahead of others too much, thread barriers are imposed after each update.', 'timestamp': 'Fri Dec 22 09:33:08 +0000 2023'}, {'annotations': [], 'text': '(B) Checkerboard size and switching frequencies determine the distribution of waiting times between consecutive updates of the same locations.', 'timestamp': 'Fri Dec 22 09:33:09 +0000 2023'}, {'annotations': [], 'text': '(C) Small checkerboard areas or infrequent switches can alter the waiting time distribution substantially compared to the single-update scheme.', 'timestamp': 'Fri Dec 22 09:33:10 +0000 2023'}, {'annotations': [], 'text': 'Update frequency histograms were generated by simulating one MCS (256 updates) on a 16x16 grid.', 'timestamp': 'Fri Dec 22 09:33:11 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'MCSs', 'obj': 'cell_type', 'prob': 0.5453883409500122, 'span': {'begin': 76, 'end': 80}}], 'text': 'Waiting time histograms were simulated on the same grid using 100 simulated MCSs.', 'timestamp': 'Fri Dec 22 09:33:12 +0000 2023'}, {'annotations': [], 'text': 'cell boundary type: 60% 1 @- dark-light % of cell boundaries simulation based on Graner et al.!', 'timestamp': 'Fri Dec 22 09:33:13 +0000 2023'}, {'annotations': [], 'text': 'is implemented with two strategies to prevent race conditions for cell size variables.', 'timestamp': 'Fri Dec 22 09:33:14 +0000 2023'}, {'annotations': [], 'text': 'The mutex locking strategy proposed by Tapia et al.', 'timestamp': 'Fri Dec 22 09:33:15 +0000 2023'}, {'annotations': [], 'text': '! leads to severe starvation in our tests.', 'timestamp': 'Fri Dec 22 09:33:16 +0000 2023'}, {'annotations': [], 'text': 'Using an atomic update approach recovers the desired sorting behavior, closely matching sorting speed of the original CPM.', 'timestamp': 'Fri Dec 22 09:33:17 +0000 2023'}, {'annotations': [], 'text': 'The simulation was run with a 4x4 area per thread and 2x2 checkerboards, but other sizes were tested and did not make a significant difference for this simulation.', 'timestamp': 'Fri Dec 22 09:33:19 +0000 2023'}, {'annotations': [], 'text': 'Sorting trends of different CPM implementations are averaged over 5 simulations.', 'timestamp': 'Fri Dec 22 09:33:20 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPMs', 'obj': 'DNA', 'prob': 0.4661286175251007, 'span': {'begin': 86, 'end': 90}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPMs', 'obj': 'DNA', 'prob': 0.4117409586906433, 'span': {'begin': 101, 'end': 105}}], 'text': 'Simulation parameters can be found Aligning the lattice update statistics of parallel CPMs to serial CPMs Our novel parallel execution scheme guarantees simulation data consistency, but drastically changes the way the simulation lattice is updated.', 'timestamp': 'Fri Dec 22 09:33:21 +0000 2023'}, {'annotations': [], 'text': 'We investigated statistical properties of copy attempts on the simulation lattice for different lattice positions; specifically, we analysed the relative ordering of copy attempts, frequency of copy attempts, and waiting times between copy attempts.', 'timestamp': 'Fri Dec 22 09:33:22 +0000 2023'}, {'annotations': [], 'text': 'This allowed us to quantify the divergence of parallel lattice update kinetics from the serial CPM.', 'timestamp': 'Fri Dec 22 09:33:23 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cells', 'obj': 'cell_type', 'prob': 0.3679133951663971, 'span': {'begin': 79, 'end': 84}}], 'text': 'Such divergence could lead to changed simulation behavior in complex models of cells and tissues.', 'timestamp': 'Fri Dec 22 09:33:24 +0000 2023'}, {'annotations': [], 'text': 'Without guardrails, performing independent CPM updates in parallel would allow different parts of a simulated tissue to have been alive for different lengths of simulated time (Figure 3A).', 'timestamp': 'Fri Dec 22 09:33:25 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.4616106152534485, 'span': {'begin': 7, 'end': 11}}], 'text': 'Modern GPUs use thread barriers to synchronize parallel threads without strong computational performance implications.', 'timestamp': 'Fri Dec 22 09:33:26 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'checkerboard region', 'obj': 'DNA', 'prob': 0.7476058006286621, 'span': {'begin': 85, 'end': 104}}], 'text': 'We set thread barriers after ever single copy attempt, so for every currently active checkerboard region there is only a maximum time divergence of one copy attempt between active regions.', 'timestamp': 'Fri Dec 22 09:33:28 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'active region', 'obj': 'DNA', 'prob': 0.49735385179519653, 'span': {'begin': 97, 'end': 110}}], 'text': 'Our parallel execution scheme allows us to freely choose the size of subregions and frequency of active region switching (Figure 3B), without impacting data consistency guarantees.', 'timestamp': 'Fri Dec 22 09:33:29 +0000 2023'}, {'annotations': [], 'text': 'We measure frequency and waiting times between copy attempts per position on the lattice in the standard, serial implementation of the CPM (Figure 3C, left column).', 'timestamp': 'Fri Dec 22 09:33:30 +0000 2023'}, {'annotations': [], 'text': 'For our initial parameter choice, we found that underlying update statistics are drastically changed (Figure 3C, middle column): waiting times between copy attempts became much more predictable due to the much more even distribution of copy attempts compared to the serial CPM.', 'timestamp': 'Fri Dec 22 09:33:31 +0000 2023'}, {'annotations': [], 'text': 'Using large parallel subsections with frequent active region switching allows non-parallel update statistics to be closely recapitulated in our parallel scheme (Figure 3C, right column).', 'timestamp': 'Fri Dec 22 09:33:32 +0000 2023'}, {'annotations': [], 'text': 'We thus find that parallelisation parameters have major influence on copy attempt statistics on the lattice.', 'timestamp': 'Fri Dec 22 09:33:33 +0000 2023'}, {'annotations': [], 'text': 'By switching active regions of the checkerboard frequently, and avoiding too small areas per thread, we can obtain statistics that closely resemble the serial CPM.', 'timestamp': 'Fri Dec 22 09:33:34 +0000 2023'}, {'annotations': [], 'text': 'Emergent cell and tissue behavior is maintained in a parallel CPM Our goal is to build parallel CPMs that are much faster than previous implementations but maintain cell and tissue kinetics consistent with existing serial CPMs.', 'timestamp': 'Fri Dec 22 09:33:35 +0000 2023'}, {'annotations': [], 'text': 'We hypothesized that aligning the waiting times between update attempts of our parallel CPM to the serial version should suffice to achieve such consis- tency.', 'timestamp': 'Fri Dec 22 09:33:36 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.46646010875701904, 'span': {'begin': 175, 'end': 178}}], 'text': 'To investigate this, we ran several simulation scenarios with different CPM variants, measuring emergent behaviors on the cell and tissue levels and comparing these to serial CPM versions.', 'timestamp': 'Fri Dec 22 09:33:38 +0000 2023'}, {'annotations': [], 'text': 'We first investigated the classic cell sorting simulation:!', 'timestamp': 'Fri Dec 22 09:33:39 +0000 2023'}, {'annotations': [], 'text': \"two mixed populations of cells with differential  A single update 4 copy attempts per switch, = 0.3 03 2 202 5 0.2 5 50.1 & 0.1 & 0.0 8 0.0 8 2000 6000 10000 2000 6000 10000 time (MCS) time (MCS) B  low polarisation strength 2 ee qe 4S ' high polarisation strength ' .\", 'timestamp': 'Fri Dec 22 09:33:41 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:6355'], 'is_neural_normalized': True, 'mention': 'Cc 8', 'obj': 'gene', 'prob': 0.9875783324241638, 'span': {'begin': 77, 'end': 81}}], 'text': ') f 4 i 3 08 3 08 3  0.4  0.4  8 0.0 8 0.0 8 200 400 600 800 200 400 600 800 Cc 8 copy attempts per switch, 0.25 0.50 0.75 1.00 0.25 0.50 0.75 1.00 1 copy attempt per switch, 200 400 600 800 1 copy attempt per switch, 0.25 0.50 0.75 1.00 are similar.', 'timestamp': 'Fri Dec 22 09:33:43 +0000 2023'}, {'annotations': [], 'text': '(A) Cells in a circular chemokine gradient simulation move highly regular in a serial CPM (low per-pixel entropy, shaded: 1SD); parallel simulations diverge slowly from reference behavior when copy attempt statistics are preserved (low per-pixel cross-entropy).', 'timestamp': 'Fri Dec 22 09:33:44 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:12', '71', '72', '9457', '345651', '11332', '440915', '641455'], 'is_neural_normalized': True, 'mention': 'Act', 'obj': 'gene', 'prob': 0.508391261100769, 'span': {'begin': 11, 'end': 14}}], 'text': '(B) In the Act model of intra-cell polarization, stronger polarization strength leads to higher directional autocorrelation (measured for 750 MCSs at 50 MCS intervals, averaged over 50 simulations, shading: 4 emerges when update waiting time distributions are preserved.', 'timestamp': 'Fri Dec 22 09:33:46 +0000 2023'}, {'annotations': [], 'text': '(C) t1SD); directional persistence only Phase transition simulation in 3D where synchronous motility emerges at a density around 0.5.', 'timestamp': 'Fri Dec 22 09:33:47 +0000 2023'}, {'annotations': [], 'text': 'Cell order indices after 2000 MCSs are shown (averaged over 5 simulation runs, shading: 4 ransition only occurs when update waiting time statistics are preserved.', 'timestamp': 'Fri Dec 22 09:33:48 +0000 2023'}, {'annotations': [], 'text': 'Simulation parameters can be found in Table 1 and Table 2.  adhesion (i.e., cells of the same type have stronger adhesion to each other than cells of different types) sort themselves in two homogeneous regions over time (Figure 4, grey lines).', 'timestamp': 'Fri Dec 22 09:33:49 +0000 2023'}, {'annotations': [], 'text': 'The mixing kinetics can be measured by tracking homogeneous and heterogene: based parallelisation approach proposed by Tapia et a leads to a lack of sorting behavior; the simulation concurrent copy attempt is allowed per cell un scheme allows expected sorting behavior to re remained consistent for all parallelisation paramete: ous cell neighbours over time.!', 'timestamp': 'Fri Dec 22 09:33:51 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:63915'], 'is_neural_normalized': True, 'mention': 'mutex 1', 'obj': 'gene', 'prob': 0.8991247415542603, 'span': {'begin': 16, 'end': 23}}], 'text': 'Strikingly, the mutex 1.18  while still being able to run a simulation  encounters so-called starvation, because only one er this scheme (Figure 4, red lines).', 'timestamp': 'Fri Dec 22 09:33:52 +0000 2023'}, {'annotations': [], 'text': 'Our less restrictive -emerge (Figure 4, black lines).', 'timestamp': 'Fri Dec 22 09:33:53 +0000 2023'}, {'annotations': [], 'text': 'The sor rs we examined (Supplementary Figure $1).', 'timestamp': 'Fri Dec 22 09:33:54 +0000 2023'}, {'annotations': [], 'text': 'We next examined three motility model variants of the CPM, where additional biases are placed on o obtain cells that move persis function of random membrane fluctuations.', 'timestamp': 'Fri Dec 22 09:33:55 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Bec', 'obj': 'cell_type', 'prob': 0.4774242639541626, 'span': {'begin': 0, 'end': 3}}], 'text': 'Bec: need to roughly maintain their target size, copy sections for movement to occur.', 'timestamp': 'Fri Dec 22 09:33:57 +0000 2023'}, {'annotations': [], 'text': 'We therefore s by changes to the waiting times between copy at ikely to succeed when aligned with the underlyi bias is fixed in this model  the field does not change simulations with the same initial conditions closely contains a cell in multiple simulations, where low between simulations, and low cross-entropy indica: with reference serial runs.', 'timestamp': 'Fri Dec 22 09:33:58 +0000 2023'}, {'annotations': [], 'text': 'In simulations with increased quickly (Figure 5A, middle column); w. rises much slower (Figure 5A, right column).', 'timestamp': 'Fri Dec 22 09:33:59 +0000 2023'}, {'annotations': [], 'text': 'Next, we analysed a model that simulates localize border of a cell makes further copy attempts involving those times between copy attempts are very important in this mo in a small neighbourhood successfully expand at the same time; this is less y model we consider adds a chemotactic vector field to the CPM; copy attempts are more ng vector field (see Methods).', 'timestamp': 'Fri Dec 22 09:34:01 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'motile cells', 'obj': 'cell_type', 'prob': 0.9053431153297424, 'span': {'begin': 169, 'end': 181}}], 'text': 'Because the directional his alignment using pixel-wise (cross-)entropy: we measure the probability that a pixe ently; in the standard CPM movement only occurs as a ause motile cells have a leading and a retrac attempts need to sufficiently alternate between these cel peculated that movement models might be more affecte over time  cells move in a highly regular fashion, an align even at a pixel level (Figure 5A, left column).', 'timestamp': 'Fri Dec 22 09:34:02 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cross-entropy cell', 'obj': 'DNA', 'prob': 0.4990772604942322, 'span': {'begin': 174, 'end': 192}}], 'text': 'pixel probabilities indicated high agreement es a high level of agreement of parallel simulations ged lattice update statistics, cross-entropy pdate statistics are preserved cross-entropy cell polarization, where recent pixel updates at the pixels more likely to succeed.', 'timestamp': 'Fri Dec 22 09:34:03 +0000 2023'}, {'annotations': [], 'text': 'The waiting el, as cells start to move when many pixels likely when copy attempts are spaced out evenly.', 'timestamp': 'Fri Dec 22 09:34:04 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'parame cell', 'obj': 'cell_line', 'prob': 0.7146728038787842, 'span': {'begin': 24, 'end': 35}}], 'text': 'A polarization strength parame cell; varying the polarization strength interpolates motility be motility modes (Figure 5B, lef waiting time statistics this full range of behaviors they re-emerge when update sta Finally, we examined a simulatio with this vector, and the vector motility model involves informat ion tracked on the to the simulation grid, placing additional demands on a para!', 'timestamp': 'Fri Dec 22 09:34:05 +0000 2023'}, {'annotations': [], 'text': 'density cell collective exhibit Brownian movement, as density is increased;??', 'timestamp': 'Fri Dec 22 09:34:06 +0000 2023'}, {'annotations': [], 'text': 'this characteristic phase left column).', 'timestamp': 'Fri Dec 22 09:34:08 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'preferred', 'obj': 'DNA', 'prob': 0.5637134313583374, 'span': {'begin': 186, 'end': 195}}], 'text': 'In a parallel CPM this phase transition disa, (Figure 5C, middle column), but re-emerges when update Taken together, these results demonstrate that a strict updat n of a tightly packe a preferred movement direction vector; copy attempts for a ce s how much polarization bias is added to a tween Brownian and directionally persistent el simulations with changed update frequency and oes not appear (Figure 5B, middle column), but (Figure 5B, right column).', 'timestamp': 'Fri Dec 22 09:34:09 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'immune cells', 'obj': 'cell_type', 'prob': 0.928694486618042, 'span': {'begin': 11, 'end': 23}}], 'text': 'lective of immune cells, where each cell has 1 are more likely to succeed when they align ated to align with recent movement of the cell.', 'timestamp': 'Fri Dec 22 09:34:10 +0000 2023'}, {'annotations': [], 'text': 'This attice, and additional information tracked external el CPM.', 'timestamp': 'Fri Dec 22 09:34:11 +0000 2023'}, {'annotations': [], 'text': 'In a similar simulation model, low transitioning sharply to ordered collective movement ransition is reproduced in the CPM (Figure 5C, pears when update statistics are changed statistics are preserved (Figure 5C, right e scheme preventing parallel updates of the same cell is not necessary for maintaining valid CPM behavior and can even lead to severe starvation where simulations do not progress anymore.', 'timestamp': 'Fri Dec 22 09:34:12 +0000 2023'}, None, {'annotations': [], 'text': 'GPU-based CPM parallelisation accelerates models by several orders of mag- Having confirmed that our parallel CPM indeed pro duced correct model behavior in a range of examples, we next investigated the performance improvements resulting from our implementation choices when compared to a serial CPM.', 'timestamp': 'Fri Dec 22 09:34:15 +0000 2023'}, {'annotations': [], 'text': '% 1000 2 s 8 g 8 & 19547x13805x Be 3 3 10 3 2 3 < 1 g 24 a & 2 2 eg ol  kernel specialisation ( ae ) 128 5127 20487 8192!', 'timestamp': 'Fri Dec 22 09:34:16 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'pre', 'obj': 'DNA', 'prob': 0.4685356020927429, 'span': {'begin': 372, 'end': 375}}], 'text': 'implementation shared memory @ @ e@ D E area/thread: @ 4x4 A8x8 mi 16x16 >  03 +0 3 10000 3 9786x 2 S 1000 g = 02 = 0.9 > & 400   g 8 & 2 8 308 3 oo 10   2902x ZO 8 8  1 s G 07 642 256 10242 implementation 1 4 16 64 1 4 16 64 0.0 0.1 0.20.3 dimension updates per switch updates per switch simulation time Performance benefits are obtained by a shared memory scheme, and a pre-compiled kernel specialization scheme, based on a sorting simulation extended to densely fill simulation space (Figure 4, all sorting benchmarks are average of 5 simulations of 100 MCSs, 100 MCS burn-in).', 'timestamp': 'Fri Dec 22 09:34:18 +0000 2023'}, {'annotations': [], 'text': '(B) Distributing the lattice over 4 GPUs speeds up the simulation further.', 'timestamp': 'Fri Dec 22 09:34:19 +0000 2023'}, {'annotations': [], 'text': '(C) Performance of sorting simulations are significantly higher relative to two serial implementations; relative speedup shown for 2048x2048 simulations.', 'timestamp': 'Fri Dec 22 09:34:20 +0000 2023'}, {'annotations': [], 'text': '(D) Performance of 3D cell collective simulations is also improved (Figure 5C, average of 2 simulations of 100 MCSs, 100 MCS burn-in); relative speedup shown for 256 simulations.', 'timestamp': 'Fri Dec 22 09:34:21 +0000 2023'}, {'annotations': [], 'text': '(E) Trade-offs between speed and accuracy for chemokine field simulations (Figure 5A, accuracy: 1 - cross-entropy of parallel ensemble relative to serial ensemble, 250 simulations per ensemble).', 'timestamp': 'Fri Dec 22 09:34:22 +0000 2023'}, {'annotations': [], 'text': 'CPM parameter settings can be found in Table 1 and Table 2.', 'timestamp': 'Fri Dec 22 09:34:24 +0000 2023'}, {'annotations': [], 'text': 'We first examined the performance gains resulting from our use of a shared memory caching scheme (Figure 2D, Methods) using a 2D sorting simulation benchmark (Figure 4).', 'timestamp': 'Fri Dec 22 09:34:25 +0000 2023'}, {'annotations': [], 'text': 'On its own, shared memory caching achieved only minimal speedup (Figure 6A).', 'timestamp': 'Fri Dec 22 09:34:26 +0000 2023'}, {'annotations': [], 'text': 'However, for further performance improvement, we pre-generated different versions of our code based on which CPM extensions were in use (Methods).', 'timestamp': 'Fri Dec 22 09:34:27 +0000 2023'}, {'annotations': [], 'text': 'Combining these specialized kernels with the caching scheme achieved a significant 10-fold speedup, more than expected from either optimization alone (Figure 6A).', 'timestamp': 'Fri Dec 22 09:34:28 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.47674185037612915, 'span': {'begin': 18, 'end': 21}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.427679181098938, 'span': {'begin': 146, 'end': 150}}], 'text': 'Testing our multi-GPU scheme (Figure 2E, Methods) in the same scenario achieved a further 2.3x speedup by distributing the simulation across four GPUs (Figure 6B).', 'timestamp': 'Fri Dec 22 09:34:29 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.8114858865737915, 'span': {'begin': 32, 'end': 35}}], 'text': 'Hence, careful consideration of GPU-specific implementation choices resulted in significant performance benefits.', 'timestamp': 'Fri Dec 22 09:34:30 +0000 2023'}, {'annotations': [], 'text': 'We next compared our parallel CPM (on a single GPU) to our own optimized CPU implementation (Methods), and to the state-of-the-art software Morpheus.', 'timestamp': 'Fri Dec 22 09:34:32 +0000 2023'}, {'annotations': [], 'text': '! Using the same 2D cell sorting bench- mark, we observed a {3,500-fold speedup compared to our own CPU implementation, and a j25,000-fold improvement relative to Morpheus (Figure 6C).', 'timestamp': 'Fri Dec 22 09:34:33 +0000 2023'}, {'annotations': [], 'text': 'On the 3D cell collective simulation (Figure 5C), a {,600-fold speedup was achieved relative to serial CPMs (Figure 6D).', 'timestamp': 'Fri Dec 22 09:34:34 +0000 2023'}, {'annotations': [], 'text': 'It is important to note that our benchmark hardware did not permit use of shared memory caching for the 3D model; future hardware with more shared memory may improve 3D performance.', 'timestamp': 'Fri Dec 22 09:34:35 +0000 2023'}, {'annotations': [], 'text': 'As shown in Figure 3 and Figure 5, checkerboard size and switching frequency influences model behav- ior: larger divisions and more frequent switching leads to behavior concordant with serial CPMs.', 'timestamp': 'Fri Dec 22 09:34:36 +0000 2023'}, {'annotations': [], 'text': 'These update patterns also interact with caching, further impacting performance.', 'timestamp': 'Fri Dec 22 09:34:37 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'parallelisation parameters', 'obj': 'DNA', 'prob': 0.479205846786499, 'span': {'begin': 49, 'end': 75}}], 'text': 'As the sorting simulation has low sensitivity to parallelisation parameters (Supplementary Figure $1), we instead benchmarked our 2D chemotaxis simulation (Figure 5A).', 'timestamp': 'Fri Dec 22 09:34:38 +0000 2023'}, {'annotations': [], 'text': 'We found that larger subdivisions indeed slowed simulations down considerably, whereas frequent switching led to a less marked slowdown (Figure 6E).', 'timestamp': 'Fri Dec 22 09:34:39 +0000 2023'}, {'annotations': [], 'text': 'These results suggested a trade-off between performance and model accuracy.', 'timestamp': 'Fri Dec 22 09:34:41 +0000 2023'}, {'annotations': [], 'text': 'Hence, our parallel CPM represents a significant speed improvement over even well optimized serial CPMs  such as Morpheus.', 'timestamp': 'Fri Dec 22 09:34:42 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.7166396975517273, 'span': {'begin': 57, 'end': 60}}], 'text': 'This improvement was to a substantial extent a result of GPU-specific optimizations and led to an overall speedup of several orders of magnitude.', 'timestamp': 'Fri Dec 22 09:34:43 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.7044103145599365, 'span': {'begin': 33, 'end': 39}}], 'text': 'Building an image-based model of T cell surveillance of the liver Our parallel CPM was motivated by the need for tissue simulations at realistic scale with realistic cell morphology.', 'timestamp': 'Fri Dec 22 09:34:44 +0000 2023'}, {'annotations': [], 'text': 'While our approach increases engineering complexity of the model, it can also simplify model building, as it enables direct use of imaging data to initialize the simulation environment.', 'timestamp': 'Fri Dec 22 09:34:45 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9112301468849182, 'span': {'begin': 112, 'end': 119}}], 'text': 'To determine the feasibility of directly using real-world imaging data in our system, we constructed a model of T cells moving in the sinusoidal regions of the liver.', 'timestamp': 'Fri Dec 22 09:34:46 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Tissue resident memory T cells', 'obj': 'cell_type', 'prob': 0.734682023525238, 'span': {'begin': 0, 'end': 30}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trms', 'obj': 'cell_line', 'prob': 0.7562732696533203, 'span': {'begin': 32, 'end': 36}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trms', 'obj': 'cell_type', 'prob': 0.721453070640564, 'span': {'begin': 153, 'end': 157}}, {'id': ['mesh:D008288', 'mim:609148'], 'is_neural_normalized': False, 'mention': 'malaria', 'obj': 'disease', 'prob': 0.9999243021011353, 'span': {'begin': 204, 'end': 211}}], 'text': 'Tissue resident memory T cells (Trms) are confined to specific organs, where they sample their envi- ronment for signs of disease+  in particular, liver Trms are key in providing sterilising immunity for malaria.', 'timestamp': 'Fri Dec 22 09:34:47 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trms', 'obj': 'cell_type', 'prob': 0.662162184715271, 'span': {'begin': 2, 'end': 6}}, {'id': ['mesh:D008288', 'mim:609148'], 'is_neural_normalized': False, 'mention': 'malaria', 'obj': 'disease', 'prob': 0.9999676942825317, 'span': {'begin': 73, 'end': 80}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trms', 'obj': 'cell_type', 'prob': 0.7950738668441772, 'span': {'begin': 115, 'end': 119}}], 'text': '? Trms display a distinct motility pattern in the liver, and efficacy of malaria clearing depends on the number of Trms present.', 'timestamp': 'Fri Dec 22 09:34:48 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'liver sinusoids', 'obj': 'cell_type', 'prob': 0.518401026725769, 'span': {'begin': 48, 'end': 63}}, {'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trms', 'obj': 'gene', 'prob': 0.6249759197235107, 'span': {'begin': 84, 'end': 88}}, {'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trms', 'obj': 'gene', 'prob': 0.7536947727203369, 'span': {'begin': 190, 'end': 194}}], 'text': '? We constructed a 3D simulation environment of liver sinusoids  capil- laries that Trms reside in and migrate through  and combined it with a polarized cell motility model to represent the Trms (Figure 7B).', 'timestamp': 'Fri Dec 22 09:34:50 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trm', 'obj': 'cell_type', 'prob': 0.712886393070221, 'span': {'begin': 0, 'end': 3}}], 'text': 'Trm motility parameters were derived from a LN model constructed in parallel, and our parameter choice will be explained in more detail with the LN data.', 'timestamp': 'Fri Dec 22 09:34:51 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trms', 'obj': 'DNA', 'prob': 0.45767539739608765, 'span': {'begin': 125, 'end': 129}}], 'text': 'These parameter settings lead to motility modes also observed in intravital liver data: intermittent movement and pausing of Trms.', 'timestamp': 'Fri Dec 22 09:34:52 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9147524833679199, 'span': {'begin': 110, 'end': 117}}], 'text': 'The goal of our simulation was to predict how much of the liver volume would be explored by a given number of T cells within a set amount of time; this same question was addressed by Fernandez-Ruiz et al.', 'timestamp': 'Fri Dec 22 09:34:53 +0000 2023'}, {'annotations': [], 'text': 'using a more abstract mathematical model.', 'timestamp': 'Fri Dec 22 09:34:55 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trm population', 'obj': 'cell_line', 'prob': 0.6963381767272949, 'span': {'begin': 61, 'end': 75}}, {'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trm', 'obj': 'gene', 'prob': 0.9847688674926758, 'span': {'begin': 61, 'end': 64}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'exploring cells', 'obj': 'cell_type', 'prob': 0.8306978940963745, 'span': {'begin': 105, 'end': 120}}], 'text': '? As can be expected, we found that liver exploration by the Trm population increased with the number of exploring cells, and with the amount of exploration time (Figure 7C).', 'timestamp': 'Fri Dec 22 09:34:56 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Trm', 'obj': 'cell_line', 'prob': 0.8929936289787292, 'span': {'begin': 99, 'end': 102}}], 'text': 'Remarkably, the sinusoid volume explored by an individual cell increased dramatically as the total Trm population was raized from 1,000 to 10,000 cells (Figure 7D).', 'timestamp': 'Fri Dec 22 09:34:58 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'stationary cells', 'obj': 'cell_line', 'prob': 0.9268476366996765, 'span': {'begin': 204, 'end': 220}}], 'text': 'Further investigation revealed the source of this phase transition: cells in our simulations can be in an unpolarized state, in which they do not move persistently; collisions can trigger polarization of stationary cells.', 'timestamp': 'Fri Dec 22 09:34:59 +0000 2023'}, {'annotations': [], 'text': 'Since such collisions occur more frequently at higher densities, this causes a larger fraction of cells to be motile (Figure 7E).', 'timestamp': 'Fri Dec 22 09:35:00 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trms', 'obj': 'gene', 'prob': 0.954503059387207, 'span': {'begin': 9, 'end': 13}}, {'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trm', 'obj': 'gene', 'prob': 0.9888854026794434, 'span': {'begin': 52, 'end': 55}}, {'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trms', 'obj': 'gene', 'prob': 0.8798872828483582, 'span': {'begin': 163, 'end': 167}}], 'text': 'Thus, if Trms can help each other to move, a larger Trm population would screen the environment more efficiently than expected from the Notably, the mean speed of Trms peak at ~10.5 um/min (Figure 7F), matching average speeds in our lymph node simulations  but only at densities where the arrest coefficient is close to 0.', 'timestamp': 'Fri Dec 22 09:35:02 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9174125790596008, 'span': {'begin': 86, 'end': 93}}], 'text': 'In two-photon experiments, liver Trm speed is reported to be significantly lower than T cells in lymph nodes; reproduce this in our simulations with moderate cell density, which leads to both a realistic slowdown, and higher arrest coefficient.', 'timestamp': 'Fri Dec 22 09:35:03 +0000 2023'}, {'annotations': [], 'text': 'The simulated liver Trm speed and arrest coefficient closely matches previ- ously reported values in literature,??', 'timestamp': 'Fri Dec 22 09:35:04 +0000 2023'}, {'annotations': [], 'text': 'where they were used to quantify observed characteristic patrolling These liver experiments demonstrated that our system allowed us to combine 3D imaging data with a relatively realistic model of polarizing cell motility.', 'timestamp': 'Fri Dec 22 09:35:06 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:55687'], 'is_neural_normalized': True, 'mention': 'Trm', 'obj': 'gene', 'prob': 0.9828422665596008, 'span': {'begin': 69, 'end': 72}}], 'text': 'The resulting simulation produced an interesting phase transition of Trm screening efficiency.', 'timestamp': 'Fri Dec 22 09:35:07 +0000 2023'}, {'annotations': [], 'text': 'While we do not claim that such an effect is present in the real liver  more detailed tuning of the motility model to match the characteristics of real sinusoidal Trm cells would be out of scope for this simple proof of concept  this does illustrate how detailed models such as the CPM generate emergent behavior that is not an immediately obvious consequence of the model ingredients and can generate interesting biological hypotheses worthy of further exploration.', 'timestamp': 'Fri Dec 22 09:35:08 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cell collective', 'obj': 'cell_type', 'prob': 0.7148670554161072, 'span': {'begin': 2, 'end': 19}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9082027673721313, 'span': {'begin': 102, 'end': 109}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'presenting cells', 'obj': 'cell_type', 'prob': 0.7791904807090759, 'span': {'begin': 136, 'end': 152}}], 'text': 'A T cell collective moves robustly in a realistic, organ-scale model of the lymph In LNs, millions of T cells move in search of antigen-presenting cells,?', 'timestamp': 'Fri Dec 22 09:35:10 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cognate', 'obj': 'DNA', 'prob': 0.5446880459785461, 'span': {'begin': 18, 'end': 25}}], 'text': '! and finding the cognate antigen is a key early step in adaptive immune responses.*!', 'timestamp': 'Fri Dec 22 09:35:11 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.7237823009490967, 'span': {'begin': 0, 'end': 6}}], 'text': 'T cell motility in the LN is shaped by the high deformability of the cells?', 'timestamp': 'Fri Dec 22 09:35:12 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'fibroblastic reticular cell network', 'obj': 'cell_type', 'prob': 0.8136028051376343, 'span': {'begin': 63, 'end': 98}}], 'text': 'and their interactions with a complex stromal environment, the fibroblastic reticular cell network.', 'timestamp': 'Fri Dec 22 09:35:13 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'moving cells', 'obj': 'cell_type', 'prob': 0.8712626099586487, 'span': {'begin': 69, 'end': 81}}], 'text': '*3:4 Given the LNs unique structure as an organ consisting mainly of moving cells, it presents a challenging test case for any tissue simulation system.', 'timestamp': 'Fri Dec 22 09:35:15 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'LN T cell', 'obj': 'cell_type', 'prob': 0.8523593544960022, 'span': {'begin': 23, 'end': 32}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'LN', 'obj': 'cell_type', 'prob': nan, 'span': {'begin': 166, 'end': 168}}], 'text': 'Indeed, many models of LN T cell dynamics already exist, but they were often limited in either cell count and simulation volume, or in realism of cell morphology and LN environment.', 'timestamp': 'Fri Dec 22 09:35:16 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.8682790994644165, 'span': {'begin': 13, 'end': 19}}], 'text': 'For example, T cell interactions with a synthetic stromal structure have been investigated using 3D CPM simulations of thousands of cells, while collective motility  at a more realistic scale of hundreds of thousands of cells has been investigated with non-morphological agent-based models.', 'timestamp': 'Fri Dec 22 09:35:17 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.8839187622070312, 'span': {'begin': 137, 'end': 144}}], 'text': '*:3 We were therefore interested in our systems ability to simulate a realistically large collective of millions of realistically shaped T cells.', 'timestamp': 'Fri Dec 22 09:35:19 +0000 2023'}, {'annotations': [], 'text': 'As in the previous use case, we used microscopy data to initialize  and parameterize  our model.', 'timestamp': 'Fri Dec 22 09:35:21 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'LN', 'obj': 'cell_type', 'prob': nan, 'span': {'begin': 17, 'end': 19}}], 'text': 'We constructed a LN simulation using the work of Beltman et al.9 as a starting point.', 'timestamp': 'Fri Dec 22 09:35:22 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:12', '71', '72', '9457', '345651', '11332', '440915', '641455'], 'is_neural_normalized': True, 'mention': 'Act', 'obj': 'gene', 'prob': 0.9963853359222412, 'span': {'begin': 166, 'end': 169}}], 'text': 'To improve realism, we used an imaged LN section to initialize the stromal structure (Figure 8A), we used the Act-CPM extension for cell motility, and we fit the two Act motility parameters directly to intravital imaging data (Figure 8B-E).', 'timestamp': 'Fri Dec 22 09:35:23 +0000 2023'}, {'annotations': [], 'text': 'When fitting these parameters, we aimed to reproduce the intermittent stop-and-go motion observed in real data whilst also quantitatively approximate the real speed distri- bution.', 'timestamp': 'Fri Dec 22 09:35:24 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.8217685222625732, 'span': {'begin': 131, 'end': 138}}], 'text': 'Our system allowed fitting to be performed on realistically large simulations (256um diameter simulation volume, containing 50,135 T cells, 20 minutes simulated time).', 'timestamp': 'Fri Dec 22 09:35:25 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell collective', 'obj': 'cell_type', 'prob': 0.4304350018501282, 'span': {'begin': 138, 'end': 153}}], 'text': 'Crucially, we did not use periodic boundaries which, in dense simulations, can lead to artificial global synchronisation of motion in the cell collective.', 'timestamp': 'Fri Dec 22 09:35:26 +0000 2023'}, {'annotations': [], 'text': 'We used the KL divergence to compare simulated to real speed distributions (Figure 8B).', 'timestamp': 'Fri Dec 22 09:35:27 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.9070225954055786, 'span': {'begin': 22, 'end': 29}}], 'text': 'Increased adhesion of T cells to the stromal structure made the speed distribution more right-skewed (Figure 8C, top), which was more consistent with real data (Figure 8C bottom).', 'timestamp': 'Fri Dec 22 09:35:29 +0000 2023'}, {'annotations': [], 'text': 'For most examined parameter settings, we found the pronounced variations in cell speed along cell trajectories that are also observed in real data (Figure 8D).', 'timestamp': 'Fri Dec 22 09:35:30 +0000 2023'}, {'annotations': [], 'text': 'When we performed grid searches of motility parameters at two different stromal adhesion strengths (Figure 8E), both grid searches had broadly similar fit landscapes, and the overall lowest KL divergence values were very similar.', 'timestamp': 'Fri Dec 22 09:35:31 +0000 2023'}, {'annotations': [], 'text': 'Qualitatively, we found more right-skewed and bimodal speed distributions in high-adhesion than low-adhesion simulations for a subset of the motility parameters (Supplementary Figure $2).', 'timestamp': 'Fri Dec 22 09:35:32 +0000 2023'}, {'annotations': [], 'text': 'Based on these fitting results, we chose the best fit motility parameters of the high adhesion setting for our further simulations.', 'timestamp': 'Fri Dec 22 09:35:33 +0000 2023'}, {'annotations': [{'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.6771363615989685, 'span': {'begin': 44, 'end': 50}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.842300534248352, 'span': {'begin': 137, 'end': 144}}], 'text': 'We next investigated organ-scale collective T cell motility by scaling up our simulated volume to 1mm diameter, now containing 3,359,752 T cells (Figure 9A).', 'timestamp': 'Fri Dec 22 09:35:35 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cell zone', 'obj': 'cell_type', 'prob': 0.6303771138191223, 'span': {'begin': 117, 'end': 128}}], 'text': 'While this is not entirely realistic  a real LN has a more complex compartmental structure, while we simulate only a T cell zone  our main intention was to determine the spatial range of collective behavior that can now be investigated with our system.', 'timestamp': 'Fri Dec 22 09:35:36 +0000 2023'}, {'annotations': [], 'text': 'We simulated one virtual day in 1.9 days of wall clock time  where serial implementations were projected o have taken a full 1.4 years (Figure 1A).', 'timestamp': 'Fri Dec 22 09:35:37 +0000 2023'}, {'annotations': [], 'text': 'We used a mixing simulation of two differentially labelled populations as an easily observable proxy for collective search (Figure 9B).', 'timestamp': 'Fri Dec 22 09:35:38 +0000 2023'}, {'annotations': [], 'text': 'In the simulation, it takes about 1 hour to reach a state where 50% of the cells are next to at least one cell of the other population (Figure 9C).', 'timestamp': 'Fri Dec 22 09:35:39 +0000 2023'}, {'annotations': [], 'text': 'While stromal adhesion influences intra-cellular stop-and-go motility, it has minimal influence on collective motion: high and low stromal adhesion simulations have broadly similar mixing speeds.', 'timestamp': 'Fri Dec 22 09:35:44 +0000 2023'}, {'annotations': [], 'text': 'While collective motility is slightly ordered at cell-scale distances, no ordering persists at organ-scale (Figure 9D).', 'timestamp': 'Fri Dec 22 09:35:45 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'LN', 'obj': 'cell_type', 'prob': nan, 'span': {'begin': 35, 'end': 37}}], 'text': 'In summary, we used an organ-scale LN simulation as a challenging test case: an environment known for complex, large scale collective motility dynamics.', 'timestamp': 'Fri Dec 22 09:35:46 +0000 2023'}, {'annotations': [], 'text': 'Our system allowed us to integrate of microscopy data directly into our simulation, and to fit motility parameters directly on two-photon data.', 'timestamp': 'Fri Dec 22 09:35:48 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'fitted cell population', 'obj': 'cell_line', 'prob': 0.9426925182342529, 'span': {'begin': 36, 'end': 58}}], 'text': 'We investigated the behavior of the fitted cell population at organ-scale (millions of cells).', 'timestamp': 'Fri Dec 22 09:35:49 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'large', 'obj': 'cell_line', 'prob': 0.5885519981384277, 'span': {'begin': 51, 'end': 56}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell collectives', 'obj': 'cell_line', 'prob': 0.4378315508365631, 'span': {'begin': 57, 'end': 73}}], 'text': 'We hope that this will inspire further research on large cell collectives  integrating various sources of empirical data, at a high level of realism, and at realistic scale.', 'timestamp': 'Fri Dec 22 09:35:50 +0000 2023'}, {'annotations': [], 'text': 'We have implemented a parallel version of the CPM that, in some experiments, was 3 to 4 orders of magnitude faster on GPUs than a state-of-the-art serial CPM.', 'timestamp': 'Fri Dec 22 09:35:51 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5519843697547913, 'span': {'begin': 127, 'end': 130}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6493641138076782, 'span': {'begin': 158, 'end': 161}}], 'text': 'To achieve this, we designed novel schemes for lattice work splitting and synchronization, maintaining global state, efficient GPU memory transfer, and multi-GPU execution.', 'timestamp': 'Fri Dec 22 09:35:52 +0000 2023'}, {'annotations': [], 'text': 'We found that preservation of local lattice update statistics was sufficient to maintain high-level system dynamics, despite removing the sequential nature of the original In our work we aimed to ensure that important properties of CPM simulation trajectories were main- tained.', 'timestamp': 'Fri Dec 22 09:35:53 +0000 2023'}, {'annotations': [], 'text': 'Some previous CPM parallelisation efforts instead focused on equilibrium state distributions as important quantities.', 'timestamp': 'Fri Dec 22 09:35:54 +0000 2023'}, {'annotations': [], 'text': 'For example, Chen et al.!?', 'timestamp': 'Fri Dec 22 09:35:55 +0000 2023'}, {'annotations': [], 'text': 'argue that their parallel CPM maintains certain equi- librium properties.', 'timestamp': 'Fri Dec 22 09:35:56 +0000 2023'}, {'annotations': [], 'text': 'While such arguments are frequently applied in the analysis of energy-based models  in physics, Voss-Bhme et al.', 'timestamp': 'Fri Dec 22 09:35:58 +0000 2023'}, {'annotations': [], 'text': 'showe While we have shown that the same CPM in our benchmark scenarios, thi parameters that were originally fitte to re-fit their simulation parameters that these are less relevant for the CPM because the Metropolis algorithm used in the CPM drastically alters longtime behavior of the mode range of behaviors is present in our parallel CPM as in a serial s would still need to be carefully checked when re-using simulation on a serial implementation.', 'timestamp': 'Fri Dec 22 09:36:00 +0000 2023'}, {'annotations': [], 'text': 'Specifically, researchers might want when using our system instead of a serial is sensible in general when moving between different implementations of the s as the CPM where implementation details might anyway differ version  advice that subtly.', 'timestamp': 'Fri Dec 22 09:36:02 +0000 2023'}, {'annotations': [], 'text': 'Furthermore, the observed relation between model kinetics and performance (Figure 6E) also suggests it might be fruitful to explore more aggressive parallelisation settings that might yield further speed-up whilst maintaining desired simulated behavior.', 'timestamp': 'Fri Dec 22 09:36:03 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.49542269110679626, 'span': {'begin': 153, 'end': 156}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5996119976043701, 'span': {'begin': 219, 'end': 222}}], 'text': 'Through accelerating the CPM on accessible GPU hardware, we have increased the range of feasible simulation sizes  because our scheme fully utilises all GPU cores, simulation size is now mostly limited by the available GPU memory.', 'timestamp': 'Fri Dec 22 09:36:04 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6101958155632019, 'span': {'begin': 14, 'end': 17}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5318412780761719, 'span': {'begin': 46, 'end': 49}}], 'text': 'As memory per GPU increases, our scheme wi ti-GPU prototype scheme will allow for further increase of simulation volume.', 'timestamp': 'Fri Dec 22 09:36:05 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6792986392974854, 'span': {'begin': 21, 'end': 24}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5124723315238953, 'span': {'begin': 76, 'end': 79}}], 'text': 'Our to include inter-GPU communication over MPI,* allowing ding block for a GPU cluster-based version of the CPM  her to lcm?', 'timestamp': 'Fri Dec 22 09:36:06 +0000 2023'}, {'annotations': [], 'text': 'scale.', 'timestamp': 'Fri Dec 22 09:36:07 +0000 2023'}, {'annotations': [], 'text': 'multi-GPU scheme could be further extende our system to be us cell shape into acco 14, 15, 35, 36 ed as a fundamental bui ur Imm?', 'timestamp': 'Fri Dec 22 09:36:08 +0000 2023'}, {'annotations': [], 'text': 'simulations fur 1 easily scale to larger ular simulations are often performed using more coarse-grained agent-based model Our high-performance CPM now allows researchers to perform simulations that much more detailed and allow the investigation of new questions that require taking unt, such as migration in complex environments or morphogenesis.', 'timestamp': 'Fri Dec 22 09:36:10 +0000 2023'}, {'annotations': [], 'text': 'We hope that ions of liver and LN environments will serve as inspiration for many such spatially resolved tissue and organ simulations in the future.', 'timestamp': 'Fri Dec 22 09:36:11 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'cell_line', 'prob': 0.4618610143661499, 'span': {'begin': 80, 'end': 83}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.8818099498748779, 'span': {'begin': 169, 'end': 176}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'E i F cells', 'obj': 'cell_line', 'prob': 0.9323649406433105, 'span': {'begin': 194, 'end': 205}}], 'text': 'Similar approaches to the one shown here could be useful for parallelisation of CPM variants or other sequential lattice-based models.  A B mmm 100 yum explored space @ T cells S 10 1000 100000 E i F cells @ 0.6 E so 10 1000 100000 10 1000 100000 a motile state.', 'timestamp': 'Fri Dec 22 09:36:12 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'liver', 'obj': 'cell_type', 'prob': 0.4845820665359497, 'span': {'begin': 33, 'end': 38}}], 'text': '(A) A 1024x1024x182 um scan of a liver sinusoid is directly imported into our simulation environment (64ym slice shown).', 'timestamp': 'Fri Dec 22 09:36:13 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.7642619013786316, 'span': {'begin': 14, 'end': 21}}], 'text': '(B) Simulated T cells explore liver sinusoids (5000 cell simulation shown).', 'timestamp': 'Fri Dec 22 09:36:14 +0000 2023'}, {'annotations': [], 'text': '(C) Simulations at different cell densities explore the environment at different speeds (single simulation of 24hr, with 1 min snapshots).', 'timestamp': 'Fri Dec 22 09:36:15 +0000 2023'}, {'annotations': [], 'text': '(D) An increase of explored volume per cell occurs when cell density rises from 1,000 to 10,000 cells; 100,000 cells fill ~1/3 of sinusoid volume, causing exploration efficiency to saturate.', 'timestamp': 'Fri Dec 22 09:36:16 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated cells', 'obj': 'cell_type', 'prob': 0.8102577924728394, 'span': {'begin': 31, 'end': 46}}], 'text': '(E) At low cell densities most simulated cells are in an un-polarized, non-motile state; at higher densities collisions cause cells to transition in to a polarized, motile state (1 simulation of 30 min per density, arrest coefficient: fraction cells ; 2um/min,??', 'timestamp': 'Fri Dec 22 09:36:18 +0000 2023'}, {'annotations': [], 'text': 'shaded: Wilson score interval based on the number of 2-minute subtracks).', 'timestamp': 'Fri Dec 22 09:36:21 +0000 2023'}, {'annotations': [], 'text': 'CPM parameter settings can be found in Table 3.', 'timestamp': 'Fri Dec 22 09:36:22 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.8364054560661316, 'span': {'begin': 35, 'end': 42}}], 'text': '(F) Average speed of the simulated T cells at different densities.', 'timestamp': 'Fri Dec 22 09:36:28 +0000 2023'}, {'annotations': [], 'text': '1-15 ym 16-30 pm 31-45 im 46-60 pm 61-75 um 76-90 um =: 91-110 um B simufation =e D jowadhesion E  towadhesion low adhesion 3 3 a 5 0 10 20 30 0 10 20 30 SOHm =m 3 speed (\\\\um/min) speed (um/min) high adhesion & high adhesion high adhesion 8 0.40 @ 0.25 ee eh cc 5, = 0.20 a 3 0.00 ge 3 os 10 20 30 adhesion o 4 4.', 'timestamp': 'Fri Dec 22 09:36:31 +0000 2023'}, {'annotations': [], 'text': ' % polarisation decay 10 30 speed (um/min) speed (um/min) data from Mandl et al.?!.', 'timestamp': 'Fri Dec 22 09:36:37 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': '50,185 CPM cells', 'obj': 'cell_line', 'prob': 0.780194878578186, 'span': {'begin': 17, 'end': 33}}], 'text': 'to the motion of 50,185 CPM cells moving in a spherical structure (256um diameter).', 'timestamp': 'Fri Dec 22 09:36:38 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'fibroblastic', 'obj': 'cell_line', 'prob': 0.7321947813034058, 'span': {'begin': 53, 'end': 65}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'fibroblastic reticular cells', 'obj': 'cell_type', 'prob': 0.8869133591651917, 'span': {'begin': 53, 'end': 81}}, {'id': ['NCBITaxon:10090'], 'is_neural_normalized': False, 'mention': 'murine', 'obj': 'species', 'prob': 0.9583553671836853, 'span': {'begin': 120, 'end': 126}}], 'text': '(A) A 130x130x110 um* scan of the stromal structure (fibroblastic reticular cells) in the paracortical T cell zone of a murine LN.', 'timestamp': 'Fri Dec 22 09:36:39 +0000 2023'}, {'annotations': [], 'text': 'The images show average intensity projections of diffe ranges.', 'timestamp': 'Fri Dec 22 09:36:40 +0000 2023'}, {'annotations': [], 'text': 'After segmentation and mirroring, these data are used to fill a spherical simulation vo Speeds observed in simulated data are used to construct a probability distribution (upper panel, solid), which we compare to the empirical distribution (upper panel, dashed) using the Kullback-Leibler (KL) divergence metric.', 'timestamp': 'Fri Dec 22 09:36:42 +0000 2023'}, {'annotations': [], 'text': 'To this end, we bin the speed distributions, compute the divergence in each bin, and sum the values for each bin (lower panel).', 'timestamp': 'Fri Dec 22 09:36:44 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.7954138517379761, 'span': {'begin': 26, 'end': 33}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'stromal cells', 'obj': 'cell_type', 'prob': 0.7822971343994141, 'span': {'begin': 37, 'end': 50}}, {'id': ['CL:0000084'], 'is_neural_normalized': False, 'mention': 'T cell', 'obj': 'cell_type', 'prob': 0.8926588892936707, 'span': {'begin': 179, 'end': 185}}], 'text': '(C) Increased adhesion of T cells to stromal cells causes the speed distributions to be more right-skewed (top) and match real data more closely (bot: Increased adhesion can slow T cell paths non-uniformly by creating longer and/or more frequent pauses (red) but does not prevent rapid motion altogether.', 'timestamp': 'Fri Dec 22 09:36:45 +0000 2023'}, {'annotations': [], 'text': 'Two representative tracks are shown.', 'timestamp': 'Fri Dec 22 09:36:46 +0000 2023'}, {'annotations': [], 'text': '( search of motility parameters for cell polarization reveals which parameter ranges best fit the intravital data.', 'timestamp': 'Fri Dec 22 09:36:47 +0000 2023'}, {'annotations': [], 'text': 'Both low and high adhesion simulations have similar fit landscapes; the best fit is high adhesion simulation, although the difference to the low-adhesion best fit is small (1 simulation per parameter combination, best fit: red box/red line).', 'timestamp': 'Fri Dec 22 09:36:48 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated cells', 'obj': 'cell_type', 'prob': 0.6339921355247498, 'span': {'begin': 50, 'end': 65}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'T cells', 'obj': 'cell_type', 'prob': 0.6601049900054932, 'span': {'begin': 246, 'end': 253}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'stromal cells', 'obj': 'cell_type', 'prob': 0.9052724242210388, 'span': {'begin': 258, 'end': 271}}], 'text': 'CPM parameter settings can be found in  3,359,752 simulated cells O mins 40 mins 80 mins 120 mins D 06 5 0.6 0 30 60 90 120 30 100 300 1000 time (in min) scale (in um) high adhesion low adhesion LN simulation at 1mm diameter containing 3,359,752 T cells and stromal cells (grey; Figure 8A).', 'timestamp': 'Fri Dec 22 09:36:49 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'simulated cells', 'obj': 'cell_type', 'prob': 0.6027023792266846, 'span': {'begin': 43, 'end': 58}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'red', 'obj': 'cell_type', 'prob': 0.5895609259605408, 'span': {'begin': 74, 'end': 77}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'green cells', 'obj': 'cell_type', 'prob': 0.5083209872245789, 'span': {'begin': 82, 'end': 93}}], 'text': 'For clearer visualization, only 10% of the simulated cells are shown, and red and green cells are equally param- eterized although coloured differently.', 'timestamp': 'Fri Dec 22 09:36:50 +0000 2023'}, {'annotations': [], 'text': '(B) The simulation is initialized with two equally parameterized populations each starting at one half of the sphere.', 'timestamp': 'Fri Dec 22 09:36:51 +0000 2023'}, {'annotations': [], 'text': 'Mixing is quantified by tracking the fraction of cells having at least one neighbour of the other population.', 'timestamp': 'Fri Dec 22 09:36:52 +0000 2023'}, {'annotations': [], 'text': '(C) Global mixing is slower in the high-adhesion simulation (red) than at low adhesion (black), although the difference is minor.', 'timestamp': 'Fri Dec 22 09:36:53 +0000 2023'}, {'annotations': [], 'text': 'Both simulations reach 50% mixing in ~1h.', 'timestamp': 'Fri Dec 22 09:36:55 +0000 2023'}, {'annotations': [], 'text': '(D) Cells do not synchronize their directionality much at longer spatial scales, and this absence of long-range synchronization is the same for low and high adhesion.', 'timestamp': 'Fri Dec 22 09:36:56 +0000 2023'}, {'annotations': [], 'text': 'CPM parameter set- tings can be found in Table 3.', 'timestamp': 'Fri Dec 22 09:36:57 +0000 2023'}, {'annotations': [], 'text': 'Table 1: Model settings for 2D simulations.', 'timestamp': 'Fri Dec 22 09:36:58 +0000 2023'}, {'annotations': [], 'text': 'For all 2D simulations we use Viarget = 2500, Atarget = 600, Aperimeter = 2; Jcett,ecm = 10 and T = 10.', 'timestamp': 'Fri Dec 22 09:37:00 +0000 2023'}, {'annotations': [], 'text': 'The vector field in the 2D circular chemotaxis simulations (Figure 5A) consists of vectors at a 90 degree angle to radial vectors from points in the field to the simulation center, with all vectors in the field normalized to length 1.', 'timestamp': 'Fri Dec 22 09:37:01 +0000 2023'}, {'annotations': [], 'text': '| Avotume | Aaet | Maxact | Achemotaxis Fig 1B 20, 30, 40 | 200, 250, 300, 350, 400 50  Fig 5A, Fig 6E | 20 =  10 Fig 5B 20 50-800, in increments of 50 | 50  Table 2: Model settings for 3D simulations.', 'timestamp': 'Fri Dec 22 09:37:02 +0000 2023'}, {'annotations': [], 'text': 'For all simulations Viargee = 150, Avolume = 25, Atarget = 1500, Aperimeter = 0.2, Scett,ce = 10 and T = 20.', 'timestamp': 'Fri Dec 22 09:37:04 +0000 2023'}, {'annotations': [], 'text': 'The time resolution is 1 second per Monte | uw _| At | Apersistence | Jcell,BCM Fig 1A  f|[ 10 Fig 5C, Fig 6D | 0.95 | 15 | 40 0 Table 3: Model settings for liver and lymph node simulations.', 'timestamp': 'Fri Dec 22 09:37:05 +0000 2023'}, {'annotations': [], 'text': 'For all simulations Viarge, = 150, Avolume = 25, Atarget = 1500, Aperimeter = 0-2, Jcett,cett = 10, Jeet,ncm = 0, and T = 20.', 'timestamp': 'Fri Dec 22 09:37:06 +0000 2023'}, {'annotations': [], 'text': 'The time resolution is 1 second per Monte Carlo step.', 'timestamp': 'Fri Dec 22 09:37:07 +0000 2023'}, {'annotations': [], 'text': '| Aact | Maxact Teell, FRC Fig 7 2600 10 0 Fig 8B | 2600 10 0 Fig 8C | 2600 10 -400, -300, -200, -100, 0 Fig 8D | 2600 10 -400, 10 Fig 8E | 200-3000, in increments of 200 | 10, 15, 20, 25, 30, 40, 50, 60, 70 | -3000, 10 Fig 9C | 2600 10 -800, 0 Fig 9D | 2600 10 -800 This research was supported by Young Investigator Grant 10620 from KWF Kankerbestrijding and HFSP program grant RGP0053/2020 (to JT).', 'timestamp': 'Fri Dec 22 09:37:08 +0000 2023'}, {'annotations': [], 'text': 'JT was also supported by NWO Vidi grant VI.Vidi.192.084.', 'timestamp': 'Fri Dec 22 09:37:10 +0000 2023'}, {'annotations': [], 'text': 'We thank Franka Buytenhuijs for giving valuable feedback on this manuscript, and we thank L.W.A.', 'timestamp': 'Fri Dec 22 09:37:11 +0000 2023'}, {'annotations': [], 'text': 'Stuurman and Pedro Aroca Lara for testing early versions of this simulation system.', 'timestamp': 'Fri Dec 22 09:37:12 +0000 2023'}, {'annotations': [], 'text': 'We also gratefully acknowledge the support of NVIDIA Corporation with the donation of the Titan V GPU used for this research.', 'timestamp': 'Fri Dec 22 09:37:13 +0000 2023'}, {'annotations': [], 'text': 'The funders had no role in design, execution, and interpretation of the research reported in SS and JT conceived the study.', 'timestamp': 'Fri Dec 22 09:37:14 +0000 2023'}, {'annotations': [], 'text': 'SS designed the parallel CPM algorithms.', 'timestamp': 'Fri Dec 22 09:37:15 +0000 2023'}, {'annotations': [], 'text': 'SS performed the research and analysed the data.', 'timestamp': 'Fri Dec 22 09:37:17 +0000 2023'}, {'annotations': [], 'text': 'SS and JT wrote the manuscript.', 'timestamp': 'Fri Dec 22 09:37:18 +0000 2023'}, {'annotations': [], 'text': 'JT supervized the project.', 'timestamp': 'Fri Dec 22 09:37:19 +0000 2023'}, {'annotations': [], 'text': 'SM provided critical feedback and provided the liver and lymph node imaging data.', 'timestamp': 'Fri Dec 22 09:37:20 +0000 2023'}, {'annotations': [], 'text': 'SD performed image acquisition of the liver and LN data.', 'timestamp': 'Fri Dec 22 09:37:21 +0000 2023'}, {'annotations': [], 'text': 'All authors critically revized the manuscript for intellectual content.', 'timestamp': 'Fri Dec 22 09:37:22 +0000 2023'}, {'annotations': [], 'text': 'The authors declare no competing interests.', 'timestamp': 'Fri Dec 22 09:37:23 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'biological cells', 'obj': 'cell_type', 'prob': 0.8549009561538696, 'span': {'begin': 114, 'end': 130}}], 'text': 'Materials and Methods Cellular Potts model and model variants The CPM is a lattice based formalism for simulating biological cells.', 'timestamp': 'Fri Dec 22 09:37:25 +0000 2023'}, {'annotations': [], 'text': 'Simulations proceeds with a Markov chain Monte Carlo process, where a single position at a time can transition to one of the states of its neighbours.', 'timestamp': 'Fri Dec 22 09:37:26 +0000 2023'}, {'annotations': [], 'text': 'The transition probability is based on the change of a Hamiltonian energy function.', 'timestamp': 'Fri Dec 22 09:37:27 +0000 2023'}, {'annotations': [{'id': ['CHEBI:29979'], 'is_neural_normalized': False, 'mention': 'H', 'obj': 'drug', 'prob': 0.999607503414154, 'span': {'begin': 4, 'end': 5}}, {'id': ['CHEBI:30455'], 'is_neural_normalized': False, 'mention': 'Te', 'obj': 'drug', 'prob': 0.9907642602920532, 'span': {'begin': 46, 'end': 48}}, {'id': ['CHEBI:30455'], 'is_neural_normalized': False, 'mention': 'Te', 'obj': 'drug', 'prob': 0.999890923500061, 'span': {'begin': 65, 'end': 67}}], 'text': 'The H =~ Ip(ox),r(0))(1  84:,0;) +  > Avotume(Te)(V (0)  Veargot(Te))?', 'timestamp': 'Fri Dec 22 09:37:28 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lattice sites', 'obj': 'DNA', 'prob': 0.5186437368392944, 'span': {'begin': 204, 'end': 217}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell type', 'obj': 'cell_type', 'prob': 0.5623244047164917, 'span': {'begin': 333, 'end': 342}}], 'text': '(1) as its Hamiltonian, with i, 7 being possible neighbour locations in the lattice, 7; the cell present at location i, T is the type of a particular cell, 5 is the Kronecker delta, V(c) is the number of lattice sites a cell o occupies, and Ayolume(T) how strongly a volume constraint contributes to the Hamiltonian for a particular cell type.', 'timestamp': 'Fri Dec 22 09:37:29 +0000 2023'}, {'annotations': [], 'text': 'In all our simulations a Moore neighbourhood was used.', 'timestamp': 'Fri Dec 22 09:37:30 +0000 2023'}, {'annotations': [], 'text': 'Transitions stochastically succeed based on the change in energy AH, using a Boltzmann probability distribution 1 if AH <0, P(AH) = {laws if AH > 0.', 'timestamp': 'Fri Dec 22 09:37:31 +0000 2023'}, {'annotations': [], 'text': '(2) By adding extra terms to the Hamiltonian, constraints can be added to cells and new types of behavior introduced to the model.', 'timestamp': 'Fri Dec 22 09:37:33 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'Atarget', 'obj': 'DNA', 'prob': 0.5894106030464172, 'span': {'begin': 150, 'end': 157}}], 'text': 'When cell shape needs to be parameterized beyond just cell size, a perimeter constraint* is often introduced, Hperimeter = Ss Aperimeter (To )(A(a) _ Atarget (To).', 'timestamp': 'Fri Dec 22 09:37:34 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lattice sites', 'obj': 'DNA', 'prob': 0.6114188432693481, 'span': {'begin': 58, 'end': 71}}], 'text': '(3) Here A(c) represents the number of interfaces between lattice sites of a cell with sites not part of the We also implemented several motility models.', 'timestamp': 'Fri Dec 22 09:37:35 +0000 2023'}, {'annotations': [], 'text': 'The first allows a vector field to bias the energy change AH during a copy attempt AH chemotaxis = Achemotaxis (To )#i+7 * Ci.', 'timestamp': 'Fri Dec 22 09:37:37 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'position 7', 'obj': 'DNA', 'prob': 0.4702039659023285, 'span': {'begin': 138, 'end': 148}}], 'text': '(4) Here #;_,; is the unit vector pointing in the direction of a copy attempt from position i to 7, and C; the value of a vector field at position 7.', 'timestamp': 'Fri Dec 22 09:37:38 +0000 2023'}, {'annotations': [], 'text': 'In this paper, C; is always a field of unit vectors.', 'timestamp': 'Fri Dec 22 09:37:39 +0000 2023'}, {'annotations': [], 'text': 'For the model where cells exhibit persistent movement based on one vector per cell, we add AHpersistence = Apersistence (To )#i+5 * Po,t (5) to the Hamiltonian delta.', 'timestamp': 'Fri Dec 22 09:37:41 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'MCS', 'obj': 'DNA', 'prob': 0.6348241567611694, 'span': {'begin': 103, 'end': 106}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell type T', 'obj': 'cell_type', 'prob': 0.6872813105583191, 'span': {'begin': 233, 'end': 244}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'MCSs', 'obj': 'DNA', 'prob': 0.8925271034240723, 'span': {'begin': 319, 'end': 323}}], 'text': 'The vector ,,, is the persistence direction of cell o at time t, which is updated iteratively at every MCS with ey Xo ~~ X, tAt a Dot = (1 = pr) S = + flr Pot1s (6) Pot = 7 (7) Here ju, is the strength of directional persistence for cell type T, Xot is the centroid location of cell o at time t and At is the number of MCSs we look back for to determine the current direction of a cell.', 'timestamp': 'Fri Dec 22 09:37:42 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:12', '71', '72', '9457', '345651', '11332', '440915', '641455'], 'is_neural_normalized': True, 'mention': 'Act', 'obj': 'gene', 'prob': 0.909369707107544, 'span': {'begin': 26, 'end': 29}}], 'text': 'Finally, we implement the Act model that allows for localized polarization.', 'timestamp': 'Fri Dec 22 09:37:44 +0000 2023'}, {'annotations': [], 'text': 'For each position i we track an activity value each time said location is updated, here labelled as Act(i).', 'timestamp': 'Fri Dec 22 09:37:45 +0000 2023'}, {'annotations': [], 'text': 'This localized memory is initialized at Maxact, and decays by 1 each timestep until it reaches 0.', 'timestamp': 'Fri Dec 22 09:37:46 +0000 2023'}, None, {'annotations': [], 'text': ' GMact (J)), (8)  making copies that originate from areas of a cell with a high activity value more likely.', 'timestamp': 'Fri Dec 22 09:37:47 +0000 2023'}, {'annotations': [], 'text': 'The activity is calculated for a full Moore neighbourhood using a geometric mean GMaci(i) = ( T] Act(n)) Ol.', 'timestamp': 'Fri Dec 22 09:37:48 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.4282142221927643, 'span': {'begin': 112, 'end': 115}}], 'text': '(9) N(i) are all neighbours of pixel or voxel i. Optimized CPU implementation of CPM Besides a high-performance GPU implementation, we also developed our own CPU implementation.', 'timestamp': 'Fri Dec 22 09:37:50 +0000 2023'}, {'annotations': [], 'text': 'This served both as a reference implementation for model behavior and internal implementation details for us, and as a reference performance benchmark.', 'timestamp': 'Fri Dec 22 09:37:51 +0000 2023'}, {'annotations': [], 'text': 'We strived to reduce the major performance bottlenecks we could see in profiling.', 'timestamp': 'Fri Dec 22 09:37:52 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lattice sites', 'obj': 'DNA', 'prob': 0.7207123637199402, 'span': {'begin': 26, 'end': 39}}], 'text': 'The implementation tracks lattice sites on cell boundaries, and only tries copy attempts involving such positions; a common optimisation, which is also implemented in Morpheus.', 'timestamp': 'Fri Dec 22 09:37:53 +0000 2023'}, {'annotations': [], 'text': 'The tracking implementation uses a high-performance hashmap implementation.*!', 'timestamp': 'Fri Dec 22 09:37:54 +0000 2023'}, {'annotations': [], 'text': 'We use a custom high-performance pseudorandom number generator, xoshiro256.4?_ The lattice data is stored with a Z- order space filling curve, allowing data that is close in the simulation space also to be close in the linear memory space, maximizing cache efficiency.', 'timestamp': 'Fri Dec 22 09:37:55 +0000 2023'}, {'annotations': [], 'text': 'Memory addresses are calculated using intrinsic bit interleaving instructions from the x86 bit manipulation instruction set extension.', 'timestamp': 'Fri Dec 22 09:37:56 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.4974348545074463, 'span': {'begin': 4, 'end': 7}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CUDA', 'obj': 'DNA', 'prob': 0.6939306259155273, 'span': {'begin': 92, 'end': 96}}], 'text': 'Our CPM was implemented using CUDA; concepts equivalent to ones described here exist in non-CUDA GPU hardware.', 'timestamp': 'Fri Dec 22 09:37:58 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6183270215988159, 'span': {'begin': 2, 'end': 5}}], 'text': 'A GPU consists of multiple streaming multiprocessors (SMs).', 'timestamp': 'Fri Dec 22 09:37:59 +0000 2023'}, {'annotations': [], 'text': 'Each SM runs multiple threads in parallel; contrasting with CPU threads, threads on the same SM execute in lockstep, in a Single Instruction Multiple Threads approach.', 'timestamp': 'Fri Dec 22 09:38:00 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5207992792129517, 'span': {'begin': 9, 'end': 12}}], 'text': 'A single GPU process  called a kernel  can consist of up to millions of threads.', 'timestamp': 'Fri Dec 22 09:38:01 +0000 2023'}, {'annotations': [], 'text': 'GPUs have relatively slow global memory available to all threads, and much faster shared memory local to SMs.', 'timestamp': 'Fri Dec 22 09:38:02 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.46887579560279846, 'span': {'begin': 74, 'end': 77}}], 'text': 'Global memory transfer speed is often the main performance bottleneck for GPU programs.', 'timestamp': 'Fri Dec 22 09:38:04 +0000 2023'}, {'annotations': [], 'text': 'Threads are bundled in thread blocks (TBs); each TB is bound to a specific SM.', 'timestamp': 'Fri Dec 22 09:38:05 +0000 2023'}, {'annotations': [], 'text': 'By caching data on their SM in shared memory, TBs can greatly reduce global memory transfer bottlenecks.', 'timestamp': 'Fri Dec 22 09:38:06 +0000 2023'}, {'annotations': [], 'text': 'Memory transfer bottlenecks can be further reduced by two strategies: read coalescing and high SM occupancy.', 'timestamp': 'Fri Dec 22 09:38:07 +0000 2023'}, {'annotations': [], 'text': 'Read coalescing has sequential threads in a TB do contiguous reads of global memory; contiguous reads are needed to fully utilise the memory bus bandwidth.', 'timestamp': 'Fri Dec 22 09:38:08 +0000 2023'}, {'annotations': [], 'text': 'High occupancy is achieved when a high number of threads are scheduled to execute per SM; the SM scheduler will switch out threads when they are blocked on memory transfers, ensuring a high number of in-flight memory instructions at We propose a scheme of memory transfer, local SM execution and synchronization (Figure 2D).', 'timestamp': 'Fri Dec 22 09:38:14 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell ids', 'obj': 'DNA', 'prob': 0.7226570844650269, 'span': {'begin': 37, 'end': 45}}], 'text': 'We focus on caching just the grid of cell ids, as this data is always needed, and can be used to quickly reject copy attempts internal to a cell.', 'timestamp': 'Fri Dec 22 09:38:21 +0000 2023'}, {'annotations': [], 'text': 'The simulation grid is divided in subsections, with regions larger than earlier checkerboard subdivisions; each subsection is assigned to a TB, subsection size is chosen so active regions fit in shared memory.', 'timestamp': 'Fri Dec 22 09:38:25 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'TB', 'obj': 'DNA', 'prob': nan, 'span': {'begin': 5, 'end': 7}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'checkerboard sections', 'obj': 'DNA', 'prob': 0.531246542930603, 'span': {'begin': 25, 'end': 46}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'coalesced reads', 'obj': 'RNA', 'prob': 0.4433795213699341, 'span': {'begin': 123, 'end': 138}}], 'text': 'Each TB determines which checkerboard sections are active in its region; active sections are cached in shared memory using coalesced reads.', 'timestamp': 'Fri Dec 22 09:38:27 +0000 2023'}, {'annotations': [], 'text': 'After caching  intended to speed up reads  the simulation is run, and successful copy attempts write new values to both shared and global memory.', 'timestamp': 'Fri Dec 22 09:38:33 +0000 2023'}, {'annotations': [], 'text': 'Alternatively, changes could be cached in shared memory and written back coalesced to global memory; in our testing, an immediate shared/global write yielded better performance.', 'timestamp': 'Fri Dec 22 09:38:35 +0000 2023'}, {'annotations': [], 'text': 'If a TB is done with simulation execution, it immediately starts a memory request for the next active region, ensuring maximum occupancy.', 'timestamp': 'Fri Dec 22 09:38:37 +0000 2023'}, {'annotations': [], 'text': 'All threads in the kernel are synchronized after new memory transfers are finished, ensuring maximum overlap of code execution and memory transfer.', 'timestamp': 'Fri Dec 22 09:38:42 +0000 2023'}, {'annotations': [], 'text': 'CPM literature offers a wide range of model variants and extensions, of which we implement a mod- est subset.', 'timestamp': 'Fri Dec 22 09:38:46 +0000 2023'}, {'annotations': [], 'text': 'This comes with potential performance issues: even when extensions are disabled through branching statements, on GPUs such branching code still incurs a significant performance loss compared to CPUs.', 'timestamp': 'Fri Dec 22 09:38:47 +0000 2023'}, {'annotations': [], 'text': 'We therefore pre-generated specialized versions of the kernel for a range of different extensions.', 'timestamp': 'Fri Dec 22 09:38:51 +0000 2023'}, {'annotations': [], 'text': 'We use a custom high-performance random number generator on the GPU.', 'timestamp': 'Fri Dec 22 09:39:01 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'CPM', 'obj': 'DNA', 'prob': 0.4960204064846039, 'span': {'begin': 40, 'end': 43}}], 'text': '*%  Composite variable tracking Several CPM extensions introduce additional global variables.', 'timestamp': 'Fri Dec 22 09:39:02 +0000 2023'}, {'annotations': [], 'text': 'Some of these variable need to be kept in-syne with each other  one variable being slightly out of date with others is not tolerated.', 'timestamp': 'Fri Dec 22 09:39:03 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'lattice', 'obj': 'DNA', 'prob': 0.4912048578262329, 'span': {'begin': 136, 'end': 143}}], 'text': 'One example is tracking of a center-of-mass centroid; for 2D this can be tracked as a sum of x and y values, and corresponding count of lattice locations occupied (Figure 2C).', 'timestamp': 'Fri Dec 22 09:39:04 +0000 2023'}, {'annotations': [], 'text': 'These are integer values: tracking centroid locations as floating-point would cause values to drift due to rounding errors.', 'timestamp': 'Fri Dec 22 09:39:05 +0000 2023'}, {'annotations': [], 'text': 'The centroid value {a,y} = = X {asum, Ysum }- (10) Online updates change both a cells centroid position and size.', 'timestamp': 'Fri Dec 22 09:39:07 +0000 2023'}, {'annotations': [], 'text': 'When a new lattice position {Xposition, Yposition } is added to a cell, its centroids x value can be updated with Loia X | lola + Lposition Cell sizes are stored interleaved with sum values; sizes and sums are both stored as 32-bit unsigned integers.', 'timestamp': 'Fri Dec 22 09:39:08 +0000 2023'}, {'annotations': [], 'text': 'The interleaved storage enables use of 64-bit operations to update both: a copy-and-swap (CAS) operation is used, which atomically updates sum and cell size values simultaneously.', 'timestamp': 'Fri Dec 22 09:39:10 +0000 2023'}, {'annotations': [], 'text': 'This scheme also handles periodic boundaries, by adding or subtracting dimension x |e| to the sum values before executing the CAS as needed, to keep centroids within simulation bounds.', 'timestamp': 'Fri Dec 22 09:39:11 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.4150868058204651, 'span': {'begin': 51, 'end': 55}}], 'text': 'We devized a scheme to execute our CPM on multiple GPUs (Figure 2E); this scheme is intended as a prototype, and as such not as optimized as other parts of our CPM.', 'timestamp': 'Fri Dec 22 09:39:12 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5184324383735657, 'span': {'begin': 73, 'end': 76}}], 'text': 'Simulation space is cut into equal regions, each region is assigned to a GPU.', 'timestamp': 'Fri Dec 22 09:39:13 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'cell ids', 'obj': 'DNA', 'prob': 0.6058043241500854, 'span': {'begin': 36, 'end': 44}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5956013798713684, 'span': {'begin': 63, 'end': 66}}], 'text': 'A regions lattice data, such as the cell ids, are sent to each GPU, with an additional border layer.', 'timestamp': 'Fri Dec 22 09:39:14 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5548716187477112, 'span': {'begin': 56, 'end': 59}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.7119873762130737, 'span': {'begin': 100, 'end': 103}}], 'text': 'Global variables such as cell sizes are tracked on each GPU as deltas of the value since last inter-GPU data synchronisation.', 'timestamp': 'Fri Dec 22 09:39:15 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPUs', 'obj': 'DNA', 'prob': 0.6114341616630554, 'span': {'begin': 52, 'end': 56}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.5769087076187134, 'span': {'begin': 120, 'end': 123}}], 'text': 'Every four MCSs, these deltas are gathered from all GPUs, deltas are integrated, and updated values distributed to each GPU.', 'timestamp': 'Fri Dec 22 09:39:16 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6335692405700684, 'span': {'begin': 6, 'end': 9}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6302101016044617, 'span': {'begin': 63, 'end': 66}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.7061171531677246, 'span': {'begin': 88, 'end': 91}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.7283400297164917, 'span': {'begin': 159, 'end': 162}}, {'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.7916373014450073, 'span': {'begin': 166, 'end': 169}}], 'text': 'Multi-GPU speedup is mostly hampered by frequent or slow inter-GPU communication; inter-GPU synchronization after every MCS negates most speedup, and a faster GPU-to-GPU interlink would likely increase performance.', 'timestamp': 'Fri Dec 22 09:39:17 +0000 2023'}, {'annotations': [], 'text': 'The CPU-based simulations, using our custom optimized CPU implementation and Morpheus 2.2.6 (the last non-parallelized version of Morpheus), and most GPU simulations were run on a system with a Nvidia Titan V, an AMD 2950x CPU and 128 GB of RAM at 2133 MHz.', 'timestamp': 'Fri Dec 22 09:39:19 +0000 2023'}, {'annotations': [{'id': ['CUI-less'], 'is_neural_normalized': False, 'mention': 'GPU', 'obj': 'DNA', 'prob': 0.6108351349830627, 'span': {'begin': 42, 'end': 45}}], 'text': 'The only exceptions to this are the multi-GPU simulations (Figure 2E), which were run on a system with four Nvidia Geforce RTX 2080 Ti GPUs, an Intel i9-9820x CPU at 3.3 GHz, and 128 GB of RAM at 2133 MHz.', 'timestamp': 'Fri Dec 22 09:39:20 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:6363'], 'is_neural_normalized': False, 'mention': 'CCL19', 'obj': 'gene', 'prob': 0.9917254447937012, 'span': {'begin': 76, 'end': 81}}, {'id': ['NCBIGene:1969', '10306'], 'is_neural_normalized': True, 'mention': 'EYFP', 'obj': 'gene', 'prob': 0.9889999628067017, 'span': {'begin': 82, 'end': 86}}, {'id': ['NCBITaxon:10095'], 'is_neural_normalized': False, 'mention': 'mice', 'obj': 'species', 'prob': 0.9509281516075134, 'span': {'begin': 87, 'end': 91}}, {'id': ['mesh:C003043'], 'is_neural_normalized': False, 'mention': 'paraformaldehyde', 'obj': 'drug', 'prob': 0.9993519186973572, 'span': {'begin': 110, 'end': 126}}, {'id': ['CHEBI:60594'], 'is_neural_normalized': False, 'mention': 'PFA', 'obj': 'drug', 'prob': 0.9998096227645874, 'span': {'begin': 128, 'end': 131}}], 'text': 'Imaging of thick lymph node sections Naive inguinal LNs were harvested from CCL19-EYFP mice and fixed with 4% paraformaldehyde (PFA, Electron Microscopy Sciences) for 1 hour at 4C.', 'timestamp': 'Fri Dec 22 09:39:23 +0000 2023'}, {'annotations': [], 'text': 'Following a brief wash with PBS, LNs were embedded in 2% agarose and 250-300 ym tissue sections were generated using a vibratome (Leica).', 'timestamp': 'Fri Dec 22 09:39:24 +0000 2023'}, {'annotations': [{'id': ['NCBIGene:3198'], 'is_neural_normalized': True, 'mention': 'BSA', 'obj': 'gene', 'prob': 0.5133447051048279, 'span': {'begin': 84, 'end': 87}}, {'id': ['mesh:D017830'], 'is_neural_normalized': False, 'mention': 'Triton X-100', 'obj': 'drug', 'prob': 0.939365804195404, 'span': {'begin': 123, 'end': 135}}, {'id': ['NCBITaxon:9986'], 'is_neural_normalized': False, 'mention': 'rabbit', 'obj': 'species', 'prob': 0.974775493144989, 'span': {'begin': 196, 'end': 202}}, {'id': ['NCBIGene:2994'], 'is_neural_normalized': True, 'mention': 'anti-GFP Ab', 'obj': 'gene', 'prob': 0.9535024762153625, 'span': {'begin': 203, 'end': 214}}, {'id': ['mesh:D011136'], 'is_neural_normalized': False, 'mention': 'Tween-20', 'obj': 'drug', 'prob': 0.9272661805152893, 'span': {'begin': 321, 'end': 329}}], 'text': 'LN sections were subsequently immersed in a Special Wash Buffer (Baxter water, 0.2% BSA, 10% 10X PBS, 0.2% Tween- 20, 0.2% Triton X-100, 0.2% 10% SDS) at 4C overnight, followed by staining with a rabbit anti-GFP Ab (Thermo Scientific, 1:100, diluted in blocking buffer made with Baxter water, 0.2% BSA, 10% 10X PBS, 0.1% Tween-20, 0.2% 10% SDS) at 4C for three days.', 'timestamp': 'Fri Dec 22 09:39:26 +0000 2023'}, {'annotations': [{'id': ['mesh:D005990'], 'is_neural_normalized': False, 'mention': 'glycerol', 'obj': 'drug', 'prob': 0.9732889533042908, 'span': {'begin': 79, 'end': 87}}], 'text': 'After incubation, LN sections were thoroughly washed in PBS and mounted in 85% glycerol in PBS with glass coverslips.', 'timestamp': 'Fri Dec 22 09:39:27 +0000 2023'}, {'annotations': [], 'text': 'Thick LN sections were imaged using a LSM880 AiryscanFast confocal microscope (Carl Zeiss) with a x63/1.4NA oil immersion lens and images processed with ZEN 2010 software (Carl Zeiss).', 'timestamp': 'Fri Dec 22 09:39:28 +0000 2023'}]\n",
      "Wall time: 8min 49s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "entity_list = []\n",
    "print(len(sentences))\n",
    "\n",
    "i=0\n",
    "# The last sentence is invalid\n",
    "for s in sentences[:-1]:\n",
    "    print(\"doing sentence : \" + str(i))\n",
    "    entity_list.append(query_plain2(s))\n",
    "    i += 1\n",
    "\n",
    "print(\"all calls done\")\n",
    "print(entity_list)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d906999a",
   "metadata": {},
   "source": [
    "We use filtered_entity_list because there are some cases where the bern2 API returns nothing because of an error "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6fbc8fa9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'text': 'The CPM!',\n",
       "  'text_sha256': '07740a8ca042666ca553659e99794d2f88fbfe35966c94277de2477830442547'},\n",
       " {'text': 'is a powerful simulation method that is known for its ability to simulate a wide range of biological scenarios.',\n",
       "  'text_sha256': '1b7d80918c88b7fe4e28e3f333b6d232f49c9ec61a21cab46045e3ce8282d4f7'},\n",
       " {'entities': [{'entity_id': 'single cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'single cells'}],\n",
       "  'text': 'When modelling tissue environments, many researchers wish to consider both he complex morphology of single cells, and the behavior of these cells as part of larger collectives.',\n",
       "  'text_sha256': 'c1a776d22bb692de4f55202d391f385dc9a9b9097389e13ce3d6fdab49e6c5d2'},\n",
       " {'text': 'The appeal of the CPM lies partly in its ability to integrate both these aspects in a single model.',\n",
       "  'text_sha256': '90d7a11080f5e7bab41b6bbdd1c90ad090a86f6a9a8e6961e84943e2d4af8181'},\n",
       " {'entities': [{'entity_id': 'tumor',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'tumor'}],\n",
       "  'text': 'As such, it has been applied to diverse problems including tumor growth, vascularization, wound healing,* and morphogenesis.>* Many extensions and hybrid model variants are available in the literature 0, for example, add support for cell motility, cell signalling,! and finite-element based mechanical interactions.!!',\n",
       "  'text_sha256': '61b874f4c1e62af79801f8a6e8df7da0a6c3f023db61e684b2d760d7adac7688'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': 'Thus, the CPM can be described as a generic platform upon which a wide range of models can be built, possibly explaining its popularity in the tissue modelling field.',\n",
       "  'text_sha256': 'b808851ca99964749f3ce136e0e4b3bde9d3ba3b6ba8c24231c780cad131cee8'},\n",
       " {'text': 'For computational reasons, CPM models are often built using a small number of cells (perhaps in the housands) in simplified two-dimensional (2D) geometries.',\n",
       "  'text_sha256': 'b24cc58b3a8b7e4bdd09c0c2a66f70a4523cb03b46ca0f8bfa37b90627a8c120'},\n",
       " {'text': 'Some research questions in tissue modelling however demand a realistically large number of cells, and accurate 3D representations of tissues.',\n",
       "  'text_sha256': '8f985c307692e4ecfd8fddf55b1d3b4bf36f9a60728e6e99093240adb77a68f3'},\n",
       " {'entities': [{'entity_id': 'motile cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'motile cells'}],\n",
       "  'text': 'For instance, it is widely appreciated that the efficiency of space exploration of motile cells depends criti- cally on the dimensionality of the environment, leading researchers to use 3D CPM simulations. !?',\n",
       "  'text_sha256': '3961c6d1330ff56459eacf038481ae131ee78bd81949c0e5136ba837d9104630'},\n",
       " {'entities': [{'entity_id': 'tumor',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'tumor'},\n",
       "   {'entity_id': 'tumors',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'tumors'},\n",
       "   {'entity_id': 'tumor',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'tumor'}],\n",
       "  'text': ':13 Assessment of tumor growth behavior and invasive patterns of tumors likewise changes as a function of dimension, resulting in the use of 3D CPM simulations in the tumor modelling field as well.?',\n",
       "  'text_sha256': '6402458a476479be9942d8ea4c6a13d914e6f4ee3e0207dd0979cbc470f3b351'},\n",
       " {'text': 'While there is thus a clear research demand for realistically large 3D CPM simulations, scaling up such simulations is generally considered computationally infeasible.4 1 Removing this computational barrier would allow a host of CPM based modelling research to be scaled to what is biologically realistic, instead of scaling to what technical constraints allow.',\n",
       "  'text_sha256': 'cc389063a33eaa5a7e132c389a23deaf3cbc32ca5cbe3334c501fc32963a6541'},\n",
       " {'text': 'The CPM is an on-lattice method: simulations are executed on a grid.',\n",
       "  'text_sha256': 'd433b21a1de8fa8c4333607ed31b2c1be5e19516c0d27e75ec358115dfed12c7'},\n",
       " {'text': 'A common strategy to speed up grid-based models is to cut the simulation into smaller subsections that are run in parallel.',\n",
       "  'text_sha256': 'e7bed6df93c5124cf4ade694384f3fb914c86c282e55be4c84d42d861a9dee5a'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': 'However, for the CPM, model behavior is intimately tied to locations on the grid being updated sequentially, in a single- location update Markov chain Monte Carlo process.',\n",
       "  'text_sha256': 'd479116930eb690c808971ca68c68cb886c871e0da72ef92667cc25acf5d71f6'},\n",
       " {'text': 'As such, parallelisation has been cautioned against, as it risks drastic changes in CPM system kinetics. Nevertheless, several attempts have been made to employ subsection-based parallelisation.',\n",
       "  'text_sha256': '7c57b832527a78f5c16589535ff1ed3693800b85e1c1dd47efc7dc9dbc369ca4'},\n",
       " {'text': 'Chen et al.',\n",
       "  'text_sha256': '30ebeb6e4a6fc0e251c33eed36eca304a833177e5f2e6bbe89bd820a28882c8f'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': 'proposed a checkerboard scheme that divides the CPM lattice up over a computer cluster, with sections chosen much larger than cells, preventing cell-level synchronization issues.',\n",
       "  'text_sha256': '81e8206c58f43bd15c413e1e91fe7a8642cea3897a693ce797b95ec8503e9b78'},\n",
       " {'text': '! Tapia et al.',\n",
       "  'text_sha256': '5a28ca76bab17220af83687dd4505ffd6ced395aeca6329560daaf8945af2764'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'propose a similar scheme for GPUs:! because the sections are much smaller than a cell on the GPU, they use a per-cell locking mechanism to prevent simultaneous updates of a cell.',\n",
       "  'text_sha256': 'da84a56e5f4b986c9f7ea62f4e05526fe4f075d4da11eff1d86e108a4b9c060e'},\n",
       " {'text': 'Neither approach has been extensively checked for correct model behavior.',\n",
       "  'text_sha256': 'bc5e44fd4bbb2fd91e3ce4c2e926009d63e2b8a8d34894ff9fad5184781d91da'},\n",
       " {'text': 'To our knowledge, no publicly available implementation of subsection-based parallelisation exists for the CPM.',\n",
       "  'text_sha256': 'c1709ae115f6ae7a1cde61469ad819e0cf56ee3a679b303ed2bd174110e8163a'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'The software Morpheus ! includes a non-subsection-based Central Processing Unit (CPU) parallelisation strategy; this method is less suited to fully utilise parallel accelerator hardware such as GPUs.',\n",
       "  'text_sha256': 'e1311db60d5aa2522ad8cd24c8874d903429d8eace54c816d1cc78646f9ddc2e'},\n",
       " {'text': 'In this paper, we aim to make several contributions: First, we introduce a highly optimized parallelized implementation of the CPM using several optimization strategies not previously explored in the literature.',\n",
       "  'text_sha256': 'c9bfefc7f29b6e4b74c458bf149fd04fe9509b5f4da2eb1dc31230049856a233'},\n",
       " {'text': 'Second, we conduct an empirical analysis of potential pitfalls in parallelising a model dependent on serial updates of single locations and show that our approach mitigates these issues while preserving performance benefits.',\n",
       "  'text_sha256': 'e5d30102eec2e91fdeab0bdd28751cf1244fff9a0f073ffc5f09548994dfc67a'},\n",
       " {'text': 'Third, we demonstrate that our approach retains the behavior of cell motility simulations that are particularly prone to changed model kinetics upon parallelisation.',\n",
       "  'text_sha256': 'ff2f10feaaa9a88008149cafb5a9f872ff5b1ea444f90c410bec28638776a7a4'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': 'And finally, we build two large 3D image-based simulations of T cell migration in liver and lymph nodes, demonstrating how complex tissue environments can be analysed at a larger scale and realism than was possible before.',\n",
       "  'text_sha256': '29b77947137ab13a8b30432f6876d5f85bfcda62974ea5c8cfa345e944459252'},\n",
       " {'text': 'Serial Cellular Potts Models are limited to sub-millimetre scale The CPMs serial update schedule  only one location at a time is updated on a grid  is central to its behavior, and thus not trivially changed.',\n",
       "  'text_sha256': 'ad6d81b2c6c1a54988393b7c7fb43663318a76f6475076842145216daabc322c'},\n",
       " {'text': 'We therefore first aimed to quantify the computational limitations for models based on the serial CPM: how big can our computational experiments be, and what timescale can they span, without making fundamental changes to the CPM?',\n",
       "  'text_sha256': '1b8cbbcbc1fdbd5707f204df5876707be423109a42d8370469d2b906e74e3f03'},\n",
       " {'text': 'We set up 3D CPM tissue experiments at different spatial and temporal scales and measured the required wall-clock time.',\n",
       "  'text_sha256': '05efc636afbdfb19dd78fd0d25ba4890b9985bf19ed52aa9263aec5fa52c8463'},\n",
       " {'entities': [{'entity_id': 'highly',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'highly'},\n",
       "   {'entity_id': 'highly motile cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'highly motile cells'}],\n",
       "  'text': 'We choose spatial and temporal simulation resolutions that allow reasonably accurate representations of highly motile cells within tissues.',\n",
       "  'text_sha256': 'a529526f206f6e72f7191e0f645df299d74fa876de2165944862ace113c7ef74'},\n",
       " {'entities': [{'entity_id': 'Lymphocytes',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Lymphocytes'},\n",
       "   {'entity_id': 'nucleated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'nucleated cells'},\n",
       "   {'entity_id': 'B cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'B cells'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'Lymphocytes are among the smallest nucleated cells in the mammalian body at 8-10um diameter for B cells and T cells.',\n",
       "  'text_sha256': '83e5011ebc9e7002a2f1b2f95115f4527cfa17f2f0cd6deece57c8eff45e8a0e'},\n",
       " {'text': 'At speeds of up to a few body lo | 30 values/param.',\n",
       "  'text_sha256': 'f92381b5fe1474e55df48a02d9d0414f4a925c2ad62a48816759fdc65f0ae637'},\n",
       " {'text': '2  1wk 7 5 c= a = , - [sj x e Pa  jw || 8 ow 6 cS ihr 4_- 2 L | 2 OS Tmin 10 values/param.',\n",
       "  'text_sha256': '6305fccbeb50c2fd356f1c2e096089bc564879ec0a78c56a3e5498a899560fa1'},\n",
       " {'text': 'polarisation strength replicates 1 2 3 4 5 sembles.',\n",
       "  'text_sha256': 'a5f73c80ccb0a6ecc3d8a89ebefb0285d9365c01603784ec00d01a10a7b9623e'},\n",
       " {'text': '(A) Computational cost of simulating a 256m* volume (1m per voxel) for 10 simulated minutes (1s per simulation step).',\n",
       "  'text_sha256': '1bd50ba9033ca882016c82f02e45c2e34ade65711bdecc76318ec116e118bf55'},\n",
       " {'text': 'With a high-performance CPU implementation, this takes 1.2 hours of wall-clock time.',\n",
       "  'text_sha256': '26f523ee1072b4d8ba0f6cec522ff5faa14a827bcc666ab735ebb507b5fed639'},\n",
       " {'text': 'Extrapolation to a full day of simulated time for a 1mm?',\n",
       "  'text_sha256': 'e8030b240c0a3bf7c8f7da854ac71678a461fba9ae108ec3b5fed5069b79e309'},\n",
       " {'text': 'simulation estimates 1.4 years of wall-clock time.',\n",
       "  'text_sha256': '8036bfd78a28a1b6d4c071a081bdb49f0dafad2e9f0b29d109eb0d4f1694a82c'},\n",
       " {'text': 'Time measurements (black) are averages of 5 simulations, projections are based on inear fits of simulations > 64m* in log-log space (grey).',\n",
       "  'text_sha256': 'cfcf44962ad7a4ec353be660f576a74f33120fb70f55b6475f4946491c9fac46'},\n",
       " {'entities': [{'entity_id': 'CPMs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'CPMs'}],\n",
       "  'text': '(B) Building CPMs typically requires tuning parameters.',\n",
       "  'text_sha256': '99351f8da6afdbeb5db37a51b56ae1097c13681d2da4a2f886c5b296e130b8a2'},\n",
       " {'entities': [{'entity_id': 'motile cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'motile cells'},\n",
       "   {'entity_id': 'Act',\n",
       "    'other_ids': ['NCBIGene:12',\n",
       "     '71',\n",
       "     '72',\n",
       "     '9457',\n",
       "     '345651',\n",
       "     '11332',\n",
       "     '440915',\n",
       "     '641455'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Act'}],\n",
       "  'text': 'Here we perform a multidimensional grid search for parameters that yield motile cells using the Act extension of the CPM; a modest single parameter grid search where 10 values are tried takes 5.9 seconds; larger grid searches take exponentially more time as function of parameters searched, with a five parameter search with 30 values is extrapolated to 1.1 years of wall-clock time.',\n",
       "  'text_sha256': '7169513414939bcc8d30caa57e70db4f1bd0a179eaa3f91291cea1d9df21de9d'},\n",
       " {'text': 'Time measurements (black) are averaged over 5 simulations, projections fitted on linear trends in linear-log space.',\n",
       "  'text_sha256': 'f5ca8702e0960be91d14effefa8a229eaa6fe3bf2b559f45b82764586903ae80'},\n",
       " {'text': 'CPM parameters are shown in Table 1 and Table 2. engths per minute, they are also among the fastest movers within tissues.',\n",
       "  'text_sha256': '5bb613246fb275f080847334762d3b84223bb19d0b4000e3c0efe4db15a8e950'},\n",
       " {'text': 'We chose lum?',\n",
       "  'text_sha256': '900361877d85944f2139125855d806474f3024b3d08e3f66281d75769b496358'},\n",
       " {'entities': [{'entity_id': 'murine',\n",
       "    'other_ids': ['NCBITaxon:10090'],\n",
       "    'entity_type': 'species',\n",
       "    'entity': 'murine'},\n",
       "   {'entity_id': 'tumor',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'tumor'},\n",
       "   {'entity_id': 'avascular',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'avascular'},\n",
       "   {'entity_id': 'avascular tumor spheroid',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'avascular tumor spheroid'}],\n",
       "  'text': 'per voxel and second per simulation step, or Monte Carlo Step (MCS), as our simulation resolutions, corresponding to Beltman et al.9 To create realistically large models of, for example, a complete murine Lymph Node (LN), or scale a tumor model to the size of an avascular tumor spheroid,? experiments at mm scale would be needed.',\n",
       "  'text_sha256': 'da2d7061082f53d7d5f85348fd5af1a2110e720a49eafbbc04d1730c1b506416'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': 'Certain organ scale dynamics require timescales of at least a day, such as T cell transit through LNs.',\n",
       "  'text_sha256': '28e24e4ff3e48ac31b89dcee2a04ca285268def04976f95101f97adc337ef59c'},\n",
       " {'text': '?4 More modest simulation experiments at 256um?',\n",
       "  'text_sha256': 'f7c7034ca2b5e593fffd56ef0ddbe91440c6fe4e0a4a0708e7ecf99a39f88b41'},\n",
       " {'entities': [{'entity_id': 'CPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPU'}],\n",
       "  'text': '(Figure 1A) at 10 simulated minutes require ~1h of wall-clock time on a highly optimized CPU implementation of the CPM (Methods).',\n",
       "  'text_sha256': '218227b76d90241833bb931129625621ebb2017261d34ee875becd03af776468'},\n",
       " {'entities': [{'entity_id': 'motile cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'motile cells'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'},\n",
       "   {'entity_id': 'cancer',\n",
       "    'other_ids': ['mesh:D009369'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'cancer'}],\n",
       "  'text': 'Such experiments are sufficient to see minimal movement of motile cells such as T cells and can capture organ compartments such as a LN paracortical T cell zone, or a micrometastasis for cancer modelling.',\n",
       "  'text_sha256': '486fd569275f6ac9297274271316fc34e1df45bd5db9583a2f0b1bbca2b5f33b'},\n",
       " {'text': 'For mm?',\n",
       "  'text_sha256': '8e866191c42130ee49ed68ce162126ecf7583173cc643d9102b2ff660260bb88'},\n",
       " {'entities': [{'entity_id': 'patient',\n",
       "    'other_ids': ['NCBITaxon:9606'],\n",
       "    'entity_type': 'species',\n",
       "    'entity': 'patient'}],\n",
       "  'text': 'scale, one-day experiments however, researchers would have to wait at least a year  making them impractical for all but the most patient of researchers.',\n",
       "  'text_sha256': '48a39ddd70f30b03a1ffcd6a5a40ae5d05464a6f775d298c2cfda76c347e72a9'},\n",
       " {'text': 'Beyond the need or experimental replicates, simulations can have parameters for which associated model behavior is non- obvious, requiring parameterized simulation ensembles.',\n",
       "  'text_sha256': '182fc38b74e5c216aff04c32b77cc89f44b40dd38332102ac3baa0b19d29fc4a'},\n",
       " {'entities': [{'entity_id': 'actin',\n",
       "    'other_ids': ['NCBIGene:148709', '402221'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'actin'}],\n",
       "  'text': 'In Figure 1B, we explore a cell motility extension to the CPM that coarsely models actin dynamics. A single two-dimension simulation is relatively cheap, but when exploring even a modest number of parameters the execution time goes up exponentially.',\n",
       "  'text_sha256': 'b40fd225e3166c2af1ef2deade90a39dcdf6bd21cfd56f5bf4a67566d2bc576e'},\n",
       " {'text': 'Thus, serial 3D CPM simulations at a scale of 1 mm are currently not feasible, and even 100um?',\n",
       "  'text_sha256': '0a6afeb026495bff0194f853c40a9749165e015c524c638de0d5630902cb7167'},\n",
       " {'text': 'scale simulations were extremely challenging in our experiments.',\n",
       "  'text_sha256': '66869d08fe1463c70eef2861532c84f380e7dc5070e7bf5e66447b14a6be92c2'},\n",
       " {'text': 'While ensemble runs can be accelerated with parallel simulations without changing the CPM, the high computational costs of running large ensembles urther motivates the need for a faster CPM.',\n",
       "  'text_sha256': '1986ffb681c2cfac0a1d9b4fc87f61a44b727c1f63b99ad5284f4b8edac1e583'},\n",
       " {'text': 'Designing a parallel execution scheme for the CPM To make large scale simulations practical, we propose a parallelisation scheme for the CPM that uses highly parallel GPU hardware.',\n",
       "  'text_sha256': 'e537f29db2769e0a23c64995668709919880e366b869ed464f5e00520a98b006'},\n",
       " {'text': 'A natural way divide the simulation grid in support up to thousands of 0 parallelise on-lattice models, such as the CPM, is to er sections; each section runs a local simulation in parallel.',\n",
       "  'text_sha256': '44d175e85df38d3ffaf31ed020eb37efbeb48b0c4f0bffeec2b2a76829831d80'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'As GPUs ations in parallel, these sections will be quite small.',\n",
       "  'text_sha256': 'ea4c79543b2f80e3f0c092bd408d8940e1d6dd7b51f88eaf6caaaff353b0e9dd'},\n",
       " {'entities': [{'entity_id': 'simulated',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'simulated'},\n",
       "   {'entity_id': 'simulated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'simulated cells'}],\n",
       "  'text': 'The CPM also updates information outside of the grid, such as sizes of simulated cells.',\n",
       "  'text_sha256': 'bd4bd7037387ff290f2476b2eb6d484b9919f3c42a809c020f8deed560a85ae5'},\n",
       " {'text': 'We first consider three issues when designing our parallelisation scheme: adjacent sections with active simulation interfering with each other during updates o: simultaneously, and stale in: updated by another simulati A CPM simulation proceeds copying the positions state to a ne simulation running, we run t. boundary.',\n",
       "  'text_sha256': 'f39110b6f88cc142f4d8fb2d059cb3fa6027e6dbe3adbe959fe0ebba796a24af'},\n",
       " {'text': 'Such race condi standard problem in paralle to be active at the same time.',\n",
       "  'text_sha256': 'c73f7b980649e8c81b6ca9934d537fe343de7f71ead8e702167e726fed68d429'},\n",
       " {'entities': [{'entity_id': 'simulated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'simulated cells'}],\n",
       "  'text': 'ion being used by one paralle hoosing a random position on k of simulations racing to wri run the risk of information ge m, with a straight- The CPM also tracks information involving more than one gri simulated cells.',\n",
       "  'text_sha256': 'fbee3275c1a1d54c08a64c5d521111418e750852764b5da28fd229f71d9338f8'},\n",
       " {'entities': [{'entity_id': 'single cell',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'single cell'}],\n",
       "  'text': 'We face the same race condition problem here, updates touch a single cell (Figure 2B, left column).',\n",
       "  'text_sha256': '7b57ee9ae191df7038c3701e41c7fb1bf7da86dce4396c762b956753ef9d777e'},\n",
       " {'text': 'small active subsections, making such concurrent up scheme is proposed, that rejects all but the first concurrent update attempt on a varia middle column).',\n",
       "  'text_sha256': 'e737ec1f44cc4c9e42c43a3eabdcad4cb51c64cf1eec4ac940c39a3e1a04fcb4'},\n",
       " {'text': 'We instead opt to use atomic operations (Figure 2B, right column), which equa!',\n",
       "  'text_sha256': 'e56e61049cf1e8de6affb62f0ebd397cd2ad808e7de5759a7b6ac49587bf824f'},\n",
       " {'text': 'prevents update information from getting lost, as the hardware will eventually integrate all updates designed an additional scheme that sti CPMs, global information travels instantaneously: as only one grid position is w hout incorporating information of the overwritten update.',\n",
       "  'text_sha256': 'a196575c5d08e6920a3425d50532c06a0486c02b4eeeb86065f572a131f1bdc2'},\n",
       " {'text': 'This g updates (Figure 2A).',\n",
       "  'text_sha256': 'ab6ee61a32f88ce22d1aaf303d235f043dbd8c0ec712d755855798f6dcdfbd8f'},\n",
       " {'text': 'Regions cycle perio ates more likely.',\n",
       "  'text_sha256': '9a18c54f5489a93f27a6c8075cf2c75425f8d188f38a34b67e4c51013a19f5a8'},\n",
       " {'text': 'In Tapia et a ; non-grid information such as cell sizes being updated simulation after said information got the simulation grid, and stochastically ighbour.',\n",
       "  'text_sha256': '15b08c38d40ca020184e5d9a4df84d11cdc766a6c09aa354987500914b3359a4'},\n",
       " {'text': 'If adjacent subsections of the grid each have an active parallel e to the same location on the sections ting lost, as one section overwri ion: never allow two adjacent sections rn of sections is only ever active; active location, such as the current sizes as it is possible that multiple paral rallelisation requires a high number 1.18 a per-cell locking dated at a time, date attempt has access to the results of all previous updates.',\n",
       "  'text_sha256': 'e53eb92f6bb60523f6d60a12d309724022aa6dadf04ef7e28a89945dc3965b18'},\n",
       " {'text': 'This instantaneous information access is relaxed in our proposed scheme: when information such as cell size is accessed, out-of-date due to still in-flight updates to the accessed information.',\n",
       "  'text_sha256': '34320fe83b2d2e23c7374a5c6eca6451f6875e879a5d12c225f7d0608a4df41b'},\n",
       " {'text': 'The cell-loc!',\n",
       "  'text_sha256': '8b7daa4edbf0407796f8a1edee6cef1b766e6ffe5269010bdaec443d99353eed'},\n",
       " {'text': 't al.',\n",
       "  'text_sha256': '87dc97af4a31d8750ef9800e3806f5b5185cc99f29fed9fae8a4b3efbbcd8f2a'},\n",
       " {'text': 'prevents this from happening, by outright rejecting any updates that cou information; we explored a less strict scheme, reasoning that the stochastic CPM copy attempts erate out-of-date information.',\n",
       "  'text_sha256': '3fe08783d3065590e5fa8849826a4fc68d881b38366b8b907087dc67abff8e8e'},\n",
       " {'text': 'For CPM extensions that do not tolerate their global variables ightly out sync with each other (e.g.',\n",
       "  'text_sha256': '3daf1fb9da6abfd1f4d8d68330cc665414cb8f7002b6f58ec62337ab4dc38836'},\n",
       " {'text': 'variables used in center-of-mass centroid tracking) we 1 will keep variables in syne with each other, and sits on top of our d have used out- atomics based scheme (Figure 2C, Methods).',\n",
       "  'text_sha256': '46aae3bc61626bd8c39f335288b1d8abd82c4fbe5f3915393eaa68f2ca4c27fc'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'},\n",
       "   {'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'Otherwise, due to atomic updating, our scheme still gives strong guarantees on tracked data being correct  for example, cell sizes will be tracked correctly  and we cache local da F our scheme on rea GPUs, memory access almost always prevents easy use of a GPUs caching hierarchy.',\n",
       "  'text_sha256': 'ba691953cd690239e1d1770283525f049095cb7cbd08551c25eb9efae95b40a6'},\n",
       " {'text': '! To maximize performan scheme that carefully operates in loc hardware requires consideration o: y allows us to explore the effects of allowing the use of slightly outdated information.',\n",
       "  'text_sha256': 'c58011eec51b7a307aa5e48f6efb7d2300a923738e4557cc2033bfe1cff338fd'},\n",
       " {'entities': [{'entity_id': 'CPMs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPMs'}],\n",
       "  'text': 'CPMs random access pattern ce, we designed a data transfer f several technical details.',\n",
       "  'text_sha256': 'e51b4fa547c52b2c3758c8881a0697fedf42ef7557db059fdbd22f7fef1c3112'},\n",
       " {'text': 'k-step with checkerboard work splitting (Figure 2D, Methods): a needed by the regions that are becoming active, an tions to fully utilise overlap memory transfer and GPU hardware.',\n",
       "  'text_sha256': 'ee75c581307ecaf4b2574cd7bcaa32a7f30ddb88de9f42e626f8b82e45f1594c'},\n",
       " {'text': 'To increase simulation volume, and gain further speed ups, we also prototyped a multi-GPU scheme (Figure 2E, Methods).',\n",
       "  'text_sha256': '1682c0cbf41525d0fd40134a1e724c945ffbdb4d230a7578ff380e1d6e519d7b'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': 'In short, our proposed parallel CPM sc!',\n",
       "  'text_sha256': '2a47c1da17879f9340c072b9cb189e3a4f457693ca09777150c7b8b1a3683e2d'},\n",
       " {'entities': [{'entity_id': 'heme',\n",
       "    'other_ids': ['mesh:D006418'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'heme'}],\n",
       "  'text': 'heme is tailored to make opt code.',\n",
       "  'text_sha256': '5fa575140f70a6beb5e5ab7537e39f84f81cb413a6aa1ab3505323848e77c109'},\n",
       " {'text': 'It still gives strong guarantees on simulation sta imal use of GPUs: hardware specialized e consistency  for example, a global variable tracking a cells size will always equal the number of pixels on the grid occupied by that cell.',\n",
       "  'text_sha256': '2458b97e0b4b0397ccfdc9bba7689bfadc94c899d2adbefff1d085004ad3db06'},\n",
       " {'text': 'Yet, we do all flexibility and further maximise performance gains.',\n",
       "  'text_sha256': 'db6148579622009f062ad1c313276b971b08de487cfe8f615dbf2afd47cf6b02'},\n",
       " {'text': 'ow the use of slightly outdated values of such global variables in copy attempts to gain B race condition mutex locking atomic updates (Tapia et al.)',\n",
       "  'text_sha256': 'ed21c50706216b78a5241c8f6cf4625800133d98eb4fee8e5d63c546ba6fd440'},\n",
       " {'text': '(our method) (lel=12) (lla i= 12 |} | ASP$$ $l new = 13)- - - [le] = 13 &)- - - {le] = +18) nO OO Oo Hii --(=tie---(41=19 reese ew.',\n",
       "  'text_sha256': '7e5dd25023b0a39d03b05416d9863f4538fad9931902ed0e8dfac7c26b465d42'},\n",
       " {'text': '= 13)- - - (lel := 13 )- - - { ?',\n",
       "  'text_sha256': '8b110f0a1c3805f8f105c25b219997d1051cc5b2cad2737f60dc8757bd328b75'},\n",
       " {'text': '| {l*],.4 = 13)- - -(lo = 13 w)- - - | c I*]s:= 4 I#lnou = 5  Xu Yous *= (2-8, 3.3}  Xen?',\n",
       "  'text_sha256': 'cbaa2162edfa6c549cf9aa1db969a1125beb32ffaf370660b86a48b60fccdd78'},\n",
       " {'entities': [{'entity_id': 'Yaun',\n",
       "    'other_ids': ['CHEBI:74879'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Yaun'},\n",
       "   {'entity_id': 'Xeon 2',\n",
       "    'other_ids': ['mesh:D003520'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Xeon 2'}],\n",
       "  'text': 'Voow = (3; 3} (Kumn= 11) (Yaun= 13} (lel= 4) 4 ee [Xeon 2= +4!',\n",
       "  'text_sha256': 'c4d1f1cb9c89077115125f05cd223b04f27bcc496120184d11642161ed8f0f38'},\n",
       " {'entities': [{'entity_id': 'Yeu',\n",
       "    'other_ids': ['CHEBI:74883'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Yeu'},\n",
       "   {'entity_id': 'Be',\n",
       "    'other_ids': ['CHEBI:30503'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Be'},\n",
       "   {'entity_id': 'Xon',\n",
       "    'other_ids': ['CHEBI:49957'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Xon'},\n",
       "   {'entity_id': 'Van',\n",
       "    'other_ids': ['CHEBI:27698'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Van'},\n",
       "   {'entity_id': 'Iel',\n",
       "    'other_ids': ['CHEBI:51468'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Iel'},\n",
       "   {'entity_id': 'Xeun',\n",
       "    'other_ids': ['CHEBI:49956'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Xeun'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Yeu = +2 | lel= +1 | Be] Xon= 15) (Van= 15) (Iel=5) [new = Xeun/@l = 3; Yoow = Yeun/l#] = 3) E_ per GPU region border sync ~-7--- global simulation barrier SS caput sizessyne pu |e] = 23 je| = 29 |e| = 23 |e| = 29 [Je|:= +4 Jel: +14=he!',\n",
       "  'text_sha256': 'ae59771258465c9aaef1ebfcbaba302ebcd99572cd0c9cd6db5055975b6d8672'},\n",
       " {'text': '= +0 |el:=-5 | |e] = 27 |e|=25 |e] = 27 je|=25 GPUs.',\n",
       "  'text_sha256': 'b7af249570f146ead59fef179db3e6fd1cb23da7c992a209a2787cf740533194'},\n",
       " {'text': '(A) We divide simulation space into a subgrid, with alternating regions being active in a checkerboard pattern, preventing update conflicts on the grid.',\n",
       "  'text_sha256': 'a4c14ecfc3a25150f19764f126118d3706cb18374dbefa3d06b10b0cacdb67e2'},\n",
       " {'text': '(B) We implement two strategies to prevent global variable update conflicts (e.g.',\n",
       "  'text_sha256': 'a6324bfd5c2e1eed4eece4cd08b9233e59d79a48504980e1dedd0205976705c8'},\n",
       " {'text': 'cell size tracking): a mutex locking strategy (Tapia et al.',\n",
       "  'text_sha256': '29cbebf0cc93eeaf987deebe5be9fdbcd1e1f77973e76566b167dc01399b97b0'},\n",
       " {'text': \"'8), and a novel atomic operation strategy.\",\n",
       "  'text_sha256': '3d5066b642fa35d7b8727187473b0249142c0209aabccf919fb9e613206df2b4'},\n",
       " {'text': '(C) Combining interleaved storage and atomic operations on adjacent data allows variables to be kept in sync with each other.',\n",
       "  'text_sha256': '45cfae738a33be991ac64006f79f82c6f54fc41daa7bac4959f8d7cee0e62cae'},\n",
       " {'text': '(D) A four-step execution scheme, where active checkerboard areas are transferred to fast shared memory, accelerates copy attempts.',\n",
       "  'text_sha256': 'b9170861d17a161bb54d8b954c5bb35ca7a9b0fe815fef5654574564ae2460e0'},\n",
       " {'text': 'Thread barriers synchronize active threads after transfers.',\n",
       "  'text_sha256': '4c40cce2a3c785ace293b0c2d872b827b88ebc1db9d21c514f19d51090a638db'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': '(E) A multi-GPU execution scheme, where regions of the simulation are assigned to different GPUs.',\n",
       "  'text_sha256': '1b05ef5f5abcbda05219b31c5e5389836ef851463d2b9c514768df5087d69c43'},\n",
       " {'text': 'Borders are synchronized between relevant adjacent regions, and global variables are synchronized through delta tracking.',\n",
       "  'text_sha256': 'dc73cda30f470ab5d0d942380fc45bc0f109a5cbaeeea2034216e065084ef65c'},\n",
       " {'entities': [{'entity_id': 'thread',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'thread'}],\n",
       "  'text': 'A T, : time in region time divergence thread barriers mm : time for one copy attempt - T, 5 checkerboard size thread size e ?',\n",
       "  'text_sha256': 'd3a72e1821604cae9042df7d860d67ad79d9a1a3d1c01a2762aa5d722d71b7fa'},\n",
       " {'entities': [{'entity_id': 'lh __ 0123 45',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'lh __ 0123 45'}],\n",
       "  'text': 'ee : : <@ e |e eee  eco  }  1|0@ : eee ao @ : : ee T, T. : : copy attempts before checherboard switch single update parallel update time since last copy attempt ee@ r ) 4 copy attempts per switch, 1 copy attempt per switch, number of copy attempts 4x4 area/thread, 8x8 area/thread, 2x2 area/checkerboard 4x4 area/checkerboard | | | lh __ 0123 45 0123 45 0123 4 5 copy attempts per location 0 1 2 0 1 2 0 1 2 waiting times between copy attempts (MCS) from racing ahead of others too much, thread barriers are imposed after each update.',\n",
       "  'text_sha256': '6763b3a58e814847902f23bf2773984c40dcaf5679b366bda54eeb5ff881a727'},\n",
       " {'text': '(B) Checkerboard size and switching frequencies determine the distribution of waiting times between consecutive updates of the same locations.',\n",
       "  'text_sha256': '5f37188710dc881c255248c82355cec9038edd11b170e8624776369eed3efe2e'},\n",
       " {'text': '(C) Small checkerboard areas or infrequent switches can alter the waiting time distribution substantially compared to the single-update scheme.',\n",
       "  'text_sha256': '4e55f52b23064c6ca3603293e9eda1683ecba6fe3c5497eebcb6d635d0b21b5d'},\n",
       " {'text': 'Update frequency histograms were generated by simulating one MCS (256 updates) on a 16x16 grid.',\n",
       "  'text_sha256': '1db5e3746102e824eb85821eb1c1ffff30d7ef4b36670ad119d843bdaee85eac'},\n",
       " {'entities': [{'entity_id': 'MCSs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'MCSs'}],\n",
       "  'text': 'Waiting time histograms were simulated on the same grid using 100 simulated MCSs.',\n",
       "  'text_sha256': '843fa39a481fe19b44fa94fe2222725c684ca8e39fd7d5aa5c09fb6548ba9999'},\n",
       " {'text': 'cell boundary type: 60% 1 @- dark-light % of cell boundaries simulation based on Graner et al.!',\n",
       "  'text_sha256': 'cf7f1a1b1a9a5a991e1bb9d2409a83b44caf58cf06af09e6cc6b7ccaf269e12a'},\n",
       " {'text': 'is implemented with two strategies to prevent race conditions for cell size variables.',\n",
       "  'text_sha256': '53f1777cce3f1fe10374b0da1b400fed7fa711afbb52f6689cc7275894f72c9d'},\n",
       " {'text': 'The mutex locking strategy proposed by Tapia et al.',\n",
       "  'text_sha256': '69c4bee1d9c3e478b981b483ba7a5cc2ae1e7c49ebd66c3941f7772e589ca6a4'},\n",
       " {'text': '! leads to severe starvation in our tests.',\n",
       "  'text_sha256': '2838cffb953f48b4237f4400f2876951e1ce3b8eec00c578360ebce1bed3144f'},\n",
       " {'text': 'Using an atomic update approach recovers the desired sorting behavior, closely matching sorting speed of the original CPM.',\n",
       "  'text_sha256': 'c5de6ed5a2fff189dd9cf1a8ef7ed19938040c174d10712a25d8174f49f92e16'},\n",
       " {'text': 'The simulation was run with a 4x4 area per thread and 2x2 checkerboards, but other sizes were tested and did not make a significant difference for this simulation.',\n",
       "  'text_sha256': '695b7a7eeaf2d6efd2685350e8a8a728dd05d3ce7aff0bc7e82279ee20252afa'},\n",
       " {'text': 'Sorting trends of different CPM implementations are averaged over 5 simulations.',\n",
       "  'text_sha256': '6e135630d3a6cd0cef7e87c98186f27e2f10db4b2144fa9ed13e888679a29aa2'},\n",
       " {'entities': [{'entity_id': 'CPMs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPMs'},\n",
       "   {'entity_id': 'CPMs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPMs'}],\n",
       "  'text': 'Simulation parameters can be found Aligning the lattice update statistics of parallel CPMs to serial CPMs Our novel parallel execution scheme guarantees simulation data consistency, but drastically changes the way the simulation lattice is updated.',\n",
       "  'text_sha256': '097e8796fa2cf8309637f51462787f77ebe3f36af8358cdaec15f41d850bb02a'},\n",
       " {'text': 'We investigated statistical properties of copy attempts on the simulation lattice for different lattice positions; specifically, we analysed the relative ordering of copy attempts, frequency of copy attempts, and waiting times between copy attempts.',\n",
       "  'text_sha256': '23b0e8c566fc94db4fdca3675e7f38f25d0f11edd241511c4b6b870e5e6c8b58'},\n",
       " {'text': 'This allowed us to quantify the divergence of parallel lattice update kinetics from the serial CPM.',\n",
       "  'text_sha256': '80384869b8b71acd4a207d91e6800e910835f69d6fd01c70c11c903d06509316'},\n",
       " {'entities': [{'entity_id': 'cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'cells'}],\n",
       "  'text': 'Such divergence could lead to changed simulation behavior in complex models of cells and tissues.',\n",
       "  'text_sha256': '43d32d9285ddecef76c112fae5fe014323a4d02f405bc8e11ea3020562255d55'},\n",
       " {'text': 'Without guardrails, performing independent CPM updates in parallel would allow different parts of a simulated tissue to have been alive for different lengths of simulated time (Figure 3A).',\n",
       "  'text_sha256': '6bdd04ad2715041c3e6f076fab33a05c9545d9ab97e6a7fe65fa81b06ec70ebf'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'Modern GPUs use thread barriers to synchronize parallel threads without strong computational performance implications.',\n",
       "  'text_sha256': '2a1fd7dcff3e2ecbf6382581078a05dc1c01da3a706e41870262f6f833416d5c'},\n",
       " {'entities': [{'entity_id': 'checkerboard region',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'checkerboard region'}],\n",
       "  'text': 'We set thread barriers after ever single copy attempt, so for every currently active checkerboard region there is only a maximum time divergence of one copy attempt between active regions.',\n",
       "  'text_sha256': '36e54dc30f8533959631bf6463a4486cdea3f9c982489c4521395beaea7825af'},\n",
       " {'entities': [{'entity_id': 'active region',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'active region'}],\n",
       "  'text': 'Our parallel execution scheme allows us to freely choose the size of subregions and frequency of active region switching (Figure 3B), without impacting data consistency guarantees.',\n",
       "  'text_sha256': 'a67be9aac4b236018ba729ecb6b95eb045a29c42593a48d0da04e62d5971d2a0'},\n",
       " {'text': 'We measure frequency and waiting times between copy attempts per position on the lattice in the standard, serial implementation of the CPM (Figure 3C, left column).',\n",
       "  'text_sha256': '2fd572ad0c9eb32a6f8022335b2184783e9948b1f83ef1f3749983d0c03b4d24'},\n",
       " {'text': 'For our initial parameter choice, we found that underlying update statistics are drastically changed (Figure 3C, middle column): waiting times between copy attempts became much more predictable due to the much more even distribution of copy attempts compared to the serial CPM.',\n",
       "  'text_sha256': '9ce3d7f63ff6e7ef25b8d369282a8d4e45e13bdbb2116f74d552319ba4e39426'},\n",
       " {'text': 'Using large parallel subsections with frequent active region switching allows non-parallel update statistics to be closely recapitulated in our parallel scheme (Figure 3C, right column).',\n",
       "  'text_sha256': 'aeaecc1a8d538887918f9a117a887ed4847e20e2e73cbb6bef10521a297f0f84'},\n",
       " {'text': 'We thus find that parallelisation parameters have major influence on copy attempt statistics on the lattice.',\n",
       "  'text_sha256': 'cf59ae3c097c80dc387a2cda03ef7967c86784cee38278e0c1795bc22c3b5c57'},\n",
       " {'text': 'By switching active regions of the checkerboard frequently, and avoiding too small areas per thread, we can obtain statistics that closely resemble the serial CPM.',\n",
       "  'text_sha256': '483e2648a222f2bfc17a0488a70044d754f3a9447e45272d077a2de17a8a1b69'},\n",
       " {'text': 'Emergent cell and tissue behavior is maintained in a parallel CPM Our goal is to build parallel CPMs that are much faster than previous implementations but maintain cell and tissue kinetics consistent with existing serial CPMs.',\n",
       "  'text_sha256': '96a5d197934ec6a6e061a8e269ea3f737ceab88f3ac7cce49b73125eef9392fa'},\n",
       " {'text': 'We hypothesized that aligning the waiting times between update attempts of our parallel CPM to the serial version should suffice to achieve such consis- tency.',\n",
       "  'text_sha256': 'ec744ae7c19d051ae86b0e5af177c61d68670640061526d9ba78b53762aa1442'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': 'To investigate this, we ran several simulation scenarios with different CPM variants, measuring emergent behaviors on the cell and tissue levels and comparing these to serial CPM versions.',\n",
       "  'text_sha256': 'b66f4bf93acfaa63d7066913382c2b2fa8bcc977a40738fbaa82931e244abe25'},\n",
       " {'text': 'We first investigated the classic cell sorting simulation:!',\n",
       "  'text_sha256': '35c59626b8ded6fab524128a45f56e8a53afba6e00fbe5a1dcb9b56fe04ac59b'},\n",
       " {'text': \"two mixed populations of cells with differential  A single update 4 copy attempts per switch, = 0.3 03 2 202 5 0.2 5 50.1 & 0.1 & 0.0 8 0.0 8 2000 6000 10000 2000 6000 10000 time (MCS) time (MCS) B  low polarisation strength 2 ee qe 4S ' high polarisation strength ' .\",\n",
       "  'text_sha256': 'ad99743928fd9b207bbdf33e87b9231862de25387c32de484e6eb503beb17683'},\n",
       " {'entities': [{'entity_id': 'Cc 8',\n",
       "    'other_ids': ['NCBIGene:6355'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Cc 8'}],\n",
       "  'text': ') f 4 i 3 08 3 08 3  0.4  0.4  8 0.0 8 0.0 8 200 400 600 800 200 400 600 800 Cc 8 copy attempts per switch, 0.25 0.50 0.75 1.00 0.25 0.50 0.75 1.00 1 copy attempt per switch, 200 400 600 800 1 copy attempt per switch, 0.25 0.50 0.75 1.00 are similar.',\n",
       "  'text_sha256': 'ccfdab49ac955f4fdbec4b337e1b1b72272a4c1cb47a16cc09640abec4c7b9fb'},\n",
       " {'text': '(A) Cells in a circular chemokine gradient simulation move highly regular in a serial CPM (low per-pixel entropy, shaded: 1SD); parallel simulations diverge slowly from reference behavior when copy attempt statistics are preserved (low per-pixel cross-entropy).',\n",
       "  'text_sha256': '5d8e07a2745416bfebd54bdc6786f855f91dc50fc2cf9f2e3c6da5f14e3dbc87'},\n",
       " {'entities': [{'entity_id': 'Act',\n",
       "    'other_ids': ['NCBIGene:12',\n",
       "     '71',\n",
       "     '72',\n",
       "     '9457',\n",
       "     '345651',\n",
       "     '11332',\n",
       "     '440915',\n",
       "     '641455'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Act'}],\n",
       "  'text': '(B) In the Act model of intra-cell polarization, stronger polarization strength leads to higher directional autocorrelation (measured for 750 MCSs at 50 MCS intervals, averaged over 50 simulations, shading: 4 emerges when update waiting time distributions are preserved.',\n",
       "  'text_sha256': '54cff8cf00ddc5284a17e43da31f2eee1f387484fa52cc1078dff124651750e4'},\n",
       " {'text': '(C) t1SD); directional persistence only Phase transition simulation in 3D where synchronous motility emerges at a density around 0.5.',\n",
       "  'text_sha256': 'd8de9bbfbde4abb5bb4929d4feaec33baba50018138def331fbf0dd869bbd922'},\n",
       " {'text': 'Cell order indices after 2000 MCSs are shown (averaged over 5 simulation runs, shading: 4 ransition only occurs when update waiting time statistics are preserved.',\n",
       "  'text_sha256': '8cd044d8712b298b512f8ba0a3cc69cd64902ed2a1fef1892da4bf50dcbca051'},\n",
       " {'text': 'Simulation parameters can be found in Table 1 and Table 2.  adhesion (i.e., cells of the same type have stronger adhesion to each other than cells of different types) sort themselves in two homogeneous regions over time (Figure 4, grey lines).',\n",
       "  'text_sha256': '9d359b307c7b543951b45d2a8d2ffe9f79b8cdd8d5ce8299ab46f41cd29ae31d'},\n",
       " {'text': 'The mixing kinetics can be measured by tracking homogeneous and heterogene: based parallelisation approach proposed by Tapia et a leads to a lack of sorting behavior; the simulation concurrent copy attempt is allowed per cell un scheme allows expected sorting behavior to re remained consistent for all parallelisation paramete: ous cell neighbours over time.!',\n",
       "  'text_sha256': '9d246a9f0503f8b5d57104d0e504e4f4c85d16fb9ff893d8f409a20b52c2c7e5'},\n",
       " {'entities': [{'entity_id': 'mutex 1',\n",
       "    'other_ids': ['NCBIGene:63915'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'mutex 1'}],\n",
       "  'text': 'Strikingly, the mutex 1.18  while still being able to run a simulation  encounters so-called starvation, because only one er this scheme (Figure 4, red lines).',\n",
       "  'text_sha256': 'a78313108f923398d03491f407431541fc4cae73f3c21292bbdbf045bff0a8f8'},\n",
       " {'text': 'Our less restrictive -emerge (Figure 4, black lines).',\n",
       "  'text_sha256': '434ab525830dae2de70d20c6cb0affae79305996804d3c6080f3bab2bc063fd0'},\n",
       " {'text': 'The sor rs we examined (Supplementary Figure $1).',\n",
       "  'text_sha256': 'e7f35c63665a2530e6384dd5e9ee20f30648215fb46abe953bfcf33d4f9fbf74'},\n",
       " {'text': 'We next examined three motility model variants of the CPM, where additional biases are placed on o obtain cells that move persis function of random membrane fluctuations.',\n",
       "  'text_sha256': 'ad196d3570da2135f6cebb4d75a88ce293acdab0f63ce29bc3e8fb57c43a6619'},\n",
       " {'entities': [{'entity_id': 'Bec',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Bec'}],\n",
       "  'text': 'Bec: need to roughly maintain their target size, copy sections for movement to occur.',\n",
       "  'text_sha256': 'a52dd9d98ad18dafdbff1bf99e97fe22b65fae9e7c8b8ee84fe2470021ec59dc'},\n",
       " {'text': 'We therefore s by changes to the waiting times between copy at ikely to succeed when aligned with the underlyi bias is fixed in this model  the field does not change simulations with the same initial conditions closely contains a cell in multiple simulations, where low between simulations, and low cross-entropy indica: with reference serial runs.',\n",
       "  'text_sha256': 'bcd13dbd22cdaee466d28598f8ec29dcceda3af65daf6ca23746ee8b6dd9541f'},\n",
       " {'text': 'In simulations with increased quickly (Figure 5A, middle column); w. rises much slower (Figure 5A, right column).',\n",
       "  'text_sha256': '709038e6a4a2ac3755298b94435a85ef5d83c8686a9077a9151aec15ee9d37a0'},\n",
       " {'text': 'Next, we analysed a model that simulates localize border of a cell makes further copy attempts involving those times between copy attempts are very important in this mo in a small neighbourhood successfully expand at the same time; this is less y model we consider adds a chemotactic vector field to the CPM; copy attempts are more ng vector field (see Methods).',\n",
       "  'text_sha256': 'cb5f166501035b53592751ce0fec3a2e6602662c25fcb0afdaeaf575dbc77982'},\n",
       " {'entities': [{'entity_id': 'motile cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'motile cells'}],\n",
       "  'text': 'Because the directional his alignment using pixel-wise (cross-)entropy: we measure the probability that a pixe ently; in the standard CPM movement only occurs as a ause motile cells have a leading and a retrac attempts need to sufficiently alternate between these cel peculated that movement models might be more affecte over time  cells move in a highly regular fashion, an align even at a pixel level (Figure 5A, left column).',\n",
       "  'text_sha256': '7efdc5a22b7eac2cf96571d94eb09c3f9b1e60e653280209701af0de2ec1b6c8'},\n",
       " {'entities': [{'entity_id': 'cross-entropy cell',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'cross-entropy cell'}],\n",
       "  'text': 'pixel probabilities indicated high agreement es a high level of agreement of parallel simulations ged lattice update statistics, cross-entropy pdate statistics are preserved cross-entropy cell polarization, where recent pixel updates at the pixels more likely to succeed.',\n",
       "  'text_sha256': '172f47f71067000101715ec54a1110d595bdcbae1f13ef13f5169b684fe9ce83'},\n",
       " {'text': 'The waiting el, as cells start to move when many pixels likely when copy attempts are spaced out evenly.',\n",
       "  'text_sha256': '8ead8c87a9c3866de7bf84859f4d37a9dafdd1a9306f0e1718050de20f323a49'},\n",
       " {'entities': [{'entity_id': 'parame cell',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'parame cell'}],\n",
       "  'text': 'A polarization strength parame cell; varying the polarization strength interpolates motility be motility modes (Figure 5B, lef waiting time statistics this full range of behaviors they re-emerge when update sta Finally, we examined a simulatio with this vector, and the vector motility model involves informat ion tracked on the to the simulation grid, placing additional demands on a para!',\n",
       "  'text_sha256': 'b020f3be8c64ac5659e79f9f655d2405a04293feb1ce2a2dbfa7a7ff7e382a21'},\n",
       " {'text': 'density cell collective exhibit Brownian movement, as density is increased;??',\n",
       "  'text_sha256': '4d6829567828def815a5483dc6efd108b0425e6f38a38cfb4e666ff7638f7fc2'},\n",
       " {'text': 'this characteristic phase left column).',\n",
       "  'text_sha256': 'bd9c20d78afe6197356a555a57918014acae947a912791c7f51d4eddce17ad84'},\n",
       " {'entities': [{'entity_id': 'preferred',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'preferred'}],\n",
       "  'text': 'In a parallel CPM this phase transition disa, (Figure 5C, middle column), but re-emerges when update Taken together, these results demonstrate that a strict updat n of a tightly packe a preferred movement direction vector; copy attempts for a ce s how much polarization bias is added to a tween Brownian and directionally persistent el simulations with changed update frequency and oes not appear (Figure 5B, middle column), but (Figure 5B, right column).',\n",
       "  'text_sha256': 'a8465ea183c62aa345a270615f784cce3e944c3c0f5b7f55555d1e9024bc6d0d'},\n",
       " {'entities': [{'entity_id': 'immune cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'immune cells'}],\n",
       "  'text': 'lective of immune cells, where each cell has 1 are more likely to succeed when they align ated to align with recent movement of the cell.',\n",
       "  'text_sha256': '22d072b609ee8ffba44c61cbd1b38e757185eef9b0c4709767950670b791672f'},\n",
       " {'text': 'This attice, and additional information tracked external el CPM.',\n",
       "  'text_sha256': 'f127f08648506986aa640410ce9768c4916a134c81e0607a96cbfe167e6105f7'},\n",
       " {'text': 'In a similar simulation model, low transitioning sharply to ordered collective movement ransition is reproduced in the CPM (Figure 5C, pears when update statistics are changed statistics are preserved (Figure 5C, right e scheme preventing parallel updates of the same cell is not necessary for maintaining valid CPM behavior and can even lead to severe starvation where simulations do not progress anymore.',\n",
       "  'text_sha256': 'be9c8d74cf8fe85ea30b2c74369044f4a5c271142a086bc95e80987e68300e8f'},\n",
       " {'text': 'GPU-based CPM parallelisation accelerates models by several orders of mag- Having confirmed that our parallel CPM indeed pro duced correct model behavior in a range of examples, we next investigated the performance improvements resulting from our implementation choices when compared to a serial CPM.',\n",
       "  'text_sha256': '90b90f23575a3232c4d4b3647d08d2bc1dfdd113853ab0724a49c569210e459e'},\n",
       " {'text': '% 1000 2 s 8 g 8 & 19547x13805x Be 3 3 10 3 2 3 < 1 g 24 a & 2 2 eg ol  kernel specialisation ( ae ) 128 5127 20487 8192!',\n",
       "  'text_sha256': '4768ec939d1be834a4a9b89aebf9fe7e85ce3fc467dc40bb00ab7aa00a3c490a'},\n",
       " {'entities': [{'entity_id': 'pre',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'pre'}],\n",
       "  'text': 'implementation shared memory @ @ e@ D E area/thread: @ 4x4 A8x8 mi 16x16 >  03 +0 3 10000 3 9786x 2 S 1000 g = 02 = 0.9 > & 400   g 8 & 2 8 308 3 oo 10   2902x ZO 8 8  1 s G 07 642 256 10242 implementation 1 4 16 64 1 4 16 64 0.0 0.1 0.20.3 dimension updates per switch updates per switch simulation time Performance benefits are obtained by a shared memory scheme, and a pre-compiled kernel specialization scheme, based on a sorting simulation extended to densely fill simulation space (Figure 4, all sorting benchmarks are average of 5 simulations of 100 MCSs, 100 MCS burn-in).',\n",
       "  'text_sha256': 'ab9eb4994d3132a8de8d5f8d00554d1f9c30f5b302a7737135d0499f409d5ebf'},\n",
       " {'text': '(B) Distributing the lattice over 4 GPUs speeds up the simulation further.',\n",
       "  'text_sha256': '3ede5ac6ff3b554d4bd664102426e99771388e59a9c2bd407252889534d83209'},\n",
       " {'text': '(C) Performance of sorting simulations are significantly higher relative to two serial implementations; relative speedup shown for 2048x2048 simulations.',\n",
       "  'text_sha256': 'dc01ba0dcd4d0161dd15a53fe8a48a5a40c7ca72b25e223836dedec2a2d939d2'},\n",
       " {'text': '(D) Performance of 3D cell collective simulations is also improved (Figure 5C, average of 2 simulations of 100 MCSs, 100 MCS burn-in); relative speedup shown for 256 simulations.',\n",
       "  'text_sha256': 'd1accf8f765725206426b268d9744bf4eadb0e7621095698e11cfd567937f5b1'},\n",
       " {'text': '(E) Trade-offs between speed and accuracy for chemokine field simulations (Figure 5A, accuracy: 1 - cross-entropy of parallel ensemble relative to serial ensemble, 250 simulations per ensemble).',\n",
       "  'text_sha256': '99c453a5245c44b393fdfb421c0ae1a79928410f64f1f759829e70de12478c9b'},\n",
       " {'text': 'CPM parameter settings can be found in Table 1 and Table 2.',\n",
       "  'text_sha256': '59d830e7456d450356448d5ff8930f13450087d93d12921f9a6cf549614a0778'},\n",
       " {'text': 'We first examined the performance gains resulting from our use of a shared memory caching scheme (Figure 2D, Methods) using a 2D sorting simulation benchmark (Figure 4).',\n",
       "  'text_sha256': '8a7d195758f364f926216421824ffdf48a13949e0815e7ec6fb1cf85630e3f93'},\n",
       " {'text': 'On its own, shared memory caching achieved only minimal speedup (Figure 6A).',\n",
       "  'text_sha256': 'e8f2fbff654c1cd8b6a569fee5bf0d1e582d655cda3615fdd29ad144c5effd2d'},\n",
       " {'text': 'However, for further performance improvement, we pre-generated different versions of our code based on which CPM extensions were in use (Methods).',\n",
       "  'text_sha256': '2f2ab9920e61d81ae300f22d8a8e7da0627dea0313500852a23dae1d23daf9a5'},\n",
       " {'text': 'Combining these specialized kernels with the caching scheme achieved a significant 10-fold speedup, more than expected from either optimization alone (Figure 6A).',\n",
       "  'text_sha256': '464ca9029c7425c62370f73feafb71ecb45fa35b8d94be68670d8d63447d735d'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'Testing our multi-GPU scheme (Figure 2E, Methods) in the same scenario achieved a further 2.3x speedup by distributing the simulation across four GPUs (Figure 6B).',\n",
       "  'text_sha256': '980c40aac2abb9f7871648c00b71399d4d8f361ccef2c9a38e2cca18588493d2'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Hence, careful consideration of GPU-specific implementation choices resulted in significant performance benefits.',\n",
       "  'text_sha256': '7b2f221877363a4dd880effedeb08b2023e51e05d66821e55f3a76249598d21e'},\n",
       " {'text': 'We next compared our parallel CPM (on a single GPU) to our own optimized CPU implementation (Methods), and to the state-of-the-art software Morpheus.',\n",
       "  'text_sha256': '394e5e62b51d8e174ac29b0cee9e11d0f3f955d0c5e3730094fd406b330f5c17'},\n",
       " {'text': '! Using the same 2D cell sorting bench- mark, we observed a {3,500-fold speedup compared to our own CPU implementation, and a j25,000-fold improvement relative to Morpheus (Figure 6C).',\n",
       "  'text_sha256': 'b5326d462992e39750812ffaa7316351d2f41651e5b33d36209529c4627e5b5a'},\n",
       " {'text': 'On the 3D cell collective simulation (Figure 5C), a {,600-fold speedup was achieved relative to serial CPMs (Figure 6D).',\n",
       "  'text_sha256': '382da07c28d14fa1bee858e2536be00b1da1ea0649a44a73f781c94dc15440c2'},\n",
       " {'text': 'It is important to note that our benchmark hardware did not permit use of shared memory caching for the 3D model; future hardware with more shared memory may improve 3D performance.',\n",
       "  'text_sha256': '526ab96aad8c075d6b86fef649e19036cb1345dcb5473dbb3a6ac3db52f6e23f'},\n",
       " {'text': 'As shown in Figure 3 and Figure 5, checkerboard size and switching frequency influences model behav- ior: larger divisions and more frequent switching leads to behavior concordant with serial CPMs.',\n",
       "  'text_sha256': '8ee60725ff508a8a608bf55f6af90ebcc260b45a35b3d253e2d4779dab73cf67'},\n",
       " {'text': 'These update patterns also interact with caching, further impacting performance.',\n",
       "  'text_sha256': '078f7395acc41154dd601548a28b458001a7c497786d787e9ab5f354c23233e7'},\n",
       " {'entities': [{'entity_id': 'parallelisation parameters',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'parallelisation parameters'}],\n",
       "  'text': 'As the sorting simulation has low sensitivity to parallelisation parameters (Supplementary Figure $1), we instead benchmarked our 2D chemotaxis simulation (Figure 5A).',\n",
       "  'text_sha256': '24726a5d9d7530e5dfb74b2ef77be397271af1abc5841cfed599cba255b7167e'},\n",
       " {'text': 'We found that larger subdivisions indeed slowed simulations down considerably, whereas frequent switching led to a less marked slowdown (Figure 6E).',\n",
       "  'text_sha256': 'be066b5af6deba65853d77efba7d68157c2fbeb41c2b62049b0d9d3cd7393b7a'},\n",
       " {'text': 'These results suggested a trade-off between performance and model accuracy.',\n",
       "  'text_sha256': '7bc96f8390bd7d19cbd8016e01b5214a6de0ed67864115bdb8a0a341d4203e1c'},\n",
       " {'text': 'Hence, our parallel CPM represents a significant speed improvement over even well optimized serial CPMs  such as Morpheus.',\n",
       "  'text_sha256': '9a738ec6c499a78ec84cb4d206d8c5eec3771cae62f1be83793596a6e3d4359e'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'This improvement was to a substantial extent a result of GPU-specific optimizations and led to an overall speedup of several orders of magnitude.',\n",
       "  'text_sha256': 'c119c3f1f9959359df857a11312712683b1ba473bda0ad5facfab81b4dd1f1a4'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': 'Building an image-based model of T cell surveillance of the liver Our parallel CPM was motivated by the need for tissue simulations at realistic scale with realistic cell morphology.',\n",
       "  'text_sha256': 'fa811e5fd53f7e144701c53bd724fcbd08512e2558cc8623dd98691efe60bbea'},\n",
       " {'text': 'While our approach increases engineering complexity of the model, it can also simplify model building, as it enables direct use of imaging data to initialize the simulation environment.',\n",
       "  'text_sha256': '629046b9fe63557f82ddc8d622d4b2fedf2915872b1309a0b7e678a35a1d1d21'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'To determine the feasibility of directly using real-world imaging data in our system, we constructed a model of T cells moving in the sinusoidal regions of the liver.',\n",
       "  'text_sha256': '639c95ac645f73a7c8daa7c15d11ad7bc8886ab80798530162fa57f159581e2a'},\n",
       " {'entities': [{'entity_id': 'Tissue resident memory T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Tissue resident memory T cells'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'Trms'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Trms'},\n",
       "   {'entity_id': 'malaria',\n",
       "    'other_ids': ['mesh:D008288', 'mim:609148'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'malaria'}],\n",
       "  'text': 'Tissue resident memory T cells (Trms) are confined to specific organs, where they sample their envi- ronment for signs of disease+  in particular, liver Trms are key in providing sterilising immunity for malaria.',\n",
       "  'text_sha256': '434c9feb9b7ce29d8e4edf8d980046226f86c6c133e5bb51a44ec28b31d4e5d6'},\n",
       " {'entities': [{'entity_id': 'Trms',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Trms'},\n",
       "   {'entity_id': 'malaria',\n",
       "    'other_ids': ['mesh:D008288', 'mim:609148'],\n",
       "    'entity_type': 'disease',\n",
       "    'entity': 'malaria'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Trms'}],\n",
       "  'text': '? Trms display a distinct motility pattern in the liver, and efficacy of malaria clearing depends on the number of Trms present.',\n",
       "  'text_sha256': '5b439bfb384924b8cb12f3c04a766e09d689a00871dc5a277693631591e4320b'},\n",
       " {'entities': [{'entity_id': 'liver sinusoids',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'liver sinusoids'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trms'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trms'}],\n",
       "  'text': '? We constructed a 3D simulation environment of liver sinusoids  capil- laries that Trms reside in and migrate through  and combined it with a polarized cell motility model to represent the Trms (Figure 7B).',\n",
       "  'text_sha256': '8327115ade4cdc048a729ab1b3d4b12d86b7341efab9be1935ad292c3df8f81f'},\n",
       " {'entities': [{'entity_id': 'Trm',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'Trm'}],\n",
       "  'text': 'Trm motility parameters were derived from a LN model constructed in parallel, and our parameter choice will be explained in more detail with the LN data.',\n",
       "  'text_sha256': '175685ee4533c12f43fe99817f65f220863af5e6afaeaa804a8c910885e997fe'},\n",
       " {'entities': [{'entity_id': 'Trms',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'Trms'}],\n",
       "  'text': 'These parameter settings lead to motility modes also observed in intravital liver data: intermittent movement and pausing of Trms.',\n",
       "  'text_sha256': '61533d2da9fc7064fad62af5091606f7227903911b1e1c216ae4c960542b1a98'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'The goal of our simulation was to predict how much of the liver volume would be explored by a given number of T cells within a set amount of time; this same question was addressed by Fernandez-Ruiz et al.',\n",
       "  'text_sha256': 'cbee0bd166a481ccbee3156d698b69d213cb47a5bd6934772a5f4680e55f5a1c'},\n",
       " {'text': 'using a more abstract mathematical model.',\n",
       "  'text_sha256': 'ffd2441b4259d02c400b1e08fe2a854509c72d2457b839970fcf84eb65a37336'},\n",
       " {'entities': [{'entity_id': 'Trm population',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'Trm population'},\n",
       "   {'entity_id': 'Trm',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trm'},\n",
       "   {'entity_id': 'exploring cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'exploring cells'}],\n",
       "  'text': '? As can be expected, we found that liver exploration by the Trm population increased with the number of exploring cells, and with the amount of exploration time (Figure 7C).',\n",
       "  'text_sha256': '10d9a750f17ffc9d2faf10d41ebf31d2f01934a64333630db6fc10e12634e1a0'},\n",
       " {'entities': [{'entity_id': 'Trm',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'Trm'}],\n",
       "  'text': 'Remarkably, the sinusoid volume explored by an individual cell increased dramatically as the total Trm population was raized from 1,000 to 10,000 cells (Figure 7D).',\n",
       "  'text_sha256': '838bfeacf2f77ff095ba566174172f9f375291442fc493c0210a78179af9cdb1'},\n",
       " {'entities': [{'entity_id': 'stationary cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'stationary cells'}],\n",
       "  'text': 'Further investigation revealed the source of this phase transition: cells in our simulations can be in an unpolarized state, in which they do not move persistently; collisions can trigger polarization of stationary cells.',\n",
       "  'text_sha256': 'fad7c8287a5ed087adf68ee83cdf9ddd136719427eff8102c80e7a774dd4ea45'},\n",
       " {'text': 'Since such collisions occur more frequently at higher densities, this causes a larger fraction of cells to be motile (Figure 7E).',\n",
       "  'text_sha256': '9be2472260cbe055b2e88b26a8d37bbc14a4e9db588a834d600e9976acad4e4d'},\n",
       " {'entities': [{'entity_id': 'Trms',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trms'},\n",
       "   {'entity_id': 'Trm',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trm'},\n",
       "   {'entity_id': 'Trms',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trms'}],\n",
       "  'text': 'Thus, if Trms can help each other to move, a larger Trm population would screen the environment more efficiently than expected from the Notably, the mean speed of Trms peak at ~10.5 um/min (Figure 7F), matching average speeds in our lymph node simulations  but only at densities where the arrest coefficient is close to 0.',\n",
       "  'text_sha256': '157928dacf537c7dce7dc02c86f3c3f43a333425b048acfa3ffb3db6253f381a'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'In two-photon experiments, liver Trm speed is reported to be significantly lower than T cells in lymph nodes; reproduce this in our simulations with moderate cell density, which leads to both a realistic slowdown, and higher arrest coefficient.',\n",
       "  'text_sha256': '0d9c8137dd4d60b8d1ca3073231542f847ab2b62397eae27e70328e1afe00dc9'},\n",
       " {'text': 'The simulated liver Trm speed and arrest coefficient closely matches previ- ously reported values in literature,??',\n",
       "  'text_sha256': 'af0b868b47ca4e3f7c72fcc812c373ca2f37d9bffb0b1e4ce760ea4bf4bcd7b0'},\n",
       " {'text': 'where they were used to quantify observed characteristic patrolling These liver experiments demonstrated that our system allowed us to combine 3D imaging data with a relatively realistic model of polarizing cell motility.',\n",
       "  'text_sha256': '5cd80cd4453901fb97eff1b8f33f433314475385ebcf553084876b65527f886e'},\n",
       " {'entities': [{'entity_id': 'Trm',\n",
       "    'other_ids': ['NCBIGene:55687'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Trm'}],\n",
       "  'text': 'The resulting simulation produced an interesting phase transition of Trm screening efficiency.',\n",
       "  'text_sha256': 'c7cda64d85eaa419d0386e6733fa7eeb6d5d5cdafb06033a4aac7357939059c9'},\n",
       " {'text': 'While we do not claim that such an effect is present in the real liver  more detailed tuning of the motility model to match the characteristics of real sinusoidal Trm cells would be out of scope for this simple proof of concept  this does illustrate how detailed models such as the CPM generate emergent behavior that is not an immediately obvious consequence of the model ingredients and can generate interesting biological hypotheses worthy of further exploration.',\n",
       "  'text_sha256': '8091264e1a5ec1daf45a6367e40cc81acbf1b8f256b5fb7f51a9f45b7f7ae7c4'},\n",
       " {'entities': [{'entity_id': 'T cell collective',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell collective'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'},\n",
       "   {'entity_id': 'presenting cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'presenting cells'}],\n",
       "  'text': 'A T cell collective moves robustly in a realistic, organ-scale model of the lymph In LNs, millions of T cells move in search of antigen-presenting cells,?',\n",
       "  'text_sha256': '0dc8dc6d258b92df132ac157ffd0ab78075fb0b1ed67d5ba973a4fa0d104b89b'},\n",
       " {'entities': [{'entity_id': 'cognate',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'cognate'}],\n",
       "  'text': '! and finding the cognate antigen is a key early step in adaptive immune responses.*!',\n",
       "  'text_sha256': 'ca3311f052740c223bcadab865b40151588e0c152dd1fd31ab0ab290dca5f5f3'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': 'T cell motility in the LN is shaped by the high deformability of the cells?',\n",
       "  'text_sha256': 'ff8bf50bc8027f7013502ae782daf63a362bafef537afea8480217d4dc8470db'},\n",
       " {'entities': [{'entity_id': 'fibroblastic reticular cell network',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'fibroblastic reticular cell network'}],\n",
       "  'text': 'and their interactions with a complex stromal environment, the fibroblastic reticular cell network.',\n",
       "  'text_sha256': '7dc14ea814893cdb4c2f6bac2e5605d7fdf1a634da8038b5d46490448714d177'},\n",
       " {'entities': [{'entity_id': 'moving cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'moving cells'}],\n",
       "  'text': '*3:4 Given the LNs unique structure as an organ consisting mainly of moving cells, it presents a challenging test case for any tissue simulation system.',\n",
       "  'text_sha256': '3be3e29b1a1b75a936d6bcc79acd0c0d06c202f7ce951ac8357b76139988c566'},\n",
       " {'entities': [{'entity_id': 'LN T cell',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'LN T cell'},\n",
       "   {'entity_id': 'LN',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'LN'}],\n",
       "  'text': 'Indeed, many models of LN T cell dynamics already exist, but they were often limited in either cell count and simulation volume, or in realism of cell morphology and LN environment.',\n",
       "  'text_sha256': 'd9ee5c44f1b57b8f7851bc57c60e18483a1f6e6d2307eb9589ba2737aca89ade'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': 'For example, T cell interactions with a synthetic stromal structure have been investigated using 3D CPM simulations of thousands of cells, while collective motility  at a more realistic scale of hundreds of thousands of cells has been investigated with non-morphological agent-based models.',\n",
       "  'text_sha256': 'bbe28e1676583b46632b2713c86f0e892768a224ee9dd301ce169a67b2eddf89'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': '*:3 We were therefore interested in our systems ability to simulate a realistically large collective of millions of realistically shaped T cells.',\n",
       "  'text_sha256': '52a59ecd5cf1f2c6fedb70a5e4dd33eb44ee14e4afeec8f7fa1aa3365480b191'},\n",
       " {'text': 'As in the previous use case, we used microscopy data to initialize  and parameterize  our model.',\n",
       "  'text_sha256': '15b90667f7dbfd299614d5a12535ddbbf8fa1fda056fbbaa6f56322fa9aae2e2'},\n",
       " {'entities': [{'entity_id': 'LN',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'LN'}],\n",
       "  'text': 'We constructed a LN simulation using the work of Beltman et al.9 as a starting point.',\n",
       "  'text_sha256': 'ef664670f0a5b09aeb95ea7e4e07892127c00bf4e07a8e9a39f9f341265c7a85'},\n",
       " {'entities': [{'entity_id': 'Act',\n",
       "    'other_ids': ['NCBIGene:12',\n",
       "     '71',\n",
       "     '72',\n",
       "     '9457',\n",
       "     '345651',\n",
       "     '11332',\n",
       "     '440915',\n",
       "     '641455'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Act'}],\n",
       "  'text': 'To improve realism, we used an imaged LN section to initialize the stromal structure (Figure 8A), we used the Act-CPM extension for cell motility, and we fit the two Act motility parameters directly to intravital imaging data (Figure 8B-E).',\n",
       "  'text_sha256': '7c1ad57ea2ad753809e994a18f47d0fe43bac551cdbf09921f85b2129e6c853a'},\n",
       " {'text': 'When fitting these parameters, we aimed to reproduce the intermittent stop-and-go motion observed in real data whilst also quantitatively approximate the real speed distri- bution.',\n",
       "  'text_sha256': '753f3e18abe2fab257269d197c1e0f3ad8880cfc44195c0ea74a091d3593660a'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'Our system allowed fitting to be performed on realistically large simulations (256um diameter simulation volume, containing 50,135 T cells, 20 minutes simulated time).',\n",
       "  'text_sha256': '84ff13645bee3358ed0c6f583aec64f7b8b761def56f7123139c313f60f48848'},\n",
       " {'entities': [{'entity_id': 'cell collective',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'cell collective'}],\n",
       "  'text': 'Crucially, we did not use periodic boundaries which, in dense simulations, can lead to artificial global synchronisation of motion in the cell collective.',\n",
       "  'text_sha256': '1a9e4e92b5ae3c8699be35801a5af9b9d88967cb7c39d1514d0150073f8386a0'},\n",
       " {'text': 'We used the KL divergence to compare simulated to real speed distributions (Figure 8B).',\n",
       "  'text_sha256': 'a81893f47aa65ac2c5606ae7448cbf631d0a2c59ab5db2b49f060bc1e897de2f'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'Increased adhesion of T cells to the stromal structure made the speed distribution more right-skewed (Figure 8C, top), which was more consistent with real data (Figure 8C bottom).',\n",
       "  'text_sha256': '603628d5254e89671c73ef8b615b1dac2996d328c19159d4b33c85f8a5d3cc61'},\n",
       " {'text': 'For most examined parameter settings, we found the pronounced variations in cell speed along cell trajectories that are also observed in real data (Figure 8D).',\n",
       "  'text_sha256': '2c7fa44a385eff86bf0cf19bf012f0ed4cb1f5fa1781d44f12609d933fb72510'},\n",
       " {'text': 'When we performed grid searches of motility parameters at two different stromal adhesion strengths (Figure 8E), both grid searches had broadly similar fit landscapes, and the overall lowest KL divergence values were very similar.',\n",
       "  'text_sha256': '32c6465bd6c7c32d4f689ed7f80cdaa0f4aacf0aa76459e8f096377fc3a2a4fd'},\n",
       " {'text': 'Qualitatively, we found more right-skewed and bimodal speed distributions in high-adhesion than low-adhesion simulations for a subset of the motility parameters (Supplementary Figure $2).',\n",
       "  'text_sha256': '2dc7929e6492ec503f32fcca349fb18a156c37e47514db98e0c5633952a60b83'},\n",
       " {'text': 'Based on these fitting results, we chose the best fit motility parameters of the high adhesion setting for our further simulations.',\n",
       "  'text_sha256': '46a738f4bfc8f31f08a0f7dbc58153a6841d8d8d99fe6042cbb8b50ab5e88f2c'},\n",
       " {'entities': [{'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': 'We next investigated organ-scale collective T cell motility by scaling up our simulated volume to 1mm diameter, now containing 3,359,752 T cells (Figure 9A).',\n",
       "  'text_sha256': '18e60f06cd54619eb3f7f2a65b4ef5d265611a1cf2555921a9264e26ce4c1a3a'},\n",
       " {'entities': [{'entity_id': 'T cell zone',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell zone'}],\n",
       "  'text': 'While this is not entirely realistic  a real LN has a more complex compartmental structure, while we simulate only a T cell zone  our main intention was to determine the spatial range of collective behavior that can now be investigated with our system.',\n",
       "  'text_sha256': 'a3bef868fb478bc4fa4ac67c8de6b70c1a7da1519d913e233ae5f24a6e21cc28'},\n",
       " {'text': 'We simulated one virtual day in 1.9 days of wall clock time  where serial implementations were projected o have taken a full 1.4 years (Figure 1A).',\n",
       "  'text_sha256': '321ddb77c7f97ea071a8aff8b53bf078f2241b754c5caeed21b0350e607de5d2'},\n",
       " {'text': 'We used a mixing simulation of two differentially labelled populations as an easily observable proxy for collective search (Figure 9B).',\n",
       "  'text_sha256': 'f3038938e7b664d1ea8ed79a01eeea1dc4940970644dd93c09de595b75c1ef6e'},\n",
       " {'text': 'In the simulation, it takes about 1 hour to reach a state where 50% of the cells are next to at least one cell of the other population (Figure 9C).',\n",
       "  'text_sha256': 'd82df2932e45757a0d80c8ce64bc932e6e48613ac99fe1068ccfbe24bec4739c'},\n",
       " {'text': 'While stromal adhesion influences intra-cellular stop-and-go motility, it has minimal influence on collective motion: high and low stromal adhesion simulations have broadly similar mixing speeds.',\n",
       "  'text_sha256': '5bfcd002c2c0f2e87b88712508a6bf6d02c109c69642cc2cf2ce4872d57483e8'},\n",
       " {'text': 'While collective motility is slightly ordered at cell-scale distances, no ordering persists at organ-scale (Figure 9D).',\n",
       "  'text_sha256': 'c4c36b913002ee3d21fdd551976d340a1f2dfe95378ec67f6ba4eea8433ef027'},\n",
       " {'entities': [{'entity_id': 'LN',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'LN'}],\n",
       "  'text': 'In summary, we used an organ-scale LN simulation as a challenging test case: an environment known for complex, large scale collective motility dynamics.',\n",
       "  'text_sha256': '4e9630054ffdce96d15393275f096fdd4f3d162f4c74e49cde3137517a90ba5a'},\n",
       " {'text': 'Our system allowed us to integrate of microscopy data directly into our simulation, and to fit motility parameters directly on two-photon data.',\n",
       "  'text_sha256': 'd57bd2e9952beda4ab7cc3e7e82103b616417c47f73bcdc90d5a4b8dd38f3bae'},\n",
       " {'entities': [{'entity_id': 'fitted cell population',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'fitted cell population'}],\n",
       "  'text': 'We investigated the behavior of the fitted cell population at organ-scale (millions of cells).',\n",
       "  'text_sha256': '85c2ddc512c5ca2cd153b5df0c120de4c412284461bdf81213c933051d2f214f'},\n",
       " {'entities': [{'entity_id': 'large',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'large'},\n",
       "   {'entity_id': 'cell collectives',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'cell collectives'}],\n",
       "  'text': 'We hope that this will inspire further research on large cell collectives  integrating various sources of empirical data, at a high level of realism, and at realistic scale.',\n",
       "  'text_sha256': '2606b63914a1e8def2e6e294b030195b9192670d20b33030b3ce5576889e9e93'},\n",
       " {'text': 'We have implemented a parallel version of the CPM that, in some experiments, was 3 to 4 orders of magnitude faster on GPUs than a state-of-the-art serial CPM.',\n",
       "  'text_sha256': 'afded2fdee08bfbe88e77c311ff67e4d824ac4c266787eecb583c70cb513307d'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'To achieve this, we designed novel schemes for lattice work splitting and synchronization, maintaining global state, efficient GPU memory transfer, and multi-GPU execution.',\n",
       "  'text_sha256': 'ed7399482fd230853c2a99861d079158eefc2a24e40233d02bc444469cba7cb8'},\n",
       " {'text': 'We found that preservation of local lattice update statistics was sufficient to maintain high-level system dynamics, despite removing the sequential nature of the original In our work we aimed to ensure that important properties of CPM simulation trajectories were main- tained.',\n",
       "  'text_sha256': '3bbba0318e5ad5f2b9ca1e38d93aa53be35ce0fe1ff93a6fea235cca833c5ab3'},\n",
       " {'text': 'Some previous CPM parallelisation efforts instead focused on equilibrium state distributions as important quantities.',\n",
       "  'text_sha256': '62ba4cbea371aa546eb6be6a50a11ec18409d415d75df67c3d15ad0128302bfe'},\n",
       " {'text': 'For example, Chen et al.!?',\n",
       "  'text_sha256': '1a19c40e15240cd5541eb38ad30aa1cdaaa2239811746dace0693252c18f3696'},\n",
       " {'text': 'argue that their parallel CPM maintains certain equi- librium properties.',\n",
       "  'text_sha256': '8c3407a3ba100de163ca6489ba5c77baee84005c349fb6956d08c5ccb4bb00c9'},\n",
       " {'text': 'While such arguments are frequently applied in the analysis of energy-based models  in physics, Voss-Bhme et al.',\n",
       "  'text_sha256': '9c435c2c4ff86983059b8d9612b72fbc5397668d7100e3d3687ddf30dce37a2e'},\n",
       " {'text': 'showe While we have shown that the same CPM in our benchmark scenarios, thi parameters that were originally fitte to re-fit their simulation parameters that these are less relevant for the CPM because the Metropolis algorithm used in the CPM drastically alters longtime behavior of the mode range of behaviors is present in our parallel CPM as in a serial s would still need to be carefully checked when re-using simulation on a serial implementation.',\n",
       "  'text_sha256': 'e54763a414d716b0fada6b269258fbbe4297e9ca5f0da87ebaf817a43ce95408'},\n",
       " {'text': 'Specifically, researchers might want when using our system instead of a serial is sensible in general when moving between different implementations of the s as the CPM where implementation details might anyway differ version  advice that subtly.',\n",
       "  'text_sha256': 'd6e0787ab3ab1dfa2599b46613e1c1d3f7f75cd8f112e6445e82f7a2c253106c'},\n",
       " {'text': 'Furthermore, the observed relation between model kinetics and performance (Figure 6E) also suggests it might be fruitful to explore more aggressive parallelisation settings that might yield further speed-up whilst maintaining desired simulated behavior.',\n",
       "  'text_sha256': 'e9d0ece8fcc458c14df98ca177f9144fa43efebeb14fc7bd8c0c1f2a93d907c0'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Through accelerating the CPM on accessible GPU hardware, we have increased the range of feasible simulation sizes  because our scheme fully utilises all GPU cores, simulation size is now mostly limited by the available GPU memory.',\n",
       "  'text_sha256': '2188492b1181bd920833b9d78927909245ea3cede18b48c0b1c40e9a9cb5790c'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'As memory per GPU increases, our scheme wi ti-GPU prototype scheme will allow for further increase of simulation volume.',\n",
       "  'text_sha256': '4346c007aec8f939278916e2502b9ad0a2f93e3ae971f91735629ca6f53d9ab3'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Our to include inter-GPU communication over MPI,* allowing ding block for a GPU cluster-based version of the CPM  her to lcm?',\n",
       "  'text_sha256': 'de977e9f4072abb5c0cc73e259e0df8ff0c30599eb3c1f263c3161bedb05644a'},\n",
       " {'text': 'scale.',\n",
       "  'text_sha256': '83064bdc95f54de4ddea5238cbe26035464f11b5e4a7418527fc32bd8252063c'},\n",
       " {'text': 'multi-GPU scheme could be further extende our system to be us cell shape into acco 14, 15, 35, 36 ed as a fundamental bui ur Imm?',\n",
       "  'text_sha256': 'a073d660e23c83c40fffbc56ca22379cfa082ec15a9a9cb43d87fa766ea52e5f'},\n",
       " {'text': 'simulations fur 1 easily scale to larger ular simulations are often performed using more coarse-grained agent-based model Our high-performance CPM now allows researchers to perform simulations that much more detailed and allow the investigation of new questions that require taking unt, such as migration in complex environments or morphogenesis.',\n",
       "  'text_sha256': 'a5a082d1d804cc39fa14fabb161ba9f456fb2f26426e8571bc75cc8e8fc30fbd'},\n",
       " {'text': 'We hope that ions of liver and LN environments will serve as inspiration for many such spatially resolved tissue and organ simulations in the future.',\n",
       "  'text_sha256': 'a264488539017667129354d9870da6e85eea945603adace728dbcde2839b784e'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'CPM'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'},\n",
       "   {'entity_id': 'E i F cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'E i F cells'}],\n",
       "  'text': 'Similar approaches to the one shown here could be useful for parallelisation of CPM variants or other sequential lattice-based models.  A B mmm 100 yum explored space @ T cells S 10 1000 100000 E i F cells @ 0.6 E so 10 1000 100000 10 1000 100000 a motile state.',\n",
       "  'text_sha256': 'dcaf42980bd9c13dbf0de14a8d9700e1581c23480de9feaa8563858af13f1b57'},\n",
       " {'entities': [{'entity_id': 'liver',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'liver'}],\n",
       "  'text': '(A) A 1024x1024x182 um scan of a liver sinusoid is directly imported into our simulation environment (64ym slice shown).',\n",
       "  'text_sha256': '9d4437de8ddba44c85f062d2d07ff89abc9d10d1a88f9dc23ef4dd2cf4f961c0'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': '(B) Simulated T cells explore liver sinusoids (5000 cell simulation shown).',\n",
       "  'text_sha256': '443181646513fb548777993b0d215da2165054260e475570ed8269cced6ec362'},\n",
       " {'text': '(C) Simulations at different cell densities explore the environment at different speeds (single simulation of 24hr, with 1 min snapshots).',\n",
       "  'text_sha256': '3219c20147f1f38267ad26576a8ae7cf40426a6b22acf8d7734724e8f2017066'},\n",
       " {'text': '(D) An increase of explored volume per cell occurs when cell density rises from 1,000 to 10,000 cells; 100,000 cells fill ~1/3 of sinusoid volume, causing exploration efficiency to saturate.',\n",
       "  'text_sha256': '67b333a8dd27143a2e2ac2e78aa8cce3db822b9133f1f574f01c5dd2ee6dce75'},\n",
       " {'entities': [{'entity_id': 'simulated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'simulated cells'}],\n",
       "  'text': '(E) At low cell densities most simulated cells are in an un-polarized, non-motile state; at higher densities collisions cause cells to transition in to a polarized, motile state (1 simulation of 30 min per density, arrest coefficient: fraction cells ; 2um/min,??',\n",
       "  'text_sha256': 'aae772075004587ca4699d2bfeda05a5c63d45de396230d1c18bd9b4618c14f0'},\n",
       " {'text': 'shaded: Wilson score interval based on the number of 2-minute subtracks).',\n",
       "  'text_sha256': '8a2a8c7d54e3f2d28be4813e779e58df9d25238310512121babfbb9b1e51a9ea'},\n",
       " {'text': 'CPM parameter settings can be found in Table 3.',\n",
       "  'text_sha256': 'a642e47508e06fe6c0c00367e34865bfa17e977be0a9ea7af5a2e6810e4246fd'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'}],\n",
       "  'text': '(F) Average speed of the simulated T cells at different densities.',\n",
       "  'text_sha256': '5014627e6ce815b0b805d8e5241c428f9491f3ff1bb44262f3062333997c779c'},\n",
       " {'text': '1-15 ym 16-30 pm 31-45 im 46-60 pm 61-75 um 76-90 um =: 91-110 um B simufation =e D jowadhesion E  towadhesion low adhesion 3 3 a 5 0 10 20 30 0 10 20 30 SOHm =m 3 speed (\\\\um/min) speed (um/min) high adhesion & high adhesion high adhesion 8 0.40 @ 0.25 ee eh cc 5, = 0.20 a 3 0.00 ge 3 os 10 20 30 adhesion o 4 4.',\n",
       "  'text_sha256': 'a2ca3085a55851ad713eae26938c9bda085111eca9c2fe0a7ce24a54edb1fb9d'},\n",
       " {'text': ' % polarisation decay 10 30 speed (um/min) speed (um/min) data from Mandl et al.?!.',\n",
       "  'text_sha256': '1cbcfbaf15931102e55e632ffcc3664fb49e4ca9aa4511c2050793fc5046cd46'},\n",
       " {'entities': [{'entity_id': '50,185 CPM cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': '50,185 CPM cells'}],\n",
       "  'text': 'to the motion of 50,185 CPM cells moving in a spherical structure (256um diameter).',\n",
       "  'text_sha256': '9fe36390e1cd65636a68bece30dc4e0f88ebbf5f98a1c400cb4a90420a91799b'},\n",
       " {'entities': [{'entity_id': 'fibroblastic',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_line',\n",
       "    'entity': 'fibroblastic'},\n",
       "   {'entity_id': 'fibroblastic reticular cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'fibroblastic reticular cells'},\n",
       "   {'entity_id': 'murine',\n",
       "    'other_ids': ['NCBITaxon:10090'],\n",
       "    'entity_type': 'species',\n",
       "    'entity': 'murine'}],\n",
       "  'text': '(A) A 130x130x110 um* scan of the stromal structure (fibroblastic reticular cells) in the paracortical T cell zone of a murine LN.',\n",
       "  'text_sha256': '15d6999bcea87747d2b68493d2387a39214728ceed2054cf71c3135e115eb411'},\n",
       " {'text': 'The images show average intensity projections of diffe ranges.',\n",
       "  'text_sha256': '1243ce046595471915d9cea5f8ee1b04546535c8119f84c429bcd3b1739f0f2a'},\n",
       " {'text': 'After segmentation and mirroring, these data are used to fill a spherical simulation vo Speeds observed in simulated data are used to construct a probability distribution (upper panel, solid), which we compare to the empirical distribution (upper panel, dashed) using the Kullback-Leibler (KL) divergence metric.',\n",
       "  'text_sha256': '2b9643ab64544d9c3b76edda4c245d6f693e6cb22931414e63c2f42157ef4cf2'},\n",
       " {'text': 'To this end, we bin the speed distributions, compute the divergence in each bin, and sum the values for each bin (lower panel).',\n",
       "  'text_sha256': '97367a86cdfc0bf5436e0cd4f29da298d2a02995aad9292d6d490826e4164431'},\n",
       " {'entities': [{'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'},\n",
       "   {'entity_id': 'stromal cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'stromal cells'},\n",
       "   {'entity_id': 'T cell',\n",
       "    'other_ids': ['CL:0000084'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cell'}],\n",
       "  'text': '(C) Increased adhesion of T cells to stromal cells causes the speed distributions to be more right-skewed (top) and match real data more closely (bot: Increased adhesion can slow T cell paths non-uniformly by creating longer and/or more frequent pauses (red) but does not prevent rapid motion altogether.',\n",
       "  'text_sha256': 'daf5619109b27dc8fa0245070510cae82a859358b415e987ad5781fe805c4be5'},\n",
       " {'text': 'Two representative tracks are shown.',\n",
       "  'text_sha256': 'fd89916dbe3041621dc96b4bfe1f066f41ad2025d8ac70a1c83a90e92320d814'},\n",
       " {'text': '( search of motility parameters for cell polarization reveals which parameter ranges best fit the intravital data.',\n",
       "  'text_sha256': '2550ccd3d804c18d73c0db99fe7d366f6b78776a00b6814af941a2906d2bd003'},\n",
       " {'text': 'Both low and high adhesion simulations have similar fit landscapes; the best fit is high adhesion simulation, although the difference to the low-adhesion best fit is small (1 simulation per parameter combination, best fit: red box/red line).',\n",
       "  'text_sha256': 'aaa580c57a572ba98348fecf26855414d13287cb6c8ae644cd08a94f4ac6c420'},\n",
       " {'entities': [{'entity_id': 'simulated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'simulated cells'},\n",
       "   {'entity_id': 'T cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'T cells'},\n",
       "   {'entity_id': 'stromal cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'stromal cells'}],\n",
       "  'text': 'CPM parameter settings can be found in  3,359,752 simulated cells O mins 40 mins 80 mins 120 mins D 06 5 0.6 0 30 60 90 120 30 100 300 1000 time (in min) scale (in um) high adhesion low adhesion LN simulation at 1mm diameter containing 3,359,752 T cells and stromal cells (grey; Figure 8A).',\n",
       "  'text_sha256': '5b371a9c2f70bfcfb855f4a275567f153c14669706b2b35bbfe1045f6ec29147'},\n",
       " {'entities': [{'entity_id': 'simulated cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'simulated cells'},\n",
       "   {'entity_id': 'red',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'red'},\n",
       "   {'entity_id': 'green cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'green cells'}],\n",
       "  'text': 'For clearer visualization, only 10% of the simulated cells are shown, and red and green cells are equally param- eterized although coloured differently.',\n",
       "  'text_sha256': 'f9b90cbfd6277981c0e4a017041cdc91368be0262411415df9d4070ace783eb7'},\n",
       " {'text': '(B) The simulation is initialized with two equally parameterized populations each starting at one half of the sphere.',\n",
       "  'text_sha256': '98870d81e81c2589c971a9972c956a243c0b5c7ea3733cc574314645bee86fdc'},\n",
       " {'text': 'Mixing is quantified by tracking the fraction of cells having at least one neighbour of the other population.',\n",
       "  'text_sha256': 'ba3d75527cd7754e85529fcea44a1e7e88c58ce903c9a412c5a03c56dbdb1b6f'},\n",
       " {'text': '(C) Global mixing is slower in the high-adhesion simulation (red) than at low adhesion (black), although the difference is minor.',\n",
       "  'text_sha256': '21bad2d8aa4cd0247b90c2417e91134886cca2148eaf6a043dc33311119c70e9'},\n",
       " {'text': 'Both simulations reach 50% mixing in ~1h.',\n",
       "  'text_sha256': '8fb26989111243ecb8b3f6db42eef524a769892da6eb1b0abe4b2bf449415fd7'},\n",
       " {'text': '(D) Cells do not synchronize their directionality much at longer spatial scales, and this absence of long-range synchronization is the same for low and high adhesion.',\n",
       "  'text_sha256': '7bf097e0252674ae716951798cd88f70ab9d6780c62c3bf153399529f53dc295'},\n",
       " {'text': 'CPM parameter set- tings can be found in Table 3.',\n",
       "  'text_sha256': '353eecd29f200d928bdf69417c0486ec9b81f8705de1fe1dee0d62c1747df4d9'},\n",
       " {'text': 'Table 1: Model settings for 2D simulations.',\n",
       "  'text_sha256': '19c0d84277a2b5fa665d417871e214f5468e1040fc6f24c71e4457c0c95152c0'},\n",
       " {'text': 'For all 2D simulations we use Viarget = 2500, Atarget = 600, Aperimeter = 2; Jcett,ecm = 10 and T = 10.',\n",
       "  'text_sha256': 'baaff089278145d4a643ec2abb7273fc50adb27b19ff53cf633fd05247626755'},\n",
       " {'text': 'The vector field in the 2D circular chemotaxis simulations (Figure 5A) consists of vectors at a 90 degree angle to radial vectors from points in the field to the simulation center, with all vectors in the field normalized to length 1.',\n",
       "  'text_sha256': '4ec28d56c552ca7fcfd22dad1d2bb695ff4d86b8754f8042b6641928f1079b6c'},\n",
       " {'text': '| Avotume | Aaet | Maxact | Achemotaxis Fig 1B 20, 30, 40 | 200, 250, 300, 350, 400 50  Fig 5A, Fig 6E | 20 =  10 Fig 5B 20 50-800, in increments of 50 | 50  Table 2: Model settings for 3D simulations.',\n",
       "  'text_sha256': 'e86ab6cbd35a7dbab12e8f474989122bd68cdb68969b45bc1cccf4c27b251409'},\n",
       " {'text': 'For all simulations Viargee = 150, Avolume = 25, Atarget = 1500, Aperimeter = 0.2, Scett,ce = 10 and T = 20.',\n",
       "  'text_sha256': '3ff63ab34b86424d669bf5ae2124ade3a24a9844ddde1379d0d316d3927d50d9'},\n",
       " {'text': 'The time resolution is 1 second per Monte | uw _| At | Apersistence | Jcell,BCM Fig 1A  f|[ 10 Fig 5C, Fig 6D | 0.95 | 15 | 40 0 Table 3: Model settings for liver and lymph node simulations.',\n",
       "  'text_sha256': '610ba8319042493f29af2cffd470ee5b5b51cc2dad0c99a50a46d0935f61747d'},\n",
       " {'text': 'For all simulations Viarge, = 150, Avolume = 25, Atarget = 1500, Aperimeter = 0-2, Jcett,cett = 10, Jeet,ncm = 0, and T = 20.',\n",
       "  'text_sha256': '065c34f27c273979d38e80869a0d851edd8e47ff171c12b59eb4facdc928d9b2'},\n",
       " {'text': 'The time resolution is 1 second per Monte Carlo step.',\n",
       "  'text_sha256': 'ab5b3b00cf1eaf715d8d2c4f12529cf0ced2791116ae3cfe2a9f526b17ad6293'},\n",
       " {'text': '| Aact | Maxact Teell, FRC Fig 7 2600 10 0 Fig 8B | 2600 10 0 Fig 8C | 2600 10 -400, -300, -200, -100, 0 Fig 8D | 2600 10 -400, 10 Fig 8E | 200-3000, in increments of 200 | 10, 15, 20, 25, 30, 40, 50, 60, 70 | -3000, 10 Fig 9C | 2600 10 -800, 0 Fig 9D | 2600 10 -800 This research was supported by Young Investigator Grant 10620 from KWF Kankerbestrijding and HFSP program grant RGP0053/2020 (to JT).',\n",
       "  'text_sha256': 'b8387cece666b1ead46308f8109cef45675d54f72f2fe63a072438fe479a6e8a'},\n",
       " {'text': 'JT was also supported by NWO Vidi grant VI.Vidi.192.084.',\n",
       "  'text_sha256': 'f5f8c7d47f5bf558b922e346114d9561327094012abdd08b2dba082cff355498'},\n",
       " {'text': 'We thank Franka Buytenhuijs for giving valuable feedback on this manuscript, and we thank L.W.A.',\n",
       "  'text_sha256': '44a8a262b750a1cacb00f852b097b0b55d3e60ee34a960224a5d150537899a2e'},\n",
       " {'text': 'Stuurman and Pedro Aroca Lara for testing early versions of this simulation system.',\n",
       "  'text_sha256': 'c697a27b878e09d48a470a3d2ab875e30afbb8e7e0236a936608dcc254eb2d75'},\n",
       " {'text': 'We also gratefully acknowledge the support of NVIDIA Corporation with the donation of the Titan V GPU used for this research.',\n",
       "  'text_sha256': '301d50493c6cf30050889430b7cb77d1823737797f1d4bf3283db2f88bebabe2'},\n",
       " {'text': 'The funders had no role in design, execution, and interpretation of the research reported in SS and JT conceived the study.',\n",
       "  'text_sha256': '30280baf0c279b9c0c9eefc26f4440e22d9b2a3eecf4857207a1be87dcb6c220'},\n",
       " {'text': 'SS designed the parallel CPM algorithms.',\n",
       "  'text_sha256': 'f769fe96a1bd95558abaeee4fac08f3c5fa059afd171f3e63c34c246f85aae8a'},\n",
       " {'text': 'SS performed the research and analysed the data.',\n",
       "  'text_sha256': '639e91d8bb6113c3d0d1fdc26783469c7752e53f8bb38cb88702884b5a0ac22e'},\n",
       " {'text': 'SS and JT wrote the manuscript.',\n",
       "  'text_sha256': 'ab98a272f51a11362246b66500e3cb893ee3f66cbbaf4f7da5316aaa58f00353'},\n",
       " {'text': 'JT supervized the project.',\n",
       "  'text_sha256': 'f047d8930225373ba7efd3503bd3e7f7a75be7805da77066bf140a1f5d933930'},\n",
       " {'text': 'SM provided critical feedback and provided the liver and lymph node imaging data.',\n",
       "  'text_sha256': '4b83c9fe669289e0b641eef7a32abe142cf40a7136f37da93982c96db723f8f1'},\n",
       " {'text': 'SD performed image acquisition of the liver and LN data.',\n",
       "  'text_sha256': '69fc46b5184403fd2121dfc07a9b1974f8b97eddd70ab337aaff8c0629637334'},\n",
       " {'text': 'All authors critically revized the manuscript for intellectual content.',\n",
       "  'text_sha256': '0f4199e54dfa488eed7a81f42a6069a64a24fd14dd72a6b2f086611cc98fb2bc'},\n",
       " {'text': 'The authors declare no competing interests.',\n",
       "  'text_sha256': '9ba9761df944344679e71f50f98a25479ec835067d14e2f3d453b7d21c5f9a4f'},\n",
       " {'entities': [{'entity_id': 'biological cells',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'biological cells'}],\n",
       "  'text': 'Materials and Methods Cellular Potts model and model variants The CPM is a lattice based formalism for simulating biological cells.',\n",
       "  'text_sha256': '3d05de28a1ebfd54be713e3448f60ec4573112de7c464f8df94fa8d249947599'},\n",
       " {'text': 'Simulations proceeds with a Markov chain Monte Carlo process, where a single position at a time can transition to one of the states of its neighbours.',\n",
       "  'text_sha256': '3d104c16b19bc2d0bb1330c0a3681680a2b6b28004b836e8469b89953de589b8'},\n",
       " {'text': 'The transition probability is based on the change of a Hamiltonian energy function.',\n",
       "  'text_sha256': '5462fb58a7b01ea672d5cdbe21bcd823e5762da4ca0bee0992e971f473d5b2d3'},\n",
       " {'entities': [{'entity_id': 'H',\n",
       "    'other_ids': ['CHEBI:29979'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'H'},\n",
       "   {'entity_id': 'Te',\n",
       "    'other_ids': ['CHEBI:30455'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Te'},\n",
       "   {'entity_id': 'Te',\n",
       "    'other_ids': ['CHEBI:30455'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Te'}],\n",
       "  'text': 'The H =~ Ip(ox),r(0))(1  84:,0;) +  > Avotume(Te)(V (0)  Veargot(Te))?',\n",
       "  'text_sha256': '37478cbdc8d07cbb426e6b2b6eb6d590bceab062675ef234c75c1af741ea55b3'},\n",
       " {'entities': [{'entity_id': 'lattice sites',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'lattice sites'},\n",
       "   {'entity_id': 'cell type',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'cell type'}],\n",
       "  'text': '(1) as its Hamiltonian, with i, 7 being possible neighbour locations in the lattice, 7; the cell present at location i, T is the type of a particular cell, 5 is the Kronecker delta, V(c) is the number of lattice sites a cell o occupies, and Ayolume(T) how strongly a volume constraint contributes to the Hamiltonian for a particular cell type.',\n",
       "  'text_sha256': 'ea8e0a9642c9f1ee60d8e6045c5903c5335588d961fcb92feaaf13dfaa942d69'},\n",
       " {'text': 'In all our simulations a Moore neighbourhood was used.',\n",
       "  'text_sha256': 'cc1bc6d015dd4dee2b5eac2959f1fce8b9f1f5af618b2b35ec04f7e1de5a00b6'},\n",
       " {'text': 'Transitions stochastically succeed based on the change in energy AH, using a Boltzmann probability distribution 1 if AH <0, P(AH) = {laws if AH > 0.',\n",
       "  'text_sha256': '9e1b1c7c53006bcf2c239ee18dd55944fc2a371a4e887b4b9c21e48f36a26e1c'},\n",
       " {'text': '(2) By adding extra terms to the Hamiltonian, constraints can be added to cells and new types of behavior introduced to the model.',\n",
       "  'text_sha256': '11a4ded8170cdcad86678e7c4f0b21e2a042b2169b4d7dd6912960283ffd8f2a'},\n",
       " {'entities': [{'entity_id': 'Atarget',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'Atarget'}],\n",
       "  'text': 'When cell shape needs to be parameterized beyond just cell size, a perimeter constraint* is often introduced, Hperimeter = Ss Aperimeter (To )(A(a) _ Atarget (To).',\n",
       "  'text_sha256': 'a47f013da78bb050805f164612a5315c6c7b43ae5be3f058cdc42cca9e96c651'},\n",
       " {'entities': [{'entity_id': 'lattice sites',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'lattice sites'}],\n",
       "  'text': '(3) Here A(c) represents the number of interfaces between lattice sites of a cell with sites not part of the We also implemented several motility models.',\n",
       "  'text_sha256': '17d43370f47ae6e28b5a812fcfa7054f79ed1364164ff5ada6c6708c7dde824c'},\n",
       " {'text': 'The first allows a vector field to bias the energy change AH during a copy attempt AH chemotaxis = Achemotaxis (To )#i+7 * Ci.',\n",
       "  'text_sha256': '94adf44ba809cc9b9e64e059e46320393621e158210444e1206c4f792bc429ff'},\n",
       " {'entities': [{'entity_id': 'position 7',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'position 7'}],\n",
       "  'text': '(4) Here #;_,; is the unit vector pointing in the direction of a copy attempt from position i to 7, and C; the value of a vector field at position 7.',\n",
       "  'text_sha256': '9a2aefa25f7c4521d5bd9a1fd7ac37c3dc72021fd4192b1584d852ab50da700f'},\n",
       " {'text': 'In this paper, C; is always a field of unit vectors.',\n",
       "  'text_sha256': '83f9cf5407f7fad4eb8ee6ec856d58f06731b170b8e66a826977e23ce13d05f4'},\n",
       " {'text': 'For the model where cells exhibit persistent movement based on one vector per cell, we add AHpersistence = Apersistence (To )#i+5 * Po,t (5) to the Hamiltonian delta.',\n",
       "  'text_sha256': '29607f79e18255ec4ec2b1ce1694f981f4a990b8cc6ae5021a54e3580da205f4'},\n",
       " {'entities': [{'entity_id': 'MCS',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'MCS'},\n",
       "   {'entity_id': 'cell type T',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'cell_type',\n",
       "    'entity': 'cell type T'},\n",
       "   {'entity_id': 'MCSs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'MCSs'}],\n",
       "  'text': 'The vector ,,, is the persistence direction of cell o at time t, which is updated iteratively at every MCS with ey Xo ~~ X, tAt a Dot = (1 = pr) S = + flr Pot1s (6) Pot = 7 (7) Here ju, is the strength of directional persistence for cell type T, Xot is the centroid location of cell o at time t and At is the number of MCSs we look back for to determine the current direction of a cell.',\n",
       "  'text_sha256': '6aabf06e36c0a99a7b029a584e3dccd565e802edfd896f09ad26b09eb0560ba6'},\n",
       " {'entities': [{'entity_id': 'Act',\n",
       "    'other_ids': ['NCBIGene:12',\n",
       "     '71',\n",
       "     '72',\n",
       "     '9457',\n",
       "     '345651',\n",
       "     '11332',\n",
       "     '440915',\n",
       "     '641455'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'Act'}],\n",
       "  'text': 'Finally, we implement the Act model that allows for localized polarization.',\n",
       "  'text_sha256': 'edcdd4527f84dfa2689e8cf1bb65527b8bc0da225a021a0026d4030bde76afd8'},\n",
       " {'text': 'For each position i we track an activity value each time said location is updated, here labelled as Act(i).',\n",
       "  'text_sha256': '2c03ceb18b41163de4549f0da96e04309de7ca7eb2c8c6f2c358bdc2034db5cb'},\n",
       " {'text': 'This localized memory is initialized at Maxact, and decays by 1 each timestep until it reaches 0.',\n",
       "  'text_sha256': 'ea49756ea3eaf081ba1ebf1496e9c1ff44f213cf758a2fb949f9de4d90cca9f9'},\n",
       " {'text': ' GMact (J)), (8)  making copies that originate from areas of a cell with a high activity value more likely.',\n",
       "  'text_sha256': '1a8ee9f1389dd695e5f32e96873507420249726ac8e745aa1cb0ef97bfdea9e4'},\n",
       " {'text': 'The activity is calculated for a full Moore neighbourhood using a geometric mean GMaci(i) = ( T] Act(n)) Ol.',\n",
       "  'text_sha256': 'cceb57a494fa2ba8585f3db7e27b75c351ccdc4acdc67da3626992927844f4cd'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': '(9) N(i) are all neighbours of pixel or voxel i. Optimized CPU implementation of CPM Besides a high-performance GPU implementation, we also developed our own CPU implementation.',\n",
       "  'text_sha256': 'e3559361fcb8ca45526ea5100b5fa1c8eb744d974ad0c07be26baa6a9057a63b'},\n",
       " {'text': 'This served both as a reference implementation for model behavior and internal implementation details for us, and as a reference performance benchmark.',\n",
       "  'text_sha256': '24734bd78ab14ca02391069aa3401fbdf0a7bfb0ae75f38b17052b79338eaca9'},\n",
       " {'text': 'We strived to reduce the major performance bottlenecks we could see in profiling.',\n",
       "  'text_sha256': 'ba9a7af3fa4524fb7007a9f08bd0f4b9b4ef7a9525acb1107071729deb27692c'},\n",
       " {'entities': [{'entity_id': 'lattice sites',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'lattice sites'}],\n",
       "  'text': 'The implementation tracks lattice sites on cell boundaries, and only tries copy attempts involving such positions; a common optimisation, which is also implemented in Morpheus.',\n",
       "  'text_sha256': '9b727f9f2e9e6c12bb28f6127f8ac9560ceed8d28763c0e5d374cafebf7db67f'},\n",
       " {'text': 'The tracking implementation uses a high-performance hashmap implementation.*!',\n",
       "  'text_sha256': '2f8177561ee6b6a51fc7e75a3dccca970a009a9a66b564060c0099f6a8890a82'},\n",
       " {'text': 'We use a custom high-performance pseudorandom number generator, xoshiro256.4?_ The lattice data is stored with a Z- order space filling curve, allowing data that is close in the simulation space also to be close in the linear memory space, maximizing cache efficiency.',\n",
       "  'text_sha256': '9ffdefed43dbe2e9f91c378732346572ea311675b02a96098cb8a7929b1689ef'},\n",
       " {'text': 'Memory addresses are calculated using intrinsic bit interleaving instructions from the x86 bit manipulation instruction set extension.',\n",
       "  'text_sha256': '7cadb466e7896859c2fb77984231f6dc852ab533df0ef4bcfe765898ab10d3cf'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'},\n",
       "   {'entity_id': 'CUDA',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CUDA'}],\n",
       "  'text': 'Our CPM was implemented using CUDA; concepts equivalent to ones described here exist in non-CUDA GPU hardware.',\n",
       "  'text_sha256': 'f2e21984ad3f232457607a3c19a2eba546e1279ea01a5ba34e04ee5414921f5a'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'A GPU consists of multiple streaming multiprocessors (SMs).',\n",
       "  'text_sha256': '0052d1450b5a7df376e39163003670ade2b2179bdf3868440206013248fc2339'},\n",
       " {'text': 'Each SM runs multiple threads in parallel; contrasting with CPU threads, threads on the same SM execute in lockstep, in a Single Instruction Multiple Threads approach.',\n",
       "  'text_sha256': 'ae72172122369ae5f0506c79f65c3a6f79b282d162564411a5e736b86ec7665f'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'A single GPU process  called a kernel  can consist of up to millions of threads.',\n",
       "  'text_sha256': 'a8fc5f3bcf124e6e97a874dc50c8d098719a0331e7c6d67297da100e02d7fe4b'},\n",
       " {'text': 'GPUs have relatively slow global memory available to all threads, and much faster shared memory local to SMs.',\n",
       "  'text_sha256': 'f3ecc2499298db748b2caffc10da77b9d731d1fd72ea10bb1604dce5468af0d0'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Global memory transfer speed is often the main performance bottleneck for GPU programs.',\n",
       "  'text_sha256': 'b225db501eb7835dafdeaf7096a7571f5294dd740c6b4776853c8720e0bd7385'},\n",
       " {'text': 'Threads are bundled in thread blocks (TBs); each TB is bound to a specific SM.',\n",
       "  'text_sha256': '450bcff8f1fbbfaed7c2fbc09dfa89a444b3eab4df3897a53069b8cddac0fd57'},\n",
       " {'text': 'By caching data on their SM in shared memory, TBs can greatly reduce global memory transfer bottlenecks.',\n",
       "  'text_sha256': '857c0ca9560af9e106d761b62ffc0e86ba50cfd659495dd22ca1e7aed80e8625'},\n",
       " {'text': 'Memory transfer bottlenecks can be further reduced by two strategies: read coalescing and high SM occupancy.',\n",
       "  'text_sha256': 'a0c8ef1927b7cbd265ea235060294f7ca4f06e334f2b8b40759d8100e5a843a4'},\n",
       " {'text': 'Read coalescing has sequential threads in a TB do contiguous reads of global memory; contiguous reads are needed to fully utilise the memory bus bandwidth.',\n",
       "  'text_sha256': 'ae53b87bfa432a391267a4a906bbc8d63b54509cbcda880efb2a872a0ad30f67'},\n",
       " {'text': 'High occupancy is achieved when a high number of threads are scheduled to execute per SM; the SM scheduler will switch out threads when they are blocked on memory transfers, ensuring a high number of in-flight memory instructions at We propose a scheme of memory transfer, local SM execution and synchronization (Figure 2D).',\n",
       "  'text_sha256': 'fb75ac74a79dd9bccbde7b626a37beee6c3d0f6ffa184ad2cd98983f05eb34c4'},\n",
       " {'entities': [{'entity_id': 'cell ids',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'cell ids'}],\n",
       "  'text': 'We focus on caching just the grid of cell ids, as this data is always needed, and can be used to quickly reject copy attempts internal to a cell.',\n",
       "  'text_sha256': '163d2a0ff179328358d459e821dc8eebe3b2a294ce293c6af609336bff3af62e'},\n",
       " {'text': 'The simulation grid is divided in subsections, with regions larger than earlier checkerboard subdivisions; each subsection is assigned to a TB, subsection size is chosen so active regions fit in shared memory.',\n",
       "  'text_sha256': 'b9143a2198eab3a651eea11c2c95a2c5db71424284468bfb3714e0478adfcd1d'},\n",
       " {'entities': [{'entity_id': 'TB',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'TB'},\n",
       "   {'entity_id': 'checkerboard sections',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'checkerboard sections'},\n",
       "   {'entity_id': 'coalesced reads',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'RNA',\n",
       "    'entity': 'coalesced reads'}],\n",
       "  'text': 'Each TB determines which checkerboard sections are active in its region; active sections are cached in shared memory using coalesced reads.',\n",
       "  'text_sha256': '620fea1aea5e36382f59d5ccea3ab9bc8690a6dfbfad1ea33d6c2785a3003ac1'},\n",
       " {'text': 'After caching  intended to speed up reads  the simulation is run, and successful copy attempts write new values to both shared and global memory.',\n",
       "  'text_sha256': '5c3c6a101d06d630cacacf229cf997a93058c14428a80bc5977da15298054eac'},\n",
       " {'text': 'Alternatively, changes could be cached in shared memory and written back coalesced to global memory; in our testing, an immediate shared/global write yielded better performance.',\n",
       "  'text_sha256': '1d3eaf34d3aa79a06049ece5152369a7cd510e35d4c68aaa5404caf76a547e9b'},\n",
       " {'text': 'If a TB is done with simulation execution, it immediately starts a memory request for the next active region, ensuring maximum occupancy.',\n",
       "  'text_sha256': 'c1cdebb3171395b620b69233952c87ddaf46505f66741f2d119ae05364bd79af'},\n",
       " {'text': 'All threads in the kernel are synchronized after new memory transfers are finished, ensuring maximum overlap of code execution and memory transfer.',\n",
       "  'text_sha256': '10bcd45c2d9943345ab87a91c7897abbdc9992cf06ca39c5bb6030239aea596b'},\n",
       " {'text': 'CPM literature offers a wide range of model variants and extensions, of which we implement a mod- est subset.',\n",
       "  'text_sha256': 'b7aed7e83e3a2b28ffa8fcc2aa40e0dff23acf5763bc5f5909a09afbc440fcdc'},\n",
       " {'text': 'This comes with potential performance issues: even when extensions are disabled through branching statements, on GPUs such branching code still incurs a significant performance loss compared to CPUs.',\n",
       "  'text_sha256': 'bd87c27f095c6fb0b34d46284b79de9720cc5bc41ca04110777829793a7644e8'},\n",
       " {'text': 'We therefore pre-generated specialized versions of the kernel for a range of different extensions.',\n",
       "  'text_sha256': 'b51cee87b40a15f0ad8349608315e807b2e852b66bf27dee3237cdf7e9c7d127'},\n",
       " {'text': 'We use a custom high-performance random number generator on the GPU.',\n",
       "  'text_sha256': 'c7db9af564a0d2a7fc4968b8dc23abd8604f57bfaea944a33b59f081650d55d5'},\n",
       " {'entities': [{'entity_id': 'CPM',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'CPM'}],\n",
       "  'text': '*%  Composite variable tracking Several CPM extensions introduce additional global variables.',\n",
       "  'text_sha256': 'a4f9fd710387025f801eb911198bcc7f9f9a2daad7588d0305fcac606c487869'},\n",
       " {'text': 'Some of these variable need to be kept in-syne with each other  one variable being slightly out of date with others is not tolerated.',\n",
       "  'text_sha256': 'a202b60efe0663e9e9599937ecffd9a5f34fc47a0ccf71f0dd760f2b35d63e7a'},\n",
       " {'entities': [{'entity_id': 'lattice',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'lattice'}],\n",
       "  'text': 'One example is tracking of a center-of-mass centroid; for 2D this can be tracked as a sum of x and y values, and corresponding count of lattice locations occupied (Figure 2C).',\n",
       "  'text_sha256': 'e1c31902729e4f88c1415ebb1610e2fa92b00384ecc991622adf80762f5b337f'},\n",
       " {'text': 'These are integer values: tracking centroid locations as floating-point would cause values to drift due to rounding errors.',\n",
       "  'text_sha256': 'f9f0e54582e1ff38a040ba5cbc606d2dd98506b389f5d93d46727b235ed2cf3d'},\n",
       " {'text': 'The centroid value {a,y} = = X {asum, Ysum }- (10) Online updates change both a cells centroid position and size.',\n",
       "  'text_sha256': 'cd03a4d57248700058789d8d03b4d0e8da090c74ad25f81ea518ae54dd2c8a24'},\n",
       " {'text': 'When a new lattice position {Xposition, Yposition } is added to a cell, its centroids x value can be updated with Loia X | lola + Lposition Cell sizes are stored interleaved with sum values; sizes and sums are both stored as 32-bit unsigned integers.',\n",
       "  'text_sha256': '9fc518261f2b549e532f235a01836d43d7c688214009a59bb0e7de7fde68e2a4'},\n",
       " {'text': 'The interleaved storage enables use of 64-bit operations to update both: a copy-and-swap (CAS) operation is used, which atomically updates sum and cell size values simultaneously.',\n",
       "  'text_sha256': 'ff7a3f98a972adbdb844cfaac5578f5dfc37429e48db5e7f8eb75c62532675ce'},\n",
       " {'text': 'This scheme also handles periodic boundaries, by adding or subtracting dimension x |e| to the sum values before executing the CAS as needed, to keep centroids within simulation bounds.',\n",
       "  'text_sha256': 'b4dc3a677b0af964e102367b11ab5e77e3ceb3d77981cf19ce538b376c60bf91'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'}],\n",
       "  'text': 'We devized a scheme to execute our CPM on multiple GPUs (Figure 2E); this scheme is intended as a prototype, and as such not as optimized as other parts of our CPM.',\n",
       "  'text_sha256': 'bd6ecc3eabf55825a784a77c64891ad765d6a374b044ec32c8909658536cf4b3'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Simulation space is cut into equal regions, each region is assigned to a GPU.',\n",
       "  'text_sha256': '40508a6eb34edcad6d33864dcfa92165b9b0019def586d8e8d40aaf53ef9d2da'},\n",
       " {'entities': [{'entity_id': 'cell ids',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'cell ids'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'A regions lattice data, such as the cell ids, are sent to each GPU, with an additional border layer.',\n",
       "  'text_sha256': '506cc2b2b26e8709d1770a11340aaf87a1a07d61fe2dc688c3a8b5de7bd2044f'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Global variables such as cell sizes are tracked on each GPU as deltas of the value since last inter-GPU data synchronisation.',\n",
       "  'text_sha256': '2226bf114c1d42ae80ea652deea924a8bcdd1cc878908121cda7fce3844922bb'},\n",
       " {'entities': [{'entity_id': 'GPUs',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPUs'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Every four MCSs, these deltas are gathered from all GPUs, deltas are integrated, and updated values distributed to each GPU.',\n",
       "  'text_sha256': 'ca506f059405e970b384b3536485e576032b1eee46d60dcb76ed855cf4914c9e'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'},\n",
       "   {'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'Multi-GPU speedup is mostly hampered by frequent or slow inter-GPU communication; inter-GPU synchronization after every MCS negates most speedup, and a faster GPU-to-GPU interlink would likely increase performance.',\n",
       "  'text_sha256': 'f05976bfd3f817b4b7bd4d399c5dc4647805373c9a74842e7d0b965bbe0cf36c'},\n",
       " {'text': 'The CPU-based simulations, using our custom optimized CPU implementation and Morpheus 2.2.6 (the last non-parallelized version of Morpheus), and most GPU simulations were run on a system with a Nvidia Titan V, an AMD 2950x CPU and 128 GB of RAM at 2133 MHz.',\n",
       "  'text_sha256': '54bb2469d4d436919f9aaaf726ea6c0017962266760e002f95bb16e0855689e7'},\n",
       " {'entities': [{'entity_id': 'GPU',\n",
       "    'other_ids': ['CUI-less'],\n",
       "    'entity_type': 'DNA',\n",
       "    'entity': 'GPU'}],\n",
       "  'text': 'The only exceptions to this are the multi-GPU simulations (Figure 2E), which were run on a system with four Nvidia Geforce RTX 2080 Ti GPUs, an Intel i9-9820x CPU at 3.3 GHz, and 128 GB of RAM at 2133 MHz.',\n",
       "  'text_sha256': '7cea095280ef21cc28c1a1c394275844fdc493d317ae24cd9dc8c152d9f8e84e'},\n",
       " {'entities': [{'entity_id': 'CCL19',\n",
       "    'other_ids': ['NCBIGene:6363'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'CCL19'},\n",
       "   {'entity_id': 'EYFP',\n",
       "    'other_ids': ['NCBIGene:1969', '10306'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'EYFP'},\n",
       "   {'entity_id': 'mice',\n",
       "    'other_ids': ['NCBITaxon:10095'],\n",
       "    'entity_type': 'species',\n",
       "    'entity': 'mice'},\n",
       "   {'entity_id': 'paraformaldehyde',\n",
       "    'other_ids': ['mesh:C003043'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'paraformaldehyde'},\n",
       "   {'entity_id': 'PFA',\n",
       "    'other_ids': ['CHEBI:60594'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'PFA'}],\n",
       "  'text': 'Imaging of thick lymph node sections Naive inguinal LNs were harvested from CCL19-EYFP mice and fixed with 4% paraformaldehyde (PFA, Electron Microscopy Sciences) for 1 hour at 4C.',\n",
       "  'text_sha256': '346a47e2d40d55e5e63b4363981ad07a5e154b27d7f5737e30f6c3dd8cdb3d19'},\n",
       " {'text': 'Following a brief wash with PBS, LNs were embedded in 2% agarose and 250-300 ym tissue sections were generated using a vibratome (Leica).',\n",
       "  'text_sha256': 'e2738bc1474e6ff78520c6f45249f8c1a1ec8881494b4058a87df91a06d96ae5'},\n",
       " {'entities': [{'entity_id': 'BSA',\n",
       "    'other_ids': ['NCBIGene:3198'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'BSA'},\n",
       "   {'entity_id': 'Triton X-100',\n",
       "    'other_ids': ['mesh:D017830'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Triton X-100'},\n",
       "   {'entity_id': 'rabbit',\n",
       "    'other_ids': ['NCBITaxon:9986'],\n",
       "    'entity_type': 'species',\n",
       "    'entity': 'rabbit'},\n",
       "   {'entity_id': 'anti-GFP Ab',\n",
       "    'other_ids': ['NCBIGene:2994'],\n",
       "    'entity_type': 'gene',\n",
       "    'entity': 'anti-GFP Ab'},\n",
       "   {'entity_id': 'Tween-20',\n",
       "    'other_ids': ['mesh:D011136'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'Tween-20'}],\n",
       "  'text': 'LN sections were subsequently immersed in a Special Wash Buffer (Baxter water, 0.2% BSA, 10% 10X PBS, 0.2% Tween- 20, 0.2% Triton X-100, 0.2% 10% SDS) at 4C overnight, followed by staining with a rabbit anti-GFP Ab (Thermo Scientific, 1:100, diluted in blocking buffer made with Baxter water, 0.2% BSA, 10% 10X PBS, 0.1% Tween-20, 0.2% 10% SDS) at 4C for three days.',\n",
       "  'text_sha256': '56645c6f9aacc46ba6e34478937e3354c8e8c560310aac771e2c95efb73c0a60'},\n",
       " {'entities': [{'entity_id': 'glycerol',\n",
       "    'other_ids': ['mesh:D005990'],\n",
       "    'entity_type': 'drug',\n",
       "    'entity': 'glycerol'}],\n",
       "  'text': 'After incubation, LN sections were thoroughly washed in PBS and mounted in 85% glycerol in PBS with glass coverslips.',\n",
       "  'text_sha256': '6800e146053e75fc37d7d5977b10577c2cf12af3bda97f8e3cbee7d49f2c642d'},\n",
       " {'text': 'Thick LN sections were imaged using a LSM880 AiryscanFast confocal microscope (Carl Zeiss) with a x63/1.4NA oil immersion lens and images processed with ZEN 2010 software (Carl Zeiss).',\n",
       "  'text_sha256': 'c15036ccac5e66c03d8a625d09b1d6ad1d0aec95251759dbf060eec8de5c012e'}]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed_entities = []\n",
    "\n",
    "filtered_entity_list = [entities for entities in entity_list if entities is not None]\n",
    "\n",
    "for entities in filtered_entity_list:\n",
    "    e = []\n",
    "    \n",
    "    # If there are no entities in the text\n",
    "    if not entities.get('annotations'):\n",
    "        parsed_entities.append({\n",
    "            'text': entities['text'],\n",
    "            'text_sha256': hashlib.sha256(entities['text'].encode('utf-8')).hexdigest()\n",
    "        })\n",
    "        continue\n",
    "    \n",
    "    for entity in entities['annotations']:\n",
    "        other_ids = [id for id in entity['id'] if not id.startswith(\"BERN\")]\n",
    "        entity_type = entity['obj']\n",
    "        entity_name = entities['text'][entity['span']['begin']:entity['span']['end']]\n",
    "        \n",
    "        try:\n",
    "            entity_id = [id for id in entity['id'] if id.startswith(\"BERN\")][0]\n",
    "        except IndexError:\n",
    "            entity_id = entity_name\n",
    "        \n",
    "        e.append({\n",
    "            'entity_id': entity_id,\n",
    "            'other_ids': other_ids,\n",
    "            'entity_type': entity_type,\n",
    "            'entity': entity_name\n",
    "        })\n",
    "\n",
    "    parsed_entities.append({\n",
    "        'entities': e,\n",
    "        'text': entities['text'],\n",
    "        'text_sha256': hashlib.sha256(entities['text'].encode('utf-8')).hexdigest()\n",
    "    })\n",
    "    \n",
    "parsed_entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "07b26efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "host = 'bolt://52.91.239.25:7687'\n",
    "user = 'neo4j'\n",
    "password = 'house-round-forearm'\n",
    "driver = GraphDatabase.driver(host,auth=(user, password))\n",
    "def neo4j_query(query, params=None):\n",
    "    with driver.session() as session:\n",
    "        result = session.run(query, params)\n",
    "        return pd.DataFrame([r.values() for r in result], \n",
    "columns=result.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3e41f537",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"arX1v:2312.09317v1 [q-bio.TO] 14 Dec 2023\\n\\nA parallelized cellular Potts model that enables simulations at\\ntissue scale\\n\\nShabaz Sultan!?, Sapna Devi?, Scott N. Mueller?, Johannes Textor!?\\n\\n! Data Science group, Institute for Computing and Information Sciences, Radboud University, Nijmegen,\\nThe Netherlands\\n\\n? Department of Medical BioSciences, Radboud University Medical Center, Nijmegen, The Netherlands\\n\\n3 Department of Microbiology and Immunology, The University of Melbourne, at the Peter Doherty\\nInstitute for Infection and Immunity, Melbourne, Victoria, Australia\\n\\nSummary\\n\\nThe Cellular Potts Model (CPM) is a widely used simulation paradigm for systems of interacting cells that\\nhas been used to study scenarios ranging from plant development to morphogenesis, tumor growth and\\ncell migration. Despite their wide use, CPM simulations are considered too computationally intensive\\nfor three-dimensional (3D) models at organ scale. CPMs have been difficult to parallelise because of\\ntheir inherently sequential update scheme. Here, we present a Graphical Processing Unit (GPU)-based\\nparallelisation scheme that preserves local update statistics and is up to 3-4 orders of magnitude faster\\nthan serial implementations. We show several examples where our scheme preserves simulation behaviors\\nthat are drastically altered by existing parallelisation methods. We use our framework to construct tissue-\\nscale models of liver and lymph node environments containing millions of cells that are directly based on\\nmicroscopy-imaged tissue structures. Thus, our GPU-based CPM framework enables in silico studies of\\nmulticellular systems of unprecedented scale.\\n Introduction\\n\\nThe CPM! is a powerful simulation method that is known for its ability to simulate a wide range\\nof biological scenarios. When modelling tissue environments, many researchers wish to consider both\\nhe complex morphology of single cells, and the behavior of these cells as part of larger collectives.\\nThe appeal of the CPM lies partly in its ability to integrate both these aspects in a single model.\\nAs such, it has been applied to diverse problems including tumor growth,” vascularization,® wound\\nhealing,* and morphogenesis.>*® Many extensions and hybrid model variants are available in the literature\\n‘0, for example, add support for cell motility,” cell signalling,!° and finite-element based mechanical\\ninteractions.!! Thus, the CPM can be described as a generic platform upon which a wide range of models\\ncan be built, possibly explaining its popularity in the tissue modelling field.\\n\\nFor computational reasons, CPM models are often built using a small number of cells (perhaps in the\\nhousands) in simplified two-dimensional (2D) geometries. Some research questions in tissue modelling\\n\\nhowever demand a realistically large number of cells, and accurate 3D representations of tissues. For\\ninstance, it is widely appreciated that the efficiency of space exploration of motile cells depends criti-\\ncally on the dimensionality of the environment, leading researchers to use 3D CPM simulations.® !?:13\\nAssessment of tumor growth behavior and invasive patterns of tumors likewise changes as a function of\\ndimension, resulting in the use of 3D CPM simulations in the tumor modelling field as well.? While there\\nis thus a clear research demand for realistically large 3D CPM simulations, scaling up such simulations is\\ngenerally considered computationally infeasible.4 1° Removing this computational barrier would allow\\na host of CPM based modelling research to be scaled to what is biologically realistic, instead of scaling\\nto what technical constraints allow.\\n\\nThe CPM is an on-lattice method: simulations are executed on a grid. A common strategy to speed up\\ngrid-based models is to cut the simulation into smaller subsections that are run in parallel. However, for\\nthe CPM, model behavior is intimately tied to locations on the grid being updated sequentially, in a single-\\nlocation update Markov chain Monte Carlo process. As such, parallelisation has been cautioned against,\\nas it risks drastic changes in CPM “system kinetics”.° Nevertheless, several attempts have been made to\\nemploy subsection-based parallelisation. Chen et al. proposed a checkerboard scheme that divides the\\nCPM lattice up over a computer cluster, with sections chosen much larger than cells, preventing cell-level\\nsynchronization issues.!” Tapia et al. propose a similar scheme for GPUs:!® because the sections are much\\nsmaller than a cell on the GPU, they use a per-cell locking mechanism to prevent simultaneous updates of\\na cell. Neither approach has been extensively checked for correct model behavior. To our knowledge, no\\npublicly available implementation of subsection-based parallelisation exists for the CPM. The software\\n“Morpheus” !® includes a non-subsection-based Central Processing Unit (CPU) parallelisation strategy;\\nthis method is less suited to fully utilise parallel accelerator hardware such as GPUs.\\n\\nIn this paper, we aim to make several contributions: First, we introduce a highly optimized parallelized\\nimplementation of the CPM using several optimization strategies not previously explored in the literature.\\nSecond, we conduct an empirical analysis of potential pitfalls in parallelising a model dependent on serial\\nupdates of single locations and show that our approach mitigates these issues while preserving performance\\nbenefits. Third, we demonstrate that our approach retains the behavior of cell motility simulations that\\nare particularly prone to changed model kinetics upon parallelisation. And finally, we build two large 3D\\n\\nimage-based simulations of T cell migration in liver and lymph nodes, demonstrating how complex tissue\\nenvironments can be analysed at a larger scale and realism than was possible before.\\n\\nResults\\n\\nSerial Cellular Potts Models are limited to sub-millimetre scale\\n\\nThe CPM’s serial update schedule — only one location at a time is updated on a grid — is central to\\nits behavior, and thus not trivially changed. We therefore first aimed to quantify the computational\\nlimitations for models based on the serial CPM: how big can our computational experiments be, and\\nwhat timescale can they span, without making fundamental changes to the CPM?\\n\\nWe set up 3D CPM tissue experiments at different spatial and temporal scales and measured the required\\nwall-clock time. We choose spatial and temporal simulation resolutions that allow reasonably accurate\\nrepresentations of highly motile cells within tissues. Lymphocytes are among the smallest nucleated cells\\nin the mammalian body at 8-10um diameter for B cells and T cells. At speeds of up to a few body\\n oO\\n£\\nx\\n(2)\\n2\\n2\\noS\\n=\\n64° 256° 1024°\\nsimulation volume (um’)\\nB\\nial lyr °\\nlo | 30 values/param.\\n> - oO\\n2 € 1wk 7\\n5 c= a = , -\\n[sj x e Pa\\n£ jw || 8 ow\\n6 cS ihr 4—_-\\n2 L | 2\\noO ras o\\n° y =\\nOS Tmin 10 values/param.\\nL 4\\npolarisation strength replicates 1 2 3 4 5\\n\\nparameters\\n\\nFigure 1: Serial CPM implementations do not scale well to large simulations and big en-\\nsembles. (A) Computational cost of simulating a 256°m* volume (1m® per voxel) for 10 simulated\\nminutes (1s per simulation step). With a high-performance CPU implementation, this takes 1.2 hours of\\nwall-clock time. Extrapolation to a full day of simulated time for a 1mm? simulation estimates 1.4 years\\nof wall-clock time. Time measurements (black) are averages of 5 simulations, projections are based on\\ninear fits of simulations > 64°m* in log-log space (grey). (B) Building CPMs typically requires tuning\\nparameters. Here we perform a multidimensional grid search for parameters that yield motile cells using\\nthe Act extension of the CPM;’ a modest single parameter grid search where 10 values are tried takes\\n5.9 seconds; larger grid searches take exponentially more time as function of parameters searched, with a\\nfive parameter search with 30 values is extrapolated to 1.1 years of wall-clock time. Time measurements\\n(black) are averaged over 5 simulations, projections fitted on linear trends in linear-log space. CPM\\nparameters are shown in Table 1 and Table 2.\\n\\nengths per minute, they are also among the fastest movers within tissues. We chose lum? per voxel and\\nsecond per simulation step, or Monte Carlo Step (MCS), as our simulation resolutions, corresponding\\nto Beltman et al.9\\n\\nTo create realistically large models of, for example, a complete murine Lymph Node (LN), or scale a\\ntumor model to the size of an avascular tumor spheroid,?° experiments at mm® scale would be needed.\\nCertain organ scale dynamics require timescales of at least a day, such as T cell transit through LNs.?4\\nMore modest simulation experiments at 256°um? (Figure 1A) at 10 simulated minutes require ~1h of\\nwall-clock time on a highly optimized CPU implementation of the CPM (Methods). Such experiments are\\nsufficient to see minimal movement of motile cells such as T cells and can capture organ compartments\\nsuch as a LN paracortical T cell zone, or a micrometastasis for cancer modelling. For mm? scale, one-day\\nexperiments however, researchers would have to wait at least a year — making them impractical for all\\nbut the most patient of researchers.\\n\\nA single experiment is rarely sufficient, in both empirical and simulation research. Beyond the need\\n‘or experimental replicates, simulations can have parameters for which associated model behavior is non-\\nobvious, requiring parameterized simulation ensembles. In Figure 1B, we explore a cell motility extension\\nto the CPM that coarsely models actin dynamics.” A single two-dimension simulation is relatively cheap,\\nbut when exploring even a modest number of parameters the execution time goes up exponentially.\\n\\nThus, serial 3D CPM simulations at a scale of 1 mm® are currently not feasible, and even 100°um? scale\\n\\nsimulations were extremely challenging in our experiments. While ensemble runs can be accelerated with\\nparallel simulations without changing the CPM, the high computational costs of running large ensembles\\n‘urther motivates the need for a faster CPM.\\n\\n Designing a parallel execution scheme for the CPM\\n\\nTo make large scale simulations practical, we propose a parallelisation scheme for the CPM that uses\\n\\nhighly parallel GPU hardware. A natural way\\n\\ndivide the simulation grid in\\nsupport up to thousands of\\n\\nsmal\\ncalcu\\n\\n‘0 parallelise on-lattice models, such as the CPM, is to\\ner sections; each section runs a local simulation in parallel. As GPUs\\nations in parallel, these sections will be quite small. The CPM also\\n\\nupdates information outside of the grid, such as sizes of simulated cells. We first consider three issues\\nwhen designing our parallelisation scheme: adjacent sections with active simulation interfering with\\n\\neach other during updates o:\\nsimultaneously, and stale in:\\nupdated by another simulati\\n\\nA CPM simulation proceeds\\n\\ncopying the position’s state to a ne\\n\\nsimulation running, we run t.\\nboundary. Such “race condi\\nalmost simultaneous update\\nstandard problem in paralle\\n\\nf the\\non.\\nby cl\\n\\nhe ris\\nions”\\n— wi\\n\\nisatio\\n\\nto be active at the same time. As\\n\\nsections are sufficiently spaced to prevent overlappin,\\nbetween active and dormant state.\\n\\normal\\n\\nsimulation gri\\nion being used by one paralle’\\n\\nhoosing a random position on\\n\\nk of simulations “racing” to wri\\nrun the risk of information ge\\n\\nforward solu\\noard” patte\\n\\nm, with a straight-\\naced out “checker\\n\\nThe CPM also tracks information involving more than one gri\\n\\nsimulated cells. We face the same “race condition” problem here,\\nupdates touch a single cell (Figure 2B, left column).\\nsmall active subsections, making such concurrent up\\nscheme is proposed, that rejects all but the first concurrent update attempt on a varia\\nmiddle column). We instead opt to use atomic operations (Figure 2B, right column), which equa!\\nprevents update information from getting lost, as the hardware will eventually integrate all updates\\n\\nsuch operations.\\n\\nIn serial\\nevery W\\n\\nmay be\\nTapia e\\nof-date\\nmay to.\\nbeing s\\n\\ndesigned an additional scheme that sti\\n\\nGPU pai\\n\\nCPMs, global information travels instantaneously: as only one grid position is w\\n\\nhout incorporating information of the overwritten update. This\\n\\ng updates (Figure 2A). Regions cycle perio\\n\\nates more likely. In Tapia et a\\n\\n; non-grid information such as cell sizes being updated\\n\\nsimulation after said information got\\n\\nthe simulation grid, and stochastically\\n\\nighbour. If adjacent subsections of the grid each have an active parallel\\n\\ne to the same location on the sections’\\nting lost, as one section overwri\\n\\nion: never allow two adjacent sections\\nrn of sections is only ever active; active\\nically\\n\\nof\\nel\\nof\\n\\nlocation, such as the current sizes\\nas it is possible that multiple paral\\nrallelisation requires a high number\\n1.18 a per-cell locking\\nble (Figure 2B,\\nly\\nor\\n\\ndated at a time,\\n\\ndate attempt has access to the results of all previous updates. This instantaneous information\\naccess is relaxed in our proposed scheme: when information such as cell size is accessed,\\nout-of-date due to still in-flight updates to the accessed information. The cell-loc!\\nt al. prevents this from happening, by outright rejecting any updates that cou\\ninformation; we explored a less strict scheme, reasoning that the stochastic CPM copy attempts\\nerate out-of-date information. For CPM extensions that do not tolerate their global variables\\nightly out sync with each other (e.g. variables used in center-of-mass “centroid” tracking) we\\n1 will keep variables in syne with each other, and sits on top of our\\n\\nthis information\\ning scheme from\\nd have used out-\\n\\natomics based scheme (Figure 2C, Methods). Otherwise, due to atomic updating, our scheme still gives\\nstrong guarantees on tracked data being correct — for example, cell sizes will be tracked correctly — and\\n\\nthe added flexibili\\n\\nImplementation o\\n\\nwe cache local da\\nsimulation calcula\\n\\nF our scheme on rea\\nGPUs, memory access almost always\\nprevents easy use of a GPU’s caching hierarchy.!® To maximize performan\\nscheme that carefully operates in loc\\n\\nbottlenecks performance —\\n\\nhardware requires consideration o:\\n\\ny allows us to explore the effects of allowing the use of slightly outdated information.\\n\\nOn\\nCPM’s random access pattern\\nce, we designed a data transfer\\n\\nf several technical details.\\nbut the\\n\\nk-step with checkerboard work splitting (Figure 2D, Methods):\\na needed by the regions that are becoming active, an\\ntions to fully utilise\\n\\noverlap memory transfer and\\n\\nGPU hardware. To increase simulation volume, and gain further\\n\\nspeed ups, we also prototyped a multi-GPU scheme (Figure 2E, Methods).\\n\\nIn short, our proposed parallel CPM sc!\\n\\nin running paralle\\n\\nheme is tailored to make opt\\n\\ncode. It still gives strong guarantees on simulation sta‘\\n\\nimal use of GPUs: hardware specialized\\n\\ne consistency — for example, a\\n\\nglobal variable tracking a cell’s size will always equal the number of pixels on the grid occupied by that\\n\\ncell. Yet, we do all\\n\\nflexibility and further maximise performance gains.\\n\\now the use of slightly outdated values of such global variables in copy attempts to gain\\n ba)\\n\\n?-\\n\\nB race condition mutex locking atomic updates\\n(Tapia et al.) (our method)\\n(lel=12) (lla i= 12 |}\\n\\nold old *\\n\\n| ASP—$$ $l new = 13)- - - [le] = 13 &)- - - {le] = +18)\\nnO OO Oo Hii --(=tie---(41=19\\n@©008000\\neee0\\n\\nreese ew. = 13)- - - (lel := 13 )- - - {\\n? | ——{l*],.4 = 13)- - -(lo = 13 w)- - - |\\n\\nc I*]s:= 4 I#lnou = 5\\n© Xu Yous *= (2-8, 3.3} © Xen? Voow = (3; 3} (Kumn= 11) (Yaun= 13} (lel= 4)\\n5\\n4 ee [Xeon 2= +4! Yeu = +2 | lel= +1 | Be]\\nXon= 15) (Van= 15) (Iel=5)\\n[new = Xeun/@l = 3; Yoow = Yeun/l#] = 3)\\n\\nE_ per GPU region border sync\\n\\n~-7--- global simulation barrier\\nSS caput sizessyne pu\\n\\n2\\n|e] = 23 je| = 29 |e| = 23 |e| = 29\\n[Je|:= +4 Jel: +14=he! = +0 |el:=-5 |\\n\\n|e] = 27 |e|=25 |e] = 27 je|=25\\n\\nFigure 2: Designing a parallel CPM based on checkerboarding and atomic updates for\\nGPUs. (A) We divide simulation space into a subgrid, with alternating regions being active in a\\n“checkerboard” pattern, preventing update conflicts on the grid. (B) We implement two strategies to\\nprevent global variable update conflicts (e.g. cell size tracking): a mutex locking strategy (Tapia et\\nal.'8), and a novel atomic operation strategy. (C) Combining interleaved storage and atomic operations\\non adjacent data allows variables to be kept in sync with each other. (D) A four-step execution scheme,\\nwhere active checkerboard areas are transferred to fast shared memory, accelerates copy attempts. Thread\\nbarriers synchronize active threads after transfers. (E) A multi-GPU execution scheme, where regions\\nof the simulation are assigned to different GPUs. Borders are synchronized between relevant adjacent\\nregions, and global variables are synchronized through delta tracking.\\n\\nglobal memory (slow)\\n\\non\\n A T, : time in region time divergence thread barriers\\n\\nmm : time for one copy attempt ———-—— T,\\nid ‘. :\\n0 Oe\\ne\\n\\ne%\\nSys\\n\\nT,\\n5 checkerboard size thread size\\n\\n' loa! —~“—_.\\n\\ne ? ee : :\\n<@ e ‘|e eee ‘ eco\\né } ° 1|0@ : eee\\nao @ : : ee\\n\\nT, T. : :\\n\\n<>\\ncopy attempts before checherboard switch\\n\\nsingle update parallel update\\noo\\n\\neoeoeeen\\ntime since last copy attempt\\nee®@ r ) 4 copy attempts per switch, 1 copy attempt per switch,\\nnumber of copy attempts 4x4 area/thread, 8x8 area/thread,\\n2x2 area/checkerboard 4x4 area/checkerboard\\n| | | lh __\\n0123 45 0123 45 0123 4 5\\n\\ncopy attempts per location\\nbinomial distribution —\\n\\n0 1 2 0 1 2 0 1 2\\nwaiting times between copy attempts (MCS)\\n\\ngeometric distribution —\\n\\nFigure 3: Checkerboarding can drastically alter local update statistics. (A) To prevent threads\\nfrom racing ahead of others too much, thread barriers are imposed after each update. (B) Checkerboard\\nsize and switching frequencies determine the distribution of waiting times between consecutive updates\\nof the same locations. (C) Small checkerboard areas or infrequent switches can alter the waiting time\\ndistribution substantially compared to the single-update scheme. Update frequency histograms were\\ngenerated by simulating one MCS (256 updates) on a 16x16 grid. Waiting time histograms were simulated\\non the same grid using 100 simulated MCSs.\\n cell boundary type:\\n\\n@-@ dark-dark\\n\\n60% 1 @-© dark-light\\n-© light-light\\n\\n40%\\n\\n20%\\n\\n% of cell boundaries\\n\\n0%\\n\\n10 1000 100000\\ntime (MCS)\\n\\nFigure 4: Atomic updates lead to correct behavior in a simulation of cell sorting. A cell sorting\\nsimulation based on Graner et al.! is implemented with two strategies to prevent race conditions for cell\\nsize variables. The mutex locking strategy proposed by Tapia et al.!® leads to severe starvation in our\\ntests. Using an atomic update approach recovers the desired sorting behavior, closely matching sorting\\nspeed of the original CPM. The simulation was run with a 4x4 area per thread and 2x2 checkerboards,\\nbut other sizes were tested and did not make a significant difference for this simulation. Sorting trends\\nof different CPM implementations are averaged over 5 simulations. Simulation parameters can be found\\nin Table 1.\\n\\nAligning the lattice update statistics of parallel CPMs to serial CPMs\\n\\nOur novel parallel execution scheme guarantees simulation data consistency, but drastically changes the\\nway the simulation lattice is updated. We investigated statistical properties of copy attempts on the\\nsimulation lattice for different lattice positions; specifically, we analysed the relative ordering of copy\\nattempts, frequency of copy attempts, and waiting times between copy attempts. This allowed us to\\nquantify the divergence of parallel lattice update kinetics from the serial CPM. Such divergence could\\nlead to changed simulation behavior in complex models of cells and tissues.\\n\\nWithout guardrails, performing independent CPM updates in parallel would allow different parts of a\\nsimulated tissue to have been alive for different lengths of simulated time (Figure 3A). Modern GPUs use\\nthread barriers to synchronize parallel threads without strong computational performance implications.\\nWe set thread barriers after ever single copy attempt, so for every currently active checkerboard region\\nthere is only a maximum time divergence of one copy attempt between active regions.\\n\\nOur parallel execution scheme allows us to freely choose the size of subregions and frequency of active\\nregion switching (Figure 3B), without impacting data consistency guarantees. We measure frequency\\nand waiting times between copy attempts per position on the lattice in the standard, serial implementation\\nof the CPM (Figure 3C, left column). For our initial parameter choice, we found that underlying update\\nstatistics are drastically changed (Figure 3C, middle column): waiting times between copy attempts\\nbecame much more predictable due to the much more even distribution of copy attempts compared to\\nthe serial CPM. Using large parallel subsections with frequent active region switching allows non-parallel\\nupdate statistics to be closely recapitulated in our parallel scheme (Figure 3C, right column).\\n\\nWe thus find that parallelisation parameters have major influence on copy attempt statistics on the\\nlattice. By switching active regions of the checkerboard frequently, and avoiding too small areas per\\nthread, we can obtain statistics that closely resemble the serial CPM.\\n\\nEmergent cell and tissue behavior is maintained in a parallel CPM\\n\\nOur goal is to build parallel CPMs that are much faster than previous implementations but maintain cell\\nand tissue kinetics consistent with existing serial CPMs. We hypothesized that aligning the waiting times\\nbetween update attempts of our parallel CPM to the serial version should suffice to achieve such consis-\\ntency. To investigate this, we ran several simulation scenarios with different CPM variants, measuring\\nemergent behaviors on the cell and tissue levels and comparing these to serial CPM versions.\\n\\nWe first investigated the classic cell sorting simulation:! two mixed populations of cells with differential\\n A single update\\n4 copy attempts per switch,\\n4x4 area/thread,\\ntime 2x2 area/checkerboard\\nTeoo00\\ne\\ne\\ne\\ne\\ne\\nTy\\nay ay\\n= 0.3 £03 2\\n202 5 0.2 5\\n50.1 & 0.1 &\\na a\\n0.0 8 0.0 8\\n2000 6000 10000 2000 6000 10000\\ntime (MCS) time (MCS)\\nB » low polarisation strength\\n! I\\ni i\\n& ied |\\n2 ee qe 4S\\n' high polarisation strength '\\n. «) f 4 i\\nc c c\\n& & &\\n3 08 3 08 3\\n£ 0.4 £ 0.4 £\\n8 8 8\\n8 0.0 8 0.0 8\\n& & &\\n200 400 600 800 200 400 600 800\\npolarisation polarisation\\nCc 8 copy attempts per switch,\\n4x4x4 area/thread,\\n2x2x2 area/checkerboard\\n\\nparallel update\\n\\nlow density crowd\\n\\n“Ip\\n\\n~ “it ze\\n\\n+\\n\\n0.8\\n& 0.6 Gy\\n204 z\\n© 0.2 fe)\\n0.0 A\\n0.25 0.50 0.75 1.00 0.25 0.50 0.75 1.00\\ndensity density\\n\\neso\\noun\\n\\n1 copy attempt per switch,\\n8x8 area/thread,\\n4x4 area/checkerboard\\n\\nS\\nwo\\n\\noo\\n\\n2000 6000 10000\\ntime (MCS)\\n\\n0.8\\n\\n00 » awa\\n0.0\\n\\n200 400 600 800\\npolarisation\\n\\n1 copy attempt per switch,\\n8x8x8 area/thread,\\n4x4x4 area/checkerboard\\n\\n0.25 0.50 0.75 1.00\\ndensity\\n\\nFigure 5: Motility behavior matches serial CPM behavior when parallel CPM local statistics\\nare similar. (A) Cells in a circular chemokine gradient simulation move highly regular in a serial CPM\\n(low per-pixel entropy, shaded: 1SD); parallel simulations diverge slowly from reference behavior when\\ncopy attempt statistics are preserved (low per-pixel cross-entropy). (B) In the Act model of intra-cell\\npolarization,” stronger polarization strength leads to higher directional autocorrelation (measured for 750\\n\\nMCSs at 50 MCS intervals, averaged over 50 simulations, shading: 4\\n\\nemerges when update waiting time distributions are preserved. (C)\\n\\nt1SD); directional persistence only\\nPhase transition simulation in 3D\\n\\nwhere synchronous motility emerges at a density around 0.5. Cell order indices after 2000 MCSs are\\n\\nshown (averaged over 5 simulation runs, shading: 4\\n\\nt1SD). The phase\\n\\nransition only occurs when update\\n\\nwaiting time statistics are preserved. Simulation parameters can be found in Table 1 and Table 2.\\n adhesion (i.e., cells of the same type have stronger\\n\\nadhesion to each other than cells of different types)\\n\\nsort themselves in two homogeneous regions over time (Figure 4, grey lines). The mixing kinetics can be\\n\\nmeasured by tracking homogeneous and heterogene:\\n\\nbased parallelisation approach proposed by Tapia et a\\n\\nleads to a lack of sorting behavior; the simulation\\nconcurrent copy attempt is allowed per cell un\\nscheme allows expected sorting behavior to re\\n\\nremained consistent for all parallelisation paramete:\\n\\n‘ous cell neighbours over time.! Strikingly, the mutex\\n1.18 — while still being able to run a simulation —\\nencounters so-called “starvation”, because only one\\n\\ner this scheme (Figure 4, red lines). Our less restrictive\\n-emerge (Figure 4, black lines). The sor\\n\\ning behavior\\nrs we examined (Supplementary Figure $1).\\n\\nWe next examined three motility model variants of the CPM, where additional biases are placed on\\n\\no obtain cells that move persis\\nfunction of random membrane fluctuations. Bec:\\nneed to roughly maintain their target size, copy\\nsections for movement to occur. We therefore s\\nby changes to the waiting times between copy at\\n\\ncopy attempts\\n\\nThe first motili\\nikely to succeed when aligned with the underlyi\\nbias is fixed in this model — the field does not change\\nsimulations with the same initial conditions closely\\nWe quantified\\ncontains a cell in multiple simulations, where low\\nbetween simulations, and low cross-entropy indica:\\nwith reference serial runs. In simulations with\\nincreased quickly (Figure 5A, middle column); w.\\nrises much slower (Figure 5A, right column).\\n\\nNext, we analysed a model that simulates localize\\n\\nborder of a cell makes further copy attempts involving those\\ntimes between copy attempts are very important in this mo\\nin a small neighbourhood successfully expand at the same time; this is less\\n\\ny model we consider adds a chemotactic vector field to the CPM; copy attempts are more\\nng vector field (see Methods). Because the directional\\n\\nhis alignment using pixel-wise (cross-)entropy: we measure the probability that a pixe\\nentropy of\\n\\nhighly chan\\nhen the w\\n\\nently; in the standard CPM movement only occurs as a\\nause motile cells have a leading and a retrac\\nattempts need to sufficiently alternate between these cel\\npeculated that movement models might be more affecte\\nempts.\\n\\ning edge an\\n\\nover time — cells move in a highly regular fashion, an\\nalign even at a pixel level (Figure 5A, left column).\\n\\npixel probabilities indicated high agreement\\nes a high level of agreement of parallel simulations\\nged lattice update statistics, cross-entropy\\npdate statistics are preserved cross-entropy\\n\\ncell polarization,’ where recent pixel updates at the\\npixels more likely to succeed. The waiting\\nel, as cells start to move when many pixels\\nlikely when copy attempts are\\n\\nspaced out evenly. A polarization strength parame\\n\\ncell; varying the polarization strength interpolates motility be\\ncolumn). In paral\\n\\nmotility modes (Figure 5B, lef\\nwaiting time statistics this full range of behaviors\\nthey re-emerge when update sta’\\n\\nFinally, we examined a simulatio\\n\\nwith this vector, and the vector\\nmotility model involves informat\\n\\nis periodically upd\\nion tracked on the\\n\\nto the simulation grid, placing additional demands on a para!\\n\\ndensity cell collective exhibit Brownian movement,\\nas density is increased;?? this characteristic phase\\n\\nleft column). In a parallel CPM this phase transition disa,\\n(Figure 5C, middle column), but re-emerges when update\\n\\ncolumn).\\n\\nTaken together, these results demonstrate that a strict updat\\n\\nistics are preserved\\n\\nn of a tightly packe\\na preferred movement direction vector; copy attempts for a ce\\n\\ns how much polarization bias is added to a\\ntween Brownian and directionally persistent\\nel simulations with changed update frequency and\\noes not appear (Figure 5B, middle column), but\\n(Figure 5B, right column).\\n\\n3D col\\n\\ner contro\\n\\nlective of immune cells,° where each cell has\\n1 are more likely to succeed when they align\\nated to align with recent movement of the cell. This\\nattice, and additional information tracked external\\nel CPM. In a similar simulation model, low\\ntransitioning sharply to ordered collective movement\\nransition is reproduced in the CPM (Figure 5C,\\npears when update statistics are changed\\nstatistics are preserved (Figure 5C, right\\n\\ne scheme preventing parallel updates of the\\n\\nsame cell is not necessary for maintaining valid CPM behavior and can even lead to severe “starvation”\\n\\nwhere simulations do not progress anymore. On\\n\\nhe other hand, preserving waiting times between copy\\n\\nattempts involving the same pixel can be very important to maintain model behavior, and was critical\\n\\nin all three motility scenarios we examined.\\n\\nGPU-based CPM parallelisation accelerates models by several orders of mag-\\n\\nnitude\\n\\nHaving confirmed that our parallel CPM indeed pro\\n\\nduced correct model behavior in a range of examples,\\n\\nwe next investigated the performance improvements resulting from our implementation choices when\\n\\ncompared to a serial CPM.\\n % 1000 2\\n10 g 2\\n< 100 oO\\ns 8 g 8 & 19547x13805x\\nBe 3 3 10 3\\n®\\n2 3 < 1 g\\n24 a & 2\\n& 2 =\\n2 eg ol °\\n= 0.01 a\\nkernel specialisation ( ae ) 128° 5127 20487 8192! implementation\\nGPUs dimension\\nshared memory @ @ e@\\nD E area/thread: @ 4x4 A8x8 mi 16x16\\n> ° 03 +0\\n3 10000 3 9786x 2\\n®\\nS 1000 g = 02 = 0.9 >\\n& 400 ® § g 8\\n& 2 8 308 3\\noo «10 ® — 2902x ZO 8 8\\n£ 1 s G 07\\n0.0\\n642 256° 10242 implementation 1 4 16 64 1 4 16 64 0.0 0.1 0.20.3\\ndimension updates per switch updates per switch simulation time\\n\\nFigure 6: Parallel CPMs outperform serial implementations by orders of magnitude. (A)\\nPerformance benefits are obtained by a shared memory scheme, and a pre-compiled kernel specialization\\nscheme, based on a sorting simulation extended to densely fill simulation space (Figure 4, all sorting\\nbenchmarks are average of 5 simulations of 100 MCSs, 100 MCS burn-in). (B) Distributing the lattice\\nover 4 GPUs speeds up the simulation further. (C) Performance of sorting simulations are significantly\\nhigher relative to two serial implementations; relative speedup shown for 2048x2048 simulations. (D)\\nPerformance of 3D cell collective simulations is also improved (Figure 5C, average of 2 simulations of\\n100 MCSs, 100 MCS burn-in); relative speedup shown for 256° simulations. (E) Trade-offs between\\nspeed and accuracy for chemokine field simulations (Figure 5A, accuracy: 1 - cross-entropy of parallel\\nensemble relative to serial ensemble, 250 simulations per ensemble). CPM parameter settings can be\\nfound in Table 1 and Table 2.\\n\\nWe first examined the performance gains resulting from our use of a shared memory caching scheme\\n(Figure 2D, Methods) using a 2D sorting simulation benchmark (Figure 4). On its own, shared memory\\ncaching achieved only minimal speedup (Figure 6A). However, for further performance improvement,\\nwe pre-generated different versions of our code based on which CPM extensions were in use (Methods).\\nCombining these specialized kernels with the caching scheme achieved a significant 10-fold speedup, more\\nthan expected from either optimization alone (Figure 6A). Testing our multi-GPU scheme (Figure 2E,\\nMethods) in the same scenario achieved a further 2.3x speedup by distributing the simulation across\\nfour GPUs (Figure 6B). Hence, careful consideration of GPU-specific implementation choices resulted\\nin significant performance benefits.\\n\\nWe next compared our parallel CPM (on a single GPU) to our own optimized CPU implementation\\n(Methods), and to the state-of-the-art software “Morpheus”.!® Using the same 2D cell sorting bench-\\nmark, we observed a {3,500-fold speedup compared to our own CPU implementation, and a j25,000-fold\\nimprovement relative to Morpheus (Figure 6C). On the 3D cell collective simulation (Figure 5C), a\\n{,600-fold speedup was achieved relative to serial CPMs (Figure 6D). It is important to note that our\\nbenchmark hardware did not permit use of shared memory caching for the 3D model; future hardware\\nwith more shared memory may improve 3D performance.\\n\\nAs shown in Figure 3 and Figure 5, checkerboard size and switching frequency influences model behav-\\nior: larger divisions and more frequent switching leads to behavior concordant with serial CPMs. These\\nupdate patterns also interact with caching, further impacting performance. As the sorting simulation has\\nlow sensitivity to parallelisation parameters (Supplementary Figure $1), we instead benchmarked our\\n2D chemotaxis simulation (Figure 5A). We found that larger subdivisions indeed slowed simulations\\ndown considerably, whereas frequent switching led to a less marked slowdown (Figure 6E). These results\\nsuggested a trade-off between performance and model accuracy.\\n\\nHence, our parallel CPM represents a significant speed improvement over even well optimized serial CPMs\\n\\n10\\n such as Morpheus. This improvement was to a substantial extent a result of GPU-specific optimizations\\nand led to an overall speedup of several orders of magnitude.\\n\\nBuilding an image-based model of T cell surveillance of the liver\\n\\nOur parallel CPM was motivated by the need for tissue simulations at realistic scale with realistic cell\\nmorphology. While our approach increases engineering complexity of the model, it can also simplify\\nmodel building, as it enables direct use of imaging data to initialize the simulation environment. To\\ndetermine the feasibility of directly using real-world imaging data in our system, we constructed a model\\nof T cells moving in the sinusoidal regions of the liver.\\n\\nTissue resident memory T cells (Trms) are confined to specific organs, where they sample their envi-\\nronment for signs of disease”+ — in particular, liver Trms are key in providing sterilising immunity for\\nmalaria.?° Trms display a distinct motility pattern in the liver, and efficacy of malaria clearing depends\\non the number of Trms present.?® We constructed a 3D simulation environment of liver sinusoids — capil-\\nlaries that Trms reside in and migrate through — and combined it with a polarized cell motility model” to\\nrepresent the Trms (Figure 7B). Trm motility parameters were derived from a LN model constructed in\\n\\nparallel, and our parameter choice will be explained in more detail with the LN data. These parameter\\nsettings lead to motility modes also observed in intravital liver data: intermittent movement and pausing\\nof Trms. The goal of our simulation was to predict how much of the liver volume would be explored by a\\ngiven number of T cells within a set amount of time; this same question was addressed by Fernandez-Ruiz\\net al. using a more abstract mathematical model.?°\\n\\nAs can be expected, we found that liver exploration by the Trm population increased with the number of\\n\\nexploring cells, and with the amount of exploration time (Figure 7C). Remarkably, the sinusoid volume\\nexplored by an individual cell increased dramatically as the total Trm population was raized from 1,000\\nto 10,000 cells (Figure 7D). Further investigation revealed the source of this phase transition: cells in\\nour simulations can be in an unpolarized state, in which they do not move persistently; collisions can\\ntrigger polarization of stationary cells. Since such collisions occur more frequently at higher densities,\\nthis causes a larger fraction of cells to be motile (Figure 7E). Thus, if Trms can help each other to\\nmove, a larger Trm population would screen the environment more efficiently than expected from the\\npopulation size alone.\\n\\nNotably, the mean speed of Trms peak at ~10.5 um/min (Figure 7F), matching average speeds in our\\nlymph node simulations — but only at densities where the arrest coefficient is close to 0. In two-photon\\nexperiments, liver Trm speed is reported to be significantly lower than T cells in lymph nodes;\\nreproduce this in our simulations with moderate cell density, which leads to both a realistic slowdown,\\n\\nand higher arrest coefficient. The simulated liver Trm speed and arrest coefficient closely matches previ-\\nously reported values in literature,?? where they were used to quantify observed characteristic patrolling\\nbehavior of Trms.\\n\\nThese liver experiments demonstrated that our system allowed us to combine 3D imaging data with a\\nrelatively realistic model of polarizing cell motility. The resulting simulation produced an interesting\\nphase transition of Trm screening efficiency. While we do not claim that such an effect is present in the\\n\\nreal liver — more detailed tuning of the motility model to match the characteristics of real sinusoidal Trm\\ncells would be out of scope for this simple proof of concept — this does illustrate how detailed models such\\nas the CPM generate emergent behavior that is not an immediately obvious consequence of the model\\ningredients and can generate interesting biological hypotheses worthy of further exploration.\\n\\nA T cell collective moves robustly in a realistic, organ-scale model of the lymph\\nnode\\n\\nIn LNs, millions of T cells move in search of antigen-presenting cells,?!°° and finding the cognate antigen\\n\\nis a key early step in adaptive immune responses.*! T cell motility in the LN is shaped by the high\\ndeformability of the cells®? and their interactions with a complex stromal environment, the fibroblastic\\nreticular cell network.*3:°4 Given the LN’s unique structure as an organ consisting mainly of moving\\ncells, it presents a challenging test case for any tissue simulation system. Indeed, many models of LN\\nT cell dynamics already exist, but they were often limited in either cell count and simulation volume, or in\\nrealism of cell morphology and LN environment. For example, T cell interactions with a synthetic stromal\\nstructure have been investigated using 3D CPM simulations of thousands of cells,° while collective motility\\n\\n11\\n at a more realistic scale of hundreds of thousands of cells has been investigated with non-morphological\\nagent-based models.*°:3° We were therefore interested in our system’s ability to simulate a realistically\\nlarge collective of millions of realistically shaped T cells. As in the previous use case, we used microscopy\\ndata to initialize — and parameterize — our model.\\n\\nWe constructed a LN simulation using the work of Beltman et al.9 as a starting point. To improve\\nrealism, we used an imaged LN section to initialize the stromal structure (Figure 8A), we used the\\nAct-CPM extension for cell motility,” and we fit the two Act motility parameters directly to intravital\\nimaging data (Figure 8B-E). When fitting these parameters, we aimed to reproduce the intermittent\\nstop-and-go motion®” observed in real data whilst also quantitatively approximate the real speed distri-\\nbution. Our system allowed fitting to be performed on realistically large simulations (256um diameter\\nsimulation volume, containing 50,135 T cells, 20 minutes simulated time). Crucially, we did not use\\nperiodic boundaries which, in dense simulations, can lead to artificial global synchronisation of motion\\nin the cell collective.\\n\\nWe used the KL divergence to compare simulated to real speed distributions (Figure 8B). Increased\\nadhesion of T cells to the stromal structure made the speed distribution more right-skewed (Figure 8C,\\ntop), which was more consistent with real data (Figure 8C bottom). For most examined parameter\\nsettings, we found the pronounced variations in cell speed along cell trajectories that are also observed\\nin real data (Figure 8D). When we performed grid searches of motility parameters at two different\\nstromal adhesion strengths (Figure 8E), both grid searches had broadly similar fit landscapes, and the\\noverall lowest KL divergence values were very similar. Qualitatively, we found more right-skewed and\\nbimodal speed distributions in high-adhesion than low-adhesion simulations for a subset of the motility\\nparameters (Supplementary Figure $2). Based on these fitting results, we chose the best fit motility\\nparameters of the high adhesion setting for our further simulations.\\n\\nWe next investigated organ-scale collective T cell motility by scaling up our simulated volume to 1mm\\ndiameter, now containing 3,359,752 T cells (Figure 9A). While this is not entirely realistic — a real LN\\nhas a more complex compartmental structure, while we simulate only a T cell zone — our main intention\\nwas to determine the spatial range of collective behavior that can now be investigated with our system.\\nWe simulated one virtual day in 1.9 days of wall clock time — where serial implementations were projected\\n‘o have taken a full 1.4 years (Figure 1A).\\n\\nWe used a mixing simulation of two differentially labelled populations as an easily observable proxy for\\ncollective search (Figure 9B). In the simulation, it takes about 1 hour to reach a state where 50% of the\\ncells are next to at least one cell of the other population (Figure 9C). While stromal adhesion influences\\nintra-cellular stop-and-go motility, it has minimal influence on collective motion: high and low stromal\\nadhesion simulations have broadly similar mixing speeds. While collective motility is slightly ordered at\\ncell-scale distances, no ordering persists at organ-scale (Figure 9D).\\n\\nIn summary, we used an organ-scale LN simulation as a challenging test case: an environment known for\\ncomplex, large scale collective motility dynamics. Our system allowed us to integrate of microscopy data\\ndirectly into our simulation, and to fit motility parameters directly on two-photon data. We investigated\\nthe behavior of the fitted cell population at organ-scale (millions of cells). We hope that this will inspire\\nfurther research on large cell collectives — integrating various sources of empirical data, at a high level of\\nrealism, and at realistic scale.\\n\\nDiscussion\\n\\nWe have implemented a parallel version of the CPM that, in some experiments, was 3 to 4 orders\\nof magnitude faster on GPUs than a state-of-the-art serial CPM. To achieve this, we designed novel\\nschemes for lattice work splitting and synchronization, maintaining global state, efficient GPU memory\\ntransfer, and multi-GPU execution. We found that preservation of local lattice update statistics was\\nsufficient to maintain high-level system dynamics, despite removing the sequential nature of the original\\nCPM algorithm.\\n\\nIn our work we aimed to ensure that important properties of CPM simulation trajectories were main-\\ntained. Some previous CPM parallelisation efforts instead focused on equilibrium state distributions as\\nimportant quantities. For example, Chen et al.!? argue that their parallel CPM maintains certain equi-\\nlibrium properties. While such arguments are frequently applied in the analysis of energy-based models\\n\\n12\\n in physics, Voss-Béhme et al. showe\\n\\nWhile we have shown that the same\\nCPM in our benchmark scenarios, thi\\nparameters that were originally fitte\\nto re-fit their simulation parameters\\n\\nespecially with com\\n\\nplex models such\\n\\nthat these are less relevant for the CPM because the Metropolis\\nalgorithm used in the CPM drastically alters longtime behavior of the mode\\n\\nrange of behaviors is present in our parallel CPM as in a serial\\ns would still need to be carefully checked when re-using simulation\\non a serial implementation. Specifically, researchers might want\\nwhen using our system instead of a serial\\nis sensible in general when moving between different implementations of the s\\nas the CPM where implementation details might anyway differ\\n\\nversion — advice that\\name model formalism,\\n\\nsubtly. Furthermore, the observed relation between model kinetics and performance (Figure 6E) also\\nsuggests it might be fruitful to explore more aggressive parallelisation settings that might yield further\\n\\nspeed-up whilst maintaining desired simulated behavior.\\n\\nThrough accelerating the CPM on accessible GPU hardware, we have increased the range of feasible\\nsimulation sizes — because our scheme fully utilises all GPU cores, simulation size is now mostly limited\\n\\nby the available GPU memory. As memory per GPU increases, our scheme wi\\nti-GPU prototype scheme will allow for further increase of simulation volume. Our\\nto include inter-GPU communication over MPI,*° allowing\\nding block for a GPU cluster-based version of the CPM —\\nher to lcm? scale.\\n\\nsimulations; our mul\\n\\nmulti-GPU scheme could be further extende\\n\\nour system to be us\\npotentially scaling o\\n\\nLarge-scale multicel\\nenvironments.\\nare morphologically\\ncell shape into acco\\nour example simula’\\n\\n14, 15, 35, 36\\n\\ned as a fundamental bui\\n\\nur Imm? simulations fur\\n\\n1 easily scale to larger\\n\\nular simulations are often performed using more coarse-grained agent-based model\\nOur high-performance CPM now allows researchers to perform simulations that\\nmuch more detailed and allow the investigation of new questions that require taking\\nunt, such as migration in complex environments or morphogenesis. We hope that\\nions of liver and LN environments will serve as inspiration for many such spatially\\n\\nresolved tissue and organ simulations in the future. Similar approaches to the one shown here could be\\nuseful for parallelisation of CPM variants or other sequential lattice-based models.®\\n\\n13\\n A B mmm 100 yum explored space @ T cells\\n\\nfe)\\n\\n3 g 16\\n\\nSs Sa\\n\\na ao\\n\\ng 38 1.0\\n\\noO oO\\n\\n°° oa\\n\\n3 % 0.5\\n\\ncy Cy\\n\\nS 10 1000 100000\\nE i F cells\\n\\n_ 1 10.0\\n\\n5 0.8 =\\n\\n2 € 75\\n\\n@ 0.6 E so\\n\\n204 3\\n\\n3 3\\n\\n0.2 Q 25\\n\\n0.0 0.0\\n10 1000 100000 10 1000 100000\\ncells cells\\n\\nFigure 7: T cells in a liver sinusoid simulation exhibit a transition from a non-motile to\\na motile state. (A) A 1024x1024x182 um® scan of a liver sinusoid is directly imported into our\\nsimulation environment (64ym slice shown). (B) Simulated T cells explore liver sinusoids (5000 cell\\nsimulation shown). (C) Simulations at different cell densities explore the environment at different speeds\\n(single simulation of 24hr, with 1 min snapshots). (D) An increase of explored volume per cell occurs\\nwhen cell density rises from 1,000 to 10,000 cells; 100,000 cells fill ~1/3 of sinusoid volume, causing\\nexploration efficiency to saturate. (E) At low cell densities most simulated cells are in an un-polarized,\\nnon-motile state; at higher densities collisions cause cells to transition in to a polarized, motile state (1\\nsimulation of 30 min per density, arrest coefficient: fraction cells ; 2um/min,?? shaded: Wilson score\\ninterval based on the number of 2-minute subtracks). CPM parameter settings can be found in Table 3.\\n(F) Average speed of the simulated T cells at different densities.\\n\\n14\\n 1-15 ym 16-30 pm 31-45 im 46-60 pm 61-75 um 76-90 um =: 91-110 um\\n\\nB simufation =e D jowadhesion E — towadhesion low adhesion\\n2 2 ey\\n5 5 Ey\\n: :\\n8 8\\na a s\\n2 2 Da\\n3 3 a 5\\n0 10 20 30 0 10 20 30 SOHm =m 3\\nspeed (\\\\um/min) speed (um/min) high adhesion & high adhesion high adhesion\\noO\\nEy\\no oS\\n8 0.40 @ 0.25 ee eh cc\\n5, = 0.20 a\\n® 0.05 20.15\\nES $ 0.10\\n3 0.00 ge 3 os\\nZ 0.00\\n10 20 30 adhesion o 4 4. ° % polarisation decay 10 30\\nspeed (um/min) speed (um/min)\\n\\nKL divergence\\n\\nFigure 8: Reproducing in vivo T cell motility characteristics in a CPM. We compare intravital\\ndata from Mandl et al.?!. to the motion of 50,185 CPM cells moving in a spherical structure (256um\\ndiameter). (A) A 130x130x110 um* scan of the stromal structure (fibroblastic reticular cells) in the\\n\\nparacortical T cell zone of a murine LN. The images show average intensity projections of diffe\\nranges. After segmentation and mirroring, these data are used to fill a spherical simulation vo\\n\\nrent depth\\nlume. (B)\\n\\nSpeeds observed in simulated data are used to construct a probability distribution (upper panel, solid),\\nwhich we compare to the empirical distribution (upper panel, dashed) using the Kullback-Leibler (KL)\\ndivergence metric. To this end, we bin the speed distributions, compute the divergence in each bin, and\\nsum the values for each bin (lower panel). (C) Increased adhesion of T cells to stromal cells causes\\n\\nthe speed distributions to be more right-skewed (top) and match real data more closely (bot:\\n\\nom). (D)\\n\\nIncreased adhesion can slow T cell paths non-uniformly by creating longer and/or more frequent pauses\\n\\n(red) but does not prevent rapid motion altogether. Two representative tracks are shown. (\\n\\nE) A grid\\n\\nsearch of motility parameters for cell polarization reveals which parameter ranges best fit the intravital\\n\\ndata. Both low and high adhesion simulations have similar fit landscapes; the best fit is\\n\\nound in a\\n\\nhigh adhesion simulation, although the difference to the low-adhesion best fit is small (1 simulation per\\n\\nparameter combination, best fit: red box/red line). CPM parameter settings can be found in\\n\\n15\\n\\nTable 3.\\n 3,359,752 simulated cells\\n1pm simulation resolution\\n\\nO mins 40 mins 80 mins 120 mins\\n\\nLe)\\ni=]\\n\\n0.8\\nD 06 5 0.6\\n& 0.4 304\\nE02 20.2\\n0.0 0.0\\n\\n0 30 60 90 120 30 100 300 1000\\n\\ntime (in min) scale (in um)\\n\\nhigh adhesion low adhesion\\n\\nFigure 9: Mixing of motile T cell collectives in a realistic-scale LN simulation. (A) Spherical\\nLN simulation at 1mm diameter containing 3,359,752 T cells and stromal cells (grey; Figure 8A). For\\nclearer visualization, only 10% of the simulated cells are shown, and red and green cells are equally param-\\neterized although coloured differently. (B) The simulation is initialized with two equally parameterized\\npopulations each starting at one half of the sphere. Mixing is quantified by tracking the fraction of cells\\nhaving at least one neighbour of the other population. (C) Global mixing is slower in the high-adhesion\\nsimulation (red) than at low adhesion (black), although the difference is minor. Both simulations reach\\n50% mixing in ~1h. (D) Cells do not synchronize their directionality much at longer spatial scales, and\\nthis absence of long-range synchronization is the same for low and high adhesion. CPM parameter set-\\ntings can be found in Table 3.\\n\\n16\\n Tables\\n\\nTable 1: Model settings for 2D simulations. For all 2D simulations we use Viarget = 2500, Atarget =\\n600, Aperimeter = 2; Jcett,ecm = 10 and T = 10. The vector field in the 2D circular chemotaxis simulations\\n(Figure 5A) consists of vectors at a 90 degree angle to radial vectors from points in the field to the\\nsimulation center, with all vectors in the field normalized to length 1.\\n\\n| Avotume | Aaet | Maxact | Achemotaxis\\nFig 1B 20, 30, 40 | 200, 250, 300, 350, 400 50 —\\nFig 5A, Fig 6E | 20 = — 10\\nFig 5B 20 50-800, in increments of 50 | 50 —\\n\\nTable 2: Model settings for 3D simulations. For all simulations Viargee = 150, Avolume = 25,\\nAtarget = 1500, Aperimeter = 0.2, Scett,ce = 10 and T = 20. The time resolution is 1 second per Monte\\nCarlo step.\\n\\n| uw _| At | Apersistence | Jcell,BCM\\nFig 1A — f—|[— 10\\nFig 5C, Fig 6D | 0.95 | 15 | 40 0\\n\\nTable 3: Model settings for liver and lymph node simulations. For all simulations Viarge, = 150,\\nAvolume = 25, Atarget = 1500, Aperimeter = 0-2, Jcett,cett = 10, Jeet,ncm = 0, and T = 20. The time\\nresolution is 1 second per Monte Carlo step.\\n| Aact | Maxact Teell, FRC\\n\\nFig 7 2600 10 0\\n\\nFig 8B | 2600 10 0\\n\\nFig 8C | 2600 10 -400, -300, -200, -100, 0\\n\\nFig 8D | 2600 10 -400, 10\\n\\nFig 8E | 200-3000, in increments of 200 | 10, 15, 20, 25, 30, 40, 50, 60, 70 | -3000, 10\\n\\nFig 9C | 2600 10 -800, 0\\n\\nFig 9D | 2600 10 -800\\n\\n17\\n Acknowledgments\\n\\nThis research was supported by Young Investigator Grant 10620 from KWF Kankerbestrijding and HFSP\\nprogram grant RGP0053/2020 (to JT). JT was also supported by NWO Vidi grant VI.Vidi.192.084.\\nWe thank Franka Buytenhuijs for giving valuable feedback on this manuscript, and we thank L.W.A.\\nStuurman and Pedro Aroca Lara for testing early versions of this simulation system. We also gratefully\\nacknowledge the support of NVIDIA Corporation with the donation of the Titan V GPU used for this\\nresearch. The funders had no role in design, execution, and interpretation of the research reported in\\nthis manuscript.\\n\\nAuthor contributions\\n\\nSS and JT conceived the study. SS designed the parallel CPM algorithms. SS performed the research and\\nanalysed the data. SS and JT wrote the manuscript. JT supervized the project. SM provided critical\\nfeedback and provided the liver and lymph node imaging data. SD performed image acquisition of the\\nliver and LN data. All authors critically revized the manuscript for intellectual content.\\n\\nDeclaration of interests\\n\\nThe authors declare no competing interests.\\n\\n18\\n Materials and Methods\\n\\nCellular Potts model and model variants\\n\\nThe CPM is a lattice based formalism for simulating biological cells. Simulations proceeds with a Markov\\nchain Monte Carlo process, where a single position at a time can transition to one of the states of its\\nneighbours. The transition probability is based on the change of a Hamiltonian energy function. The\\noriginal CPM has\\n\\nH =~ Ip(ox),r(0))(1 — 84:,0;) + ¥ > Avotume(Te)(V (0) — Veargot(Te))? (1)\\nig o\\n\\nas its Hamiltonian, with i, 7 being possible neighbour locations in the lattice, 7; the cell present at location\\ni, T is the type of a particular cell, 5 is the Kronecker delta, V(c) is the number of lattice sites a cell o\\noccupies, and Ayolume(T) how strongly a volume constraint contributes to the Hamiltonian for a particular\\ncell type. In all our simulations a Moore neighbourhood was used. Transitions stochastically succeed\\nbased on the change in energy AH, using a Boltzmann probability distribution\\n\\n1 if AH <0,\\nP(AH) = {laws if AH > 0. (2)\\n\\nBy adding extra terms to the Hamiltonian, constraints can be added to cells and new types of behavior\\nintroduced to the model. When cell shape needs to be parameterized beyond just cell size, a perimeter\\nconstraint*® is often introduced,\\n\\nHperimeter = Ss Aperimeter (To )(A(a) _ Atarget (To). (3)\\n\\nHere A(c) represents the number of interfaces between lattice sites of a cell with sites not part of the\\nsame cell.\\n\\nWe also implemented several motility models. The first allows a vector field to bias the energy change\\nAH during a copy attempt\\n\\nAH chemotaxis = Achemotaxis (To )#i+7 * Ci. (4)\\n\\nHere #;_,; is the unit vector pointing in the direction of a copy attempt from position i to 7, and C; the\\nvalue of a vector field at position 7. In this paper, C; is always a field of unit vectors.\\n\\nFor the model where cells exhibit persistent movement based on one vector per cell, we add\\nAHpersistence = Apersistence (To )#i+5 * Po,t (5)\\n\\nto the Hamiltonian delta. The vector ,,, is the persistence direction of cell o at time t, which is updated\\niteratively at every MCS with\\n\\ney Xo ~~ X, t—At a\\nDot = (1 = pr) —S =— + flr Pot—1s (6)\\n|Xot — Xot-atll\\n~ Dot\\nPot = 7° (7)\\n\\\\|Do.tll\\n\\nHere ju, is the strength of directional persistence for cell type T, Xot is the centroid location of cell o at\\ntime t and At is the number of MCSs we look back for to determine the current direction of a cell.\\nFinally, we implement the Act model’ that allows for localized polarization. For each position i we track\\nan activity value each time said location is updated, here labelled as Act(i). This localized ‘memory’ is\\ninitialized at Maxact, and decays by 1 each timestep until it reaches 0. The Hamiltonian delta is adapted\\nby\\n\\nact (To)\\nMaxact (To)\\n\\nAHact (GMact (?) — GMact (J)), (8)\\n\\n19\\n making copies that originate from areas of a cell with a high activity value more likely. The activity is\\ncalculated for a full Moore neighbourhood using a geometric mean\\n\\nGMaci(i) = ( T] Act(n)) Ol. (9)\\n\\nn€N(i)\\n\\nN(i) are all neighbours of pixel or voxel i.\\n\\nOptimized CPU implementation of CPM\\n\\nBesides a high-performance GPU implementation, we also developed our own CPU implementation. This\\nserved both as a reference implementation for model behavior and internal implementation details for\\nus, and as a reference performance benchmark. We strived to reduce the major performance bottlenecks\\nwe could see in profiling. The implementation tracks lattice sites on cell boundaries, and only tries\\ncopy attempts involving such positions; a common optimisation, which is also implemented in Morpheus.\\nThe tracking implementation uses a high-performance hashmap implementation.*! We use a custom\\nhigh-performance pseudorandom number generator, xoshiro256.4?_ The lattice data is stored with a Z-\\norder space filling curve, allowing data that is close in the simulation space also to be close in the\\nlinear memory space, maximizing cache efficiency. Memory addresses are calculated using intrinsic bit\\ninterleaving instructions from the x86 bit manipulation instruction set extension.\\n\\nGPU implementation details\\n\\nOur CPM was implemented using CUDA; concepts equivalent to ones described here exist in non-CUDA\\nGPU hardware. A GPU consists of multiple streaming multiprocessors (SMs). Each SM runs multiple\\nthreads in parallel; contrasting with CPU threads, threads on the same SM execute in lockstep, in a\\nSingle Instruction Multiple Threads approach. A single GPU process — called a kernel — can consist of\\nup to millions of threads.\\n\\nGPUs have relatively slow global memory available to all threads, and much faster shared memory local\\nto SMs. Global memory transfer speed is often the main performance bottleneck for GPU programs.\\nThreads are bundled in thread blocks (TBs); each TB is bound to a specific SM. By caching data on\\ntheir SM in shared memory, TBs can greatly reduce global memory transfer bottlenecks.\\n\\nMemory transfer bottlenecks can be further reduced by two strategies: read coalescing and high SM\\noccupancy. Read coalescing has sequential threads in a TB do contiguous reads of global memory;\\ncontiguous reads are needed to fully utilise the memory bus bandwidth. High occupancy is achieved when\\na high number of threads are scheduled to execute per SM; the SM scheduler will switch out threads\\nwhen they are blocked on memory transfers, ensuring a high number of in-flight memory instructions at\\na time.\\n\\nWe propose a scheme of memory transfer, local SM execution and synchronization (Figure 2D). We\\nfocus on caching just the grid of cell ids, as this data is always needed, and can be used to quickly reject\\ncopy attempts internal to a cell. The simulation grid is divided in subsections, with regions larger than\\nearlier checkerboard subdivisions; each subsection is assigned to a TB, subsection size is chosen so active\\nregions fit in shared memory. Each TB determines which checkerboard sections are active in its region;\\nactive sections are cached in shared memory using coalesced reads.\\n\\nAfter caching — intended to speed up reads — the simulation is run, and successful copy attempts write\\nnew values to both shared and global memory. Alternatively, changes could be cached in shared memory\\nand written back coalesced to global memory; in our testing, an immediate shared/global write yielded\\nbetter performance. If a TB is done with simulation execution, it immediately starts a memory request\\nfor the next active region, ensuring maximum occupancy. All threads in the kernel are synchronized after\\nnew memory transfers are finished, ensuring maximum overlap of code execution and memory transfer.\\n\\nCPM literature offers a wide range of model variants and extensions, of which we implement a mod-\\nest subset. This comes with potential performance issues: even when extensions are disabled through\\nbranching statements, on GPUs such branching code still incurs a significant performance loss compared\\nto CPUs. We therefore pre-generated specialized versions of the kernel for a range of different extensions.\\n\\nWe use a custom high-performance random number generator on the GPU.*%\\n\\n20\\n Composite variable tracking\\n\\nSeveral CPM extensions introduce additional global variables. Some of these variable need to be kept\\nin-syne with each other — one variable being slightly out of date with others is not tolerated. One\\nexample is tracking of a center-of-mass centroid; for 2D this can be tracked as a sum of x and y values,\\nand corresponding count of lattice locations occupied (Figure 2C). These are integer values: tracking\\ncentroid locations as floating-point would cause values to drift due to rounding errors. The centroid value\\nis calculated by\\n\\n1\\n{a,y} = = X {asum, Ysum }- (10)\\n\\nOnline updates change both a cell’s centroid position and size. When a new lattice position {Xposition, Yposition }\\nis added to a cell, its centroid’s x value can be updated with\\n\\nLoia X |® lola + Lposition\\n\\n(11)\\n\\nZnew lel\\nnew\\n\\nCell sizes are stored interleaved with sum values; sizes and sums are both stored as 32-bit unsigned\\nintegers. The interleaved storage enables use of 64-bit operations to update both: a copy-and-swap\\n(CAS) operation is used, which atomically updates sum and cell size values simultaneously. This scheme\\nalso handles periodic boundaries, by adding or subtracting dimension x |e| to the sum values before\\nexecuting the CAS as needed, to keep centroids within simulation bounds.\\n\\nMulti-GPU execution\\n\\nWe devized a scheme to execute our CPM on multiple GPUs (Figure 2E); this scheme is intended as a\\nprototype, and as such not as optimized as other parts of our CPM. Simulation space is cut into equal\\nregions, each region is assigned to a GPU. A region’s lattice data, such as the cell ids, are sent to each\\nGPU, with an additional border layer. Global variables such as cell sizes are tracked on each GPU as\\ndeltas of the value since last inter-GPU data synchronisation. Every four MCSs, these deltas are gathered\\nfrom all GPUs, deltas are integrated, and updated values distributed to each GPU. Multi-GPU speedup\\nis mostly hampered by frequent or slow inter-GPU communication; inter-GPU synchronization after every\\nMCS negates most speedup, and a faster GPU-to-GPU interlink would likely increase performance.\\n\\nHardware\\n\\nThe CPU-based simulations, using our custom optimized CPU implementation and Morpheus 2.2.6 (the\\nlast non-parallelized version of Morpheus), and most GPU simulations were run on a system with a\\nNvidia Titan V, an AMD 2950x CPU and 128 GB of RAM at 2133 MHz. The only exceptions to this\\nare the multi-GPU simulations (Figure 2E), which were run on a system with four Nvidia Geforce RTX\\n2080 Ti GPUs, an Intel i9-9820x CPU at 3.3 GHz, and 128 GB of RAM at 2133 MHz.\\n\\nImaging of thick lymph node sections\\n\\nNaive inguinal LNs were harvested from CCL19-EYFP mice“ and fixed with 4% paraformaldehyde (PFA,\\nElectron Microscopy Sciences) for 1 hour at 4°C. Following a brief wash with PBS, LNs were embedded in\\n2% agarose and 250-300 ym tissue sections were generated using a vibratome (Leica). LN sections were\\nsubsequently immersed in a Special Wash Buffer (Baxter water, 0.2% BSA, 10% 10X PBS, 0.2% Tween-\\n20, 0.2% Triton X-100, 0.2% 10% SDS) at 4°C overnight, followed by staining with a rabbit anti-GFP Ab\\n(Thermo Scientific, 1:100, diluted in blocking buffer made with Baxter water, 0.2% BSA, 10% 10X PBS,\\n0.1% Tween-20, 0.2% 10% SDS) at 4°C for three days. After incubation, LN sections were thoroughly\\nwashed in PBS and mounted in 85% glycerol in PBS with glass coverslips. Thick LN sections were imaged\\nusing a LSM880 AiryscanFast confocal microscope (Carl Zeiss) with a x63/1.4NA oil immersion lens and\\nimages processed with ZEN 2010 software (Carl Zeiss).\\n\\nImaging of liver sinusoids\\n\\nFor whole-liver imaging, AlexaFluor 647 conjugated anti-CD31 rat monoclonal antibody (Biolegend) was\\nintravenously injected into wild-type C57BL/6 mice at 5ug for 15 minutes prior to harvesting the liver.\\n\\n21\\n\""
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "article_txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d2da5df0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "authors = [\"Shabaz Sultan\", \"Sapna Devi\", \"Scott N. Mueller\", \"Johannes Textor\"]\n",
    "title = \"A parallelized cellular Potts model that enables simulations at tissue scale\"\n",
    "neo4j_query(\"\"\"\n",
    "MERGE (b:Article {title: $title})\n",
    "WITH b\n",
    "UNWIND $authors AS authorName\n",
    "MERGE (a:Author {name: authorName})\n",
    "MERGE (a)-[:WROTE]->(b)\n",
    "\"\"\", {'title': title, 'authors': authors})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4b576704",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neo4j_query(\"\"\"\n",
    "MATCH (a:Article {title: $title})\n",
    "WITH a\n",
    "OPTIONAL MATCH (a)-[:HAS_SENTENCE]->(s:Sentence)\n",
    "WHERE s IS NULL\n",
    "WITH a\n",
    "UNWIND $data as row\n",
    "MERGE (s:Sentence {id: row.text_sha256})\n",
    "SET s.text = row.text\n",
    "MERGE (a)-[:HAS_SENTENCE]->(s)\n",
    "WITH s, row.entities as entities\n",
    "UNWIND entities as entity\n",
    "MERGE (e:Entity {id: entity.entity_id})\n",
    "ON CREATE SET e.other_ids = entity.other_ids,\n",
    "              e.name = entity.entity,\n",
    "              e.type = entity.entity_type\n",
    "MERGE (s)-[m:MENTIONS]->(e)\n",
    "ON CREATE SET m.count = 1\n",
    "ON MATCH SET m.count = m.count + 1\n",
    "\"\"\", {'data': parsed_entities, 'title': title})"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e4a09e27",
   "metadata": {},
   "source": [
    "Query from inside the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6da61036",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Such divergence could lead to changed simulati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The most important method, that is the main fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The mesenchymal stem cell-based therapeutic me...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              result\n",
       "0  Such divergence could lead to changed simulati...\n",
       "1  The most important method, that is the main fo...\n",
       "2  The mesenchymal stem cell-based therapeutic me..."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neo4j_query(\"\"\"\n",
    "MATCH (e:Entity)<-[:MENTIONS]-(s:Sentence)\n",
    "WHERE e.name = \"cells\"\n",
    "RETURN s.text as result\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b6199e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = RelTaggerModel.from_pretrained(\"fractalego/fewrel-zero-shot\")\n",
    "print(\"model loaded\")\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"bert-large-uncased-whole-word-masking-finetuned-squad\")\n",
    "print(\"token loaded\")\n",
    "relations = ['associated', 'interacts']\n",
    "extractor = RelationExtractor(model, tokenizer, relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cfe291f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
